<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>The moiraine R package user manual - 13&nbsp; Comparing the integration results</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./evaluation.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Comparing the integration results</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The moiraine R package user manual</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/PlantandFoodResearch/moiraine" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="https://twitter.com/intent/tweet?url=%7Curl%7C" title="Twitter" class="sidebar-tool px-1"><i class="bi bi-twitter"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./example_dataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">The example dataset</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Data preparation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_import.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Importing data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inspecting_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Inspecting the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./modifying_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocessing.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data pre-processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prefiltering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Dataset pre-filtering</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Unsupervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Integration with sPLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./so2pls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Integration with sO2PLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mofa.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Integration with MOFA</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Supervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./diablo.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Integration with DIABLO</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Results interpretation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./interpretation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evaluation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./comparison.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Comparing the integration results</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="toc-section-number">13.1</span>  Introduction</a></li>
  <li><a href="#extracting-integration-results" id="toc-extracting-integration-results" class="nav-link" data-scroll-target="#extracting-integration-results"><span class="toc-section-number">13.2</span>  Extracting integration results</a></li>
  <li><a href="#clustered-correlation-heatmaps" id="toc-clustered-correlation-heatmaps" class="nav-link" data-scroll-target="#clustered-correlation-heatmaps"><span class="toc-section-number">13.3</span>  Clustered correlation heatmaps</a></li>
  <li><a href="#a-note-about-missing-features" id="toc-a-note-about-missing-features" class="nav-link" data-scroll-target="#a-note-about-missing-features"><span class="toc-section-number">13.4</span>  A note about missing features</a></li>
  <li><a href="#correlation-matrices" id="toc-correlation-matrices" class="nav-link" data-scroll-target="#correlation-matrices"><span class="toc-section-number">13.5</span>  Correlation matrices</a></li>
  <li><a href="#pairwise-comparisons" id="toc-pairwise-comparisons" class="nav-link" data-scroll-target="#pairwise-comparisons"><span class="toc-section-number">13.6</span>  Pairwise comparisons</a></li>
  <li><a href="#comparing-samples-score" id="toc-comparing-samples-score" class="nav-link" data-scroll-target="#comparing-samples-score"><span class="toc-section-number">13.7</span>  Comparing samples score</a></li>
  <li><a href="#comparing-features-weight" id="toc-comparing-features-weight" class="nav-link" data-scroll-target="#comparing-features-weight"><span class="toc-section-number">13.8</span>  Comparing features weight</a></li>
  <li>
<a href="#sec-comparison-consensus" id="toc-sec-comparison-consensus" class="nav-link" data-scroll-target="#sec-comparison-consensus"><span class="toc-section-number">13.9</span>  Feature consensus importance</a>
  <ul class="collapse">
<li><a href="#possible-metrics" id="toc-possible-metrics" class="nav-link" data-scroll-target="#possible-metrics"><span class="toc-section-number">13.9.1</span>  Possible metrics</a></li>
  <li><a href="#computing-features-consensus-importance-score" id="toc-computing-features-consensus-importance-score" class="nav-link" data-scroll-target="#computing-features-consensus-importance-score"><span class="toc-section-number">13.9.2</span>  Computing features’ consensus importance score</a></li>
  </ul>
</li>
  <li><a href="#recap-targets-list" id="toc-recap-targets-list" class="nav-link" data-scroll-target="#recap-targets-list"><span class="toc-section-number">13.10</span>  Recap – targets list</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-comparison" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Comparing the integration results</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/targets/">targets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PlantandFoodResearch/moiraine">moiraine</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For data-frames manipulation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For colour palettes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As different integration methods use different statistical approaches and make different assumptions about the data, it can be very interesting to apply them to a same multi-omics dataset and compare the results. In this chapter, we will see how the <code>moiraine</code> package can be used to compare the results from different integration methods.</p>
<p>As a reminder, here is what the <code>_targets.R</code> script should look like so far:</p>
<details><summary><code>_targets.R</code> script
</summary><div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(targets)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tarchetypes)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(moiraine)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">tar_option_set</span>(</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">packages =</span> <span class="fu">c</span>(</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"moiraine"</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MOFA2"</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"mixOmics"</span>,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"readr"</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"tibble"</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"tidyr"</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"dplyr"</span>,</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ggplot2"</span>,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">"patchwork"</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="fu">list</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Importing data ---------------------------------------------------------------</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Data import using a target factory</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_dataset_csv_factory</span>(</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">c</span>(</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/genomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/transcriptomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/metabolomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>)</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="fu">c</span>(<span class="st">"marker"</span>, <span class="st">"gene_id"</span>, <span class="st">"sample_id"</span>),</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">features_as_rowss =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>),</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="fu">c</span>(<span class="st">"geno"</span>, <span class="st">"transcripto"</span>, <span class="st">"metabo"</span>)</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Genomics features metadata file</span></span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>    fmetadata_file_geno,</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system.file</span>(<span class="st">"extdata/genomics_features_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"file"</span></span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Genomics features metadata import</span></span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>    fmetadata_geno,</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">import_fmetadata_csv</span>(</span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>      fmetadata_file_geno,</span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_id =</span> <span class="st">"marker"</span>,</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_types =</span> <span class="fu">c</span>(<span class="st">"chromosome"</span> <span class="ot">=</span> <span class="st">"c"</span>)</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Metabolomics features metadata import</span></span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_fmetadata_csv_factory</span>(</span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">c</span>(</span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/metabolomics_features_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>)</span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="fu">c</span>(<span class="st">"feature_id"</span>),</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="fu">c</span>(<span class="st">"metabo"</span>)</span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-61"><a href="#cb2-61" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Transcriptomics features metadata import</span></span>
<span id="cb2-62"><a href="#cb2-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_fmetadata_gff_factory</span>(</span>
<span id="cb2-63"><a href="#cb2-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">system.file</span>(<span class="st">"extdata/bos_taurus_gene_model.gff3"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-64"><a href="#cb2-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature_types =</span> <span class="st">"genes"</span>,</span>
<span id="cb2-65"><a href="#cb2-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">add_fieldss =</span> <span class="fu">c</span>(<span class="st">"Name"</span>, <span class="st">"description"</span>),</span>
<span id="cb2-66"><a href="#cb2-66" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="st">"transcripto"</span></span>
<span id="cb2-67"><a href="#cb2-67" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-68"><a href="#cb2-68" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-69"><a href="#cb2-69" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Samples metadata import</span></span>
<span id="cb2-70"><a href="#cb2-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_smetadata_csv_factory</span>(</span>
<span id="cb2-71"><a href="#cb2-71" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">system.file</span>(<span class="st">"extdata/samples_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-72"><a href="#cb2-72" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="st">"animal_id"</span>,</span>
<span id="cb2-73"><a href="#cb2-73" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="st">"all"</span></span>
<span id="cb2-74"><a href="#cb2-74" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating omics sets for each dataset</span></span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a>  <span class="fu">create_omics_set_factory</span>(</span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a>    <span class="at">datasets =</span> <span class="fu">c</span>(data_geno, data_transcripto, data_metabo),</span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a>    <span class="at">omics_types =</span> <span class="fu">c</span>(<span class="st">"genomics"</span>, <span class="st">"transcriptomics"</span>, <span class="st">"metabolomics"</span>),</span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a>    <span class="at">features_metadatas =</span> <span class="fu">c</span>(fmetadata_geno, fmetadata_transcripto, fmetadata_metabo),</span>
<span id="cb2-81"><a href="#cb2-81" aria-hidden="true" tabindex="-1"></a>    <span class="at">samples_metadatas =</span> <span class="fu">c</span>(smetadata_all, smetadata_all, smetadata_all)</span>
<span id="cb2-82"><a href="#cb2-82" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-83"><a href="#cb2-83" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-84"><a href="#cb2-84" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating the MultiDataSet object</span></span>
<span id="cb2-85"><a href="#cb2-85" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-86"><a href="#cb2-86" aria-hidden="true" tabindex="-1"></a>    mo_set,</span>
<span id="cb2-87"><a href="#cb2-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">create_multiomics_set</span>(</span>
<span id="cb2-88"><a href="#cb2-88" aria-hidden="true" tabindex="-1"></a>      <span class="fu">list</span>(set_geno,</span>
<span id="cb2-89"><a href="#cb2-89" aria-hidden="true" tabindex="-1"></a>           set_transcripto,</span>
<span id="cb2-90"><a href="#cb2-90" aria-hidden="true" tabindex="-1"></a>           set_metabo)</span>
<span id="cb2-91"><a href="#cb2-91" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-92"><a href="#cb2-92" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-93"><a href="#cb2-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-94"><a href="#cb2-94" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspecting the `MultiDataSet` object -----------------------------------------</span></span>
<span id="cb2-95"><a href="#cb2-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-96"><a href="#cb2-96" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating a density plot for each dataset</span></span>
<span id="cb2-97"><a href="#cb2-97" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-98"><a href="#cb2-98" aria-hidden="true" tabindex="-1"></a>    density_plots,</span>
<span id="cb2-99"><a href="#cb2-99" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_density_data</span>(</span>
<span id="cb2-100"><a href="#cb2-100" aria-hidden="true" tabindex="-1"></a>      mo_set,</span>
<span id="cb2-101"><a href="#cb2-101" aria-hidden="true" tabindex="-1"></a>      <span class="at">combined =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-102"><a href="#cb2-102" aria-hidden="true" tabindex="-1"></a>      <span class="at">scales =</span> <span class="st">"free"</span></span>
<span id="cb2-103"><a href="#cb2-103" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-104"><a href="#cb2-104" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-105"><a href="#cb2-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-106"><a href="#cb2-106" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the relationship between features mean and standard deviation</span></span>
<span id="cb2-107"><a href="#cb2-107" aria-hidden="true" tabindex="-1"></a>  <span class="do">## for each dataset</span></span>
<span id="cb2-108"><a href="#cb2-108" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-109"><a href="#cb2-109" aria-hidden="true" tabindex="-1"></a>    mean_sd_plots,</span>
<span id="cb2-110"><a href="#cb2-110" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_meansd_data</span>(mo_set)</span>
<span id="cb2-111"><a href="#cb2-111" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-112"><a href="#cb2-112" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-113"><a href="#cb2-113" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Assessing missing values</span></span>
<span id="cb2-114"><a href="#cb2-114" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-115"><a href="#cb2-115" aria-hidden="true" tabindex="-1"></a>    n_missing_values,</span>
<span id="cb2-116"><a href="#cb2-116" aria-hidden="true" tabindex="-1"></a>    <span class="fu">check_missing_values</span>(mo_set)</span>
<span id="cb2-117"><a href="#cb2-117" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-118"><a href="#cb2-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-119"><a href="#cb2-119" aria-hidden="true" tabindex="-1"></a><span class="co"># Modifying the `MultiDataSet` object ------------------------------------------</span></span>
<span id="cb2-120"><a href="#cb2-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-121"><a href="#cb2-121" aria-hidden="true" tabindex="-1"></a>  <span class="do">## RNAseq differential expression results file</span></span>
<span id="cb2-122"><a href="#cb2-122" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-123"><a href="#cb2-123" aria-hidden="true" tabindex="-1"></a>    rnaseq_de_res_file,</span>
<span id="cb2-124"><a href="#cb2-124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system.file</span>(</span>
<span id="cb2-125"><a href="#cb2-125" aria-hidden="true" tabindex="-1"></a>      <span class="st">"extdata/transcriptomics_de_results.csv"</span>,</span>
<span id="cb2-126"><a href="#cb2-126" aria-hidden="true" tabindex="-1"></a>      <span class="at">package =</span> <span class="st">"moiraine"</span></span>
<span id="cb2-127"><a href="#cb2-127" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-128"><a href="#cb2-128" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"file"</span></span>
<span id="cb2-129"><a href="#cb2-129" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-130"><a href="#cb2-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-131"><a href="#cb2-131" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Reading the RNAseq differential expression results</span></span>
<span id="cb2-132"><a href="#cb2-132" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-133"><a href="#cb2-133" aria-hidden="true" tabindex="-1"></a>    rnaseq_de_res_df,</span>
<span id="cb2-134"><a href="#cb2-134" aria-hidden="true" tabindex="-1"></a>    <span class="fu">read_csv</span>(rnaseq_de_res_file) <span class="sc">|&gt;</span></span>
<span id="cb2-135"><a href="#cb2-135" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rename</span>(<span class="at">feature_id =</span> gene_id) <span class="sc">|&gt;</span></span>
<span id="cb2-136"><a href="#cb2-136" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"rnaseq"</span>)</span>
<span id="cb2-137"><a href="#cb2-137" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-138"><a href="#cb2-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-139"><a href="#cb2-139" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Adding the differential expression results to the MultiDataSet object</span></span>
<span id="cb2-140"><a href="#cb2-140" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-141"><a href="#cb2-141" aria-hidden="true" tabindex="-1"></a>    mo_set_de,</span>
<span id="cb2-142"><a href="#cb2-142" aria-hidden="true" tabindex="-1"></a>    <span class="fu">add_features_metadata</span>(mo_set, rnaseq_de_res_df)</span>
<span id="cb2-143"><a href="#cb2-143" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-144"><a href="#cb2-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-145"><a href="#cb2-145" aria-hidden="true" tabindex="-1"></a><span class="co"># Data pre-processing ----------------------------------------------------------</span></span>
<span id="cb2-146"><a href="#cb2-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-147"><a href="#cb2-147" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Applying transformations to the datasets</span></span>
<span id="cb2-148"><a href="#cb2-148" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transformation_datasets_factory</span>(</span>
<span id="cb2-149"><a href="#cb2-149" aria-hidden="true" tabindex="-1"></a>    mo_set_de,</span>
<span id="cb2-150"><a href="#cb2-150" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"vst-deseq2"</span>,</span>
<span id="cb2-151"><a href="#cb2-151" aria-hidden="true" tabindex="-1"></a>      <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"logx"</span>),</span>
<span id="cb2-152"><a href="#cb2-152" aria-hidden="true" tabindex="-1"></a>    <span class="at">log_bases =</span> <span class="dv">2</span>,</span>
<span id="cb2-153"><a href="#cb2-153" aria-hidden="true" tabindex="-1"></a>    <span class="at">pre_log_functions =</span> zero_to_half_min,</span>
<span id="cb2-154"><a href="#cb2-154" aria-hidden="true" tabindex="-1"></a>    <span class="at">transformed_data_name =</span> <span class="st">"mo_set_transformed"</span></span>
<span id="cb2-155"><a href="#cb2-155" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-156"><a href="#cb2-156" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-157"><a href="#cb2-157" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Density plot for each transformed dataset</span></span>
<span id="cb2-158"><a href="#cb2-158" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-159"><a href="#cb2-159" aria-hidden="true" tabindex="-1"></a>    density_plots_transformed,</span>
<span id="cb2-160"><a href="#cb2-160" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_density_data</span>(</span>
<span id="cb2-161"><a href="#cb2-161" aria-hidden="true" tabindex="-1"></a>      mo_set_transformed,</span>
<span id="cb2-162"><a href="#cb2-162" aria-hidden="true" tabindex="-1"></a>      <span class="at">combined =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-163"><a href="#cb2-163" aria-hidden="true" tabindex="-1"></a>      <span class="at">scales =</span> <span class="st">"free"</span></span>
<span id="cb2-164"><a href="#cb2-164" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-165"><a href="#cb2-165" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-166"><a href="#cb2-166" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-167"><a href="#cb2-167" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the mean-SD trend for transformed each dataset</span></span>
<span id="cb2-168"><a href="#cb2-168" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-169"><a href="#cb2-169" aria-hidden="true" tabindex="-1"></a>    mean_sd_plots_transformed,</span>
<span id="cb2-170"><a href="#cb2-170" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_meansd_data</span>(mo_set_transformed)</span>
<span id="cb2-171"><a href="#cb2-171" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-172"><a href="#cb2-172" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-173"><a href="#cb2-173" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Summary table of the transformations applied</span></span>
<span id="cb2-174"><a href="#cb2-174" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-175"><a href="#cb2-175" aria-hidden="true" tabindex="-1"></a>    transformation_summary,</span>
<span id="cb2-176"><a href="#cb2-176" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_table_transformations</span>(transformations_runs_list)</span>
<span id="cb2-177"><a href="#cb2-177" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-178"><a href="#cb2-178" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-179"><a href="#cb2-179" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Running a PCA on each dataset</span></span>
<span id="cb2-180"><a href="#cb2-180" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pca_complete_data_factory</span>(</span>
<span id="cb2-181"><a href="#cb2-181" aria-hidden="true" tabindex="-1"></a>    mo_set_transformed,</span>
<span id="cb2-182"><a href="#cb2-182" aria-hidden="true" tabindex="-1"></a>    <span class="at">complete_data_name =</span> <span class="st">"mo_set_complete"</span></span>
<span id="cb2-183"><a href="#cb2-183" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-184"><a href="#cb2-184" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-185"><a href="#cb2-185" aria-hidden="true" tabindex="-1"></a>  <span class="do">## PCA screeplots</span></span>
<span id="cb2-186"><a href="#cb2-186" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-187"><a href="#cb2-187" aria-hidden="true" tabindex="-1"></a>    pca_screeplots,</span>
<span id="cb2-188"><a href="#cb2-188" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_screeplot_pca</span>(pca_runs_list)</span>
<span id="cb2-189"><a href="#cb2-189" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-190"><a href="#cb2-190" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-191"><a href="#cb2-191" aria-hidden="true" tabindex="-1"></a>  <span class="do">## PCA sample plots</span></span>
<span id="cb2-192"><a href="#cb2-192" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-193"><a href="#cb2-193" aria-hidden="true" tabindex="-1"></a>    pca_sample_plots,</span>
<span id="cb2-194"><a href="#cb2-194" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_samples_coordinates_pca</span>(</span>
<span id="cb2-195"><a href="#cb2-195" aria-hidden="true" tabindex="-1"></a>      pca_runs_list,</span>
<span id="cb2-196"><a href="#cb2-196" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="st">"snps"</span>,</span>
<span id="cb2-197"><a href="#cb2-197" aria-hidden="true" tabindex="-1"></a>      <span class="at">pcs =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb2-198"><a href="#cb2-198" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-199"><a href="#cb2-199" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span id="cb2-200"><a href="#cb2-200" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-201"><a href="#cb2-201" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-202"><a href="#cb2-202" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-203"><a href="#cb2-203" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-204"><a href="#cb2-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-205"><a href="#cb2-205" aria-hidden="true" tabindex="-1"></a><span class="co"># Dataset pre-filtering --------------------------------------------------------</span></span>
<span id="cb2-206"><a href="#cb2-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-207"><a href="#cb2-207" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Unsupervised feature selection based on MAD score</span></span>
<span id="cb2-208"><a href="#cb2-208" aria-hidden="true" tabindex="-1"></a>  <span class="fu">feature_preselection_mad_factory</span>(</span>
<span id="cb2-209"><a href="#cb2-209" aria-hidden="true" tabindex="-1"></a>    mo_set_complete,</span>
<span id="cb2-210"><a href="#cb2-210" aria-hidden="true" tabindex="-1"></a>    <span class="at">to_keep_ns =</span> <span class="fu">c</span>(<span class="st">"snps"</span> <span class="ot">=</span> <span class="dv">1000</span>, <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="dv">1000</span>),</span>
<span id="cb2-211"><a href="#cb2-211" aria-hidden="true" tabindex="-1"></a>    <span class="at">with_ties =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-212"><a href="#cb2-212" aria-hidden="true" tabindex="-1"></a>    <span class="at">filtered_set_target_name =</span> <span class="st">"mo_presel_unsupervised"</span></span>
<span id="cb2-213"><a href="#cb2-213" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-214"><a href="#cb2-214" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-215"><a href="#cb2-215" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Diagnostic plot for MAD-based feature selection</span></span>
<span id="cb2-216"><a href="#cb2-216" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-217"><a href="#cb2-217" aria-hidden="true" tabindex="-1"></a>    preselection_mad_plot,</span>
<span id="cb2-218"><a href="#cb2-218" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_feature_preselection_mad</span>(individual_mad_values)</span>
<span id="cb2-219"><a href="#cb2-219" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-220"><a href="#cb2-220" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-221"><a href="#cb2-221" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Supervised feature selection based on bruising groups</span></span>
<span id="cb2-222"><a href="#cb2-222" aria-hidden="true" tabindex="-1"></a>  <span class="fu">feature_preselection_splsda_factory</span>(</span>
<span id="cb2-223"><a href="#cb2-223" aria-hidden="true" tabindex="-1"></a>    mo_set_complete,</span>
<span id="cb2-224"><a href="#cb2-224" aria-hidden="true" tabindex="-1"></a>    <span class="at">group =</span> <span class="st">"status"</span>,</span>
<span id="cb2-225"><a href="#cb2-225" aria-hidden="true" tabindex="-1"></a>    <span class="at">to_keep_ns =</span> <span class="fu">c</span>(<span class="st">"snps"</span> <span class="ot">=</span> <span class="dv">1000</span>, <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="dv">1000</span>),</span>
<span id="cb2-226"><a href="#cb2-226" aria-hidden="true" tabindex="-1"></a>    <span class="at">filtered_set_target_name =</span> <span class="st">"mo_presel_supervised"</span></span>
<span id="cb2-227"><a href="#cb2-227" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-228"><a href="#cb2-228" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-229"><a href="#cb2-229" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Diagnostic plot for sPLS-DA based feature selection</span></span>
<span id="cb2-230"><a href="#cb2-230" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-231"><a href="#cb2-231" aria-hidden="true" tabindex="-1"></a>    preselection_splsda_plot,</span>
<span id="cb2-232"><a href="#cb2-232" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_feature_preselection_splsda</span>(individual_splsda_perf)</span>
<span id="cb2-233"><a href="#cb2-233" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-234"><a href="#cb2-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-235"><a href="#cb2-235" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with sPLS --------------------------------------------------------</span></span>
<span id="cb2-236"><a href="#cb2-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-237"><a href="#cb2-237" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating sPLS input</span></span>
<span id="cb2-238"><a href="#cb2-238" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-239"><a href="#cb2-239" aria-hidden="true" tabindex="-1"></a>    spls_input,</span>
<span id="cb2-240"><a href="#cb2-240" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_spls</span>(</span>
<span id="cb2-241"><a href="#cb2-241" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-242"><a href="#cb2-242" aria-hidden="true" tabindex="-1"></a>      <span class="at">mode =</span> <span class="st">"canonical"</span>,</span>
<span id="cb2-243"><a href="#cb2-243" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="fu">c</span>(<span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span>)</span>
<span id="cb2-244"><a href="#cb2-244" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-245"><a href="#cb2-245" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-246"><a href="#cb2-246" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-247"><a href="#cb2-247" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Initial PLS run with no feature selection and large number of components</span></span>
<span id="cb2-248"><a href="#cb2-248" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-249"><a href="#cb2-249" aria-hidden="true" tabindex="-1"></a>    spls_novarsel,</span>
<span id="cb2-250"><a href="#cb2-250" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_run</span>(</span>
<span id="cb2-251"><a href="#cb2-251" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-252"><a href="#cb2-252" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> <span class="dv">4</span></span>
<span id="cb2-253"><a href="#cb2-253" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-254"><a href="#cb2-254" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-255"><a href="#cb2-255" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-256"><a href="#cb2-256" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of components</span></span>
<span id="cb2-257"><a href="#cb2-257" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-258"><a href="#cb2-258" aria-hidden="true" tabindex="-1"></a>    spls_perf_res,</span>
<span id="cb2-259"><a href="#cb2-259" aria-hidden="true" tabindex="-1"></a>    mixOmics<span class="sc">::</span><span class="fu">perf</span>(</span>
<span id="cb2-260"><a href="#cb2-260" aria-hidden="true" tabindex="-1"></a>      spls_novarsel,</span>
<span id="cb2-261"><a href="#cb2-261" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-262"><a href="#cb2-262" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-263"><a href="#cb2-263" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">10</span>,</span>
<span id="cb2-264"><a href="#cb2-264" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-265"><a href="#cb2-265" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-266"><a href="#cb2-266" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-267"><a href="#cb2-267" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-268"><a href="#cb2-268" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of components)</span></span>
<span id="cb2-269"><a href="#cb2-269" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Can try criterion = 'Q2.total', 'cor.tpred', 'cor.upred', 'RSS.tpred',</span></span>
<span id="cb2-270"><a href="#cb2-270" aria-hidden="true" tabindex="-1"></a>  <span class="do">## 'RSS.upred' (but avoid 'RSS' and 'PRESS')</span></span>
<span id="cb2-271"><a href="#cb2-271" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-272"><a href="#cb2-272" aria-hidden="true" tabindex="-1"></a>    spls_perf_plot,</span>
<span id="cb2-273"><a href="#cb2-273" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(spls_perf_res, <span class="at">criterion =</span> <span class="st">"Q2.total"</span>)</span>
<span id="cb2-274"><a href="#cb2-274" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-275"><a href="#cb2-275" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-276"><a href="#cb2-276" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Selected value for ncomp</span></span>
<span id="cb2-277"><a href="#cb2-277" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-278"><a href="#cb2-278" aria-hidden="true" tabindex="-1"></a>    spls_optim_ncomp,</span>
<span id="cb2-279"><a href="#cb2-279" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_get_optim_ncomp</span>(spls_perf_res, <span class="at">min_ncomp =</span> <span class="dv">2</span>)</span>
<span id="cb2-280"><a href="#cb2-280" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-281"><a href="#cb2-281" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-282"><a href="#cb2-282" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of features to retain</span></span>
<span id="cb2-283"><a href="#cb2-283" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-284"><a href="#cb2-284" aria-hidden="true" tabindex="-1"></a>    spls_tune_res,</span>
<span id="cb2-285"><a href="#cb2-285" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_tune</span>(</span>
<span id="cb2-286"><a href="#cb2-286" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-287"><a href="#cb2-287" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> spls_optim_ncomp,</span>
<span id="cb2-288"><a href="#cb2-288" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepX =</span> <span class="fu">seq</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">10</span>),</span>
<span id="cb2-289"><a href="#cb2-289" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepY =</span> <span class="fu">seq</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">10</span>),</span>
<span id="cb2-290"><a href="#cb2-290" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-291"><a href="#cb2-291" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-292"><a href="#cb2-292" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">5</span>,</span>
<span id="cb2-293"><a href="#cb2-293" aria-hidden="true" tabindex="-1"></a>      <span class="at">measure =</span> <span class="st">"cor"</span>,</span>
<span id="cb2-294"><a href="#cb2-294" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-295"><a href="#cb2-295" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-296"><a href="#cb2-296" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-297"><a href="#cb2-297" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-298"><a href="#cb2-298" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of features)</span></span>
<span id="cb2-299"><a href="#cb2-299" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-300"><a href="#cb2-300" aria-hidden="true" tabindex="-1"></a>    spls_tune_plot,</span>
<span id="cb2-301"><a href="#cb2-301" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_plot_tune</span>(spls_tune_res)</span>
<span id="cb2-302"><a href="#cb2-302" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-303"><a href="#cb2-303" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-304"><a href="#cb2-304" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Final sPLS run</span></span>
<span id="cb2-305"><a href="#cb2-305" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-306"><a href="#cb2-306" aria-hidden="true" tabindex="-1"></a>    spls_final_run,</span>
<span id="cb2-307"><a href="#cb2-307" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_run</span>(</span>
<span id="cb2-308"><a href="#cb2-308" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-309"><a href="#cb2-309" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> spls_optim_ncomp,</span>
<span id="cb2-310"><a href="#cb2-310" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepX =</span> spls_tune_res<span class="sc">$</span>choice.keepX,</span>
<span id="cb2-311"><a href="#cb2-311" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepY =</span> spls_tune_res<span class="sc">$</span>choice.keepY</span>
<span id="cb2-312"><a href="#cb2-312" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-313"><a href="#cb2-313" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-314"><a href="#cb2-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-315"><a href="#cb2-315" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with sO2PLS ------------------------------------------------------</span></span>
<span id="cb2-316"><a href="#cb2-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-317"><a href="#cb2-317" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating sO2PLS input</span></span>
<span id="cb2-318"><a href="#cb2-318" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-319"><a href="#cb2-319" aria-hidden="true" tabindex="-1"></a>    omicspls_input,</span>
<span id="cb2-320"><a href="#cb2-320" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_omicspls</span>(</span>
<span id="cb2-321"><a href="#cb2-321" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-322"><a href="#cb2-322" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="fu">c</span>(<span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span>)</span>
<span id="cb2-323"><a href="#cb2-323" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-324"><a href="#cb2-324" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-325"><a href="#cb2-325" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-326"><a href="#cb2-326" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Adjusted cross-validation for number of components</span></span>
<span id="cb2-327"><a href="#cb2-327" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-328"><a href="#cb2-328" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj,</span>
<span id="cb2-329"><a href="#cb2-329" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_o2m_adjR2</span>(</span>
<span id="cb2-330"><a href="#cb2-330" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-331"><a href="#cb2-331" aria-hidden="true" tabindex="-1"></a>      <span class="at">a =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,</span>
<span id="cb2-332"><a href="#cb2-332" aria-hidden="true" tabindex="-1"></a>      <span class="at">ax =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="dv">2</span>),</span>
<span id="cb2-333"><a href="#cb2-333" aria-hidden="true" tabindex="-1"></a>      <span class="at">ay =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="dv">2</span>),</span>
<span id="cb2-334"><a href="#cb2-334" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-335"><a href="#cb2-335" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_cores =</span> <span class="dv">6</span>,</span>
<span id="cb2-336"><a href="#cb2-336" aria-hidden="true" tabindex="-1"></a>      <span class="at">seed =</span> <span class="dv">127</span></span>
<span id="cb2-337"><a href="#cb2-337" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-338"><a href="#cb2-338" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-339"><a href="#cb2-339" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-340"><a href="#cb2-340" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj_res,</span>
<span id="cb2-341"><a href="#cb2-341" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_ncomp_adj</span>(so2pls_cv_adj)</span>
<span id="cb2-342"><a href="#cb2-342" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-343"><a href="#cb2-343" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-344"><a href="#cb2-344" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting adjusted cross-validation results</span></span>
<span id="cb2-345"><a href="#cb2-345" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-346"><a href="#cb2-346" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj_plot,</span>
<span id="cb2-347"><a href="#cb2-347" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv_adj</span>(so2pls_cv_adj)</span>
<span id="cb2-348"><a href="#cb2-348" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-349"><a href="#cb2-349" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-350"><a href="#cb2-350" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Standard cross-validation for number of components</span></span>
<span id="cb2-351"><a href="#cb2-351" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-352"><a href="#cb2-352" aria-hidden="true" tabindex="-1"></a>    so2pls_cv,</span>
<span id="cb2-353"><a href="#cb2-353" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_o2m</span>(</span>
<span id="cb2-354"><a href="#cb2-354" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-355"><a href="#cb2-355" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_adj,</span>
<span id="cb2-356"><a href="#cb2-356" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-357"><a href="#cb2-357" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_cores =</span> <span class="dv">6</span>,</span>
<span id="cb2-358"><a href="#cb2-358" aria-hidden="true" tabindex="-1"></a>      <span class="at">seed =</span> <span class="dv">356</span></span>
<span id="cb2-359"><a href="#cb2-359" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-360"><a href="#cb2-360" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-361"><a href="#cb2-361" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-362"><a href="#cb2-362" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_res,</span>
<span id="cb2-363"><a href="#cb2-363" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_ncomp</span>(so2pls_cv)</span>
<span id="cb2-364"><a href="#cb2-364" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-365"><a href="#cb2-365" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-366"><a href="#cb2-366" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting standard cross-validation results</span></span>
<span id="cb2-367"><a href="#cb2-367" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-368"><a href="#cb2-368" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_plot,</span>
<span id="cb2-369"><a href="#cb2-369" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv</span>(so2pls_cv)</span>
<span id="cb2-370"><a href="#cb2-370" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-371"><a href="#cb2-371" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-372"><a href="#cb2-372" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for sparsity parameters</span></span>
<span id="cb2-373"><a href="#cb2-373" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-374"><a href="#cb2-374" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity,</span>
<span id="cb2-375"><a href="#cb2-375" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_sparsity</span>(</span>
<span id="cb2-376"><a href="#cb2-376" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-377"><a href="#cb2-377" aria-hidden="true" tabindex="-1"></a>      <span class="at">n =</span> so2pls_cv_res[<span class="st">"n"</span>],</span>
<span id="cb2-378"><a href="#cb2-378" aria-hidden="true" tabindex="-1"></a>      <span class="at">nx =</span> so2pls_cv_res[<span class="st">"nx"</span>],</span>
<span id="cb2-379"><a href="#cb2-379" aria-hidden="true" tabindex="-1"></a>      <span class="at">ny =</span> so2pls_cv_res[<span class="st">"ny"</span>],</span>
<span id="cb2-380"><a href="#cb2-380" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-381"><a href="#cb2-381" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepx_seq =</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">5</span>, <span class="dv">30</span>, <span class="dv">5</span>), <span class="fu">seq</span>(<span class="dv">40</span>, <span class="dv">100</span>, <span class="dv">10</span>)),</span>
<span id="cb2-382"><a href="#cb2-382" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepy_seq =</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">5</span>, <span class="dv">40</span>, <span class="dv">5</span>))</span>
<span id="cb2-383"><a href="#cb2-383" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-384"><a href="#cb2-384" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-385"><a href="#cb2-385" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-386"><a href="#cb2-386" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_res,</span>
<span id="cb2-387"><a href="#cb2-387" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_keep</span>(so2pls_cv_sparsity)</span>
<span id="cb2-388"><a href="#cb2-388" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-389"><a href="#cb2-389" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-390"><a href="#cb2-390" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the results of the cross-validation for the number of features</span></span>
<span id="cb2-391"><a href="#cb2-391" aria-hidden="true" tabindex="-1"></a>  <span class="do">## to retain from each dataset for the different joint components</span></span>
<span id="cb2-392"><a href="#cb2-392" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-393"><a href="#cb2-393" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_plot,</span>
<span id="cb2-394"><a href="#cb2-394" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv_sparsity</span>(so2pls_cv_sparsity)</span>
<span id="cb2-395"><a href="#cb2-395" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-396"><a href="#cb2-396" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-397"><a href="#cb2-397" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Extracting sparsity results in table format</span></span>
<span id="cb2-398"><a href="#cb2-398" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-399"><a href="#cb2-399" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_table,</span>
<span id="cb2-400"><a href="#cb2-400" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_print_cv_sparsity</span>(so2pls_cv_sparsity_res)</span>
<span id="cb2-401"><a href="#cb2-401" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-402"><a href="#cb2-402" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-403"><a href="#cb2-403" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Final sO2PLS run</span></span>
<span id="cb2-404"><a href="#cb2-404" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-405"><a href="#cb2-405" aria-hidden="true" tabindex="-1"></a>    so2pls_final_run,</span>
<span id="cb2-406"><a href="#cb2-406" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_o2m</span>(</span>
<span id="cb2-407"><a href="#cb2-407" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-408"><a href="#cb2-408" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_res,</span>
<span id="cb2-409"><a href="#cb2-409" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_sparsity_res</span>
<span id="cb2-410"><a href="#cb2-410" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-411"><a href="#cb2-411" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-412"><a href="#cb2-412" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-413"><a href="#cb2-413" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Summary plot of percentage of variance explained</span></span>
<span id="cb2-414"><a href="#cb2-414" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-415"><a href="#cb2-415" aria-hidden="true" tabindex="-1"></a>    so2pls_summary_plot,</span>
<span id="cb2-416"><a href="#cb2-416" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_summary</span>(so2pls_final_run)</span>
<span id="cb2-417"><a href="#cb2-417" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-418"><a href="#cb2-418" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-419"><a href="#cb2-419" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Screeplot</span></span>
<span id="cb2-420"><a href="#cb2-420" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-421"><a href="#cb2-421" aria-hidden="true" tabindex="-1"></a>    so2pls_screeplot,</span>
<span id="cb2-422"><a href="#cb2-422" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_screeplot</span>(so2pls_final_run)</span>
<span id="cb2-423"><a href="#cb2-423" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-424"><a href="#cb2-424" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-425"><a href="#cb2-425" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Comparison of samples score for joint components</span></span>
<span id="cb2-426"><a href="#cb2-426" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-427"><a href="#cb2-427" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_comparison_plot,</span>
<span id="cb2-428"><a href="#cb2-428" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_compare_samples_joint_components</span>(</span>
<span id="cb2-429"><a href="#cb2-429" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-430"><a href="#cb2-430" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-431"><a href="#cb2-431" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_by =</span> <span class="st">"status"</span>,</span>
<span id="cb2-432"><a href="#cb2-432" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_by =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-433"><a href="#cb2-433" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-434"><a href="#cb2-434" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-435"><a href="#cb2-435" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-436"><a href="#cb2-436" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Coefficient plot for joint components</span></span>
<span id="cb2-437"><a href="#cb2-437" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-438"><a href="#cb2-438" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_coefficients_plot,</span>
<span id="cb2-439"><a href="#cb2-439" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_joint_components_coefficients</span>(so2pls_final_run)</span>
<span id="cb2-440"><a href="#cb2-440" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-441"><a href="#cb2-441" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-442"><a href="#cb2-442" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Joint component samples score plot</span></span>
<span id="cb2-443"><a href="#cb2-443" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-444"><a href="#cb2-444" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_samples_score_plot,</span>
<span id="cb2-445"><a href="#cb2-445" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_samples_joint_components</span>(</span>
<span id="cb2-446"><a href="#cb2-446" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-447"><a href="#cb2-447" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-448"><a href="#cb2-448" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-449"><a href="#cb2-449" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>),</span>
<span id="cb2-450"><a href="#cb2-450" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-451"><a href="#cb2-451" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-452"><a href="#cb2-452" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>)</span>
<span id="cb2-453"><a href="#cb2-453" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-454"><a href="#cb2-454" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-455"><a href="#cb2-455" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Specific components samples score plot</span></span>
<span id="cb2-456"><a href="#cb2-456" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-457"><a href="#cb2-457" aria-hidden="true" tabindex="-1"></a>    so2pls_specific_components_samples_score_plot,</span>
<span id="cb2-458"><a href="#cb2-458" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_samples_specific_components</span>(</span>
<span id="cb2-459"><a href="#cb2-459" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-460"><a href="#cb2-460" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-461"><a href="#cb2-461" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"feedlot"</span>,</span>
<span id="cb2-462"><a href="#cb2-462" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>),</span>
<span id="cb2-463"><a href="#cb2-463" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"rnaseq_batch"</span>,</span>
<span id="cb2-464"><a href="#cb2-464" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"gender"</span></span>
<span id="cb2-465"><a href="#cb2-465" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span> </span>
<span id="cb2-466"><a href="#cb2-466" aria-hidden="true" tabindex="-1"></a>      <span class="fu">map</span>(\(x) x <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>))</span>
<span id="cb2-467"><a href="#cb2-467" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-468"><a href="#cb2-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-469"><a href="#cb2-469" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with MOFA --------------------------------------------------------</span></span>
<span id="cb2-470"><a href="#cb2-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-471"><a href="#cb2-471" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating MOFA input</span></span>
<span id="cb2-472"><a href="#cb2-472" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-473"><a href="#cb2-473" aria-hidden="true" tabindex="-1"></a>    mofa_input,</span>
<span id="cb2-474"><a href="#cb2-474" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_mofa</span>(</span>
<span id="cb2-475"><a href="#cb2-475" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-476"><a href="#cb2-476" aria-hidden="true" tabindex="-1"></a>      <span class="at">options_list =</span> <span class="fu">list</span>(</span>
<span id="cb2-477"><a href="#cb2-477" aria-hidden="true" tabindex="-1"></a>        <span class="at">data_options =</span> <span class="fu">list</span>(<span class="at">scale_views =</span> <span class="cn">TRUE</span>),</span>
<span id="cb2-478"><a href="#cb2-478" aria-hidden="true" tabindex="-1"></a>        <span class="at">model_options =</span> <span class="fu">list</span>(<span class="at">likelihoods =</span> <span class="fu">c</span>(</span>
<span id="cb2-479"><a href="#cb2-479" aria-hidden="true" tabindex="-1"></a>          <span class="st">"snps"</span> <span class="ot">=</span> <span class="st">"poisson"</span>,</span>
<span id="cb2-480"><a href="#cb2-480" aria-hidden="true" tabindex="-1"></a>          <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"gaussian"</span>,</span>
<span id="cb2-481"><a href="#cb2-481" aria-hidden="true" tabindex="-1"></a>          <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"gaussian"</span>)</span>
<span id="cb2-482"><a href="#cb2-482" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb2-483"><a href="#cb2-483" aria-hidden="true" tabindex="-1"></a>        <span class="at">training_options =</span> <span class="fu">list</span>(<span class="at">seed =</span> <span class="dv">43</span>)</span>
<span id="cb2-484"><a href="#cb2-484" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb2-485"><a href="#cb2-485" aria-hidden="true" tabindex="-1"></a>      <span class="at">only_common_samples =</span> <span class="cn">FALSE</span></span>
<span id="cb2-486"><a href="#cb2-486" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-487"><a href="#cb2-487" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-488"><a href="#cb2-488" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-489"><a href="#cb2-489" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Overview plot of the samples in each dataset</span></span>
<span id="cb2-490"><a href="#cb2-490" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-491"><a href="#cb2-491" aria-hidden="true" tabindex="-1"></a>    mofa_input_plot,</span>
<span id="cb2-492"><a href="#cb2-492" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_data_overview</span>(mofa_input)</span>
<span id="cb2-493"><a href="#cb2-493" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-494"><a href="#cb2-494" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-495"><a href="#cb2-495" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Training MOFA model</span></span>
<span id="cb2-496"><a href="#cb2-496" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-497"><a href="#cb2-497" aria-hidden="true" tabindex="-1"></a>    mofa_trained,</span>
<span id="cb2-498"><a href="#cb2-498" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run_mofa</span>(</span>
<span id="cb2-499"><a href="#cb2-499" aria-hidden="true" tabindex="-1"></a>      mofa_input,</span>
<span id="cb2-500"><a href="#cb2-500" aria-hidden="true" tabindex="-1"></a>      <span class="at">save_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-501"><a href="#cb2-501" aria-hidden="true" tabindex="-1"></a>      <span class="at">use_basilisk =</span> <span class="cn">TRUE</span></span>
<span id="cb2-502"><a href="#cb2-502" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-503"><a href="#cb2-503" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-504"><a href="#cb2-504" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-505"><a href="#cb2-505" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Formatting MOFA output</span></span>
<span id="cb2-506"><a href="#cb2-506" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-507"><a href="#cb2-507" aria-hidden="true" tabindex="-1"></a>    mofa_output,</span>
<span id="cb2-508"><a href="#cb2-508" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_output</span>(mofa_trained)</span>
<span id="cb2-509"><a href="#cb2-509" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-510"><a href="#cb2-510" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-511"><a href="#cb2-511" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plots of variance explained</span></span>
<span id="cb2-512"><a href="#cb2-512" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-513"><a href="#cb2-513" aria-hidden="true" tabindex="-1"></a>    mofa_var_explained_plot,</span>
<span id="cb2-514"><a href="#cb2-514" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(</span>
<span id="cb2-515"><a href="#cb2-515" aria-hidden="true" tabindex="-1"></a>      mofa_trained,</span>
<span id="cb2-516"><a href="#cb2-516" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"view"</span>,  <span class="do">## datasets on the x-axis</span></span>
<span id="cb2-517"><a href="#cb2-517" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"factor"</span> <span class="do">## factors on the y-axis</span></span>
<span id="cb2-518"><a href="#cb2-518" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-519"><a href="#cb2-519" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-520"><a href="#cb2-520" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-521"><a href="#cb2-521" aria-hidden="true" tabindex="-1"></a>    mofa_total_var_explained_plot,</span>
<span id="cb2-522"><a href="#cb2-522" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(</span>
<span id="cb2-523"><a href="#cb2-523" aria-hidden="true" tabindex="-1"></a>      mofa_trained,</span>
<span id="cb2-524"><a href="#cb2-524" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"view"</span>,</span>
<span id="cb2-525"><a href="#cb2-525" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"factor"</span>,</span>
<span id="cb2-526"><a href="#cb2-526" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot_total =</span> <span class="cn">TRUE</span></span>
<span id="cb2-527"><a href="#cb2-527" aria-hidden="true" tabindex="-1"></a>    )[[<span class="dv">2</span>]]</span>
<span id="cb2-528"><a href="#cb2-528" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-529"><a href="#cb2-529" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-530"><a href="#cb2-530" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plot of factors correlation with covariates</span></span>
<span id="cb2-531"><a href="#cb2-531" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-532"><a href="#cb2-532" aria-hidden="true" tabindex="-1"></a>    mofa_factors_covariates_cor_plot,</span>
<span id="cb2-533"><a href="#cb2-533" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mofa_plot_cor_covariates</span>(mofa_trained)</span>
<span id="cb2-534"><a href="#cb2-534" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-535"><a href="#cb2-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-536"><a href="#cb2-536" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with DIABLO ------------------------------------------------------</span></span>
<span id="cb2-537"><a href="#cb2-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-538"><a href="#cb2-538" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating the DIABLO input</span></span>
<span id="cb2-539"><a href="#cb2-539" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-540"><a href="#cb2-540" aria-hidden="true" tabindex="-1"></a>    diablo_input,</span>
<span id="cb2-541"><a href="#cb2-541" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_mixomics_supervised</span>(</span>
<span id="cb2-542"><a href="#cb2-542" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-543"><a href="#cb2-543" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> <span class="st">"status"</span></span>
<span id="cb2-544"><a href="#cb2-544" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-545"><a href="#cb2-545" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-546"><a href="#cb2-546" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-547"><a href="#cb2-547" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Running sPLS on each dataset to construct the design matrix</span></span>
<span id="cb2-548"><a href="#cb2-548" aria-hidden="true" tabindex="-1"></a>  <span class="fu">diablo_pairwise_pls_factory</span>(diablo_input),</span>
<span id="cb2-549"><a href="#cb2-549" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-550"><a href="#cb2-550" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Initial DIABLO run with no feature selection and large number of components</span></span>
<span id="cb2-551"><a href="#cb2-551" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-552"><a href="#cb2-552" aria-hidden="true" tabindex="-1"></a>    diablo_novarsel,</span>
<span id="cb2-553"><a href="#cb2-553" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_run</span>(</span>
<span id="cb2-554"><a href="#cb2-554" aria-hidden="true" tabindex="-1"></a>      diablo_input,</span>
<span id="cb2-555"><a href="#cb2-555" aria-hidden="true" tabindex="-1"></a>      diablo_design_matrix,</span>
<span id="cb2-556"><a href="#cb2-556" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> <span class="dv">7</span></span>
<span id="cb2-557"><a href="#cb2-557" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-558"><a href="#cb2-558" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-559"><a href="#cb2-559" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-560"><a href="#cb2-560" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of components</span></span>
<span id="cb2-561"><a href="#cb2-561" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-562"><a href="#cb2-562" aria-hidden="true" tabindex="-1"></a>    diablo_perf_res,</span>
<span id="cb2-563"><a href="#cb2-563" aria-hidden="true" tabindex="-1"></a>    mixOmics<span class="sc">::</span><span class="fu">perf</span>(</span>
<span id="cb2-564"><a href="#cb2-564" aria-hidden="true" tabindex="-1"></a>      diablo_novarsel,</span>
<span id="cb2-565"><a href="#cb2-565" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-566"><a href="#cb2-566" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-567"><a href="#cb2-567" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">10</span>,</span>
<span id="cb2-568"><a href="#cb2-568" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-569"><a href="#cb2-569" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-570"><a href="#cb2-570" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-571"><a href="#cb2-571" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-572"><a href="#cb2-572" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of components)</span></span>
<span id="cb2-573"><a href="#cb2-573" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-574"><a href="#cb2-574" aria-hidden="true" tabindex="-1"></a>    diablo_perf_plot,</span>
<span id="cb2-575"><a href="#cb2-575" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_plot_perf</span>(diablo_perf_res)</span>
<span id="cb2-576"><a href="#cb2-576" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-577"><a href="#cb2-577" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-578"><a href="#cb2-578" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Selected value for ncomp</span></span>
<span id="cb2-579"><a href="#cb2-579" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-580"><a href="#cb2-580" aria-hidden="true" tabindex="-1"></a>    diablo_optim_ncomp,</span>
<span id="cb2-581"><a href="#cb2-581" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_get_optim_ncomp</span>(diablo_perf_res)</span>
<span id="cb2-582"><a href="#cb2-582" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-583"><a href="#cb2-583" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-584"><a href="#cb2-584" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of features to retain</span></span>
<span id="cb2-585"><a href="#cb2-585" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-586"><a href="#cb2-586" aria-hidden="true" tabindex="-1"></a>    diablo_tune_res,</span>
<span id="cb2-587"><a href="#cb2-587" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_tune</span>(</span>
<span id="cb2-588"><a href="#cb2-588" aria-hidden="true" tabindex="-1"></a>      diablo_input,</span>
<span id="cb2-589"><a href="#cb2-589" aria-hidden="true" tabindex="-1"></a>      diablo_design_matrix,</span>
<span id="cb2-590"><a href="#cb2-590" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> diablo_optim_ncomp,</span>
<span id="cb2-591"><a href="#cb2-591" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-592"><a href="#cb2-592" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-593"><a href="#cb2-593" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">5</span>,</span>
<span id="cb2-594"><a href="#cb2-594" aria-hidden="true" tabindex="-1"></a>      <span class="at">dist =</span> <span class="st">"centroids.dist"</span>,</span>
<span id="cb2-595"><a href="#cb2-595" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-596"><a href="#cb2-596" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-597"><a href="#cb2-597" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-598"><a href="#cb2-598" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-599"><a href="#cb2-599" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of features)</span></span>
<span id="cb2-600"><a href="#cb2-600" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-601"><a href="#cb2-601" aria-hidden="true" tabindex="-1"></a>    diablo_tune_plot,</span>
<span id="cb2-602"><a href="#cb2-602" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_plot_tune</span>(diablo_tune_res)</span>
<span id="cb2-603"><a href="#cb2-603" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-604"><a href="#cb2-604" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-605"><a href="#cb2-605" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Final DIABLO run</span></span>
<span id="cb2-606"><a href="#cb2-606" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-607"><a href="#cb2-607" aria-hidden="true" tabindex="-1"></a>    diablo_final_run,</span>
<span id="cb2-608"><a href="#cb2-608" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_run</span>(</span>
<span id="cb2-609"><a href="#cb2-609" aria-hidden="true" tabindex="-1"></a>      diablo_input,</span>
<span id="cb2-610"><a href="#cb2-610" aria-hidden="true" tabindex="-1"></a>      diablo_design_matrix,</span>
<span id="cb2-611"><a href="#cb2-611" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> diablo_optim_ncomp,</span>
<span id="cb2-612"><a href="#cb2-612" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepX =</span> diablo_tune_res<span class="sc">$</span>choice.keepX</span>
<span id="cb2-613"><a href="#cb2-613" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-614"><a href="#cb2-614" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-615"><a href="#cb2-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-616"><a href="#cb2-616" aria-hidden="true" tabindex="-1"></a><span class="co"># Interpreting the integration results -----------------------------------------</span></span>
<span id="cb2-617"><a href="#cb2-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-618"><a href="#cb2-618" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Generating standardised output</span></span>
<span id="cb2-619"><a href="#cb2-619" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-620"><a href="#cb2-620" aria-hidden="true" tabindex="-1"></a>    spls_output,</span>
<span id="cb2-621"><a href="#cb2-621" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_output</span>(spls_final_run)</span>
<span id="cb2-622"><a href="#cb2-622" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-623"><a href="#cb2-623" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-624"><a href="#cb2-624" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Percentage of variance explained</span></span>
<span id="cb2-625"><a href="#cb2-625" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-626"><a href="#cb2-626" aria-hidden="true" tabindex="-1"></a>    spls_plot_variance_explained,</span>
<span id="cb2-627"><a href="#cb2-627" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(spls_output)</span>
<span id="cb2-628"><a href="#cb2-628" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-629"><a href="#cb2-629" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-630"><a href="#cb2-630" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Samples score matrix plot</span></span>
<span id="cb2-631"><a href="#cb2-631" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-632"><a href="#cb2-632" aria-hidden="true" tabindex="-1"></a>    spls_samples_scores_plot,</span>
<span id="cb2-633"><a href="#cb2-633" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_samples_score</span>(</span>
<span id="cb2-634"><a href="#cb2-634" aria-hidden="true" tabindex="-1"></a>      spls_output,</span>
<span id="cb2-635"><a href="#cb2-635" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-636"><a href="#cb2-636" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-637"><a href="#cb2-637" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> palette_status,</span>
<span id="cb2-638"><a href="#cb2-638" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"gender"</span>,</span>
<span id="cb2-639"><a href="#cb2-639" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"feedlot"</span>,</span>
<span id="cb2-640"><a href="#cb2-640" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_lower =</span> palette_feedlot</span>
<span id="cb2-641"><a href="#cb2-641" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-642"><a href="#cb2-642" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>)</span>
<span id="cb2-643"><a href="#cb2-643" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-644"><a href="#cb2-644" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-645"><a href="#cb2-645" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Distribution of features weight</span></span>
<span id="cb2-646"><a href="#cb2-646" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-647"><a href="#cb2-647" aria-hidden="true" tabindex="-1"></a>    spls_features_weight_distribution,</span>
<span id="cb2-648"><a href="#cb2-648" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_features_weight_distr</span>(spls_output)</span>
<span id="cb2-649"><a href="#cb2-649" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-650"><a href="#cb2-650" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-651"><a href="#cb2-651" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plot of top contributing features</span></span>
<span id="cb2-652"><a href="#cb2-652" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-653"><a href="#cb2-653" aria-hidden="true" tabindex="-1"></a>    spls_top_features_plot,</span>
<span id="cb2-654"><a href="#cb2-654" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_top_features</span>(</span>
<span id="cb2-655"><a href="#cb2-655" aria-hidden="true" tabindex="-1"></a>      spls_output,</span>
<span id="cb2-656"><a href="#cb2-656" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-657"><a href="#cb2-657" aria-hidden="true" tabindex="-1"></a>      <span class="at">label_cols =</span> <span class="fu">list</span>(</span>
<span id="cb2-658"><a href="#cb2-658" aria-hidden="true" tabindex="-1"></a>        <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"Name"</span>,</span>
<span id="cb2-659"><a href="#cb2-659" aria-hidden="true" tabindex="-1"></a>        <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"name"</span></span>
<span id="cb2-660"><a href="#cb2-660" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb2-661"><a href="#cb2-661" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-662"><a href="#cb2-662" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-663"><a href="#cb2-663" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-664"><a href="#cb2-664" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Table of selected features</span></span>
<span id="cb2-665"><a href="#cb2-665" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-666"><a href="#cb2-666" aria-hidden="true" tabindex="-1"></a>    spls_selected_features,</span>
<span id="cb2-667"><a href="#cb2-667" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_selected_features</span>(spls_output)</span>
<span id="cb2-668"><a href="#cb2-668" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-669"><a href="#cb2-669" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-670"><a href="#cb2-670" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Generating standardised output</span></span>
<span id="cb2-671"><a href="#cb2-671" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-672"><a href="#cb2-672" aria-hidden="true" tabindex="-1"></a>    so2pls_output,</span>
<span id="cb2-673"><a href="#cb2-673" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_output</span>(so2pls_final_run)</span>
<span id="cb2-674"><a href="#cb2-674" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-675"><a href="#cb2-675" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-676"><a href="#cb2-676" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Percentage of variance explained</span></span>
<span id="cb2-677"><a href="#cb2-677" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-678"><a href="#cb2-678" aria-hidden="true" tabindex="-1"></a>    so2pls_plot_variance_explained,</span>
<span id="cb2-679"><a href="#cb2-679" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(so2pls_output, <span class="at">ncol =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb2-680"><a href="#cb2-680" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>, <span class="at">angle =</span> <span class="dv">30</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span>
<span id="cb2-681"><a href="#cb2-681" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-682"><a href="#cb2-682" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-683"><a href="#cb2-683" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Samples score matrix plot</span></span>
<span id="cb2-684"><a href="#cb2-684" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-685"><a href="#cb2-685" aria-hidden="true" tabindex="-1"></a>    so2pls_samples_scores_plot,</span>
<span id="cb2-686"><a href="#cb2-686" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_samples_score</span>(</span>
<span id="cb2-687"><a href="#cb2-687" aria-hidden="true" tabindex="-1"></a>      so2pls_output,</span>
<span id="cb2-688"><a href="#cb2-688" aria-hidden="true" tabindex="-1"></a>      <span class="at">latent_dimensions =</span> <span class="st">"joint component 1"</span>,</span>
<span id="cb2-689"><a href="#cb2-689" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-690"><a href="#cb2-690" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-691"><a href="#cb2-691" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> palette_status,</span>
<span id="cb2-692"><a href="#cb2-692" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"gender"</span></span>
<span id="cb2-693"><a href="#cb2-693" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-694"><a href="#cb2-694" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-695"><a href="#cb2-695" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-696"><a href="#cb2-696" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Distribution of features weight</span></span>
<span id="cb2-697"><a href="#cb2-697" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-698"><a href="#cb2-698" aria-hidden="true" tabindex="-1"></a>    so2pls_features_weight_distribution,</span>
<span id="cb2-699"><a href="#cb2-699" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_features_weight_distr</span>(so2pls_output) <span class="sc">+</span></span>
<span id="cb2-700"><a href="#cb2-700" aria-hidden="true" tabindex="-1"></a>      <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb2-701"><a href="#cb2-701" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-702"><a href="#cb2-702" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-703"><a href="#cb2-703" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plot of top contributing features</span></span>
<span id="cb2-704"><a href="#cb2-704" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-705"><a href="#cb2-705" aria-hidden="true" tabindex="-1"></a>    so2pls_top_features_plot,</span>
<span id="cb2-706"><a href="#cb2-706" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_top_features</span>(</span>
<span id="cb2-707"><a href="#cb2-707" aria-hidden="true" tabindex="-1"></a>      so2pls_output,</span>
<span id="cb2-708"><a href="#cb2-708" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-709"><a href="#cb2-709" aria-hidden="true" tabindex="-1"></a>      <span class="at">label_cols =</span> <span class="fu">list</span>(</span>
<span id="cb2-710"><a href="#cb2-710" aria-hidden="true" tabindex="-1"></a>        <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"Name"</span>,</span>
<span id="cb2-711"><a href="#cb2-711" aria-hidden="true" tabindex="-1"></a>        <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"name"</span></span>
<span id="cb2-712"><a href="#cb2-712" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb2-713"><a href="#cb2-713" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-714"><a href="#cb2-714" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-715"><a href="#cb2-715" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-716"><a href="#cb2-716" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Table of selected features</span></span>
<span id="cb2-717"><a href="#cb2-717" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-718"><a href="#cb2-718" aria-hidden="true" tabindex="-1"></a>    so2pls_selected_features,</span>
<span id="cb2-719"><a href="#cb2-719" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_selected_features</span>(</span>
<span id="cb2-720"><a href="#cb2-720" aria-hidden="true" tabindex="-1"></a>      so2pls_output,</span>
<span id="cb2-721"><a href="#cb2-721" aria-hidden="true" tabindex="-1"></a>      <span class="at">latent_dimensions =</span> <span class="st">"joint component 1"</span></span>
<span id="cb2-722"><a href="#cb2-722" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-723"><a href="#cb2-723" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-724"><a href="#cb2-724" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-725"><a href="#cb2-725" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Generating standardised output</span></span>
<span id="cb2-726"><a href="#cb2-726" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-727"><a href="#cb2-727" aria-hidden="true" tabindex="-1"></a>    mofa_output,</span>
<span id="cb2-728"><a href="#cb2-728" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_output</span>(mofa_trained)</span>
<span id="cb2-729"><a href="#cb2-729" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-730"><a href="#cb2-730" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-731"><a href="#cb2-731" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Percentage of variance explained</span></span>
<span id="cb2-732"><a href="#cb2-732" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-733"><a href="#cb2-733" aria-hidden="true" tabindex="-1"></a>    mofa_plot_variance_explained,</span>
<span id="cb2-734"><a href="#cb2-734" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(mofa_output, <span class="at">ncol =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb2-735"><a href="#cb2-735" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>, <span class="at">angle =</span> <span class="dv">30</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span>
<span id="cb2-736"><a href="#cb2-736" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-737"><a href="#cb2-737" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-738"><a href="#cb2-738" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Samples score matrix plot</span></span>
<span id="cb2-739"><a href="#cb2-739" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-740"><a href="#cb2-740" aria-hidden="true" tabindex="-1"></a>    mofa_samples_scores_plot,</span>
<span id="cb2-741"><a href="#cb2-741" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_samples_score</span>(</span>
<span id="cb2-742"><a href="#cb2-742" aria-hidden="true" tabindex="-1"></a>      mofa_output,</span>
<span id="cb2-743"><a href="#cb2-743" aria-hidden="true" tabindex="-1"></a>      <span class="at">latent_dimensions =</span> <span class="fu">paste</span>(<span class="st">"Factor"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>),</span>
<span id="cb2-744"><a href="#cb2-744" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-745"><a href="#cb2-745" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-746"><a href="#cb2-746" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> palette_status,</span>
<span id="cb2-747"><a href="#cb2-747" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"gender"</span>,</span>
<span id="cb2-748"><a href="#cb2-748" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span id="cb2-749"><a href="#cb2-749" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_lower =</span> palette_geno_comp</span>
<span id="cb2-750"><a href="#cb2-750" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-751"><a href="#cb2-751" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>)</span>
<span id="cb2-752"><a href="#cb2-752" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-753"><a href="#cb2-753" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-754"><a href="#cb2-754" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Distribution of features weight</span></span>
<span id="cb2-755"><a href="#cb2-755" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-756"><a href="#cb2-756" aria-hidden="true" tabindex="-1"></a>    mofa_features_weight_distribution,</span>
<span id="cb2-757"><a href="#cb2-757" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_features_weight_distr</span>(</span>
<span id="cb2-758"><a href="#cb2-758" aria-hidden="true" tabindex="-1"></a>      mofa_output,</span>
<span id="cb2-759"><a href="#cb2-759" aria-hidden="true" tabindex="-1"></a>      <span class="at">latent_dimensions =</span> <span class="fu">paste</span>(<span class="st">"Factor"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>)</span>
<span id="cb2-760"><a href="#cb2-760" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-761"><a href="#cb2-761" aria-hidden="true" tabindex="-1"></a>      <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb2-762"><a href="#cb2-762" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-763"><a href="#cb2-763" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-764"><a href="#cb2-764" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plot of top contributing features</span></span>
<span id="cb2-765"><a href="#cb2-765" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-766"><a href="#cb2-766" aria-hidden="true" tabindex="-1"></a>    mofa_top_features_plot,</span>
<span id="cb2-767"><a href="#cb2-767" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_top_features</span>(</span>
<span id="cb2-768"><a href="#cb2-768" aria-hidden="true" tabindex="-1"></a>      mofa_output,</span>
<span id="cb2-769"><a href="#cb2-769" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-770"><a href="#cb2-770" aria-hidden="true" tabindex="-1"></a>      <span class="at">label_cols =</span> <span class="fu">list</span>(</span>
<span id="cb2-771"><a href="#cb2-771" aria-hidden="true" tabindex="-1"></a>        <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"Name"</span>,</span>
<span id="cb2-772"><a href="#cb2-772" aria-hidden="true" tabindex="-1"></a>        <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"name"</span></span>
<span id="cb2-773"><a href="#cb2-773" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb2-774"><a href="#cb2-774" aria-hidden="true" tabindex="-1"></a>      <span class="at">latent_dimensions =</span> <span class="fu">paste</span>(<span class="st">"Factor"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)</span>
<span id="cb2-775"><a href="#cb2-775" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-776"><a href="#cb2-776" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-777"><a href="#cb2-777" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-778"><a href="#cb2-778" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Table of top contributing features</span></span>
<span id="cb2-779"><a href="#cb2-779" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-780"><a href="#cb2-780" aria-hidden="true" tabindex="-1"></a>    mofa_top_features,</span>
<span id="cb2-781"><a href="#cb2-781" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_top_features</span>(</span>
<span id="cb2-782"><a href="#cb2-782" aria-hidden="true" tabindex="-1"></a>      mofa_output,</span>
<span id="cb2-783"><a href="#cb2-783" aria-hidden="true" tabindex="-1"></a>      <span class="at">min_importance =</span> <span class="fl">0.8</span>,</span>
<span id="cb2-784"><a href="#cb2-784" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete</span>
<span id="cb2-785"><a href="#cb2-785" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-786"><a href="#cb2-786" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-787"><a href="#cb2-787" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-788"><a href="#cb2-788" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Generating standardised output</span></span>
<span id="cb2-789"><a href="#cb2-789" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-790"><a href="#cb2-790" aria-hidden="true" tabindex="-1"></a>    diablo_output,</span>
<span id="cb2-791"><a href="#cb2-791" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_output</span>(diablo_final_run)</span>
<span id="cb2-792"><a href="#cb2-792" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-793"><a href="#cb2-793" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-794"><a href="#cb2-794" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Percentage of variance explained</span></span>
<span id="cb2-795"><a href="#cb2-795" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-796"><a href="#cb2-796" aria-hidden="true" tabindex="-1"></a>    diablo_plot_variance_explained,</span>
<span id="cb2-797"><a href="#cb2-797" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(diablo_output, <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb2-798"><a href="#cb2-798" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-799"><a href="#cb2-799" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-800"><a href="#cb2-800" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Samples score matrix plot</span></span>
<span id="cb2-801"><a href="#cb2-801" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-802"><a href="#cb2-802" aria-hidden="true" tabindex="-1"></a>    diablo_samples_scores_plot,</span>
<span id="cb2-803"><a href="#cb2-803" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_samples_score</span>(</span>
<span id="cb2-804"><a href="#cb2-804" aria-hidden="true" tabindex="-1"></a>      diablo_output,</span>
<span id="cb2-805"><a href="#cb2-805" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-806"><a href="#cb2-806" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-807"><a href="#cb2-807" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> palette_status,</span>
<span id="cb2-808"><a href="#cb2-808" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"gender"</span>,</span>
<span id="cb2-809"><a href="#cb2-809" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"rnaseq_batch"</span>,</span>
<span id="cb2-810"><a href="#cb2-810" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_lower =</span> palette_rnaseq_batch</span>
<span id="cb2-811"><a href="#cb2-811" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-812"><a href="#cb2-812" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>)</span>
<span id="cb2-813"><a href="#cb2-813" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-814"><a href="#cb2-814" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-815"><a href="#cb2-815" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Distribution of features weight</span></span>
<span id="cb2-816"><a href="#cb2-816" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-817"><a href="#cb2-817" aria-hidden="true" tabindex="-1"></a>    diablo_features_weight_distribution,</span>
<span id="cb2-818"><a href="#cb2-818" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_features_weight_distr</span>(</span>
<span id="cb2-819"><a href="#cb2-819" aria-hidden="true" tabindex="-1"></a>      diablo_output</span>
<span id="cb2-820"><a href="#cb2-820" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-821"><a href="#cb2-821" aria-hidden="true" tabindex="-1"></a>      <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb2-822"><a href="#cb2-822" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-823"><a href="#cb2-823" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-824"><a href="#cb2-824" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plot of top contributing features</span></span>
<span id="cb2-825"><a href="#cb2-825" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-826"><a href="#cb2-826" aria-hidden="true" tabindex="-1"></a>    diablo_top_features_plot,</span>
<span id="cb2-827"><a href="#cb2-827" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_top_features</span>(</span>
<span id="cb2-828"><a href="#cb2-828" aria-hidden="true" tabindex="-1"></a>      diablo_output,</span>
<span id="cb2-829"><a href="#cb2-829" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-830"><a href="#cb2-830" aria-hidden="true" tabindex="-1"></a>      <span class="at">label_cols =</span> <span class="fu">list</span>(</span>
<span id="cb2-831"><a href="#cb2-831" aria-hidden="true" tabindex="-1"></a>        <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"Name"</span>,</span>
<span id="cb2-832"><a href="#cb2-832" aria-hidden="true" tabindex="-1"></a>        <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"name"</span></span>
<span id="cb2-833"><a href="#cb2-833" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb2-834"><a href="#cb2-834" aria-hidden="true" tabindex="-1"></a>      <span class="at">latent_dimensions =</span> <span class="fu">paste</span>(<span class="st">"Component"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)</span>
<span id="cb2-835"><a href="#cb2-835" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-836"><a href="#cb2-836" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-837"><a href="#cb2-837" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-838"><a href="#cb2-838" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Table of top contributing features</span></span>
<span id="cb2-839"><a href="#cb2-839" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-840"><a href="#cb2-840" aria-hidden="true" tabindex="-1"></a>    diablo_selected_features,</span>
<span id="cb2-841"><a href="#cb2-841" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_selected_features</span>(diablo_output) </span>
<span id="cb2-842"><a href="#cb2-842" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-843"><a href="#cb2-843" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-844"><a href="#cb2-844" aria-hidden="true" tabindex="-1"></a><span class="co"># Evaluating the integration results -------------------------------------------</span></span>
<span id="cb2-845"><a href="#cb2-845" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-846"><a href="#cb2-846" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Evaluating DIABLO selected features against single-omics results</span></span>
<span id="cb2-847"><a href="#cb2-847" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-848"><a href="#cb2-848" aria-hidden="true" tabindex="-1"></a>    diablo_selected_vs_singleomics_table,</span>
<span id="cb2-849"><a href="#cb2-849" aria-hidden="true" tabindex="-1"></a>    <span class="fu">evaluate_feature_selection_table</span>(</span>
<span id="cb2-850"><a href="#cb2-850" aria-hidden="true" tabindex="-1"></a>      diablo_output,</span>
<span id="cb2-851"><a href="#cb2-851" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-852"><a href="#cb2-852" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_names =</span> <span class="fu">list</span>(</span>
<span id="cb2-853"><a href="#cb2-853" aria-hidden="true" tabindex="-1"></a>        <span class="st">"snps"</span> <span class="ot">=</span> <span class="st">"qtl_type"</span>,</span>
<span id="cb2-854"><a href="#cb2-854" aria-hidden="true" tabindex="-1"></a>        <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"de_signif"</span>,</span>
<span id="cb2-855"><a href="#cb2-855" aria-hidden="true" tabindex="-1"></a>        <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"de_signif"</span></span>
<span id="cb2-856"><a href="#cb2-856" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb2-857"><a href="#cb2-857" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-858"><a href="#cb2-858" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-859"><a href="#cb2-859" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-860"><a href="#cb2-860" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting DIABLO features weight against single-omics results</span></span>
<span id="cb2-861"><a href="#cb2-861" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-862"><a href="#cb2-862" aria-hidden="true" tabindex="-1"></a>    diablo_features_weight_vs_singleomics_plot,</span>
<span id="cb2-863"><a href="#cb2-863" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_features_weight_covariate</span>(</span>
<span id="cb2-864"><a href="#cb2-864" aria-hidden="true" tabindex="-1"></a>      diablo_output, </span>
<span id="cb2-865"><a href="#cb2-865" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete, </span>
<span id="cb2-866"><a href="#cb2-866" aria-hidden="true" tabindex="-1"></a>      <span class="at">covariate =</span> <span class="fu">list</span>(</span>
<span id="cb2-867"><a href="#cb2-867" aria-hidden="true" tabindex="-1"></a>        <span class="st">"snps"</span> <span class="ot">=</span> <span class="st">"qtl_type"</span>, </span>
<span id="cb2-868"><a href="#cb2-868" aria-hidden="true" tabindex="-1"></a>        <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"de_status"</span>, </span>
<span id="cb2-869"><a href="#cb2-869" aria-hidden="true" tabindex="-1"></a>        <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"de_status"</span></span>
<span id="cb2-870"><a href="#cb2-870" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb2-871"><a href="#cb2-871" aria-hidden="true" tabindex="-1"></a>      <span class="at">remove_null_weight =</span> <span class="cn">TRUE</span></span>
<span id="cb2-872"><a href="#cb2-872" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-873"><a href="#cb2-873" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-874"><a href="#cb2-874" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-875"><a href="#cb2-875" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Genes GO annotation file </span></span>
<span id="cb2-876"><a href="#cb2-876" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-877"><a href="#cb2-877" aria-hidden="true" tabindex="-1"></a>    rnaseq_go_terms_file,</span>
<span id="cb2-878"><a href="#cb2-878" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system.file</span>(</span>
<span id="cb2-879"><a href="#cb2-879" aria-hidden="true" tabindex="-1"></a>      <span class="st">"extdata/transcriptomics_go_annotation.csv"</span>, </span>
<span id="cb2-880"><a href="#cb2-880" aria-hidden="true" tabindex="-1"></a>      <span class="at">package =</span> <span class="st">"moiraine"</span></span>
<span id="cb2-881"><a href="#cb2-881" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-882"><a href="#cb2-882" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"file"</span></span>
<span id="cb2-883"><a href="#cb2-883" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-884"><a href="#cb2-884" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-885"><a href="#cb2-885" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Genes GO annotation data-frame </span></span>
<span id="cb2-886"><a href="#cb2-886" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-887"><a href="#cb2-887" aria-hidden="true" tabindex="-1"></a>    rnaseq_go_df,</span>
<span id="cb2-888"><a href="#cb2-888" aria-hidden="true" tabindex="-1"></a>    <span class="fu">read_csv</span>(rnaseq_go_terms_file) <span class="sc">|&gt;</span> </span>
<span id="cb2-889"><a href="#cb2-889" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>(go_domain <span class="sc">==</span> <span class="st">"Biological process"</span>)</span>
<span id="cb2-890"><a href="#cb2-890" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-891"><a href="#cb2-891" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-892"><a href="#cb2-892" aria-hidden="true" tabindex="-1"></a>  <span class="do">## GO term sets</span></span>
<span id="cb2-893"><a href="#cb2-893" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-894"><a href="#cb2-894" aria-hidden="true" tabindex="-1"></a>    go_sets,</span>
<span id="cb2-895"><a href="#cb2-895" aria-hidden="true" tabindex="-1"></a>    <span class="fu">make_feature_sets_from_df</span>(</span>
<span id="cb2-896"><a href="#cb2-896" aria-hidden="true" tabindex="-1"></a>      rnaseq_go_df,</span>
<span id="cb2-897"><a href="#cb2-897" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_id =</span> <span class="st">"gene_id"</span>,</span>
<span id="cb2-898"><a href="#cb2-898" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_set =</span> <span class="st">"go_id"</span></span>
<span id="cb2-899"><a href="#cb2-899" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-900"><a href="#cb2-900" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-901"><a href="#cb2-901" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-902"><a href="#cb2-902" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Filtering GO term sets against measured features</span></span>
<span id="cb2-903"><a href="#cb2-903" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-904"><a href="#cb2-904" aria-hidden="true" tabindex="-1"></a>    go_sets_filtered,</span>
<span id="cb2-905"><a href="#cb2-905" aria-hidden="true" tabindex="-1"></a>    <span class="fu">reduce_feature_sets_data</span>(go_sets, mo_set_complete)</span>
<span id="cb2-906"><a href="#cb2-906" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-907"><a href="#cb2-907" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-908"><a href="#cb2-908" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Checking genes GO term sets against datasets</span></span>
<span id="cb2-909"><a href="#cb2-909" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-910"><a href="#cb2-910" aria-hidden="true" tabindex="-1"></a>    go_sets_check,</span>
<span id="cb2-911"><a href="#cb2-911" aria-hidden="true" tabindex="-1"></a>    <span class="fu">check_feature_sets</span>(</span>
<span id="cb2-912"><a href="#cb2-912" aria-hidden="true" tabindex="-1"></a>      go_sets_filtered, </span>
<span id="cb2-913"><a href="#cb2-913" aria-hidden="true" tabindex="-1"></a>      mo_set_complete,</span>
<span id="cb2-914"><a href="#cb2-914" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="st">"rnaseq"</span></span>
<span id="cb2-915"><a href="#cb2-915" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-916"><a href="#cb2-916" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-917"><a href="#cb2-917" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-918"><a href="#cb2-918" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Table of information about GO terms</span></span>
<span id="cb2-919"><a href="#cb2-919" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-920"><a href="#cb2-920" aria-hidden="true" tabindex="-1"></a>    go_sets_info,</span>
<span id="cb2-921"><a href="#cb2-921" aria-hidden="true" tabindex="-1"></a>    rnaseq_go_df <span class="sc">|&gt;</span></span>
<span id="cb2-922"><a href="#cb2-922" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">select</span>(go_id, go_name) <span class="sc">|&gt;</span></span>
<span id="cb2-923"><a href="#cb2-923" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">distinct</span>()</span>
<span id="cb2-924"><a href="#cb2-924" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-925"><a href="#cb2-925" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-926"><a href="#cb2-926" aria-hidden="true" tabindex="-1"></a>  <span class="do">## MOFA latent components enrichment analysis</span></span>
<span id="cb2-927"><a href="#cb2-927" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-928"><a href="#cb2-928" aria-hidden="true" tabindex="-1"></a>    mofa_enrichment_results,</span>
<span id="cb2-929"><a href="#cb2-929" aria-hidden="true" tabindex="-1"></a>    <span class="fu">evaluate_method_enrichment</span>(</span>
<span id="cb2-930"><a href="#cb2-930" aria-hidden="true" tabindex="-1"></a>      mofa_output,</span>
<span id="cb2-931"><a href="#cb2-931" aria-hidden="true" tabindex="-1"></a>      go_sets_filtered,</span>
<span id="cb2-932"><a href="#cb2-932" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="st">"rnaseq"</span>,</span>
<span id="cb2-933"><a href="#cb2-933" aria-hidden="true" tabindex="-1"></a>      <span class="at">latent_dimensions =</span> <span class="fu">paste</span>(<span class="st">"Factor"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>),</span>
<span id="cb2-934"><a href="#cb2-934" aria-hidden="true" tabindex="-1"></a>      <span class="at">use_abs =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-935"><a href="#cb2-935" aria-hidden="true" tabindex="-1"></a>      <span class="at">min_set_size =</span> <span class="dv">10</span>,</span>
<span id="cb2-936"><a href="#cb2-936" aria-hidden="true" tabindex="-1"></a>      <span class="at">add_missing_features =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-937"><a href="#cb2-937" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_complete,</span>
<span id="cb2-938"><a href="#cb2-938" aria-hidden="true" tabindex="-1"></a>      <span class="at">sets_info_df =</span> go_sets_info,</span>
<span id="cb2-939"><a href="#cb2-939" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_set =</span> <span class="st">"go_id"</span></span>
<span id="cb2-940"><a href="#cb2-940" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-941"><a href="#cb2-941" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-942"><a href="#cb2-942" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-943"><a href="#cb2-943" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting features weight for GO term 'GO:0031424'</span></span>
<span id="cb2-944"><a href="#cb2-944" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-945"><a href="#cb2-945" aria-hidden="true" tabindex="-1"></a>    mofa_enrichment_go0031424_plot,</span>
<span id="cb2-946"><a href="#cb2-946" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_features_weight_set</span>(</span>
<span id="cb2-947"><a href="#cb2-947" aria-hidden="true" tabindex="-1"></a>      mofa_output,</span>
<span id="cb2-948"><a href="#cb2-948" aria-hidden="true" tabindex="-1"></a>      go_sets_filtered[[<span class="st">"GO:0031424"</span>]],</span>
<span id="cb2-949"><a href="#cb2-949" aria-hidden="true" tabindex="-1"></a>      <span class="at">set_name =</span> <span class="st">"GO:0031424 (keratinization)"</span>,</span>
<span id="cb2-950"><a href="#cb2-950" aria-hidden="true" tabindex="-1"></a>      <span class="at">features_metric =</span> <span class="st">"importance"</span>,</span>
<span id="cb2-951"><a href="#cb2-951" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="st">"rnaseq"</span>,</span>
<span id="cb2-952"><a href="#cb2-952" aria-hidden="true" tabindex="-1"></a>      <span class="at">latent_dimensions =</span> <span class="fu">paste</span>(<span class="st">"Factor"</span>, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>)),</span>
<span id="cb2-953"><a href="#cb2-953" aria-hidden="true" tabindex="-1"></a>      <span class="at">point_alpha =</span> <span class="fl">0.2</span></span>
<span id="cb2-954"><a href="#cb2-954" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-955"><a href="#cb2-955" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-956"><a href="#cb2-956" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-957"><a href="#cb2-957" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Assessing DIABLO samples clustering</span></span>
<span id="cb2-958"><a href="#cb2-958" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-959"><a href="#cb2-959" aria-hidden="true" tabindex="-1"></a>    diablo_silhouette,</span>
<span id="cb2-960"><a href="#cb2-960" aria-hidden="true" tabindex="-1"></a>    <span class="fu">compute_samples_silhouette</span>(</span>
<span id="cb2-961"><a href="#cb2-961" aria-hidden="true" tabindex="-1"></a>      diablo_output,</span>
<span id="cb2-962"><a href="#cb2-962" aria-hidden="true" tabindex="-1"></a>      mo_set_complete,</span>
<span id="cb2-963"><a href="#cb2-963" aria-hidden="true" tabindex="-1"></a>      <span class="st">"status"</span></span>
<span id="cb2-964"><a href="#cb2-964" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-965"><a href="#cb2-965" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-966"><a href="#cb2-966" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details><section id="introduction" class="level2" data-number="13.1"><h2 data-number="13.1" class="anchored" data-anchor-id="introduction">
<span class="header-section-number">13.1</span> Introduction</h2>
<p>As explained in <a href="interpretation.html#sec-interpretation-dim-reduction"><span>Section&nbsp;11.1</span></a>, the integration tools considered in <code>moiraine</code> all perform dimension reduction of the omics datasets, and therefore all construct latent dimensions which can be described in terms of their features weight and their samples score. Therefore, we can compare the results of different integration methods in terms of their latent dimensions, by comparing either the latent dimensions’ samples score, or their features weight. When comparing the samples score, we are answering the question: <em>are the different methods capturing the same trends from the datasets?</em> By comparing features weight, we are instead answering a slightly different question: <em>are the different methods highlighting similar sets of features as driving variation in the datasets?</em> Of course, these comparisons are related, as two latent dimensions that identify very different trends in the data will most likely give high importance to very different sets of features.</p>
<p>Note that, although this chapter mainly considers a comparison of different integration methods, we could also compare the results of a same integration but with different pre-processing options (e.g.&nbsp;to assess the impact of datasets normalisation or pre-filtering on the integration results). To illustrate this, we will run a second MOFA analysis, but this time using the datasets that have been pre-filtered to retain the most variable features (<a href="prefiltering.html#sec-prefiltering-unsupervised"><span>Section&nbsp;6.3.1</span></a>), as opposed to the supervised pre-filtering results that we have used in the main analysis (<a href="prefiltering.html#sec-prefiltering-supervised"><span>Section&nbsp;6.3.2</span></a>).</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Creating the input object for the MOFA pipeline</span></span>
<span>  <span class="co">## using the unsupervised preselection results</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_unsupervised_input</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_input_mofa.html">get_input_mofa</a></span><span class="op">(</span></span>
<span>      <span class="va">mo_presel_unsupervised</span>,</span>
<span>      options_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        data_options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>scale_views <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>        model_options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>          likelihoods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>            <span class="st">"snps"</span> <span class="op">=</span> <span class="st">"poisson"</span>,</span>
<span>            <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"gaussian"</span>,</span>
<span>            <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"gaussian"</span><span class="op">)</span></span>
<span>        <span class="op">)</span>,</span>
<span>        training_options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>seed <span class="op">=</span> <span class="fl">72</span><span class="op">)</span></span>
<span>      <span class="op">)</span>,</span>
<span>      only_common_samples <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Training the model with the MOFA algorithm</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_unsupervised_trained</span>,</span>
<span>    <span class="fu">run_mofa</span><span class="op">(</span></span>
<span>      <span class="va">mofa_unsupervised_input</span>,</span>
<span>      save_data <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      use_basilisk <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Formatting MOFA output</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_unsupervised_output</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">mofa_unsupervised_trained</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="extracting-integration-results" class="level2" data-number="13.2"><h2 data-number="13.2" class="anchored" data-anchor-id="extracting-integration-results">
<span class="header-section-number">13.2</span> Extracting integration results</h2>
<p>The first step in order to compare the results from different methods is to convert them to a standardised format with the <code><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output()</a></code> function, which we did in <a href="interpretation.html"><span>Chapter&nbsp;11</span></a>. The function returns an <code>output_dimension_reduction</code> object which contains information about the latent dimensions’ features weight and samples score, as well as a summary of the percentage of variance explained for each dataset. The <code>output_dimension_reduction</code> object created also has an attribute called <code>method</code>, which contains the name of the integration method used:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_output</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">diablo_output</span>, <span class="st">"method"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "DIABLO"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To facilitate comparisons, we will combine these objects into a list:</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">spls_output</span>, <span class="va">so2pls_output</span>, <span class="va">mofa_output</span>, <span class="va">diablo_output</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>This list will be the input of most of the functions that we will showcase in this chapter.</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">output_list</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can notice that we didn’t give any name to the elements of the list that we just created. When comparing results from different integration methods, the comparison functions implemented in <code>moiraine</code> automatically extract the name of the method from each object (as shown above), and use this as labels in the plots. However, when we want to compare several results from a same integration method, we need to give unique names to the different elements in the output list, e.g.:</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list_mofa</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"MOFA (supervised pref.)"</span> <span class="op">=</span> <span class="va">mofa_output</span>,</span>
<span>    <span class="st">"MOFA (unsupervised pref.)"</span> <span class="op">=</span> <span class="va">mofa_unsupervised_output</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">output_list_mofa</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="clustered-correlation-heatmaps" class="level2" data-number="13.3"><h2 data-number="13.3" class="anchored" data-anchor-id="clustered-correlation-heatmaps">
<span class="header-section-number">13.3</span> Clustered correlation heatmaps</h2>
<p>We can visualise the correlation between the samples score and features weight of the different latent dimensions constructed by the integration methods as a heatmap, using the <code><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr()</a></code> function. The main input parameter of the function is a list of <code>output_dimension_reduction</code> objects, which we constructed in the previous section:</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr</a></span><span class="op">(</span><span class="va">output_list</span>, legend_ncol <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/comparison-heatmap-corr-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>The function generates two half-heatmaps. The heatmap on the left is a visualisation of the correlation between the features weight of the different latent dimensions; the one on the right shows the correlation between their samples score. Since correlation matrices are symmetric, only one triangle of each matrix is represented. The rows and columns of the heatmaps each correspond to one of the latent dimensions generated by one of the integration methods. Their name is abbreviated (C stands for component, F for Factor, JC for joint component, RSC for rnaseq-specific component and MSC for metabolome-specific component). The method through which each latent dimension was generated is indicated next to its name as a coloured annotation. In each heatmap, the rows and columns have been ordered according to a clustering performed on the correlation matrix, so that the latent dimensions most similar (in terms of samples score or features weight) are next to each other.</p>
<p>By default, the function represents all of the latent dimensions generated by each method, but the <code>latent_dimensions</code> parameter allows us to select only the ones we are interested in comparing. It accepts a named list, where each element gives for an integration method the latent dimensions to use. The names of the list must correspond to methods name. For example, we will look at all latent dimensions from sPLS and DIABLO, but restrict to only the first four MOFA factors, and only the joint component and first two dataset-specific components for sO2PLS:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"sO2PLS"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"joint component 1"</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"metabolome specific component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>      <span class="st">"rnaseq specific component 1"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="st">"MOFA"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/comparison-heatmap-corr-subset-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>From the heatmaps, we can see that some latent dimensions constructed by the different methods seem to capture similar trends in the data. For example, MOFA factor 1, sPLS component 1, DIABLO component 1 and sO2PLS joint component 1 are all strongly correlated in terms of their samples score. Their correlation in terms of features weight is a bit lower, which is due to the fact that some methods perform features selection, therefore all non-selected features are given a weight of 0. Note that the sign of the correlation is interesting but not very important. We can also see some latent dimensions that seem correlated with respect to one metric but not the other. For example, there is a strong correlation between the samples score of MOFA factor 4 and sPLS component 2, but this is not reflected in their features weight. Again, that can be because one method performs latent selection and not the other. On the other hand, the correlation between sPLS component 3 and DIABLO component 2 is stronger when looking at their features weight than at their samples score.</p>
<p>When using the function to compare the results of the two MOFA runs, the names that we gave to the list of output will be used as method name, and these are the names to use in order to select some latent dimensions of interest. For example:</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list_mofa</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"MOFA (supervised pref.)"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Factor "</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>,</span>
<span>    <span class="st">"MOFA (unsupervised pref.)"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Factor "</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/comparison-heatmap-corr-mofa-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="a-note-about-missing-features" class="level2" data-number="13.4"><h2 data-number="13.4" class="anchored" data-anchor-id="a-note-about-missing-features">
<span class="header-section-number">13.4</span> A note about missing features</h2>
<p>Typically, when comparing the output of different integration methods, we assume that the same input data was used for each method. However, when we are instead comparing the impact of data pre-processing on the integration results, it is possible that not all features are present in the input data for each of the methods. This is for example the case when assessing the impact of different prefiltering settings (e.g.&nbsp;supervised vs unsupervised) on the results of a particular integration method. It is also the case if we are comparing the integration results when using different subset of omics data as input: in our example, both DIABLO and MOFA were used on the three omics datasets (genomics, transcriptomics and metabolomics), while sPLS and sO2PLS were run on the transcriptomics and metabolomics datasets only. If that is the case, some features will have no weight in some of the results we are comparing (as they are not in the input data analysed by the integration method). There are two approaches to deal with that:</p>
<ul>
<li><p>Approach 1: remove these features from the comparison. The correlation between the features weight of any two latent dimensions will only be calculated using the features that are assigned a weight by both latent dimensions. This is the default approach used by <code><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr()</a></code> and other comparison functions.</p></li>
<li><p>Approach 2: the features that are assigned a weight by some methods but not others are retained for the comparison, and the missing weights are replaced with zero values.</p></li>
</ul>
<p>The problem with approach 2 is that it might artificially decrease the computed correlation coefficients, which is why approach 1 is the default in this package. However, approach 2 may reveal interesting patterns and therefore was also implemented.</p>
<p>In <code><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr()</a></code> and other comparison functions where relevant, the choice between these two approaches is controlled by the <code>include_missing_features</code> parameter. The first approach is selected with <code>include_missing_features = FALSE</code>, while the second approach is selected with <code>include_missing_features = TRUE</code>.</p>
<p>Note that the first approach does not remove features that were discarded as a result of feature selection by one of the integration method, but only features that were not in the input data for the integration method. For example, sO2PLS performs feature selection for the joint components, and as a consequence the features that are not selected as assigned a weight of 0. These features are not impacted by the choice of approach.</p>
</section><section id="correlation-matrices" class="level2" data-number="13.5"><h2 data-number="13.5" class="anchored" data-anchor-id="correlation-matrices">
<span class="header-section-number">13.5</span> Correlation matrices</h2>
<p>We can obtain the correlation coefficients displayed in the heatmaps with the <code><a href="https://rdrr.io/pkg/moiraine/man/get_features_weight_correlation.html">get_features_weight_correlation()</a></code> and <code><a href="https://rdrr.io/pkg/moiraine/man/get_samples_score_correlation.html">get_samples_score_correlation()</a></code> functions:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_samples_score_correlation.html">get_samples_score_correlation</a></span><span class="op">(</span><span class="va">output_list</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                                          sPLS___Component 1 sPLS___Component 2</span></span>
<span><span class="co">#&gt; sPLS___Component 1                              1.000000000        0.004845557</span></span>
<span><span class="co">#&gt; sPLS___Component 2                              0.004845557        1.000000000</span></span>
<span><span class="co">#&gt; sO2PLS___joint component 1                      0.973652291       -0.006320588</span></span>
<span><span class="co">#&gt; sO2PLS___rnaseq specific component 1           -0.123298596        0.025496756</span></span>
<span><span class="co">#&gt; sO2PLS___metabolome specific component 1        0.033930444       -0.148278174</span></span>
<span><span class="co">#&gt;                                          sO2PLS___joint component 1</span></span>
<span><span class="co">#&gt; sPLS___Component 1                                      0.973652291</span></span>
<span><span class="co">#&gt; sPLS___Component 2                                     -0.006320588</span></span>
<span><span class="co">#&gt; sO2PLS___joint component 1                              1.000000000</span></span>
<span><span class="co">#&gt; sO2PLS___rnaseq specific component 1                   -0.007293233</span></span>
<span><span class="co">#&gt; sO2PLS___metabolome specific component 1               -0.003995068</span></span>
<span><span class="co">#&gt;                                          sO2PLS___rnaseq specific component 1</span></span>
<span><span class="co">#&gt; sPLS___Component 1                                               -0.123298596</span></span>
<span><span class="co">#&gt; sPLS___Component 2                                                0.025496756</span></span>
<span><span class="co">#&gt; sO2PLS___joint component 1                                       -0.007293233</span></span>
<span><span class="co">#&gt; sO2PLS___rnaseq specific component 1                              1.000000000</span></span>
<span><span class="co">#&gt; sO2PLS___metabolome specific component 1                          0.018259608</span></span>
<span><span class="co">#&gt;                                          sO2PLS___metabolome specific component 1</span></span>
<span><span class="co">#&gt; sPLS___Component 1                                                    0.033930444</span></span>
<span><span class="co">#&gt; sPLS___Component 2                                                   -0.148278174</span></span>
<span><span class="co">#&gt; sO2PLS___joint component 1                                           -0.003995068</span></span>
<span><span class="co">#&gt; sO2PLS___rnaseq specific component 1                                  0.018259608</span></span>
<span><span class="co">#&gt; sO2PLS___metabolome specific component 1                              1.000000000</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_features_weight_correlation.html">get_features_weight_correlation</a></span><span class="op">(</span><span class="va">output_list</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                                          sPLS___Component 1 sPLS___Component 2</span></span>
<span><span class="co">#&gt; sPLS___Component 1                              1.000000000        0.008381676</span></span>
<span><span class="co">#&gt; sPLS___Component 2                              0.008381676        1.000000000</span></span>
<span><span class="co">#&gt; sO2PLS___joint component 1                      0.702008542       -0.005438028</span></span>
<span><span class="co">#&gt; sO2PLS___rnaseq specific component 1           -0.200398661        0.012947768</span></span>
<span><span class="co">#&gt; sO2PLS___metabolome specific component 1        0.151077963       -0.180756532</span></span>
<span><span class="co">#&gt;                                          sO2PLS___joint component 1</span></span>
<span><span class="co">#&gt; sPLS___Component 1                                      0.702008542</span></span>
<span><span class="co">#&gt; sPLS___Component 2                                     -0.005438028</span></span>
<span><span class="co">#&gt; sO2PLS___joint component 1                              1.000000000</span></span>
<span><span class="co">#&gt; sO2PLS___rnaseq specific component 1                   -0.418570009</span></span>
<span><span class="co">#&gt; sO2PLS___metabolome specific component 1                0.113863838</span></span>
<span><span class="co">#&gt;                                          sO2PLS___rnaseq specific component 1</span></span>
<span><span class="co">#&gt; sPLS___Component 1                                                -0.20039866</span></span>
<span><span class="co">#&gt; sPLS___Component 2                                                 0.01294777</span></span>
<span><span class="co">#&gt; sO2PLS___joint component 1                                        -0.41857001</span></span>
<span><span class="co">#&gt; sO2PLS___rnaseq specific component 1                               1.00000000</span></span>
<span><span class="co">#&gt; sO2PLS___metabolome specific component 1                                   NA</span></span>
<span><span class="co">#&gt;                                          sO2PLS___metabolome specific component 1</span></span>
<span><span class="co">#&gt; sPLS___Component 1                                                      0.1510780</span></span>
<span><span class="co">#&gt; sPLS___Component 2                                                     -0.1807565</span></span>
<span><span class="co">#&gt; sO2PLS___joint component 1                                              0.1138638</span></span>
<span><span class="co">#&gt; sO2PLS___rnaseq specific component 1                                           NA</span></span>
<span><span class="co">#&gt; sO2PLS___metabolome specific component 1                                1.0000000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that the formatting of the row and column names is designed for easier handling in the various plotting functions.</p>
</section><section id="pairwise-comparisons" class="level2" data-number="13.6"><h2 data-number="13.6" class="anchored" data-anchor-id="pairwise-comparisons">
<span class="header-section-number">13.6</span> Pairwise comparisons</h2>
<p>While the <code><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr()</a></code> function can handle any number of methods to compare (in our example it was 4), the heatmap can become difficult to interpret. As an alternative option, it is possible to visualise the correlation between the latent dimensions of any two methods. This is what the function <code><a href="https://rdrr.io/pkg/moiraine/man/comparison_plot_correlation.html">comparison_plot_correlation()</a></code> is for. It also takes as input a list of methods’ output, but the list must only contain two elements. For example, we will compare the results of sPLS and sO2PLS:</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/comparison_plot_correlation.html">comparison_plot_correlation</a></span><span class="op">(</span><span class="va">output_list</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/comparison-plot-correlation-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>As with <code><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr()</a></code>, the function displays the correlation between the latent dimensions’ samples score on the left, and between their features weight on the right, but using correlation plots rather than heatmaps. By default, only correlation coefficients above 0.2 have their value displayed (for better clarity), but this can be customised through the <code>min_show_corr</code> argument.</p>
<p>By default, the function plots both samples score correlations and features weight correlation, but it is possible to display only one of them by setting the <code>by</code> argument to either <code>samples</code> or <code>features</code>. In addition, as for the heatmap plotting function, it is possible to focus on certain latent dimensions by passing a named list of latent dimensions name to the <code>latent_dimensions</code> argument:</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/comparison_plot_correlation.html">comparison_plot_correlation</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  by <span class="op">=</span> <span class="st">"samples"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"sO2PLS"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"joint component 1"</span>, <span class="st">"rnaseq specific component 1"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/comparison-plot-correlation-subset-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section><section id="comparing-samples-score" class="level2" data-number="13.7"><h2 data-number="13.7" class="anchored" data-anchor-id="comparing-samples-score">
<span class="header-section-number">13.7</span> Comparing samples score</h2>
<p>The correlation plots shown above provide a useful summary of the similarity between the latent dimensions constructed by different integration methods. We can then investigate further how any two latent dimensions (from two different integration methods) are related. One way to do so is to compare their samples score in a scatterplot. In <a href="interpretation.html#sec-interpretation-samples-pair"><span>Section&nbsp;11.4.2</span></a>, the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair()</a></code> function has been used to display the samples score of two latent dimensions from the result of an integration method, e.g.:</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mo_set_complete</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Component 1"</span>, <span class="st">"Component 2"</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/plot-samples-score-pair-diablo-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>This function can also be used to compare the samples score of two latent dimensions from two integration different methods. This is done by passing to the function a list of length 2 containing the output of two different methods. The name of the latent dimensions to compare are provided as a named list, where each name corresponds to either the name of the method (if the input list is not named) or the name of the element in the input list. So for example, to compare the first latent dimension of MOFA and DIABLO:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">mo_set</span><span class="op">)</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 9 rows containing missing values (`geom_point()`).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/plot-samples-score-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We can see that the samples score of the two latent dimensions are strongly correlated, showing that the two latent dimensions capture a similar trend in the data.</p>
</section><section id="comparing-features-weight" class="level2" data-number="13.8"><h2 data-number="13.8" class="anchored" data-anchor-id="comparing-features-weight">
<span class="header-section-number">13.8</span> Comparing features weight</h2>
<p>In a similar way, we can compare the features importance score of two latent dimensions from two different integration methods. In <a href="interpretation.html#sec-interpretation-features-pair"><span>Section&nbsp;11.5.2</span></a>, the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair()</a></code> function has been used to display the features weight or importance score of two latent dimensions from the result of an integration method, e.g.:</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Component 1"</span>, <span class="st">"Component 2"</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/plot-features_weight-pair-mofa-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>This function can also be used to compare the features weight of two latent dimensions from two different integration methods. This is done by passing to the function a list of length 2 containing the output of two different methods. The name of the latent dimensions to compare are provided as a named list, where each name corresponds to either the name of the method (if the input list is not named) or the name of the element in the input list. We will again compare MOFA factor 1 and DIABLO component 1:</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">mo_set</span><span class="op">)</span>,</span>
<span>  features_metric <span class="op">=</span> <span class="st">"importance"</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/plot-features-weight-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We can see that while MOFA and DIABLO identify the same metabolites as being the most important to separate healthy and infected animals, the genomic markers and genes that are given the highest importance score by MOFA are not selected with DIABLO.</p>
<p>By default, the function plots the features signed importance (i.e.&nbsp;importance score to which the sign of their weight was added), but in this case it makes sense to look at their (non-signed) importance score, which we have done by setting the <code>features_metric</code> argument to <code>'importance'</code>. Note that we can also compare their raw weight, by setting <code>features_metric</code> to <code>'weight'</code>. If that is the case, keep in mind that the weights from different latent dimensions live on different scales, so the raw values are not directly comparable (but the features ranking and weight sign are). We can also focus on a subset of datasets by passing their name to the <code>datasets</code> argument:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span>,</span>
<span>  features_metric <span class="op">=</span> <span class="st">"weight"</span>,</span>
<span>  datasets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">mo_set</span><span class="op">)</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/plot-features-weight-raw-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>By default, the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair()</a></code> function uses the geometric consensus importance metric (more details in the next section) to highlight the 5 features identified as most important by both methods. Both the number of features highlighted and the metric used can be controlled, through the <code>top_n</code> and <code>metric</code> arguments, respectively. In the following section, we will expand on the concept of consensus importance and the different metrics available.</p>
</section><section id="sec-comparison-consensus" class="level2" data-number="13.9"><h2 data-number="13.9" class="anchored" data-anchor-id="sec-comparison-consensus">
<span class="header-section-number">13.9</span> Feature consensus importance</h2>
<section id="possible-metrics" class="level3" data-number="13.9.1"><h3 data-number="13.9.1" class="anchored" data-anchor-id="possible-metrics">
<span class="header-section-number">13.9.1</span> Possible metrics</h3>
<p>When comparing the results of several integration methods, we might find that several methods uncover similar trends in the datasets, through the latent dimensions they construct. For these latent dimensions, it is interesting to assess which features are consistently selected by the different methods as driving these trends. We can do so by computing the consensus importance score of the features, which summarises the importance given to the features by the different integration methods. Different metrics can be used to compute the consensus importance score. With <span class="math inline">\(s_j^i\)</span> denoting the importance score of feature <span class="math inline">\(j\)</span> from method <span class="math inline">\(i\)</span>, <span class="math inline">\(i = 1 \dots n_i\)</span> (more specifically, from one latent dimension constructed by the integration method), the following metrics are implemented in the package:</p>
<ul>
<li><p>Geometric mean (<code>geometric</code>): calculated as <span class="math inline">\(CI_j^G = \exp\left(\frac{1}{n_i} \sum\limits_i \log(s_j^i)\right)\)</span></p></li>
<li><p>Harmonic mean (<code>harmonic</code>): calculated as <span class="math inline">\(CI_j^H = \frac{1}{\frac{1}{n_i}\sum\limits_i\frac{1}{s_j^i}}\)</span></p></li>
<li><p>L2 norm (<code>l2</code>): calculated as <span class="math inline">\(CI_j^{L2} = \sqrt{\sum\limits_i (s_j^i)^2}\)</span></p></li>
<li><p>Average (<code>average</code>): calculated as <span class="math inline">\(CI_j^A = \frac{1}{n_i} \sum\limits_i s_j^i\)</span></p></li>
<li><p>Product (<code>product</code>): calculated as <span class="math inline">\(CI_j^P = \prod\limits_i s_j^i\)</span></p></li>
<li><p>Minimum (<code>min</code>): calculated as <span class="math inline">\(CI_j^{min} = \min\limits_i(s_j^i)\)</span></p></li>
<li><p>Maximum (<code>max</code>): calculated as <span class="math inline">\(CI_j^{max} = \max\limits_i(s_j^i)\)</span></p></li>
</ul>
<p>The difference between these metrics can be visualised in the simple case where features are assigned two different importance scores, e.g.&nbsp;by two different integration methods:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/show_consensus_metrics.html">show_consensus_metrics</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/show-consensus-importance-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>In the plots, the consensus importance values have been normalised so that the highest value is 1. As we can see, metrics such as the geometric mean, harmonic mean, product or minimum will give higher consensus scores to features that are consistently assigned a high importance score across all methods, while features that have high importance score with one method but low score with the other will get a lower consensus score. Conversely, metrics such as the L2-norm or maximum prioritise features that are given a high importance score by at least one method, regardless of their importance score with other methods.</p>
</section><section id="computing-features-consensus-importance-score" class="level3" data-number="13.9.2"><h3 data-number="13.9.2" class="anchored" data-anchor-id="computing-features-consensus-importance-score">
<span class="header-section-number">13.9.2</span> Computing features’ consensus importance score</h3>
<p>The function <code><a href="https://rdrr.io/pkg/moiraine/man/compute_consensus_importance.html">compute_consensus_importance()</a></code> takes as input a list of output objects from different integration methods, as well as a named list indicating which latent dimension from each output object should be considered. By default, the geometric mean is used, but this can be changed through the <code>metric</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">consensus_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/compute_consensus_importance.html">compute_consensus_importance</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">consensus_df</span></span>
<span><span class="co">#&gt; # A tibble: 2,049 × 3</span></span>
<span><span class="co">#&gt;    dataset feature_id         importance</span></span>
<span><span class="co">#&gt;    &lt;fct&gt;   &lt;chr&gt;                   &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 snps    ARS-BFGL-NGS-27468      1    </span></span>
<span><span class="co">#&gt;  2 snps    BovineHD2300010006      0.908</span></span>
<span><span class="co">#&gt;  3 snps    BovineHD0300000351      0.650</span></span>
<span><span class="co">#&gt;  4 snps    BovineHD0900026231      0.555</span></span>
<span><span class="co">#&gt;  5 snps    BovineHD0600028258      0.471</span></span>
<span><span class="co">#&gt;  6 snps    BovineHD1100030384      0.416</span></span>
<span><span class="co">#&gt;  7 snps    BTA-118330-no-rs        0.414</span></span>
<span><span class="co">#&gt;  8 snps    BovineHD1800016801      0.409</span></span>
<span><span class="co">#&gt;  9 snps    ARS-BFGL-NGS-84112      0.341</span></span>
<span><span class="co">#&gt; 10 snps    BovineHD1900011146      0.331</span></span>
<span><span class="co">#&gt; # ℹ 2,039 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To help compare the data-frame with the plots created in the previous section, we will add the features label to the consensus importance data-frame.</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">features_labels_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_features_labels.html">get_features_labels</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>       <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">consensus_df</span> <span class="op">&lt;-</span> <span class="va">consensus_df</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">features_labels_df</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dataset"</span>, <span class="st">"feature_id"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">consensus_df</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span><span class="va">importance</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 15 × 4</span></span>
<span><span class="co">#&gt; # Groups:   dataset [3]</span></span>
<span><span class="co">#&gt;    dataset    feature_id         importance label                </span></span>
<span><span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;                   &lt;dbl&gt; &lt;chr&gt;                </span></span>
<span><span class="co">#&gt;  1 metabolome HMDB00094               1     Citric acid          </span></span>
<span><span class="co">#&gt;  2 metabolome HMDB00042               0.508 Acetic acid          </span></span>
<span><span class="co">#&gt;  3 metabolome HMDB00169               0.448 D-Mannose            </span></span>
<span><span class="co">#&gt;  4 metabolome HMDB00148               0.375 L-Glutamic acid      </span></span>
<span><span class="co">#&gt;  5 metabolome HMDB00357               0.350 3-Hydroxybutyric acid</span></span>
<span><span class="co">#&gt;  6 rnaseq     ENSBTAG00000022715      1     ENSBTAG00000022715   </span></span>
<span><span class="co">#&gt;  7 rnaseq     ENSBTAG00000050618      0.877 ENSBTAG00000050618   </span></span>
<span><span class="co">#&gt;  8 rnaseq     ENSBTAG00000051722      0.701 ENSBTAG00000051722   </span></span>
<span><span class="co">#&gt;  9 rnaseq     ENSBTAG00000049808      0.698 IL3RA                </span></span>
<span><span class="co">#&gt; 10 rnaseq     ENSBTAG00000050206      0.687 ENSBTAG00000050206   </span></span>
<span><span class="co">#&gt; 11 snps       ARS-BFGL-NGS-27468      1     ARS-BFGL-NGS-27468   </span></span>
<span><span class="co">#&gt; 12 snps       BovineHD2300010006      0.908 BovineHD2300010006   </span></span>
<span><span class="co">#&gt; 13 snps       BovineHD0300000351      0.650 BovineHD0300000351   </span></span>
<span><span class="co">#&gt; 14 snps       BovineHD0900026231      0.555 BovineHD0900026231   </span></span>
<span><span class="co">#&gt; 15 snps       BovineHD0600028258      0.471 BovineHD0600028258</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that, using the geometric mean, the feature from the metabolomics dataset that is given the highest consensus importance is citric acid. Note that the metric chosen matters less when the latent dimensions compared are in strong agreement, as it is the case here:</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/compute_consensus_importance.html">compute_consensus_importance</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"average"</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">features_labels_df</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dataset"</span>, <span class="st">"feature_id"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span><span class="va">importance</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 15 × 4</span></span>
<span><span class="co">#&gt; # Groups:   dataset [3]</span></span>
<span><span class="co">#&gt;    dataset    feature_id             importance label                 </span></span>
<span><span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;                       &lt;dbl&gt; &lt;chr&gt;                 </span></span>
<span><span class="co">#&gt;  1 metabolome HMDB00094                   1     Citric acid           </span></span>
<span><span class="co">#&gt;  2 metabolome HMDB00042                   0.508 Acetic acid           </span></span>
<span><span class="co">#&gt;  3 metabolome HMDB00169                   0.448 D-Mannose             </span></span>
<span><span class="co">#&gt;  4 metabolome HMDB00357                   0.377 3-Hydroxybutyric acid </span></span>
<span><span class="co">#&gt;  5 metabolome HMDB00148                   0.375 L-Glutamic acid       </span></span>
<span><span class="co">#&gt;  6 rnaseq     ENSBTAG00000022715          1     ENSBTAG00000022715    </span></span>
<span><span class="co">#&gt;  7 rnaseq     ENSBTAG00000050618          0.872 ENSBTAG00000050618    </span></span>
<span><span class="co">#&gt;  8 rnaseq     ENSBTAG00000051722          0.711 ENSBTAG00000051722    </span></span>
<span><span class="co">#&gt;  9 rnaseq     ENSBTAG00000049808          0.699 IL3RA                 </span></span>
<span><span class="co">#&gt; 10 rnaseq     ENSBTAG00000050206          0.685 ENSBTAG00000050206    </span></span>
<span><span class="co">#&gt; 11 snps       ARS-BFGL-NGS-27468          1     ARS-BFGL-NGS-27468    </span></span>
<span><span class="co">#&gt; 12 snps       BovineHD2300010006          0.883 BovineHD2300010006    </span></span>
<span><span class="co">#&gt; 13 snps       BovineHD0100032240          0.645 BovineHD0100032240    </span></span>
<span><span class="co">#&gt; 14 snps       Hapmap55381-rs29025399      0.645 Hapmap55381-rs29025399</span></span>
<span><span class="co">#&gt; 15 snps       BovineHD0300000351          0.625 BovineHD0300000351</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair()</a></code> function, we can specify which metric should be used for the consensus importance to highlight the top features. For example, we can illustrate the top features according to the average consensus metric:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"average"</span>,</span>
<span>  datasets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="comparison_files/figure-html/plot-features-weight-average-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section></section><section id="recap-targets-list" class="level2" data-number="13.10"><h2 data-number="13.10" class="anchored" data-anchor-id="recap-targets-list">
<span class="header-section-number">13.10</span> Recap – targets list</h2>
<p>Although we didn’t create many new target for the plots in this chapter, we can turn some plots into targets.</p>
<details><summary>
Targets list for comparing different integration methods’ output
</summary><div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Creating a list of integration methods output objects</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">output_list</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">spls_output</span>, <span class="va">so2pls_output</span>, <span class="va">mofa_output</span>, <span class="va">diablo_output</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Heatmap for comparison of integration methods output</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">comparison_methods_heatmap_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr</a></span><span class="op">(</span><span class="va">output_list</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Correlation plot for comparison of sPLS and sO2PLS results</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_so2pls_correlation_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/comparison_plot_correlation.html">comparison_plot_correlation</a></span><span class="op">(</span><span class="va">output_list</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Comparison of samples score for MOFA factor 1 and DIABLO component 1</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_so2pls_samples_score_comparison_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair</a></span><span class="op">(</span></span>
<span>      <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      colour_by <span class="op">=</span> <span class="st">"status"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Comparison of features weight for MOFA factor 1 and sO2PLS joint component 1</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_so2pls_features_weight_comparison_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>      <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>        <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Table of features' consensus importance for MOFA factor 1 and sO2PLS joint</span></span>
<span>  <span class="co">## component 1</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_so2pls_features_weight_consensus_importance</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/compute_consensus_importance.html">compute_consensus_importance</a></span><span class="op">(</span></span>
<span>      <span class="va">output_list</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_features_labels.html">get_features_labels</a></span><span class="op">(</span></span>
<span>          <span class="va">mo_set</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>               <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></span>
<span>        <span class="op">)</span>,</span>
<span>        by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"dataset"</span>, <span class="st">"feature_id"</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details>

</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./evaluation.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>