<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>The moiraine R package user manual - 10&nbsp; Integration with DIABLO</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./interpretation.html" rel="next">
<link href="./mofa.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Integration with DIABLO</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The moiraine R package user manual</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/PlantandFoodResearch/moiraine" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="https://twitter.com/intent/tweet?url=%7Curl%7C" title="Twitter" class="sidebar-tool px-1"><i class="bi bi-twitter"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./example_dataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">The example dataset</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Data preparation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_import.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Importing data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inspecting_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Inspecting the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./modifying_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocessing.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data pre-processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prefiltering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Dataset pre-filtering</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Unsupervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Integration with sPLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./so2pls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Integration with sO2PLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mofa.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Integration with MOFA</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Supervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./diablo.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Integration with DIABLO</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Results interpretation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./interpretation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evaluation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./comparison.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Comparing the integration results</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#what-is-diablo" id="toc-what-is-diablo" class="nav-link active" data-scroll-target="#what-is-diablo"><span class="toc-section-number">10.1</span>  What is DIABLO?</a></li>
  <li><a href="#creating-the-diablo-input" id="toc-creating-the-diablo-input" class="nav-link" data-scroll-target="#creating-the-diablo-input"><span class="toc-section-number">10.2</span>  Creating the DIABLO input</a></li>
  <li>
<a href="#constructing-the-design-matrix" id="toc-constructing-the-design-matrix" class="nav-link" data-scroll-target="#constructing-the-design-matrix"><span class="toc-section-number">10.3</span>  Constructing the design matrix</a>
  <ul class="collapse">
<li><a href="#predefined-design-matrices" id="toc-predefined-design-matrices" class="nav-link" data-scroll-target="#predefined-design-matrices"><span class="toc-section-number">10.3.1</span>  Predefined design matrices</a></li>
  <li><a href="#estimating-the-design-matrix-through-pairwise-pls" id="toc-estimating-the-design-matrix-through-pairwise-pls" class="nav-link" data-scroll-target="#estimating-the-design-matrix-through-pairwise-pls"><span class="toc-section-number">10.3.2</span>  Estimating the design matrix through pairwise PLS</a></li>
  </ul>
</li>
  <li><a href="#choosing-the-number-of-latent-components" id="toc-choosing-the-number-of-latent-components" class="nav-link" data-scroll-target="#choosing-the-number-of-latent-components"><span class="toc-section-number">10.4</span>  Choosing the number of latent components</a></li>
  <li><a href="#choosing-the-number-of-features-to-retain" id="toc-choosing-the-number-of-features-to-retain" class="nav-link" data-scroll-target="#choosing-the-number-of-features-to-retain"><span class="toc-section-number">10.5</span>  Choosing the number of features to retain</a></li>
  <li><a href="#final-diablo-run" id="toc-final-diablo-run" class="nav-link" data-scroll-target="#final-diablo-run"><span class="toc-section-number">10.6</span>  Final DIABLO run</a></li>
  <li>
<a href="#results-interpretation" id="toc-results-interpretation" class="nav-link" data-scroll-target="#results-interpretation"><span class="toc-section-number">10.7</span>  Results interpretation</a>
  <ul class="collapse">
<li><a href="#correlation-between-datasets" id="toc-correlation-between-datasets" class="nav-link" data-scroll-target="#correlation-between-datasets"><span class="toc-section-number">10.7.1</span>  Correlation between datasets</a></li>
  <li><a href="#samples-projection-to-the-latent-component-space" id="toc-samples-projection-to-the-latent-component-space" class="nav-link" data-scroll-target="#samples-projection-to-the-latent-component-space"><span class="toc-section-number">10.7.2</span>  Samples projection to the latent component space</a></li>
  <li><a href="#correlation-circle-plots" id="toc-correlation-circle-plots" class="nav-link" data-scroll-target="#correlation-circle-plots"><span class="toc-section-number">10.7.3</span>  Correlation circle plots</a></li>
  <li><a href="#circos-plot" id="toc-circos-plot" class="nav-link" data-scroll-target="#circos-plot"><span class="toc-section-number">10.7.4</span>  Circos plot</a></li>
  </ul>
</li>
  <li><a href="#recap-targets-list" id="toc-recap-targets-list" class="nav-link" data-scroll-target="#recap-targets-list"><span class="toc-section-number">10.8</span>  Recap – targets list</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-diablo" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Integration with DIABLO</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/targets/">targets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PlantandFoodResearch/moiraine">moiraine</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For integration with DIABLO</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.mixOmics.org">mixOmics</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For working with lists</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For data-frames manipulation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that the omics datasets have been appropriately pre-processed and pre-filtered, we are ready to perform the actual data integration step. In this chapter, we will show how to perform multi-omics data integration with the <a href="https://mixomicsteam.github.io/mixOmics-Vignette/id_06.html">DIABLO</a> method from the <a href="http://mixomics.org/"><code>mixOmics</code></a> package.</p>
<p>As a reminder, here is what the <code>_targets.R</code> script should look like so far:</p>
<details><summary><code>_targets.R</code> script
</summary><div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(targets)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tarchetypes)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(moiraine)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">tar_option_set</span>(</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">packages =</span> <span class="fu">c</span>(</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"moiraine"</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MOFA2"</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"mixOmics"</span>,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"readr"</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"tibble"</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"tidyr"</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"dplyr"</span>,</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ggplot2"</span>,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">"patchwork"</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="fu">list</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Importing data ---------------------------------------------------------------</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Data import using a target factory</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_dataset_csv_factory</span>(</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">c</span>(</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/genomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/transcriptomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/metabolomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>)</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="fu">c</span>(<span class="st">"marker"</span>, <span class="st">"gene_id"</span>, <span class="st">"sample_id"</span>),</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">features_as_rowss =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>),</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="fu">c</span>(<span class="st">"geno"</span>, <span class="st">"transcripto"</span>, <span class="st">"metabo"</span>)</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Genomics features metadata file</span></span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>    fmetadata_file_geno,</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system.file</span>(<span class="st">"extdata/genomics_features_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"file"</span></span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Genomics features metadata import</span></span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>    fmetadata_geno,</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">import_fmetadata_csv</span>(</span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>      fmetadata_file_geno,</span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_id =</span> <span class="st">"marker"</span>,</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_types =</span> <span class="fu">c</span>(<span class="st">"chromosome"</span> <span class="ot">=</span> <span class="st">"c"</span>)</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Metabolomics features metadata import</span></span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_fmetadata_csv_factory</span>(</span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">c</span>(</span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/metabolomics_features_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>)</span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="fu">c</span>(<span class="st">"feature_id"</span>),</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="fu">c</span>(<span class="st">"metabo"</span>)</span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-61"><a href="#cb2-61" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Transcriptomics features metadata import</span></span>
<span id="cb2-62"><a href="#cb2-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_fmetadata_gff_factory</span>(</span>
<span id="cb2-63"><a href="#cb2-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">system.file</span>(<span class="st">"extdata/bos_taurus_gene_model.gff3"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-64"><a href="#cb2-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature_types =</span> <span class="st">"genes"</span>,</span>
<span id="cb2-65"><a href="#cb2-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">add_fieldss =</span> <span class="fu">c</span>(<span class="st">"Name"</span>, <span class="st">"description"</span>),</span>
<span id="cb2-66"><a href="#cb2-66" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="st">"transcripto"</span></span>
<span id="cb2-67"><a href="#cb2-67" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-68"><a href="#cb2-68" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-69"><a href="#cb2-69" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Samples metadata import</span></span>
<span id="cb2-70"><a href="#cb2-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_smetadata_csv_factory</span>(</span>
<span id="cb2-71"><a href="#cb2-71" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">system.file</span>(<span class="st">"extdata/samples_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-72"><a href="#cb2-72" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="st">"animal_id"</span>,</span>
<span id="cb2-73"><a href="#cb2-73" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="st">"all"</span></span>
<span id="cb2-74"><a href="#cb2-74" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating omics sets for each dataset</span></span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a>  <span class="fu">create_omics_set_factory</span>(</span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a>    <span class="at">datasets =</span> <span class="fu">c</span>(data_geno, data_transcripto, data_metabo),</span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a>    <span class="at">omics_types =</span> <span class="fu">c</span>(<span class="st">"genomics"</span>, <span class="st">"transcriptomics"</span>, <span class="st">"metabolomics"</span>),</span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a>    <span class="at">features_metadatas =</span> <span class="fu">c</span>(fmetadata_geno, fmetadata_transcripto, fmetadata_metabo),</span>
<span id="cb2-81"><a href="#cb2-81" aria-hidden="true" tabindex="-1"></a>    <span class="at">samples_metadatas =</span> <span class="fu">c</span>(smetadata_all, smetadata_all, smetadata_all)</span>
<span id="cb2-82"><a href="#cb2-82" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-83"><a href="#cb2-83" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-84"><a href="#cb2-84" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating the MultiDataSet object</span></span>
<span id="cb2-85"><a href="#cb2-85" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-86"><a href="#cb2-86" aria-hidden="true" tabindex="-1"></a>    mo_set,</span>
<span id="cb2-87"><a href="#cb2-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">create_multiomics_set</span>(</span>
<span id="cb2-88"><a href="#cb2-88" aria-hidden="true" tabindex="-1"></a>      <span class="fu">list</span>(set_geno,</span>
<span id="cb2-89"><a href="#cb2-89" aria-hidden="true" tabindex="-1"></a>           set_transcripto,</span>
<span id="cb2-90"><a href="#cb2-90" aria-hidden="true" tabindex="-1"></a>           set_metabo)</span>
<span id="cb2-91"><a href="#cb2-91" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-92"><a href="#cb2-92" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-93"><a href="#cb2-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-94"><a href="#cb2-94" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspecting the `MultiDataSet` object -----------------------------------------</span></span>
<span id="cb2-95"><a href="#cb2-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-96"><a href="#cb2-96" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating a density plot for each dataset</span></span>
<span id="cb2-97"><a href="#cb2-97" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-98"><a href="#cb2-98" aria-hidden="true" tabindex="-1"></a>    density_plots,</span>
<span id="cb2-99"><a href="#cb2-99" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_density_data</span>(</span>
<span id="cb2-100"><a href="#cb2-100" aria-hidden="true" tabindex="-1"></a>      mo_set,</span>
<span id="cb2-101"><a href="#cb2-101" aria-hidden="true" tabindex="-1"></a>      <span class="at">combined =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-102"><a href="#cb2-102" aria-hidden="true" tabindex="-1"></a>      <span class="at">scales =</span> <span class="st">"free"</span></span>
<span id="cb2-103"><a href="#cb2-103" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-104"><a href="#cb2-104" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-105"><a href="#cb2-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-106"><a href="#cb2-106" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the relationship between features mean and standard deviation</span></span>
<span id="cb2-107"><a href="#cb2-107" aria-hidden="true" tabindex="-1"></a>  <span class="do">## for each dataset</span></span>
<span id="cb2-108"><a href="#cb2-108" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-109"><a href="#cb2-109" aria-hidden="true" tabindex="-1"></a>    mean_sd_plots,</span>
<span id="cb2-110"><a href="#cb2-110" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_meansd_data</span>(mo_set)</span>
<span id="cb2-111"><a href="#cb2-111" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-112"><a href="#cb2-112" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-113"><a href="#cb2-113" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Assessing missing values</span></span>
<span id="cb2-114"><a href="#cb2-114" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-115"><a href="#cb2-115" aria-hidden="true" tabindex="-1"></a>    n_missing_values,</span>
<span id="cb2-116"><a href="#cb2-116" aria-hidden="true" tabindex="-1"></a>    <span class="fu">check_missing_values</span>(mo_set)</span>
<span id="cb2-117"><a href="#cb2-117" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-118"><a href="#cb2-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-119"><a href="#cb2-119" aria-hidden="true" tabindex="-1"></a><span class="co"># Modifying the `MultiDataSet` object ------------------------------------------</span></span>
<span id="cb2-120"><a href="#cb2-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-121"><a href="#cb2-121" aria-hidden="true" tabindex="-1"></a>  <span class="do">## RNAseq differential expression results file</span></span>
<span id="cb2-122"><a href="#cb2-122" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-123"><a href="#cb2-123" aria-hidden="true" tabindex="-1"></a>    rnaseq_de_res_file,</span>
<span id="cb2-124"><a href="#cb2-124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system.file</span>(</span>
<span id="cb2-125"><a href="#cb2-125" aria-hidden="true" tabindex="-1"></a>      <span class="st">"extdata/transcriptomics_de_results.csv"</span>,</span>
<span id="cb2-126"><a href="#cb2-126" aria-hidden="true" tabindex="-1"></a>      <span class="at">package =</span> <span class="st">"moiraine"</span></span>
<span id="cb2-127"><a href="#cb2-127" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-128"><a href="#cb2-128" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"file"</span></span>
<span id="cb2-129"><a href="#cb2-129" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-130"><a href="#cb2-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-131"><a href="#cb2-131" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Reading the RNAseq differential expression results</span></span>
<span id="cb2-132"><a href="#cb2-132" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-133"><a href="#cb2-133" aria-hidden="true" tabindex="-1"></a>    rnaseq_de_res_df,</span>
<span id="cb2-134"><a href="#cb2-134" aria-hidden="true" tabindex="-1"></a>    <span class="fu">read_csv</span>(rnaseq_de_res_file) <span class="sc">|&gt;</span></span>
<span id="cb2-135"><a href="#cb2-135" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rename</span>(<span class="at">feature_id =</span> gene_id) <span class="sc">|&gt;</span></span>
<span id="cb2-136"><a href="#cb2-136" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"rnaseq"</span>)</span>
<span id="cb2-137"><a href="#cb2-137" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-138"><a href="#cb2-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-139"><a href="#cb2-139" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Adding the differential expression results to the MultiDataSet object</span></span>
<span id="cb2-140"><a href="#cb2-140" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-141"><a href="#cb2-141" aria-hidden="true" tabindex="-1"></a>    mo_set_de,</span>
<span id="cb2-142"><a href="#cb2-142" aria-hidden="true" tabindex="-1"></a>    <span class="fu">add_features_metadata</span>(mo_set, rnaseq_de_res_df)</span>
<span id="cb2-143"><a href="#cb2-143" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-144"><a href="#cb2-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-145"><a href="#cb2-145" aria-hidden="true" tabindex="-1"></a><span class="co"># Data pre-processing ----------------------------------------------------------</span></span>
<span id="cb2-146"><a href="#cb2-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-147"><a href="#cb2-147" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Applying transformations to the datasets</span></span>
<span id="cb2-148"><a href="#cb2-148" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transformation_datasets_factory</span>(</span>
<span id="cb2-149"><a href="#cb2-149" aria-hidden="true" tabindex="-1"></a>    mo_set_de,</span>
<span id="cb2-150"><a href="#cb2-150" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"vst-deseq2"</span>,</span>
<span id="cb2-151"><a href="#cb2-151" aria-hidden="true" tabindex="-1"></a>      <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"logx"</span>),</span>
<span id="cb2-152"><a href="#cb2-152" aria-hidden="true" tabindex="-1"></a>    <span class="at">log_bases =</span> <span class="dv">2</span>,</span>
<span id="cb2-153"><a href="#cb2-153" aria-hidden="true" tabindex="-1"></a>    <span class="at">pre_log_functions =</span> zero_to_half_min,</span>
<span id="cb2-154"><a href="#cb2-154" aria-hidden="true" tabindex="-1"></a>    <span class="at">transformed_data_name =</span> <span class="st">"mo_set_transformed"</span></span>
<span id="cb2-155"><a href="#cb2-155" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-156"><a href="#cb2-156" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-157"><a href="#cb2-157" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Density plot for each transformed dataset</span></span>
<span id="cb2-158"><a href="#cb2-158" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-159"><a href="#cb2-159" aria-hidden="true" tabindex="-1"></a>    density_plots_transformed,</span>
<span id="cb2-160"><a href="#cb2-160" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_density_data</span>(</span>
<span id="cb2-161"><a href="#cb2-161" aria-hidden="true" tabindex="-1"></a>      mo_set_transformed,</span>
<span id="cb2-162"><a href="#cb2-162" aria-hidden="true" tabindex="-1"></a>      <span class="at">combined =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-163"><a href="#cb2-163" aria-hidden="true" tabindex="-1"></a>      <span class="at">scales =</span> <span class="st">"free"</span></span>
<span id="cb2-164"><a href="#cb2-164" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-165"><a href="#cb2-165" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-166"><a href="#cb2-166" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-167"><a href="#cb2-167" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the mean-SD trend for transformed each dataset</span></span>
<span id="cb2-168"><a href="#cb2-168" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-169"><a href="#cb2-169" aria-hidden="true" tabindex="-1"></a>    mean_sd_plots_transformed,</span>
<span id="cb2-170"><a href="#cb2-170" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_meansd_data</span>(mo_set_transformed)</span>
<span id="cb2-171"><a href="#cb2-171" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-172"><a href="#cb2-172" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-173"><a href="#cb2-173" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Summary table of the transformations applied</span></span>
<span id="cb2-174"><a href="#cb2-174" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-175"><a href="#cb2-175" aria-hidden="true" tabindex="-1"></a>    transformation_summary,</span>
<span id="cb2-176"><a href="#cb2-176" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_table_transformations</span>(transformations_runs_list)</span>
<span id="cb2-177"><a href="#cb2-177" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-178"><a href="#cb2-178" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-179"><a href="#cb2-179" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Running a PCA on each dataset</span></span>
<span id="cb2-180"><a href="#cb2-180" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pca_complete_data_factory</span>(</span>
<span id="cb2-181"><a href="#cb2-181" aria-hidden="true" tabindex="-1"></a>    mo_set_transformed,</span>
<span id="cb2-182"><a href="#cb2-182" aria-hidden="true" tabindex="-1"></a>    <span class="at">complete_data_name =</span> <span class="st">"mo_set_complete"</span></span>
<span id="cb2-183"><a href="#cb2-183" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-184"><a href="#cb2-184" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-185"><a href="#cb2-185" aria-hidden="true" tabindex="-1"></a>  <span class="do">## PCA screeplots</span></span>
<span id="cb2-186"><a href="#cb2-186" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-187"><a href="#cb2-187" aria-hidden="true" tabindex="-1"></a>    pca_screeplots,</span>
<span id="cb2-188"><a href="#cb2-188" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_screeplot_pca</span>(pca_runs_list)</span>
<span id="cb2-189"><a href="#cb2-189" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-190"><a href="#cb2-190" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-191"><a href="#cb2-191" aria-hidden="true" tabindex="-1"></a>  <span class="do">## PCA sample plots</span></span>
<span id="cb2-192"><a href="#cb2-192" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-193"><a href="#cb2-193" aria-hidden="true" tabindex="-1"></a>    pca_sample_plots,</span>
<span id="cb2-194"><a href="#cb2-194" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_samples_coordinates_pca</span>(</span>
<span id="cb2-195"><a href="#cb2-195" aria-hidden="true" tabindex="-1"></a>      pca_runs_list,</span>
<span id="cb2-196"><a href="#cb2-196" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="st">"snps"</span>,</span>
<span id="cb2-197"><a href="#cb2-197" aria-hidden="true" tabindex="-1"></a>      <span class="at">pcs =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb2-198"><a href="#cb2-198" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-199"><a href="#cb2-199" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span id="cb2-200"><a href="#cb2-200" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-201"><a href="#cb2-201" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-202"><a href="#cb2-202" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-203"><a href="#cb2-203" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-204"><a href="#cb2-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-205"><a href="#cb2-205" aria-hidden="true" tabindex="-1"></a><span class="co"># Dataset pre-filtering --------------------------------------------------------</span></span>
<span id="cb2-206"><a href="#cb2-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-207"><a href="#cb2-207" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Unsupervised feature selection based on MAD score</span></span>
<span id="cb2-208"><a href="#cb2-208" aria-hidden="true" tabindex="-1"></a>  <span class="fu">feature_preselection_mad_factory</span>(</span>
<span id="cb2-209"><a href="#cb2-209" aria-hidden="true" tabindex="-1"></a>    mo_set_complete,</span>
<span id="cb2-210"><a href="#cb2-210" aria-hidden="true" tabindex="-1"></a>    <span class="at">to_keep_ns =</span> <span class="fu">c</span>(<span class="st">"snps"</span> <span class="ot">=</span> <span class="dv">1000</span>, <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="dv">1000</span>),</span>
<span id="cb2-211"><a href="#cb2-211" aria-hidden="true" tabindex="-1"></a>    <span class="at">with_ties =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-212"><a href="#cb2-212" aria-hidden="true" tabindex="-1"></a>    <span class="at">filtered_set_target_name =</span> <span class="st">"mo_presel_unsupervised"</span></span>
<span id="cb2-213"><a href="#cb2-213" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-214"><a href="#cb2-214" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-215"><a href="#cb2-215" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Diagnostic plot for MAD-based feature selection</span></span>
<span id="cb2-216"><a href="#cb2-216" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-217"><a href="#cb2-217" aria-hidden="true" tabindex="-1"></a>    preselection_mad_plot,</span>
<span id="cb2-218"><a href="#cb2-218" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_feature_preselection_mad</span>(individual_mad_values)</span>
<span id="cb2-219"><a href="#cb2-219" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-220"><a href="#cb2-220" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-221"><a href="#cb2-221" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Supervised feature selection based on bruising groups</span></span>
<span id="cb2-222"><a href="#cb2-222" aria-hidden="true" tabindex="-1"></a>  <span class="fu">feature_preselection_splsda_factory</span>(</span>
<span id="cb2-223"><a href="#cb2-223" aria-hidden="true" tabindex="-1"></a>    mo_set_complete,</span>
<span id="cb2-224"><a href="#cb2-224" aria-hidden="true" tabindex="-1"></a>    <span class="at">group =</span> <span class="st">"status"</span>,</span>
<span id="cb2-225"><a href="#cb2-225" aria-hidden="true" tabindex="-1"></a>    <span class="at">to_keep_ns =</span> <span class="fu">c</span>(<span class="st">"snps"</span> <span class="ot">=</span> <span class="dv">1000</span>, <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="dv">1000</span>),</span>
<span id="cb2-226"><a href="#cb2-226" aria-hidden="true" tabindex="-1"></a>    <span class="at">filtered_set_target_name =</span> <span class="st">"mo_presel_supervised"</span></span>
<span id="cb2-227"><a href="#cb2-227" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-228"><a href="#cb2-228" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-229"><a href="#cb2-229" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Diagnostic plot for sPLS-DA based feature selection</span></span>
<span id="cb2-230"><a href="#cb2-230" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-231"><a href="#cb2-231" aria-hidden="true" tabindex="-1"></a>    preselection_splsda_plot,</span>
<span id="cb2-232"><a href="#cb2-232" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_feature_preselection_splsda</span>(individual_splsda_perf)</span>
<span id="cb2-233"><a href="#cb2-233" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-234"><a href="#cb2-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-235"><a href="#cb2-235" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with sPLS --------------------------------------------------------</span></span>
<span id="cb2-236"><a href="#cb2-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-237"><a href="#cb2-237" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating sPLS input</span></span>
<span id="cb2-238"><a href="#cb2-238" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-239"><a href="#cb2-239" aria-hidden="true" tabindex="-1"></a>    spls_input,</span>
<span id="cb2-240"><a href="#cb2-240" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_spls</span>(</span>
<span id="cb2-241"><a href="#cb2-241" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-242"><a href="#cb2-242" aria-hidden="true" tabindex="-1"></a>      <span class="at">mode =</span> <span class="st">"canonical"</span>,</span>
<span id="cb2-243"><a href="#cb2-243" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="fu">c</span>(<span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span>)</span>
<span id="cb2-244"><a href="#cb2-244" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-245"><a href="#cb2-245" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-246"><a href="#cb2-246" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-247"><a href="#cb2-247" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Initial PLS run with no feature selection and large number of components</span></span>
<span id="cb2-248"><a href="#cb2-248" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-249"><a href="#cb2-249" aria-hidden="true" tabindex="-1"></a>    spls_novarsel,</span>
<span id="cb2-250"><a href="#cb2-250" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_run</span>(</span>
<span id="cb2-251"><a href="#cb2-251" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-252"><a href="#cb2-252" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> <span class="dv">4</span></span>
<span id="cb2-253"><a href="#cb2-253" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-254"><a href="#cb2-254" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-255"><a href="#cb2-255" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-256"><a href="#cb2-256" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of components</span></span>
<span id="cb2-257"><a href="#cb2-257" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-258"><a href="#cb2-258" aria-hidden="true" tabindex="-1"></a>    spls_perf_res,</span>
<span id="cb2-259"><a href="#cb2-259" aria-hidden="true" tabindex="-1"></a>    mixOmics<span class="sc">::</span><span class="fu">perf</span>(</span>
<span id="cb2-260"><a href="#cb2-260" aria-hidden="true" tabindex="-1"></a>      spls_novarsel,</span>
<span id="cb2-261"><a href="#cb2-261" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-262"><a href="#cb2-262" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-263"><a href="#cb2-263" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">10</span>,</span>
<span id="cb2-264"><a href="#cb2-264" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-265"><a href="#cb2-265" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-266"><a href="#cb2-266" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-267"><a href="#cb2-267" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-268"><a href="#cb2-268" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of components)</span></span>
<span id="cb2-269"><a href="#cb2-269" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Can try criterion = 'Q2.total', 'cor.tpred', 'cor.upred', 'RSS.tpred',</span></span>
<span id="cb2-270"><a href="#cb2-270" aria-hidden="true" tabindex="-1"></a>  <span class="do">## 'RSS.upred' (but avoid 'RSS' and 'PRESS')</span></span>
<span id="cb2-271"><a href="#cb2-271" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-272"><a href="#cb2-272" aria-hidden="true" tabindex="-1"></a>    spls_perf_plot,</span>
<span id="cb2-273"><a href="#cb2-273" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(spls_perf_res, <span class="at">criterion =</span> <span class="st">"Q2.total"</span>)</span>
<span id="cb2-274"><a href="#cb2-274" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-275"><a href="#cb2-275" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-276"><a href="#cb2-276" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Selected value for ncomp</span></span>
<span id="cb2-277"><a href="#cb2-277" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-278"><a href="#cb2-278" aria-hidden="true" tabindex="-1"></a>    spls_optim_ncomp,</span>
<span id="cb2-279"><a href="#cb2-279" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_get_optim_ncomp</span>(spls_perf_res, <span class="at">min_ncomp =</span> <span class="dv">2</span>)</span>
<span id="cb2-280"><a href="#cb2-280" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-281"><a href="#cb2-281" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-282"><a href="#cb2-282" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of features to retain</span></span>
<span id="cb2-283"><a href="#cb2-283" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-284"><a href="#cb2-284" aria-hidden="true" tabindex="-1"></a>    spls_tune_res,</span>
<span id="cb2-285"><a href="#cb2-285" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_tune</span>(</span>
<span id="cb2-286"><a href="#cb2-286" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-287"><a href="#cb2-287" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> spls_optim_ncomp,</span>
<span id="cb2-288"><a href="#cb2-288" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepX =</span> <span class="fu">seq</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">10</span>),</span>
<span id="cb2-289"><a href="#cb2-289" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepY =</span> <span class="fu">seq</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">10</span>),</span>
<span id="cb2-290"><a href="#cb2-290" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-291"><a href="#cb2-291" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-292"><a href="#cb2-292" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">5</span>,</span>
<span id="cb2-293"><a href="#cb2-293" aria-hidden="true" tabindex="-1"></a>      <span class="at">measure =</span> <span class="st">"cor"</span>,</span>
<span id="cb2-294"><a href="#cb2-294" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-295"><a href="#cb2-295" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-296"><a href="#cb2-296" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-297"><a href="#cb2-297" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-298"><a href="#cb2-298" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of features)</span></span>
<span id="cb2-299"><a href="#cb2-299" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-300"><a href="#cb2-300" aria-hidden="true" tabindex="-1"></a>    spls_tune_plot,</span>
<span id="cb2-301"><a href="#cb2-301" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_plot_tune</span>(spls_tune_res)</span>
<span id="cb2-302"><a href="#cb2-302" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-303"><a href="#cb2-303" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-304"><a href="#cb2-304" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Final sPLS run</span></span>
<span id="cb2-305"><a href="#cb2-305" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-306"><a href="#cb2-306" aria-hidden="true" tabindex="-1"></a>    spls_final_run,</span>
<span id="cb2-307"><a href="#cb2-307" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_run</span>(</span>
<span id="cb2-308"><a href="#cb2-308" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-309"><a href="#cb2-309" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> spls_optim_ncomp,</span>
<span id="cb2-310"><a href="#cb2-310" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepX =</span> spls_tune_res<span class="sc">$</span>choice.keepX,</span>
<span id="cb2-311"><a href="#cb2-311" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepY =</span> spls_tune_res<span class="sc">$</span>choice.keepY</span>
<span id="cb2-312"><a href="#cb2-312" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-313"><a href="#cb2-313" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-314"><a href="#cb2-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-315"><a href="#cb2-315" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with sO2PLS ------------------------------------------------------</span></span>
<span id="cb2-316"><a href="#cb2-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-317"><a href="#cb2-317" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating sO2PLS input</span></span>
<span id="cb2-318"><a href="#cb2-318" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-319"><a href="#cb2-319" aria-hidden="true" tabindex="-1"></a>    omicspls_input,</span>
<span id="cb2-320"><a href="#cb2-320" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_omicspls</span>(</span>
<span id="cb2-321"><a href="#cb2-321" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-322"><a href="#cb2-322" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="fu">c</span>(<span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span>)</span>
<span id="cb2-323"><a href="#cb2-323" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-324"><a href="#cb2-324" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-325"><a href="#cb2-325" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-326"><a href="#cb2-326" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Adjusted cross-validation for number of components</span></span>
<span id="cb2-327"><a href="#cb2-327" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-328"><a href="#cb2-328" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj,</span>
<span id="cb2-329"><a href="#cb2-329" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_o2m_adjR2</span>(</span>
<span id="cb2-330"><a href="#cb2-330" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-331"><a href="#cb2-331" aria-hidden="true" tabindex="-1"></a>      <span class="at">a =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,</span>
<span id="cb2-332"><a href="#cb2-332" aria-hidden="true" tabindex="-1"></a>      <span class="at">ax =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="dv">2</span>),</span>
<span id="cb2-333"><a href="#cb2-333" aria-hidden="true" tabindex="-1"></a>      <span class="at">ay =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="dv">2</span>),</span>
<span id="cb2-334"><a href="#cb2-334" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-335"><a href="#cb2-335" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_cores =</span> <span class="dv">6</span>,</span>
<span id="cb2-336"><a href="#cb2-336" aria-hidden="true" tabindex="-1"></a>      <span class="at">seed =</span> <span class="dv">127</span></span>
<span id="cb2-337"><a href="#cb2-337" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-338"><a href="#cb2-338" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-339"><a href="#cb2-339" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-340"><a href="#cb2-340" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj_res,</span>
<span id="cb2-341"><a href="#cb2-341" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_ncomp_adj</span>(so2pls_cv_adj)</span>
<span id="cb2-342"><a href="#cb2-342" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-343"><a href="#cb2-343" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-344"><a href="#cb2-344" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting adjusted cross-validation results</span></span>
<span id="cb2-345"><a href="#cb2-345" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-346"><a href="#cb2-346" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj_plot,</span>
<span id="cb2-347"><a href="#cb2-347" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv_adj</span>(so2pls_cv_adj)</span>
<span id="cb2-348"><a href="#cb2-348" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-349"><a href="#cb2-349" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-350"><a href="#cb2-350" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Standard cross-validation for number of components</span></span>
<span id="cb2-351"><a href="#cb2-351" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-352"><a href="#cb2-352" aria-hidden="true" tabindex="-1"></a>    so2pls_cv,</span>
<span id="cb2-353"><a href="#cb2-353" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_o2m</span>(</span>
<span id="cb2-354"><a href="#cb2-354" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-355"><a href="#cb2-355" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_adj,</span>
<span id="cb2-356"><a href="#cb2-356" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-357"><a href="#cb2-357" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_cores =</span> <span class="dv">6</span>,</span>
<span id="cb2-358"><a href="#cb2-358" aria-hidden="true" tabindex="-1"></a>      <span class="at">seed =</span> <span class="dv">356</span></span>
<span id="cb2-359"><a href="#cb2-359" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-360"><a href="#cb2-360" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-361"><a href="#cb2-361" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-362"><a href="#cb2-362" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_res,</span>
<span id="cb2-363"><a href="#cb2-363" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_ncomp</span>(so2pls_cv)</span>
<span id="cb2-364"><a href="#cb2-364" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-365"><a href="#cb2-365" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-366"><a href="#cb2-366" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting standard cross-validation results</span></span>
<span id="cb2-367"><a href="#cb2-367" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-368"><a href="#cb2-368" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_plot,</span>
<span id="cb2-369"><a href="#cb2-369" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv</span>(so2pls_cv)</span>
<span id="cb2-370"><a href="#cb2-370" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-371"><a href="#cb2-371" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-372"><a href="#cb2-372" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for sparsity parameters</span></span>
<span id="cb2-373"><a href="#cb2-373" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-374"><a href="#cb2-374" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity,</span>
<span id="cb2-375"><a href="#cb2-375" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_sparsity</span>(</span>
<span id="cb2-376"><a href="#cb2-376" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-377"><a href="#cb2-377" aria-hidden="true" tabindex="-1"></a>      <span class="at">n =</span> so2pls_cv_res[<span class="st">"n"</span>],</span>
<span id="cb2-378"><a href="#cb2-378" aria-hidden="true" tabindex="-1"></a>      <span class="at">nx =</span> so2pls_cv_res[<span class="st">"nx"</span>],</span>
<span id="cb2-379"><a href="#cb2-379" aria-hidden="true" tabindex="-1"></a>      <span class="at">ny =</span> so2pls_cv_res[<span class="st">"ny"</span>],</span>
<span id="cb2-380"><a href="#cb2-380" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-381"><a href="#cb2-381" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepx_seq =</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">5</span>, <span class="dv">30</span>, <span class="dv">5</span>), <span class="fu">seq</span>(<span class="dv">40</span>, <span class="dv">100</span>, <span class="dv">10</span>)),</span>
<span id="cb2-382"><a href="#cb2-382" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepy_seq =</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">5</span>, <span class="dv">40</span>, <span class="dv">5</span>))</span>
<span id="cb2-383"><a href="#cb2-383" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-384"><a href="#cb2-384" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-385"><a href="#cb2-385" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-386"><a href="#cb2-386" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_res,</span>
<span id="cb2-387"><a href="#cb2-387" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_keep</span>(so2pls_cv_sparsity)</span>
<span id="cb2-388"><a href="#cb2-388" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-389"><a href="#cb2-389" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-390"><a href="#cb2-390" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the results of the cross-validation for the number of features</span></span>
<span id="cb2-391"><a href="#cb2-391" aria-hidden="true" tabindex="-1"></a>  <span class="do">## to retain from each dataset for the different joint components</span></span>
<span id="cb2-392"><a href="#cb2-392" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-393"><a href="#cb2-393" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_plot,</span>
<span id="cb2-394"><a href="#cb2-394" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv_sparsity</span>(so2pls_cv_sparsity)</span>
<span id="cb2-395"><a href="#cb2-395" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-396"><a href="#cb2-396" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-397"><a href="#cb2-397" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Extracting sparsity results in table format</span></span>
<span id="cb2-398"><a href="#cb2-398" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-399"><a href="#cb2-399" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_table,</span>
<span id="cb2-400"><a href="#cb2-400" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_print_cv_sparsity</span>(so2pls_cv_sparsity_res)</span>
<span id="cb2-401"><a href="#cb2-401" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-402"><a href="#cb2-402" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-403"><a href="#cb2-403" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Final sO2PLS run</span></span>
<span id="cb2-404"><a href="#cb2-404" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-405"><a href="#cb2-405" aria-hidden="true" tabindex="-1"></a>    so2pls_final_run,</span>
<span id="cb2-406"><a href="#cb2-406" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_o2m</span>(</span>
<span id="cb2-407"><a href="#cb2-407" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-408"><a href="#cb2-408" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_res,</span>
<span id="cb2-409"><a href="#cb2-409" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_sparsity_res</span>
<span id="cb2-410"><a href="#cb2-410" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-411"><a href="#cb2-411" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-412"><a href="#cb2-412" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-413"><a href="#cb2-413" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Summary plot of percentage of variance explained</span></span>
<span id="cb2-414"><a href="#cb2-414" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-415"><a href="#cb2-415" aria-hidden="true" tabindex="-1"></a>    so2pls_summary_plot,</span>
<span id="cb2-416"><a href="#cb2-416" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_summary</span>(so2pls_final_run)</span>
<span id="cb2-417"><a href="#cb2-417" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-418"><a href="#cb2-418" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-419"><a href="#cb2-419" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Screeplot</span></span>
<span id="cb2-420"><a href="#cb2-420" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-421"><a href="#cb2-421" aria-hidden="true" tabindex="-1"></a>    so2pls_screeplot,</span>
<span id="cb2-422"><a href="#cb2-422" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_screeplot</span>(so2pls_final_run)</span>
<span id="cb2-423"><a href="#cb2-423" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-424"><a href="#cb2-424" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-425"><a href="#cb2-425" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Comparison of samples score for joint components</span></span>
<span id="cb2-426"><a href="#cb2-426" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-427"><a href="#cb2-427" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_comparison_plot,</span>
<span id="cb2-428"><a href="#cb2-428" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_compare_samples_joint_components</span>(</span>
<span id="cb2-429"><a href="#cb2-429" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-430"><a href="#cb2-430" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-431"><a href="#cb2-431" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_by =</span> <span class="st">"status"</span>,</span>
<span id="cb2-432"><a href="#cb2-432" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_by =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-433"><a href="#cb2-433" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-434"><a href="#cb2-434" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-435"><a href="#cb2-435" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-436"><a href="#cb2-436" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Coefficient plot for joint components</span></span>
<span id="cb2-437"><a href="#cb2-437" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-438"><a href="#cb2-438" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_coefficients_plot,</span>
<span id="cb2-439"><a href="#cb2-439" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_joint_components_coefficients</span>(so2pls_final_run)</span>
<span id="cb2-440"><a href="#cb2-440" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-441"><a href="#cb2-441" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-442"><a href="#cb2-442" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Joint component samples score plot</span></span>
<span id="cb2-443"><a href="#cb2-443" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-444"><a href="#cb2-444" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_samples_score_plot,</span>
<span id="cb2-445"><a href="#cb2-445" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_samples_joint_components</span>(</span>
<span id="cb2-446"><a href="#cb2-446" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-447"><a href="#cb2-447" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-448"><a href="#cb2-448" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-449"><a href="#cb2-449" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>),</span>
<span id="cb2-450"><a href="#cb2-450" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-451"><a href="#cb2-451" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-452"><a href="#cb2-452" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>)</span>
<span id="cb2-453"><a href="#cb2-453" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-454"><a href="#cb2-454" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-455"><a href="#cb2-455" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Specific components samples score plot</span></span>
<span id="cb2-456"><a href="#cb2-456" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-457"><a href="#cb2-457" aria-hidden="true" tabindex="-1"></a>    so2pls_specific_components_samples_score_plot,</span>
<span id="cb2-458"><a href="#cb2-458" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_samples_specific_components</span>(</span>
<span id="cb2-459"><a href="#cb2-459" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-460"><a href="#cb2-460" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-461"><a href="#cb2-461" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"feedlot"</span>,</span>
<span id="cb2-462"><a href="#cb2-462" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>),</span>
<span id="cb2-463"><a href="#cb2-463" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"rnaseq_batch"</span>,</span>
<span id="cb2-464"><a href="#cb2-464" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"gender"</span></span>
<span id="cb2-465"><a href="#cb2-465" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span> </span>
<span id="cb2-466"><a href="#cb2-466" aria-hidden="true" tabindex="-1"></a>      <span class="fu">map</span>(\(x) x <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>))</span>
<span id="cb2-467"><a href="#cb2-467" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-468"><a href="#cb2-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-469"><a href="#cb2-469" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with MOFA --------------------------------------------------------</span></span>
<span id="cb2-470"><a href="#cb2-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-471"><a href="#cb2-471" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating MOFA input</span></span>
<span id="cb2-472"><a href="#cb2-472" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-473"><a href="#cb2-473" aria-hidden="true" tabindex="-1"></a>    mofa_input,</span>
<span id="cb2-474"><a href="#cb2-474" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_mofa</span>(</span>
<span id="cb2-475"><a href="#cb2-475" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-476"><a href="#cb2-476" aria-hidden="true" tabindex="-1"></a>      <span class="at">options_list =</span> <span class="fu">list</span>(</span>
<span id="cb2-477"><a href="#cb2-477" aria-hidden="true" tabindex="-1"></a>        <span class="at">data_options =</span> <span class="fu">list</span>(<span class="at">scale_views =</span> <span class="cn">TRUE</span>),</span>
<span id="cb2-478"><a href="#cb2-478" aria-hidden="true" tabindex="-1"></a>        <span class="at">model_options =</span> <span class="fu">list</span>(<span class="at">likelihoods =</span> <span class="fu">c</span>(</span>
<span id="cb2-479"><a href="#cb2-479" aria-hidden="true" tabindex="-1"></a>          <span class="st">"snps"</span> <span class="ot">=</span> <span class="st">"poisson"</span>,</span>
<span id="cb2-480"><a href="#cb2-480" aria-hidden="true" tabindex="-1"></a>          <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"gaussian"</span>,</span>
<span id="cb2-481"><a href="#cb2-481" aria-hidden="true" tabindex="-1"></a>          <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"gaussian"</span>)</span>
<span id="cb2-482"><a href="#cb2-482" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb2-483"><a href="#cb2-483" aria-hidden="true" tabindex="-1"></a>        <span class="at">training_options =</span> <span class="fu">list</span>(<span class="at">seed =</span> <span class="dv">43</span>)</span>
<span id="cb2-484"><a href="#cb2-484" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb2-485"><a href="#cb2-485" aria-hidden="true" tabindex="-1"></a>      <span class="at">only_common_samples =</span> <span class="cn">FALSE</span></span>
<span id="cb2-486"><a href="#cb2-486" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-487"><a href="#cb2-487" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-488"><a href="#cb2-488" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-489"><a href="#cb2-489" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Overview plot of the samples in each dataset</span></span>
<span id="cb2-490"><a href="#cb2-490" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-491"><a href="#cb2-491" aria-hidden="true" tabindex="-1"></a>    mofa_input_plot,</span>
<span id="cb2-492"><a href="#cb2-492" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_data_overview</span>(mofa_input)</span>
<span id="cb2-493"><a href="#cb2-493" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-494"><a href="#cb2-494" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-495"><a href="#cb2-495" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Training MOFA model</span></span>
<span id="cb2-496"><a href="#cb2-496" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-497"><a href="#cb2-497" aria-hidden="true" tabindex="-1"></a>    mofa_trained,</span>
<span id="cb2-498"><a href="#cb2-498" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run_mofa</span>(</span>
<span id="cb2-499"><a href="#cb2-499" aria-hidden="true" tabindex="-1"></a>      mofa_input,</span>
<span id="cb2-500"><a href="#cb2-500" aria-hidden="true" tabindex="-1"></a>      <span class="at">save_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-501"><a href="#cb2-501" aria-hidden="true" tabindex="-1"></a>      <span class="at">use_basilisk =</span> <span class="cn">TRUE</span></span>
<span id="cb2-502"><a href="#cb2-502" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-503"><a href="#cb2-503" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-504"><a href="#cb2-504" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-505"><a href="#cb2-505" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Formatting MOFA output</span></span>
<span id="cb2-506"><a href="#cb2-506" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-507"><a href="#cb2-507" aria-hidden="true" tabindex="-1"></a>    mofa_output,</span>
<span id="cb2-508"><a href="#cb2-508" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_output</span>(mofa_trained)</span>
<span id="cb2-509"><a href="#cb2-509" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-510"><a href="#cb2-510" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-511"><a href="#cb2-511" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plots of variance explained</span></span>
<span id="cb2-512"><a href="#cb2-512" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-513"><a href="#cb2-513" aria-hidden="true" tabindex="-1"></a>    mofa_var_explained_plot,</span>
<span id="cb2-514"><a href="#cb2-514" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(</span>
<span id="cb2-515"><a href="#cb2-515" aria-hidden="true" tabindex="-1"></a>      mofa_trained,</span>
<span id="cb2-516"><a href="#cb2-516" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"view"</span>,  <span class="do">## datasets on the x-axis</span></span>
<span id="cb2-517"><a href="#cb2-517" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"factor"</span> <span class="do">## factors on the y-axis</span></span>
<span id="cb2-518"><a href="#cb2-518" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-519"><a href="#cb2-519" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-520"><a href="#cb2-520" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-521"><a href="#cb2-521" aria-hidden="true" tabindex="-1"></a>    mofa_total_var_explained_plot,</span>
<span id="cb2-522"><a href="#cb2-522" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(</span>
<span id="cb2-523"><a href="#cb2-523" aria-hidden="true" tabindex="-1"></a>      mofa_trained,</span>
<span id="cb2-524"><a href="#cb2-524" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"view"</span>,</span>
<span id="cb2-525"><a href="#cb2-525" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"factor"</span>,</span>
<span id="cb2-526"><a href="#cb2-526" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot_total =</span> <span class="cn">TRUE</span></span>
<span id="cb2-527"><a href="#cb2-527" aria-hidden="true" tabindex="-1"></a>    )[[<span class="dv">2</span>]]</span>
<span id="cb2-528"><a href="#cb2-528" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-529"><a href="#cb2-529" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-530"><a href="#cb2-530" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plot of factors correlation with covariates</span></span>
<span id="cb2-531"><a href="#cb2-531" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-532"><a href="#cb2-532" aria-hidden="true" tabindex="-1"></a>    mofa_factors_covariates_cor_plot,</span>
<span id="cb2-533"><a href="#cb2-533" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mofa_plot_cor_covariates</span>(mofa_trained)</span>
<span id="cb2-534"><a href="#cb2-534" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-535"><a href="#cb2-535" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details><section id="what-is-diablo" class="level2" data-number="10.1"><h2 data-number="10.1" class="anchored" data-anchor-id="what-is-diablo">
<span class="header-section-number">10.1</span> What is DIABLO?</h2>
<p>DIABLO (for Data Integration Analysis for Biomarker discovery using Latent Components) is a multivariate approach to perform supervised data integration. Given two or more omics datasets for which measurements are taken on the same samples, DIABLO aims at selecting correlated features across the datasets that best discriminate between the different sample groups for a categorical outcome of interest.</p>
<p>DIABLO works by iteratively constructing linear combinations of the features, called latent components, which maximise the correlation between the datasets and with the categorical outcome. In order to perform feature selection, the latent components are subjected to <span class="math inline">\(L1\)</span>-regularisation (or LASSO), i.e.&nbsp;the number of features included in the linear combination is constrained by the user. Moreover, the optimisation problem is weighted to allow the user to control the balance between maximising the correlation between omics datasets and discriminating between the outcome groups of interest.</p>
<p>DIABLO requires as input the matrices of omics measurements, all with the same samples, as well as a factor variable indicating the outcome group for each sample. While the omics datasets are automatically centred and scaled by DIABLO, proper preprocessing and normalisation is assumed to be carried out by the user. Although the DIABLO algorithm can handle the presence of missing features, it will prohibit the use of cross-validation. It is thus recommended to perform data imputation prior to running a DIABLO analysis. Importantly, DIABLO tends to perform better when the number of features in the datasets is not too large. Therefore, it is highly recommended to perform some prefiltering prior to using DIABLO.</p>
</section><section id="creating-the-diablo-input" class="level2" data-number="10.2"><h2 data-number="10.2" class="anchored" data-anchor-id="creating-the-diablo-input">
<span class="header-section-number">10.2</span> Creating the DIABLO input</h2>
<p>The first step is to transform the <code>MultiDataSet</code> object into a suitable format for the <code>mixOmics</code> package. This is done with the <code><a href="https://rdrr.io/pkg/moiraine/man/get_input_mixomics_supervised.html">get_input_mixomics_supervised()</a></code> function, which takes as input:</p>
<ul>
<li><p>a <code>MultiDataSet</code> object,</p></li>
<li><p>the name of the column in the samples metadata table that corresponds to the categorical outcome of interest,</p></li>
<li><p>optionally, the names of the datasets to include in the analysis. This is useful if we want to exclude one of more datasets from the analysis.</p></li>
</ul>
<p>In our case, we want to find differences between the healthy and diseased animals. We will use the multi-omics datasets that have gone through supervised prefiltering (i.e.&nbsp;we discarded the features least related with the disease status, as seen in <a href="prefiltering.html"><span>Chapter&nbsp;6</span></a>).</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_input</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_input_mixomics_supervised.html">get_input_mixomics_supervised</a></span><span class="op">(</span></span>
<span>    <span class="va">mo_presel_supervised</span>,</span>
<span>    group <span class="op">=</span> <span class="st">"status"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Importantly, the <code><a href="https://rdrr.io/pkg/moiraine/man/get_input_mixomics_supervised.html">get_input_mixomics_supervised()</a></code> function only retains samples that are present in all omics datasets to be analysed. It also makes sure that the column provided as categorical outcome does not contain numerical values, as DIABLO can only handle categorical outcome. If the column contains integers, they will be considered as levels of a factor.</p>
<p>The result of the function is a named list with one element per dataset to integrate, plus a <code>Y</code> element that contains the categorical outcome. The omics datasets are stored as matrices with samples as rows and features as columns; the categorical outcome is a named factor vector.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_input</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">diablo_input</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ snps      : num [1:135, 1:1000] 2 1 2 2 2 2 1 2 2 2 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:135] "R21" "Y3660" "Y3243" "R5764" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:1000] "21-25977541-C-T-rs41974686" "22-51403583-A-C-rs210306176" "24-12959068-G-T-rs381471286" "8-85224224-T-C-rs43565287" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "datatype")= chr "SnpSet"</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "package")= chr "Biobase"</span></span>
<span><span class="co">#&gt;  $ rnaseq    : num [1:135, 1:994] 3.87 4.97 4.08 3.49 3.49 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:135] "R21" "Y3660" "Y3243" "R5764" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:994] "ENSBTAG00000000020" "ENSBTAG00000000046" "ENSBTAG00000000056" "ENSBTAG00000000061" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "datatype")= chr "ExpressionSet"</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "package")= chr "Biobase"</span></span>
<span><span class="co">#&gt;  $ metabolome: num [1:135, 1:55] 3.19 3.28 3.22 3.72 3.63 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:135] "R21" "Y3660" "Y3243" "R5764" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:55] "HMDB00001" "HMDB00008" "HMDB00042" "HMDB00043" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "datatype")= chr "MetabolomeSet"</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "package")= chr "moiraine"</span></span>
<span><span class="co">#&gt;  $ Y         : Factor w/ 2 levels "BRD","Control": 2 2 2 2 1 2 2 2 2 2 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:135] "R21" "Y3660" "Y3243" "R5764" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="constructing-the-design-matrix" class="level2" data-number="10.3"><h2 data-number="10.3" class="anchored" data-anchor-id="constructing-the-design-matrix">
<span class="header-section-number">10.3</span> Constructing the design matrix</h2>
<p>DIABLO relies on a design matrix to balance its two optimisation objectives: maximising the covariance between the omics datasets, and maximising the discrimination between the outcome categories. The design matrix is a matrix with one row and one column per dataset, plus a row and a column for the <code>"Y"</code> dataset, i.e.&nbsp;the categorical outcome. The values within each cell of the matrix indicate the ratio between the two objectives for this combination of dataset. A value of 0 means that we want to prioritise outcome discrimination, while a value of 1 indicates that we want to prioritise maximising the covariance between the two corresponding datasets. All values must be between 0 and 1.</p>
<p>There are two options for constructing the design matrix, which we present below.</p>
<section id="predefined-design-matrices" class="level3" data-number="10.3.1"><h3 data-number="10.3.1" class="anchored" data-anchor-id="predefined-design-matrices">
<span class="header-section-number">10.3.1</span> Predefined design matrices</h3>
<p>A first option is to choose a strategy based on what we are trying to obtain from the integration:</p>
<ul>
<li>if we want to strike a balance between the two objectives (recommended option), we’ll constructed a “weighted full” design matrix that looks like this:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_predefined_design_matrix.html">diablo_predefined_design_matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">diablo_input</span><span class="op">)</span>, <span class="st">"weighted_full"</span><span class="op">)</span></span>
<span><span class="co">#&gt;            snps rnaseq metabolome Y</span></span>
<span><span class="co">#&gt; snps        0.0    0.1        0.1 1</span></span>
<span><span class="co">#&gt; rnaseq      0.1    0.0        0.1 1</span></span>
<span><span class="co">#&gt; metabolome  0.1    0.1        0.0 1</span></span>
<span><span class="co">#&gt; Y           1.0    1.0        1.0 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>if we want to maximise the discrimination between the outcome categories, we’ll construct a “null” design matrix that looks like this:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_predefined_design_matrix.html">diablo_predefined_design_matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">diablo_input</span><span class="op">)</span>, <span class="st">"null"</span><span class="op">)</span></span>
<span><span class="co">#&gt;            snps rnaseq metabolome Y</span></span>
<span><span class="co">#&gt; snps          0      0          0 1</span></span>
<span><span class="co">#&gt; rnaseq        0      0          0 1</span></span>
<span><span class="co">#&gt; metabolome    0      0          0 1</span></span>
<span><span class="co">#&gt; Y             1      1          1 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>if we want only to maximise the covariance between the datasets, we’ll construct a “full” design matrix that looks like this:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_predefined_design_matrix.html">diablo_predefined_design_matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">diablo_input</span><span class="op">)</span>, <span class="st">"full"</span><span class="op">)</span></span>
<span><span class="co">#&gt;            snps rnaseq metabolome Y</span></span>
<span><span class="co">#&gt; snps          0      1          1 1</span></span>
<span><span class="co">#&gt; rnaseq        1      0          1 1</span></span>
<span><span class="co">#&gt; metabolome    1      1          0 1</span></span>
<span><span class="co">#&gt; Y             1      1          1 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will show how to use these pre-defined design matrices when running DIABLO.</p>
</section><section id="estimating-the-design-matrix-through-pairwise-pls" class="level3" data-number="10.3.2"><h3 data-number="10.3.2" class="anchored" data-anchor-id="estimating-the-design-matrix-through-pairwise-pls">
<span class="header-section-number">10.3.2</span> Estimating the design matrix through pairwise PLS</h3>
<p>Alternatively, we can let the data guide the construction of the design matrix. This is achieved by assessing the correlation between each pair of datasets, through a <a href="https://mixomicsteam.github.io/Bookdown/pls.html">PLS</a> (Projection to Latent Structures) run. More specifically, the correlation between the datasets is computed as the correlation coefficient between the first component constructed for each dataset during the PLS run. Then, based on the correlation obtained between a pair of dataset, we can decide on a value to use for the design matrix. Typically, the following thresholds are recommended by the authors of the mixOmics package:</p>
<ul>
<li><p>correlation coefficient of 0.8 or above between two datasets: assign a value of 1 in the corresponding cell of the design matrix;</p></li>
<li><p>correlation coefficient below 0.8: assign a value of 0.1 in the corresponding cell of the design matrix.</p></li>
</ul>
<p>The <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_pairwise_pls_factory.html">diablo_pairwise_pls_factory()</a></code> function automates this process. It takes as input the DIABLO input object that we constructed previously:</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_pairwise_pls_factory.html">diablo_pairwise_pls_factory</a></span><span class="op">(</span><span class="va">diablo_input</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>The function works as follows:</p>
<ul>
<li>It creates a list of all possible pairs of datasets, which is stored in the <code>diablo_pairs_datasets</code> target:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">diablo_pairs_datasets</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "snps"   "rnaseq"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] "snps"       "metabolome"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] "rnaseq"     "metabolome"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>It uses dynamic branching to perform a PLS run on each pair of datasets, via the <code><a href="https://rdrr.io/pkg/moiraine/man/run_pairwise_pls.html">run_pairwise_pls()</a></code> function. The results are stored as a list in the <code>diablo_pls_runs_list</code> target. Each element of the list has a <code>datasets_name</code> attribute to indicate which datasets were analysed:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">diablo_pls_runs_list</span><span class="op">)</span>, <span class="va">attr</span>, <span class="st">"datasets_name"</span><span class="op">)</span></span>
<span><span class="co">#&gt; $diablo_pls_runs_list_05fb1980</span></span>
<span><span class="co">#&gt; [1] "snps"   "rnaseq"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $diablo_pls_runs_list_1b4ec539</span></span>
<span><span class="co">#&gt; [1] "snps"       "metabolome"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $diablo_pls_runs_list_4f48d450</span></span>
<span><span class="co">#&gt; [1] "rnaseq"     "metabolome"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>It constructs the estimated correlation matrix between the datasets, based on the results of the PLS runs, via the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_pairwise_pls_corr.html">diablo_get_pairwise_pls_corr()</a></code> function. The resulting matrix is available through the <code>diablo_pls_correlation_matrix</code> target:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">diablo_pls_correlation_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 snps    rnaseq metabolome</span></span>
<span><span class="co">#&gt; snps       1.0000000 0.6474490  0.6263349</span></span>
<span><span class="co">#&gt; rnaseq     0.6474490 1.0000000  0.8607195</span></span>
<span><span class="co">#&gt; metabolome 0.6263349 0.8607195  1.0000000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>It constructs the design matrix according to the datasets correlation matrix, through the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_generate_design_matrix.html">diablo_generate_design_matrix()</a></code> function. This function has parameters to customise how the correlation matrix should be translated into a design matrix, notably by setting the threshold to use on the correlation coefficients (default is 0.8, as recommended). These arguments can be customised in the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_pairwise_pls_factory.html">diablo_pairwise_pls_factory()</a></code> function. The resulting design function is stored in the target <code>diablo_design_matrix</code>:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">diablo_design_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt;            snps rnaseq metabolome Y</span></span>
<span><span class="co">#&gt; snps        0.0    0.1        0.1 1</span></span>
<span><span class="co">#&gt; rnaseq      0.1    0.0        1.0 1</span></span>
<span><span class="co">#&gt; metabolome  0.1    1.0        0.0 1</span></span>
<span><span class="co">#&gt; Y           1.0    1.0        1.0 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="choosing-the-number-of-latent-components" class="level2" data-number="10.4"><h2 data-number="10.4" class="anchored" data-anchor-id="choosing-the-number-of-latent-components">
<span class="header-section-number">10.4</span> Choosing the number of latent components</h2>
<p>One important parameter that must be set when performing a DIABLO analysis is the number of latent components to construct for each dataset. The optimal number of components can be estimated by cross-validation, implemented in the <code><a href="https://rdrr.io/pkg/mixOmics/man/perf.html">mixOmics::perf()</a></code> function. This function assesses the classification performance (i.e.&nbsp;how well the different outcome groups are separated) achieved by DIABLO for different numbers of latent components.</p>
<p>Choosing the optimal number of latent components to construct is a multi-step process. The first step is to run DIABLO without feature selection, setting the number of latent components to the maximum value we wish to test. We recommend to set this to the number of groups in the categorical outcome + 2, which in our case equals 4; however this can be further refined after checking the results. For this example, we will set the maximum to 7. This is done through the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_run.html">diablo_run()</a></code> function, which is a wrapper for the <code><a href="https://rdrr.io/pkg/mixOmics/man/block.splsda.html">mixOmics::block.splsda()</a></code> function. The function also requires as input the design matrix to be used; here we will use the one constructed from the PLS runs:</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_novarsel</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_run.html">diablo_run</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_input</span>, </span>
<span>    <span class="va">diablo_design_matrix</span>, </span>
<span>    ncomp <span class="op">=</span> <span class="fl">7</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Alternatively, if we want to use one of the predefined design matrices, we can pass on one of <code>'null'</code>, <code>'weighted_full'</code> or <code>'full'</code> instead of the computed <code>diablo_design_matrix</code>, e.g.:</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_novarsel</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_run.html">diablo_run</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_input</span>, </span>
<span>    <span class="st">"weighted_full"</span>, </span>
<span>    ncomp <span class="op">=</span> <span class="fl">7</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Then, we call the <code><a href="https://rdrr.io/pkg/mixOmics/man/perf.html">mixOmics::perf()</a></code> function on the result of this first DIABLO run. There are a number of parameters to set:</p>
<ul>
<li><p><code>validation</code>: the type of cross-validation to perform, M-fold (<code>"Mfold"</code>) or leave-one-out (<code>"loo"</code>). We recommend to use M-fold validation, except when the number of samples is very small.</p></li>
<li><p><code>folds</code>: for M-fold cross-validation, the number of folds to construct, i.e.&nbsp;the number of groups in which to split the samples. Each group in turn will be considered as test set while the remaining groups will be considered the training set. The value to use depends on the number of samples in the datasets. By default, 10 is a reasonable number. For leave-one-out cross-validation, this parameter is set to the number of samples (that is the principle of leave-one-out cross-validation).</p></li>
<li><p><code>nrepeat</code>: the number of times the cross-validation will be repeated. This is important for M-fold cross-validation, as the way the samples are split into groups affects the results. Therefore, by repeating the cross-validation scheme we’re averaging the results over different splits, thus reducing the impact of samples splitting. We recommend at least 10 repeats. Irrelevant for leave-one-out cross-validation, so can be left to 1.</p></li>
<li><p><code>cpus</code>: number of CPUs to use for the computation. Useful if <code>folds</code> <span class="math inline">\(\times\)</span> <code>repeats</code> is large, as this can be computationally intensive.</p></li>
</ul>
<p>Here we’ll perform a 10-fold cross validation with 10 repeats.</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_perf_res</span>,</span>
<span>  <span class="fu">mixOmics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/perf.html">perf</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_novarsel</span>,</span>
<span>    validation <span class="op">=</span> <span class="st">"Mfold"</span>,</span>
<span>    folds <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    nrepeat <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    cpus <span class="op">=</span> <span class="fl">3</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>We can visualise the results of the cross-validation with the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_perf.html">diablo_plot_perf()</a></code> function:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_perf_res</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_perf.html">diablo_plot_perf</a></span><span class="op">(</span><span class="va">diablo_perf_res</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/plot-diablo-perf-res-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>The plot displays the cross-validation results computed with several different distances and error rates:</p>
<ul>
<li><p>Distance: this refers to the prediction distance that is used to predict the samples group in the test set, based on the samples grouping in the training set. DIABLO tests the maximum, centroids and Mahalanobis distance. The authors of the package recommend using either the centroids or the Mahalanobis distance over the maximum distance when choosing the optimal number of components.</p></li>
<li><p>Error rate: this refers to the method by which the performance of the produced model is computed. DIABLO uses both the overall misclassification error rate and the balanced error rate. The authors recommend the latter, as it is is less biased towards the majority group when there is an unbalanced number of samples per group.</p></li>
</ul>
<p>The function <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_optim_ncomp.html">diablo_get_optim_ncomp()</a></code> extracts from the cross-validation results the optimal number of components to compute, given a chosen distance and error rate. The authors of the package recommend to use the results obtained with the centroids distance and the balanced error rate; these are used by default by the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_optim_ncomp.html">diablo_get_optim_ncomp()</a></code> function. In our example, the optimal number of components is:</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_optim_ncomp.html">diablo_get_optim_ncomp</a></span><span class="op">(</span><span class="va">diablo_perf_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For ease of reuse we will save this value as a target in our analysis pipeline:</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_optim_ncomp</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_optim_ncomp.html">diablo_get_optim_ncomp</a></span><span class="op">(</span><span class="va">diablo_perf_res</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="choosing-the-number-of-features-to-retain" class="level2" data-number="10.5"><h2 data-number="10.5" class="anchored" data-anchor-id="choosing-the-number-of-features-to-retain">
<span class="header-section-number">10.5</span> Choosing the number of features to retain</h2>
<p>The next parameter to set is the number of features to retain from the different datasets for each latent component. This is usually chosen by performing cross-validation on a grid of possible values. The range of values to test depends on the type of question we are trying to answer: selecting a larger number of features might lead to a better discrimination of the outcome groups, but will be hard to manually inspect for further interpretation.</p>
<p>The function <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_tune.html">diablo_tune()</a></code> provides a wrapper around the <code><a href="https://rdrr.io/pkg/mixOmics/man/tune.html">mixOmics::tune()</a></code> function that performs this cross-validation. Some of the arguments are similar to the <code><a href="https://rdrr.io/pkg/mixOmics/man/perf.html">mixOmics::perf()</a></code> function, e.g.&nbsp;<code>validation</code>, <code>folds</code>, <code>nrepeats</code> or <code>cpus</code>. In addition, we recommend setting the <code>dist</code> argument, which corresponds to the prediction distance metric used for performance assessment, to <code>"centroids.dist"</code> (or <code>"mahalanobis.dist"</code>).</p>
<p>The <code>keepX_list</code> argument controls the grid of values to be tested as possible number of features to retain from each dataset. It should be in the form of a named list, with one element per dataset, and where each element is a vector of integers corresponding to the values to test. The names of the list should correspond to the names of the datasets in the <code>MultiDataSet</code> object. If no value is provided for <code>keepX_list</code>, six values ranging from 5 to 30 (by increments of 5) are tested for each dataset.</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_tune_res</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_tune.html">diablo_tune</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_input</span>,</span>
<span>    <span class="va">diablo_design_matrix</span>,</span>
<span>    ncomp <span class="op">=</span> <span class="va">diablo_optim_ncomp</span>,</span>
<span>    validation <span class="op">=</span> <span class="st">"Mfold"</span>,</span>
<span>    folds <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    nrepeat <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    dist <span class="op">=</span> <span class="st">"centroids.dist"</span>,</span>
<span>    cpus <span class="op">=</span> <span class="fl">3</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>This step can be very time-consuming, especially if the grid of values to test is very large. For this example, it takes around 50 minutes to run.</p>
<p>The cross-validation results can be inspected with the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_tune.html">diablo_plot_tune()</a></code> function:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_tune_res</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_tune.html">diablo_plot_tune</a></span><span class="op">(</span><span class="va">diablo_tune_res</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plot-tune-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>The visualisation shows the performance of DIABLO runs with different number of features retained from each dataset. The different runs are ordered according to their performance. Here, we can see for example that it seems preferable to retain more genes and less metabolites for component 1.</p>
<p>The optimal number of features to retain from each dataset for the different latent components is stored in the cross-validation results object, and can be accessed with:</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">diablo_tune_res</span><span class="op">$</span><span class="va">choice.keepX</span></span>
<span><span class="co">#&gt; $snps</span></span>
<span><span class="co">#&gt; [1] 20 25  5 25</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $rnaseq</span></span>
<span><span class="co">#&gt; [1] 30 15 20 30</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $metabolome</span></span>
<span><span class="co">#&gt; [1] 10 10  5  5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For reporting purposes, the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_table_optim_keepX.html">diablo_table_optim_keepX()</a></code> function displays the optimal <code>keepX</code> values in a table format:</p>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_table_optim_keepX.html">diablo_table_optim_keepX</a></span><span class="op">(</span><span class="va">diablo_tune_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 3 × 6</span></span>
<span><span class="co">#&gt;   Dataset    `Component 1` `Component 2` `Component 3` `Component 4` Total</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;dbl&gt;         &lt;dbl&gt;         &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 snps                  20            25             5            25    75</span></span>
<span><span class="co">#&gt; 2 rnaseq                30            15            20            30    95</span></span>
<span><span class="co">#&gt; 3 metabolome            10            10             5             5    30</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="final-diablo-run" class="level2" data-number="10.6"><h2 data-number="10.6" class="anchored" data-anchor-id="final-diablo-run">
<span class="header-section-number">10.6</span> Final DIABLO run</h2>
<p>Once a value has been selected for all parameters, it is time to perform the final DIABLO run:</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_final_run</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_run.html">diablo_run</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_input</span>,</span>
<span>    <span class="va">diablo_design_matrix</span>,</span>
<span>    ncomp <span class="op">=</span> <span class="va">diablo_optim_ncomp</span>,</span>
<span>    keepX <span class="op">=</span> <span class="va">diablo_tune_res</span><span class="op">$</span><span class="va">choice.keepX</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_final_run</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To facilitate reporting, the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_params.html">diablo_get_params()</a></code> function extracts from the DIABLO result the parameters used (i.e.&nbsp;number of latent components computed and number of features retained from each dataset for each latent component), with HTML formatting:</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_params.html">diablo_get_params</a></span><span class="op">(</span><span class="va">diablo_final_run</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 3</span></span>
<span><span class="co">#&gt;   Parameter Description                                                    Value</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;     &lt;chr&gt;                                                          &lt;chr&gt;</span></span>
<span><span class="co">#&gt; 1 ncomp     Number of latent component                                     4    </span></span>
<span><span class="co">#&gt; 2 keepX     Number of features retained in each X for each latent compone… snps…</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="results-interpretation" class="level2" data-number="10.7"><h2 data-number="10.7" class="anchored" data-anchor-id="results-interpretation">
<span class="header-section-number">10.7</span> Results interpretation</h2>
<p>In <a href="interpretation.html"><span>Chapter&nbsp;11</span></a>, we show the different functionalities implemented in the <code>moiraine</code> package that facilitate the interpretation of the results from an integration tool. In this section, we show some of the DIABLO-specific plots that can be generated to help interpret the results of a DIABLO run.</p>
<section id="correlation-between-datasets" class="level3" data-number="10.7.1"><h3 data-number="10.7.1" class="anchored" data-anchor-id="correlation-between-datasets">
<span class="header-section-number">10.7.1</span> Correlation between datasets</h3>
<p>First, we can assess how well the latent components correlate across the datasets. The <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot.html">diablo_plot()</a></code> function is adapted from the <code><a href="https://rdrr.io/pkg/mixOmics/man/plotDiablo.html">mixOmics::plotDiablo()</a></code> function, and displays, for a given latent component (specified with the <code>ncomp</code> argument), the correlation between the samples coordinates for this latent component across the datasets. Additionally, it allows to assess how well the latent components discriminate the outcome groups in each dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">n_comp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_optim_ncomp.html">diablo_get_optim_ncomp</a></span><span class="op">(</span><span class="va">diablo_perf_res</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">walk</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="va">n_comp</span><span class="op">)</span>, </span>
<span>  \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot.html">diablo_plot</a></span><span class="op">(</span><span class="va">diablo_final_run</span>, ncomp <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Latent component"</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plot-1.png" class="img-fluid" width="768"></p>
</div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plot-2.png" class="img-fluid" width="768"></p>
</div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plot-3.png" class="img-fluid" width="768"></p>
</div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plot-4.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>For the first three latent components, the strongest correlation is observed between the transcriptomics and metabolomics components, while the lowest correlation is observed between the genomics and metabolomics components. Across all three datasets, the first latent component alone is able to separate quite clearly the control and diseased animals. Note that as each latent component maximises the correlation between the datasets, these plots inform us about co-variation across the datasets.</p>
</section><section id="samples-projection-to-the-latent-component-space" class="level3" data-number="10.7.2"><h3 data-number="10.7.2" class="anchored" data-anchor-id="samples-projection-to-the-latent-component-space">
<span class="header-section-number">10.7.2</span> Samples projection to the latent component space</h3>
<p>We can also represent the samples in the subspace spanned by the latent components for each dataset, using the <code><a href="https://rdrr.io/pkg/mixOmics/man/plotIndiv.html">mixOmics::plotIndiv()</a></code> function. For example, we can have a look at the samples coordinates for the first two latent components:</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/plotIndiv.html">plotIndiv</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_final_run</span>,</span>
<span>  comp <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,</span>
<span>  ind.names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  legend.title <span class="op">=</span> <span class="st">"Disease status"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plotIndiv-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>As noted above, based on the first two latent components, there is a clear separation of the control and BRD animals across all three datasets.</p>
<p>Ideally, we would look at all possible combinations of latent components, as follows:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">walk</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/combn.html">combn</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="va">n_comp</span><span class="op">)</span>, <span class="fl">2</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>  \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/plotIndiv.html">plotIndiv</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_final_run</span>,</span>
<span>      comp <span class="op">=</span> <span class="va">x</span>,</span>
<span>      ind.names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>      legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      legend.title <span class="op">=</span> <span class="st">"Phenotype group"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span> </span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="correlation-circle-plots" class="level3" data-number="10.7.3"><h3 data-number="10.7.3" class="anchored" data-anchor-id="correlation-circle-plots">
<span class="header-section-number">10.7.3</span> Correlation circle plots</h3>
<p>The correlation circle plots produced by the <code><a href="https://rdrr.io/pkg/mixOmics/man/plotVar.html">mixOmics::plotVar()</a></code> function displays the contribution of the selected features to the different latent components. We will focus here on the first two latent components:</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/plotVar.html">plotVar</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_final_run</span>,</span>
<span>  comp <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,</span>
<span>  var.names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="co">## If overlap = TRUE, features from the</span></span>
<span>  <span class="co">## different datasets are shown in one plot</span></span>
<span>  overlap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">16</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  cex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plotVar-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Across all three datasets, it seems that most selected features contribute to either one or the other latent component, but not both.</p>
<p>The <code><a href="https://rdrr.io/pkg/mixOmics/man/plotVar.html">plotVar()</a></code> function offers the option to show the label of the features rather than representing them as points. However, it can be more informative to use information from the feature metadata as labels, rather than using the feature IDs. For example in the transcriptomics dataset, it would be more interesting to use the name of the genes. This information is available in the datasets’ features metadata:</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mo_set_de</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_features_metadata.html">get_features_metadata</a></span><span class="op">(</span><span class="va">mo_set_de</span><span class="op">)</span><span class="op">[[</span><span class="st">"rnaseq"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    20335 obs. of  15 variables:</span></span>
<span><span class="co">#&gt;  $ feature_id : chr  "ENSBTAG00000000005" "ENSBTAG00000000008" "ENSBTAG00000000009" "ENSBTAG00000000010" ...</span></span>
<span><span class="co">#&gt;  $ chromosome : Factor w/ 178 levels "1","2","3","4",..: 17 29 18 21 8 20 20 5 5 23 ...</span></span>
<span><span class="co">#&gt;  $ start      : int  65389743 32214439 12338037 34209956 7950815 33708626 33674806 74883636 74905031 27720176 ...</span></span>
<span><span class="co">#&gt;  $ end        : int  65505336 32244810 12342272 34223394 7971600 33732944 33703223 74893752 74919112 27721739 ...</span></span>
<span><span class="co">#&gt;  $ width      : int  115594 30372 4236 13439 20786 24319 28418 10117 14082 1564 ...</span></span>
<span><span class="co">#&gt;  $ strand     : Factor w/ 3 levels "+","-","*": 1 2 1 1 2 1 1 2 2 2 ...</span></span>
<span><span class="co">#&gt;  $ Name       : chr  "GRK3" "KCNJ1" "FOXF1" "UBL7" ...</span></span>
<span><span class="co">#&gt;  $ description: chr  "G protein-coupled receptor kinase 3 [Source:VGNC Symbol;Acc:VGNC:53904]" "potassium inwardly rectifying channel subfamily J member 1 [Source:VGNC Symbol;Acc:VGNC:30453]" "forkhead box F1 [Source:VGNC Symbol;Acc:VGNC:29084]" "ubiquitin like 7 [Source:VGNC Symbol;Acc:VGNC:50128]" ...</span></span>
<span><span class="co">#&gt;  $ log_fc     : num  0.136 -0.1297 1.2714 0.3957 0.0777 ...</span></span>
<span><span class="co">#&gt;  $ log_cpm    : num  5.905 -0.744 -2.563 6.256 -2.761 ...</span></span>
<span><span class="co">#&gt;  $ f          : num  4.163 1.068 23.956 60.53 0.142 ...</span></span>
<span><span class="co">#&gt;  $ p_value    : num  4.32e-02 3.03e-01 2.73e-06 1.58e-12 7.07e-01 ...</span></span>
<span><span class="co">#&gt;  $ fdr        : num  7.17e-02 3.94e-01 9.67e-06 1.52e-11 7.77e-01 ...</span></span>
<span><span class="co">#&gt;  $ de_signif  : chr  "Not DE" "Not DE" "Not DE" "Not DE" ...</span></span>
<span><span class="co">#&gt;  $ de_status  : chr  "Not DE" "Not DE" "Not DE" "Not DE" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_var.html">diablo_plot_var()</a></code> function is a variant of <code><a href="https://rdrr.io/pkg/mixOmics/man/plotVar.html">plotVar()</a></code>, which uses columns from the features metadata to label features plot. It takes as an input the DIABLO result object as well as the <code>MultiDataSet</code> object, and a named list providing for each dataset the name of the column in the feature metadata data-frame to use as features label:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_var.html">diablo_plot_var</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_final_run</span>,</span>
<span>  <span class="va">mo_set_de</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  overlap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  comp <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plot-var-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Note that if a dataset is not present in the list passed to <code>label_cols</code> (here, that is the case of the genomics dataset), the feature IDs will be used as labels.</p>
</section><section id="circos-plot" class="level3" data-number="10.7.4"><h3 data-number="10.7.4" class="anchored" data-anchor-id="circos-plot">
<span class="header-section-number">10.7.4</span> Circos plot</h3>
<p>Lastly, it is possible to represent the correlation between features selected from different datasets, with the <code><a href="https://rdrr.io/pkg/mixOmics/man/circosPlot.html">mixOmics::circosPlot()</a></code> function. For ease of visualisation, it only displays correlations above a certain threshold (specified via the <code>cutoff</code> argument). By default, it displays the features selected for all latent components, but this can be controlled via the <code>comp</code> argument:</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/circosPlot.html">circosPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_final_run</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  size.variables <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  comp <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-circosPlot-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>As for the correlation circle plot function, the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_circos.html">diablo_plot_circos()</a></code> function generates the same plot, but allows us to use columns in the feature metadata of each dataset as feature labels:</p>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_circos.html">diablo_plot_circos</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_final_run</span>,</span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">mo_set</span><span class="op">)</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  size.variables <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  comp <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="diablo_files/figure-html/diablo-plot-circos-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>This plot is useful to identify features across the datasets with high correlations.</p>
</section></section><section id="recap-targets-list" class="level2" data-number="10.8"><h2 data-number="10.8" class="anchored" data-anchor-id="recap-targets-list">
<span class="header-section-number">10.8</span> Recap – targets list</h2>
<p>For convenience, here is the list of targets that we created in this section:</p>
<details><summary>
Targets list for DIABLO analysis
</summary><div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Creating the DIABLO input</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_input</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_input_mixomics_supervised.html">get_input_mixomics_supervised</a></span><span class="op">(</span></span>
<span>      <span class="va">mo_presel_supervised</span>,</span>
<span>      group <span class="op">=</span> <span class="st">"status"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Running sPLS on each dataset to construct the design matrix</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_pairwise_pls_factory.html">diablo_pairwise_pls_factory</a></span><span class="op">(</span><span class="va">diablo_input</span><span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Initial DIABLO run with no feature selection and large number of components</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_novarsel</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_run.html">diablo_run</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_input</span>,</span>
<span>      <span class="va">diablo_design_matrix</span>,</span>
<span>      ncomp <span class="op">=</span> <span class="fl">7</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Cross-validation for number of components</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_perf_res</span>,</span>
<span>    <span class="fu">mixOmics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/perf.html">perf</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_novarsel</span>,</span>
<span>      validation <span class="op">=</span> <span class="st">"Mfold"</span>,</span>
<span>      folds <span class="op">=</span> <span class="fl">10</span>,</span>
<span>      nrepeat <span class="op">=</span> <span class="fl">10</span>,</span>
<span>      cpus <span class="op">=</span> <span class="fl">3</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plotting cross-validation results (for number of components)</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_perf_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_perf.html">diablo_plot_perf</a></span><span class="op">(</span><span class="va">diablo_perf_res</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Selected value for ncomp</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_optim_ncomp</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_get_optim_ncomp.html">diablo_get_optim_ncomp</a></span><span class="op">(</span><span class="va">diablo_perf_res</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Cross-validation for number of features to retain</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_tune_res</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_tune.html">diablo_tune</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_input</span>,</span>
<span>      <span class="va">diablo_design_matrix</span>,</span>
<span>      ncomp <span class="op">=</span> <span class="va">diablo_optim_ncomp</span>,</span>
<span>      validation <span class="op">=</span> <span class="st">"Mfold"</span>,</span>
<span>      folds <span class="op">=</span> <span class="fl">10</span>,</span>
<span>      nrepeat <span class="op">=</span> <span class="fl">5</span>,</span>
<span>      dist <span class="op">=</span> <span class="st">"centroids.dist"</span>,</span>
<span>      cpus <span class="op">=</span> <span class="fl">3</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plotting cross-validation results (for number of features)</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_tune_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_tune.html">diablo_plot_tune</a></span><span class="op">(</span><span class="va">diablo_tune_res</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Final DIABLO run</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_final_run</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_run.html">diablo_run</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_input</span>,</span>
<span>      <span class="va">diablo_design_matrix</span>,</span>
<span>      ncomp <span class="op">=</span> <span class="va">diablo_optim_ncomp</span>,</span>
<span>      keepX <span class="op">=</span> <span class="va">diablo_tune_res</span><span class="op">$</span><span class="va">choice.keepX</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details>

</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./mofa.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Integration with MOFA</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./interpretation.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>