<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>The moiraine R package user manual - 13&nbsp; Evaluating the integration results</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./interpretation.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The moiraine R package user manual</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./example_dataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">The example dataset</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Data preparation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_import.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Importing data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inspecting_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Inspecting the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./modifying_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocessing.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data pre-processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prefiltering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Dataset pre-filtering</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Unsupervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Integration with sPLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./so2pls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Integration with sO2PLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mofa.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Integration with MOFA</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Supervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./diablo.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Integration with DIABLO</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Results interpretation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./interpretation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evaluation.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="toc-section-number">13.1</span>  Introduction</a></li>
  <li>
<a href="#evaluating-features-prioritisation" id="toc-evaluating-features-prioritisation" class="nav-link" data-scroll-target="#evaluating-features-prioritisation"><span class="toc-section-number">13.2</span>  Evaluating features prioritisation</a>
  <ul class="collapse">
<li><a href="#table-of-counts-of-selected-genes-against-a-grouping" id="toc-table-of-counts-of-selected-genes-against-a-grouping" class="nav-link" data-scroll-target="#table-of-counts-of-selected-genes-against-a-grouping"><span class="toc-section-number">13.2.1</span>  Table of counts of selected genes against a grouping</a></li>
  <li><a href="#plotting-features-weight-against-a-covariate" id="toc-plotting-features-weight-against-a-covariate" class="nav-link" data-scroll-target="#plotting-features-weight-against-a-covariate"><span class="toc-section-number">13.2.2</span>  Plotting features weight against a covariate</a></li>
  </ul>
</li>
  <li>
<a href="#features-set-enrichment" id="toc-features-set-enrichment" class="nav-link" data-scroll-target="#features-set-enrichment"><span class="toc-section-number">13.3</span>  Features set enrichment</a>
  <ul class="collapse">
<li><a href="#constructing-feature-sets" id="toc-constructing-feature-sets" class="nav-link" data-scroll-target="#constructing-feature-sets"><span class="toc-section-number">13.3.1</span>  Constructing feature sets</a></li>
  <li><a href="#filtering-feature-sets-against-the-datasets" id="toc-filtering-feature-sets-against-the-datasets" class="nav-link" data-scroll-target="#filtering-feature-sets-against-the-datasets"><span class="toc-section-number">13.3.2</span>  Filtering feature sets against the datasets</a></li>
  <li><a href="#checking-the-number-of-features-assigned-to-sets" id="toc-checking-the-number-of-features-assigned-to-sets" class="nav-link" data-scroll-target="#checking-the-number-of-features-assigned-to-sets"><span class="toc-section-number">13.3.3</span>  Checking the number of features assigned to sets</a></li>
  <li><a href="#enrichment-of-latent-dimensions" id="toc-enrichment-of-latent-dimensions" class="nav-link" data-scroll-target="#enrichment-of-latent-dimensions"><span class="toc-section-number">13.3.4</span>  Enrichment of latent dimensions</a></li>
  <li><a href="#plotting-features-weight-for-feature-sets" id="toc-plotting-features-weight-for-feature-sets" class="nav-link" data-scroll-target="#plotting-features-weight-for-feature-sets"><span class="toc-section-number">13.3.5</span>  Plotting features weight for feature sets</a></li>
  </ul>
</li>
  <li><a href="#assessing-samples-clustering" id="toc-assessing-samples-clustering" class="nav-link" data-scroll-target="#assessing-samples-clustering"><span class="toc-section-number">13.4</span>  Assessing samples clustering</a></li>
  <li><a href="#recap-targets-list" id="toc-recap-targets-list" class="nav-link" data-scroll-target="#recap-targets-list"><span class="toc-section-number">14</span>  Recap – targets list</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-evaluation" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/targets/">targets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PlantandFoodResearch/moiraine">moiraine</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For data-frames manipulation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For colour palettes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After having investigated the results of an integration method (<a href="interpretation.html"><span>Chapter&nbsp;12</span></a>), we will show in this chapter how to evaluate the integration results against single-omics analyses results or prior biological knowledge.</p>
<p>As a reminder, here is what the <code>_targets.R</code> script should look like so far:</p>
<details><summary><code>_targets.R</code> script
</summary><div class="cell">
<div class="cell">

</div>
</div>
</details><section id="introduction" class="level2" data-number="13.1"><h2 data-number="13.1" class="anchored" data-anchor-id="introduction">
<span class="header-section-number">13.1</span> Introduction</h2>
<p>Evaluating the results of a multi-omics integration analysis is not an easy task, mainly because when analysing biological data there is no ground truth about the underlying biological mechanisms at play. However, it is possible to gain some measure of confidence in our results by comparing them to what we already know about the biological system studied, or to other analyses (typically single-omics analyses) performed on the dataset. In the latter case, this can also show us what insights are gained by analysing the omics datasets together that would have been missed by looking at each omics dataset separately. Note that in this vignette, we refer to the <strong>evaluation</strong> of an integration method as comparison with prior knowledge or previous studies, while in <span class="quarto-unresolved-ref">?sec-comparison</span> we will talk about the <strong>comparison</strong> of results obtained with different integration tools.</p>
<p>When evaluating integration results, we can compare the features contribution (see <a href="interpretation.html#sec-interpretation-dim-reduction"><span>Section&nbsp;12.1</span></a>) to some prior information that we have about the features: for example, their score in a differential expression analysis, whether they were found differentially expressed or not, or the group to which they belong (e.g.&nbsp;biological pathway or chemical class). On the other hand, we can assess whether the latent dimensions created separate different sample groups (e.g.&nbsp;control vs treatment, or different time points).</p>
<p>For this vignette, we will use the results from the DIABLO (<a href="diablo.html"><span>Chapter&nbsp;11</span></a>) and MOFA (<a href="mofa.html"><span>Chapter&nbsp;10</span></a>) analyses as examples. We will use the dimension reduction output objects that we created in <a href="interpretation.html"><span>Chapter&nbsp;12</span></a>, which contains the results of the DIABLO and MOFA runs stored in a standardised format (<a href="interpretation.html#sec-interpretation-standard-output"><span>Section&nbsp;12.2</span></a>):</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_output</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mofa_output</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that for the DIABLO run, the samples score are computed for the weighted average of each latent component, rather than for the dataset-specific latent components (this is controlled through the <code>use_average_dimensions</code> argument from the <code><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output()</a></code> function).</p>
<p>We will also use the <code>MultiDataSet</code> object that contains information about the samples and features:</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mo_set_complete</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="evaluating-features-prioritisation" class="level2" data-number="13.2"><h2 data-number="13.2" class="anchored" data-anchor-id="evaluating-features-prioritisation">
<span class="header-section-number">13.2</span> Evaluating features prioritisation</h2>
<p>First, we can evaluate the integration results in terms of the features prioritisation, i.e.&nbsp;the ranking of the features for each latent dimension in terms of their importance score. We can for example assess whether the features that were given the highest scores for some latent components were also highlighted through single-omics analyses, or compare the importance score of groups of features, for example compounds from different chemical classes or genes from different gene families.</p>
<p>Here, we will compare the DIABLO results to the results of traditional single-omics analyses. In our case, a GWAS and eQTL analysis were performed to detect genomic variants associated with the disease status of the samples. The mode of action of significant markers (i.e.&nbsp;QTL, cis-eQTL or trans-eQTL). The GWAS status (high-scoring or not) of each marker was recorded in the genomics features metadata table:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_features_metadata.html">get_features_metadata</a></span><span class="op">(</span><span class="va">mo_set_complete</span><span class="op">)</span><span class="op">[[</span><span class="st">"snps"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">qtl_type</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] NA           "cis eQTL"   "trans eQTL" "QTL"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For the transcriptomics and metabolomics dataset, a differential expression (DE) analysis was performed on each of them to compare control and BRD animals. The resulting adjusted p-value, log2-fold change or t-value, status (upregulated, downregulated or not differentially expressed) and significance (differentially expressed or not) of each gene or compound are recorded in each dataset’s features metadata:</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_features_metadata.html">get_features_metadata</a></span><span class="op">(</span><span class="va">mo_set_complete</span><span class="op">)</span><span class="op">[[</span><span class="st">"rnaseq"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">feature_id</span>, <span class="va">log_fc</span><span class="op">:</span><span class="va">de_status</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                            feature_id      log_fc    log_cpm          f</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005 ENSBTAG00000000005  0.13603041  5.9051177  4.1626465</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008 ENSBTAG00000000008 -0.12965356 -0.7437052  1.0676845</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009 ENSBTAG00000000009  1.27141158 -2.5627934 23.9562951</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010 ENSBTAG00000000010  0.39567729  6.2563594 60.5303416</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011 ENSBTAG00000000011  0.07766873 -2.7608361  0.1418457</span></span>
<span><span class="co">#&gt; ENSBTAG00000000012 ENSBTAG00000000012  0.15756169  3.6628775 11.0448775</span></span>
<span><span class="co">#&gt;                         p_value          fdr de_signif de_status</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005 4.324808e-02 7.170137e-02    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008 3.032916e-01 3.938643e-01    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009 2.731328e-06 9.665375e-06    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010 1.581955e-12 1.521160e-11    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011 7.070363e-01 7.773874e-01    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000012 1.141125e-03 2.623062e-03    Not DE    Not DE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_features_metadata.html">get_features_metadata</a></span><span class="op">(</span><span class="va">mo_set_complete</span><span class="op">)</span><span class="op">[[</span><span class="st">"metabolome"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span>  </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">feature_id</span>, <span class="va">t_value</span><span class="op">:</span><span class="va">de_status</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;           feature_id     t_value      p_value         padj de_signif</span></span>
<span><span class="co">#&gt; HMDB00001  HMDB00001  -0.5557020 5.797635e-01 6.784466e-01    Not DE</span></span>
<span><span class="co">#&gt; HMDB00008  HMDB00008   0.2181562 8.276321e-01 8.925444e-01    Not DE</span></span>
<span><span class="co">#&gt; HMDB00042  HMDB00042 -12.5323491 1.753101e-24 4.821028e-23        DE</span></span>
<span><span class="co">#&gt; HMDB00043  HMDB00043  -7.9073179 7.827088e-13 3.913544e-12        DE</span></span>
<span><span class="co">#&gt; HMDB00060  HMDB00060  -0.4369834 6.628164e-01 7.439776e-01    Not DE</span></span>
<span><span class="co">#&gt; HMDB00062  HMDB00062   8.2347595 1.271549e-13 6.993518e-13        DE</span></span>
<span><span class="co">#&gt;               de_status</span></span>
<span><span class="co">#&gt; HMDB00001        Not DE</span></span>
<span><span class="co">#&gt; HMDB00008        Not DE</span></span>
<span><span class="co">#&gt; HMDB00042 downregulated</span></span>
<span><span class="co">#&gt; HMDB00043 downregulated</span></span>
<span><span class="co">#&gt; HMDB00060        Not DE</span></span>
<span><span class="co">#&gt; HMDB00062   upregulated</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="table-of-counts-of-selected-genes-against-a-grouping" class="level3" data-number="13.2.1"><h3 data-number="13.2.1" class="anchored" data-anchor-id="table-of-counts-of-selected-genes-against-a-grouping">
<span class="header-section-number">13.2.1</span> Table of counts of selected genes against a grouping</h3>
<p>As DIABLO performs feature selection, it is of interest to assess how many (e)QTLs and DE genes and compounds were retained for each latent component. We can use the <code><a href="https://rdrr.io/pkg/moiraine/man/evaluate_feature_selection_table.html">evaluate_feature_selection_table()</a></code> function to generate a table of counts comparing the features selected for each latent component to a features grouping (in our case, the results of the single-omics analyses). The features grouping is obtained from the features metadata stored in the <code>MultiDataSet</code> object. The function takes as input the integration result object, the <code>MultiDataSet</code> object and a named list giving for each dataset the name of the column from their features metadata table containing the features grouping:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">diablo_evaluation_counts_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/evaluate_feature_selection_table.html">evaluate_feature_selection_table</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"snps"</span> <span class="op">=</span> <span class="st">"qtl_type"</span>,</span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_signif"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"de_signif"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">diablo_evaluation_counts_table</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">latent_dimension</span> <span class="op">==</span> <span class="st">"Component 1"</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 9 × 6</span></span>
<span><span class="co">#&gt;   method latent_dimension dataset    feature_label selected not_selected</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;            &lt;fct&gt;      &lt;chr&gt;            &lt;int&gt;        &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 DIABLO Component 1      snps       cis eQTL             1            2</span></span>
<span><span class="co">#&gt; 2 DIABLO Component 1      snps       QTL                  1            0</span></span>
<span><span class="co">#&gt; 3 DIABLO Component 1      snps       trans eQTL           1            3</span></span>
<span><span class="co">#&gt; 4 DIABLO Component 1      snps       &lt;NA&gt;                22          970</span></span>
<span><span class="co">#&gt; 5 DIABLO Component 1      rnaseq     DE                  20           55</span></span>
<span><span class="co">#&gt; 6 DIABLO Component 1      rnaseq     Not DE              10          782</span></span>
<span><span class="co">#&gt; 7 DIABLO Component 1      rnaseq     &lt;NA&gt;                 0          127</span></span>
<span><span class="co">#&gt; 8 DIABLO Component 1      metabolome DE                  10           20</span></span>
<span><span class="co">#&gt; 9 DIABLO Component 1      metabolome Not DE               0           25</span></span>
<span></span>
<span><span class="va">diablo_evaluation_counts_table</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">latent_dimension</span> <span class="op">==</span> <span class="st">"Component 2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 9 × 6</span></span>
<span><span class="co">#&gt;   method latent_dimension dataset    feature_label selected not_selected</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;            &lt;fct&gt;      &lt;chr&gt;            &lt;int&gt;        &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 DIABLO Component 2      snps       cis eQTL             0            3</span></span>
<span><span class="co">#&gt; 2 DIABLO Component 2      snps       QTL                  0            1</span></span>
<span><span class="co">#&gt; 3 DIABLO Component 2      snps       trans eQTL           0            4</span></span>
<span><span class="co">#&gt; 4 DIABLO Component 2      snps       &lt;NA&gt;                25          967</span></span>
<span><span class="co">#&gt; 5 DIABLO Component 2      rnaseq     DE                   0           75</span></span>
<span><span class="co">#&gt; 6 DIABLO Component 2      rnaseq     Not DE              30          762</span></span>
<span><span class="co">#&gt; 7 DIABLO Component 2      rnaseq     &lt;NA&gt;                 0          127</span></span>
<span><span class="co">#&gt; 8 DIABLO Component 2      metabolome DE                   3           27</span></span>
<span><span class="co">#&gt; 9 DIABLO Component 2      metabolome Not DE               2           23</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that for the first latent component, only three out of the eight (e)QTLs were selected <a href="prefiltering.html#sec-prefiltering-supervised">^ Note that in the full genomics dataset, 36 markers were detected as QTLs or eQTLs, however only eight of them were retained in the features preselection step (&nbsp;<span>7.3.2</span></a>; two-third of the genes selected were found differentially expressed; and all ten of the selected compounds were found differentially expressed. Conversely, for the second latent component, none of the selected markers were QTLs, and none of the selected genes were differentially expressed. This makes sense, as the DIABLO analysis was run with the goal of separating the control and infected animals; therefore it makes sense that the features selected for the first latent component, which is the one best able to separate the groups, are the features detected as differentially expressed. On the contrary, the second latent component is constructed to be orthogonal to the first one, therefore the features selected will likely not be differentially expressed.</p>
<p>With this function, we can focus on only some datasets, by changing which datasets are present in the list passed to <code>col_names</code>. Similarly, we can select specific latent dimensions through the <code>latent_dimensions</code> parameter. For example, let us count the number of DE genes selected with the first latent component:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/evaluate_feature_selection_table.html">evaluate_feature_selection_table</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_signif"</span><span class="op">)</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"Component 1"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 3 × 6</span></span>
<span><span class="co">#&gt;   method latent_dimension dataset feature_label selected not_selected</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;            &lt;fct&gt;   &lt;chr&gt;            &lt;int&gt;        &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 DIABLO Component 1      rnaseq  DE                  20           55</span></span>
<span><span class="co">#&gt; 2 DIABLO Component 1      rnaseq  Not DE              10          782</span></span>
<span><span class="co">#&gt; 3 DIABLO Component 1      rnaseq  &lt;NA&gt;                NA          127</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that this function is useful for integration methods that perform feature selection, such as DIABLO or sO2PLS, but will be irrelevant for methods that do not perform feature selection (for example MOFA tends to assign a non-null weight to most features).</p>
</section><section id="plotting-features-weight-against-a-covariate" class="level3" data-number="13.2.2"><h3 data-number="13.2.2" class="anchored" data-anchor-id="plotting-features-weight-against-a-covariate">
<span class="header-section-number">13.2.2</span> Plotting features weight against a covariate</h3>
<p>We can go further and compare the weights or importance scores given to the features for each latent component to the outcome of the single-omics results. This can be done with the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_covariate.html">plot_features_weight_covariate()</a></code> function. As its name suggests, this function displays the features weight for each latent dimension against some covariate obtained from the features metadata tables. Again, we will look at the results of the GWAS/QTL and DE analyses and see whether (e)QTLs, DE genes and compounds were assigned higher weights is some of the latent components constructed with DIABLO. The input parameters for this function are similar to those for the <code><a href="https://rdrr.io/pkg/moiraine/man/evaluate_feature_selection_table.html">evaluate_feature_selection_table()</a></code> function: we need to pass the results of the integration method, the <code>MultiDataSet</code> object, and the named list giving the column in each features metadata table containing information about the features grouping. The latter is passed to the <code>covariate</code> argument of the function. If some datasets are not present in this list, they won’t be present in the plot.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_covariate.html">plot_features_weight_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>, </span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>, </span>
<span>  covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"snps"</span> <span class="op">=</span> <span class="st">"qtl_type"</span>, </span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_status"</span>, </span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"de_status"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="evaluation_files/figure-html/plot-features-weight-covariate-all-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>By default, the function plots on the y-axis the signed importance score of the features, that is, their importance score to which the sign of their weight was added. This can be controlled via the <code>features_metric</code> argument. We see that most of the points have an importance score of 0, because most of the features were not selected. We can focus on only those features with a non-null weight with the <code>remove_null_weight</code> argument:</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_covariate.html">plot_features_weight_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>, </span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>, </span>
<span>  covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"snps"</span> <span class="op">=</span> <span class="st">"qtl_type"</span>, </span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_status"</span>, </span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"de_status"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  remove_null_weight <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="evaluation_files/figure-html/plot-features-weight-covariate-nonnull-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Here, we see that for the first latent component, selected genes that were also found up-regulated were assigned on average a higher importance score that selected genes that were found non DE. Also, for the metabolomics dataset, selected compounds that were found up-regulated were assigned a positive weight while those found down-regulated were assigned a negative weight.</p>
<p>Note that with the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_covariate.html">plot_features_weight_covariate()</a></code> function, we can also plot the features against a continuous covariate, e.g.&nbsp;the log2-fold change obtained with the differential expressions:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_covariate.html">plot_features_weight_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>, </span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>, </span>
<span>  covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"log_fc"</span>, </span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"t_value"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 508 rows containing non-finite values (`stat_smooth()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 508 rows containing missing values (`geom_point()`).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="evaluation_files/figure-html/plot-features-weight-covariate-continuous-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We can see a nice trend between compounds’ t-value and their signed importance score in the differential expression analysis for the first latent component.</p>
<p>We can further change the colour or shape of the points according to some information from the features metadata (information must be passed as a named list as for the covariate).</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_covariate.html">plot_features_weight_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>, </span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>, </span>
<span>  covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"log_fc"</span>, </span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"t_value"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"padj"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_colour_viridis_c</a></span><span class="op">(</span>aesthetics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"colour"</span>, <span class="st">"fill"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 508 rows containing non-finite values (`stat_smooth()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 508 rows containing missing values (`geom_point()`).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="evaluation_files/figure-html/plot-features-weight-covariate-continuous-colour-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Note that for continuous covariates, as with the function <code><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate()</a></code>, a loess curve is fit to summarise the trend between the covariates and the features weight. If <code>colour_by</code> is used, and the corresponding variables are categorical, a different loess curve is fitted for each category. If instead the <code>colour_by</code> variables are numeric, a single loess curve will be plotted for each latent dimension and dataset.</p>
</section></section><section id="features-set-enrichment" class="level2" data-number="13.3"><h2 data-number="13.3" class="anchored" data-anchor-id="features-set-enrichment">
<span class="header-section-number">13.3</span> Features set enrichment</h2>
<p>In addition to looking at the distribution of features importance score or counting the number of selected features, we can also perform an enrichment to assess whether each of the latent dimensions constructed is enriched for some features set of interest. Features sets are groups of features that belong in a similar category: e.g.&nbsp;all features involved in a same biological pathway, or with the same GO term annotation, or even all up- or down-regulated features. They are built using prior knowledge, for example information extracted from databases, or constructed from the results of previous analyses. Importantly, a feature can belong to more than one set (for example a compound can be involved in several biological pathways).</p>
<section id="constructing-feature-sets" class="level3" data-number="13.3.1"><h3 data-number="13.3.1" class="anchored" data-anchor-id="constructing-feature-sets">
<span class="header-section-number">13.3.1</span> Constructing feature sets</h3>
<p>With the <code>moiraine</code> package, there are two ways to obtain feature sets to perform enrichment or for plots. The first one is to extract information from the features metadata tables stored in our <code>MultiDataSet</code> object; which is done with the <code><a href="https://rdrr.io/pkg/moiraine/man/make_feature_sets_from_fm.html">make_feature_sets_from_fm()</a></code> function. Note that if some features grouping is stored in the features metadata, it means that features can only belong to one set. For example, we will use the results of the single-omics analyses to group the features into sets: QTLs, cis-eQTLs, trans-eQTLs and non significant genomics markers, and up-regulated, down-regulated or non DE genes and compounds. Again, the function accepts a named list giving for each dataset the name of the column in the corresponding features metadata table to use to generate the groups, passed to the <code>col_names</code> argument of the function. It returns a named list where each element is a feature set, and contains the ID of the features in the set:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">sets_single_omics</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/make_feature_sets_from_fm.html">make_feature_sets_from_fm</a></span><span class="op">(</span></span>
<span>    <span class="va">mo_set_complete</span>,</span>
<span>    col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>      <span class="st">"snps"</span> <span class="op">=</span> <span class="st">"qtl_type"</span>, </span>
<span>      <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_status"</span>, </span>
<span>      <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"de_status"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">sets_single_omics</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 11</span></span>
<span><span class="co">#&gt;  $ NA - snps                 : chr [1:23000] "1_41768691" "10-27008241-A-C-rs42918694" "10-37505419-T-C-rs136559242" "10-49904259-G-A-rs471723345" ...</span></span>
<span><span class="co">#&gt;  $ cis eQTL                  : chr [1:31] "ARS-BFGL-BAC-13210" "ARS-BFGL-NGS-113310" "ARS-BFGL-NGS-5022" "ARS-BFGL-NGS-59463" ...</span></span>
<span><span class="co">#&gt;  $ trans eQTL                : chr [1:4] "ARS-BFGL-NGS-80280" "BovineHD0100032240" "BovineHD2300010006" "Hapmap55381-rs29025399"</span></span>
<span><span class="co">#&gt;  $ QTL                       : chr "BovineHD1800016801"</span></span>
<span><span class="co">#&gt;  $ Not DE - rnaseq           : chr [1:18630] "ENSBTAG00000000005" "ENSBTAG00000000008" "ENSBTAG00000000009" "ENSBTAG00000000010" ...</span></span>
<span><span class="co">#&gt;  $ NA - rnaseq               : chr [1:1594] "ENSBTAG00000000029" "ENSBTAG00000000071" "ENSBTAG00000000102" "ENSBTAG00000000106" ...</span></span>
<span><span class="co">#&gt;  $ upregulated - rnaseq      : chr [1:102] "ENSBTAG00000000377" "ENSBTAG00000000783" "ENSBTAG00000001051" "ENSBTAG00000001785" ...</span></span>
<span><span class="co">#&gt;  $ downregulated - rnaseq    : chr [1:9] "ENSBTAG00000001032" "ENSBTAG00000004824" "ENSBTAG00000011990" "ENSBTAG00000012403" ...</span></span>
<span><span class="co">#&gt;  $ Not DE - metabolome       : chr [1:25] "HMDB00001" "HMDB00008" "HMDB00060" "HMDB00097" ...</span></span>
<span><span class="co">#&gt;  $ downregulated - metabolome: chr [1:20] "HMDB00042" "HMDB00043" "HMDB00064" "HMDB00094" ...</span></span>
<span><span class="co">#&gt;  $ upregulated - metabolome  : chr [1:10] "HMDB00062" "HMDB00092" "HMDB00159" "HMDB00169" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default, the function keeps the different omics features separate: here for example, there is a “Not DE” set for the transcriptomics dataset, and a “Not DE” set for the metabolomics dataset. When there is no ambiguity as to which omics dataset they represent, the name of the dataset is not included in the set’s name. It is possible to merge feature sets with the same name between the omics datasets, by setting the <code>combine_omics_sets</code> argument to <code>TRUE</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">sets_single_omics_merged</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/make_feature_sets_from_fm.html">make_feature_sets_from_fm</a></span><span class="op">(</span></span>
<span>    <span class="va">mo_set_complete</span>,</span>
<span>    col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>      <span class="st">"snps"</span> <span class="op">=</span> <span class="st">"qtl_type"</span>, </span>
<span>      <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_status"</span>, </span>
<span>      <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"de_status"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    combine_omics_sets <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">sets_single_omics_merged</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 7</span></span>
<span><span class="co">#&gt;  $ NA           : chr [1:24594] "1_41768691" "10-27008241-A-C-rs42918694" "10-37505419-T-C-rs136559242" "10-49904259-G-A-rs471723345" ...</span></span>
<span><span class="co">#&gt;  $ cis eQTL     : chr [1:31] "ARS-BFGL-BAC-13210" "ARS-BFGL-NGS-113310" "ARS-BFGL-NGS-5022" "ARS-BFGL-NGS-59463" ...</span></span>
<span><span class="co">#&gt;  $ trans eQTL   : chr [1:4] "ARS-BFGL-NGS-80280" "BovineHD0100032240" "BovineHD2300010006" "Hapmap55381-rs29025399"</span></span>
<span><span class="co">#&gt;  $ QTL          : chr "BovineHD1800016801"</span></span>
<span><span class="co">#&gt;  $ Not DE       : chr [1:18655] "ENSBTAG00000000005" "ENSBTAG00000000008" "ENSBTAG00000000009" "ENSBTAG00000000010" ...</span></span>
<span><span class="co">#&gt;  $ upregulated  : chr [1:112] "ENSBTAG00000000377" "ENSBTAG00000000783" "ENSBTAG00000001051" "ENSBTAG00000001785" ...</span></span>
<span><span class="co">#&gt;  $ downregulated: chr [1:29] "ENSBTAG00000001032" "ENSBTAG00000004824" "ENSBTAG00000011990" "ENSBTAG00000012403" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now the <code>not DE</code> set contains all genes and compounds ID that have not been found differentially expressed.</p>
<p>Alternatively, feature sets can be constructed from an input data-frame. This is useful when looking at pathways or GO annotations, for which features can belong to several sets at once, as this information cannot be stored in a features metadata table. Here, we will use the GO terms assigned to the genes. Note that in this example, the feature sets will contain features from only one omics dataset, but it doesn’t have to be; instead the sets could contain features from different omics datasets.</p>
<p>We start by reading in the data-frame of GO terms. We will focus on GO terms relating to biological processes for this example:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_go_terms_file</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span></span>
<span>      <span class="st">"extdata/transcriptomics_go_annotation.csv"</span>, </span>
<span>      package <span class="op">=</span> <span class="st">"moiraine"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    format <span class="op">=</span> <span class="st">"file"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_go_df</span>,</span>
<span>    <span class="fu">read_csv</span><span class="op">(</span><span class="va">rnaseq_go_terms_file</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">go_domain</span> <span class="op">==</span> <span class="st">"Biological process"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">rnaseq_go_df</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">rnaseq_go_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 89226     4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">rnaseq_go_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 4</span></span>
<span><span class="co">#&gt;   gene_id            go_id      go_name                                go_domain</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;chr&gt;      &lt;chr&gt;                                  &lt;chr&gt;    </span></span>
<span><span class="co">#&gt; 1 ENSBTAG00000000005 GO:0006468 protein phosphorylation                Biologic…</span></span>
<span><span class="co">#&gt; 2 ENSBTAG00000000005 GO:0007165 signal transduction                    Biologic…</span></span>
<span><span class="co">#&gt; 3 ENSBTAG00000000005 GO:0031623 receptor internalization               Biologic…</span></span>
<span><span class="co">#&gt; 4 ENSBTAG00000000008 GO:0006813 potassium ion transport                Biologic…</span></span>
<span><span class="co">#&gt; 5 ENSBTAG00000000009 GO:0006355 regulation of DNA-templated transcrip… Biologic…</span></span>
<span><span class="co">#&gt; 6 ENSBTAG00000000009 GO:0006357 regulation of transcription by RNA po… Biologic…</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The data-frame contains one row per gene (<code>gene_id</code>) / GO term (<code>go_id</code>) pair. In addition, the name of each GO term and the category (biological process, molecular function or cellular component) are indicated in the <code>go_name</code> and <code>go_domain</code> columns, respectively. When reading in the file, we’ve restricted the GO terms to only those that correspond to biological processes.</p>
<p>We can turn this data-frame into a list of feature sets with the <code><a href="https://rdrr.io/pkg/moiraine/man/make_feature_sets_from_df.html">make_feature_sets_from_df()</a></code> function. The function takes as input the data-frame, as well as the name of the column in the data-frame that contains the features ID (<code>col_id</code> argument) and the name of the column in the data-frame that contains the feature sets ID or name (<code>col_set</code> argument):</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">go_sets</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/make_feature_sets_from_df.html">make_feature_sets_from_df</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_go_df</span>,</span>
<span>    col_id <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>    col_set <span class="op">=</span> <span class="st">"go_id"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">go_sets</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">go_sets</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 10745</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">go_sets</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 6</span></span>
<span><span class="co">#&gt;  $ GO:0006468: chr [1:580] "ENSBTAG00000000005" "ENSBTAG00000000013" "ENSBTAG00000000056" "ENSBTAG00000000184" ...</span></span>
<span><span class="co">#&gt;  $ GO:0007165: chr [1:737] "ENSBTAG00000000005" "ENSBTAG00000000073" "ENSBTAG00000000081" "ENSBTAG00000000105" ...</span></span>
<span><span class="co">#&gt;  $ GO:0031623: chr [1:31] "ENSBTAG00000000005" "ENSBTAG00000000308" "ENSBTAG00000001050" "ENSBTAG00000001139" ...</span></span>
<span><span class="co">#&gt;  $ GO:0006813: chr [1:96] "ENSBTAG00000000008" "ENSBTAG00000000082" "ENSBTAG00000000745" "ENSBTAG00000000973" ...</span></span>
<span><span class="co">#&gt;  $ GO:0006355: chr [1:915] "ENSBTAG00000000009" "ENSBTAG00000000040" "ENSBTAG00000000052" "ENSBTAG00000000074" ...</span></span>
<span><span class="co">#&gt;  $ GO:0006357: chr [1:627] "ENSBTAG00000000009" "ENSBTAG00000000039" "ENSBTAG00000000052" "ENSBTAG00000000154" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="filtering-feature-sets-against-the-datasets" class="level3" data-number="13.3.2"><h3 data-number="13.3.2" class="anchored" data-anchor-id="filtering-feature-sets-against-the-datasets">
<span class="header-section-number">13.3.2</span> Filtering feature sets against the datasets</h3>
<p>Before performing any enrichment, we want to make sure that the only features in the sets are the ones that were measured in the datasets. This is part of ensuring that an appropriate background or reference features set is used when performing enrichment (see for example <span class="citation" data-cites="timmons2015">Timmons, Szkop, and Gallagher (<a href="references.html#ref-timmons2015" role="doc-biblioref">2015</a>)</span>). To this end, the <code><a href="https://rdrr.io/pkg/moiraine/man/reduce_feature_sets_data.html">reduce_feature_sets_data()</a></code> function filters a list of feature sets so that only features present in a given <code>MultiDataSet</code> object are kept in the sets; it then removes empty feature sets. Note that for this filtering, we are using the <code>MultiDataSet</code> object that contains the full omics datasets, before pre-filtering:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">go_sets_filtered</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/reduce_feature_sets_data.html">reduce_feature_sets_data</a></span><span class="op">(</span><span class="va">go_sets</span>, <span class="va">mo_set_complete</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function yields the following message:</p>
<div class="cell">
<pre><code>#&gt; All features in sets are in the multi-omics dataset.</code></pre>
</div>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">go_sets_filtered</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="checking-the-number-of-features-assigned-to-sets" class="level3" data-number="13.3.3"><h3 data-number="13.3.3" class="anchored" data-anchor-id="checking-the-number-of-features-assigned-to-sets">
<span class="header-section-number">13.3.3</span> Checking the number of features assigned to sets</h3>
<p>We can then check the number of features that are assigned to at least one set with the <code><a href="https://rdrr.io/pkg/moiraine/man/check_feature_sets.html">check_feature_sets()</a></code> function, to get an idea of the coverage of our annotation. We will restrict the check to the transcriptomics dataset, since the GO annotation only covers genes:</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sets_check</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/check_feature_sets.html">check_feature_sets</a></span><span class="op">(</span></span>
<span>  <span class="va">go_sets_filtered</span>, </span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  datasets <span class="op">=</span> <span class="st">"rnaseq"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; rnaseq dataset: 14,364 of 20,335 (70.6%) features assigned to at least one set.</span></span>
<span></span>
<span><span class="va">sets_check</span></span>
<span><span class="co">#&gt; # A tibble: 1 × 5</span></span>
<span><span class="co">#&gt;   dataset n_annotated     n frac_annotated message                              </span></span>
<span><span class="co">#&gt;   &lt;fct&gt;         &lt;int&gt; &lt;int&gt;          &lt;dbl&gt; &lt;chr&gt;                                </span></span>
<span><span class="co">#&gt; 1 rnaseq        14364 20335          0.706 rnaseq dataset: 14,364 of 20,335 (70…</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function returns a tibble giving the name of the datasets, the number and fraction of features from the dataset that are present in at least one feature set (<code>n_annotated</code> and <code>frac_annotated</code> columns), the total number of features in the corresponding dataset (<code>n</code> column), as well as sentence summarising this information (<code>message</code> column) that is useful for reporting. In this example, 70.6% of the genes from the transcriptomics dataset are assigned to at least one GO term. This is important to check as the annotation coverage will impact the quality of the enrichment results that we get.</p>
</section><section id="enrichment-of-latent-dimensions" class="level3" data-number="13.3.4"><h3 data-number="13.3.4" class="anchored" data-anchor-id="enrichment-of-latent-dimensions">
<span class="header-section-number">13.3.4</span> Enrichment of latent dimensions</h3>
<p>We are now ready to perform an enrichment of the latent dimensions. There are many ways to perform functional enrichment (e.g.&nbsp;see <span class="citation" data-cites="zhao2023">Zhao and Rhee (<a href="references.html#ref-zhao2023" role="doc-biblioref">2023</a>)</span>): based on over-representation, feature ranking, or even topology when the feature sets are representing biological pathways. Here, we are not trying to provide an exhaustive list of options; however it is relatively straightforward to extract from an integration method output the features importance or list of selected features for a specific latent dimension, and use it for enrichment with the package of your choice.</p>
<p>In <code>moiraine</code>, we use the <code>gage</code> package <span class="citation" data-cites="luo2009">(<a href="references.html#ref-luo2009" role="doc-biblioref">Luo et al. 2009</a>)</span> to perform the enrichment of each latent dimension against a list of feature sets. It uses a two-sample t-test to assess whether features in a set have higher importance scores than features not in the set. The <code><a href="https://rdrr.io/pkg/moiraine/man/evaluate_method_enrichment.html">evaluate_method_enrichment()</a></code> function takes as input the results from an integration method in a standardised format (here our <code>diablo_output</code> object), and performs an enrichment for each latent dimension using the features sets provided as a list. We can focus on specific datasets or latent dimensions by passing their name to the <code>datasets</code> or <code>latent_dimensions</code> arguments, respectively. There are a number of parameters that need to be set:</p>
<ul>
<li><p><code>use_abs</code>: by default, the function uses the absolute value of the importance score to perform the enrichment, which allows it to detect features sets in which features are given strong weights that are both positive and negative. If we instead want to search for features sets in which the weight of the features are coordinated (i.e.&nbsp;either all positive or negative), we can set the <code>use_abs</code> argument to <code>FALSE</code>.</p></li>
<li><p><code>min_set_size</code>: the minimum number of features in a set needed for the set to be considered for enrichment. The default value is 5, i.e.&nbsp;sets with less than 5 features will not be tested for enrichment.</p></li>
<li><p><code>add_missing_features</code> and <code>mo_data</code>: when performing multi-omics integration, it is highly likely that the omics datasets have been pre-filtered prior to the analysis, which means that some features will not have received a weight in the results of the integration analysis. This can bias the enrichment results, as these are based on only features that have a weight. So if a features set contains 50 features, but 40 of these were discarded during the pre-filtering step, only the remaining 10 features will be considered when performing an enrichment. This might lead to finding a latent dimension enriched for this set, even though the set might not be biologically relevant since 80% of its features did not even pass the pre-filtering step. To account for that, it is possible to add to the integration results all of the features in a <code>MultiDataSet</code> object that are not present in the results. These features will be assigned a weight of 0 to emphasise that they were not found to play any role for the latent dimensions. This is done by setting <code>add_missing_features</code> to <code>TRUE</code>, and passing to <code>mo_data</code> a <code>MultiDataSet</code> object that contains the omics datasets that have not been filtered. This way, the background set used in the enrichment will be all features measured in the omics datasets, and not only the ones that passed the pre-filtering step.</p></li>
<li><p><code>sets_info_df</code> and <code>col_set</code>: the feature sets are passed as a named list, therefore the only information that we have about these sets are their names as found in the list. To facilitate the interpretation of the enrichment results, it is possible to pass a data-frame giving information about the sets, through the <code>sets_info_df</code> argument. This information will be added to the enrichment results table. In that case, the <code>col_set</code> argument must also be specified: it takes the name of the column in the data-frame passed to <code>sets_info_df</code> that contains the sets ID or named as seen in the list. Here we will create this data-frame as follows:</p></li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">go_sets_info</span>,</span>
<span>  <span class="va">rnaseq_go_df</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">go_id</span>, <span class="va">go_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">go_sets_info</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 3</span></span>
<span><span class="co">#&gt;   go_id      go_domain          go_name                                         </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt;              &lt;chr&gt;                                           </span></span>
<span><span class="co">#&gt; 1 GO:0006468 Biological process protein phosphorylation                         </span></span>
<span><span class="co">#&gt; 2 GO:0007165 Biological process signal transduction                             </span></span>
<span><span class="co">#&gt; 3 GO:0031623 Biological process receptor internalization                        </span></span>
<span><span class="co">#&gt; 4 GO:0006813 Biological process potassium ion transport                         </span></span>
<span><span class="co">#&gt; 5 GO:0006355 Biological process regulation of DNA-templated transcription       </span></span>
<span><span class="co">#&gt; 6 GO:0006357 Biological process regulation of transcription by RNA polymerase II</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can run the enrichment analysis:</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_enrichment_results</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/evaluate_method_enrichment.html">evaluate_method_enrichment</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_output</span>,</span>
<span>    <span class="va">go_sets_filtered</span>,</span>
<span>    datasets <span class="op">=</span> <span class="st">"rnaseq"</span>,</span>
<span>    use_abs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    min_set_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    add_missing_features <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>    sets_info_df <span class="op">=</span> <span class="va">go_sets_info</span>,</span>
<span>    col_set <span class="op">=</span> <span class="st">"go_id"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function returns a table giving for each latent component (<code>latent_dimension</code>) and feature set (<code>set_id</code>) the enrichment statistics (<code>stat_mean</code>), p-value (<code>pvalue</code>), adjusted p-value (<code>adj_pvalue</code>) as well as the number of features in the set that are present in the integration results (<code>set_size</code>). The table is arranged by increasing adjusted p-value, such that the significant results are at the top of the table. It is important to note that the multiple testing correction applied to the p-values occurs independently for each latent dimension, but there is no multiple testing correction performed across the latent dimensions. This can easily be done by the user, and should be considered when reporting the results.</p>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">diablo_enrichment_results</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 8</span></span>
<span><span class="co">#&gt;   latent_dimension set_id stat_mean pvalue adj_pvalue set_size go_domain go_name</span></span>
<span><span class="co">#&gt;   &lt;fct&gt;            &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;  </span></span>
<span><span class="co">#&gt; 1 Component 3      GO:00…     1.23   0.114      0.608       24 Biologic… synapt…</span></span>
<span><span class="co">#&gt; 2 Component 3      GO:00…     0.993  0.165      0.608       24 Biologic… hetero…</span></span>
<span><span class="co">#&gt; 3 Component 3      GO:00…     0.969  0.168      0.608      100 Biologic… dephos…</span></span>
<span><span class="co">#&gt; 4 Component 3      GO:00…     0.984  0.169      0.608       20 Biologic… negati…</span></span>
<span><span class="co">#&gt; 5 Component 3      GO:00…     0.968  0.169      0.608       40 Biologic… positi…</span></span>
<span><span class="co">#&gt; 6 Component 3      GO:00…     0.957  0.170      0.608      138 Biologic… protei…</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In that case, unsurprisingly, neither of the latent components were enriched for any GO term. This makes sense since DIABLO performed a strong feature selection, so very few genes had non-null weights. We can repeat the enrichment with the MOFA integration results to show more interesting results. We will run the enrichment only for the first three factors for conciseness:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_enrichment_results</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/evaluate_method_enrichment.html">evaluate_method_enrichment</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_output</span>,</span>
<span>    <span class="va">go_sets_filtered</span>,</span>
<span>    datasets <span class="op">=</span> <span class="st">"rnaseq"</span>,</span>
<span>    latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>    use_abs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    min_set_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    add_missing_features <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>    sets_info_df <span class="op">=</span> <span class="va">go_sets_info</span>,</span>
<span>    col_set <span class="op">=</span> <span class="st">"go_id"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This time, there are a couple of small p-values (even though the adjusted p-values are all quite large):</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mofa_enrichment_results</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mofa_enrichment_results</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 8</span></span>
<span><span class="co">#&gt;   latent_dimension set_id     stat_mean   pvalue adj_pvalue set_size go_domain  </span></span>
<span><span class="co">#&gt;   &lt;fct&gt;            &lt;chr&gt;          &lt;dbl&gt;    &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;      </span></span>
<span><span class="co">#&gt; 1 Factor 3         GO:0031424      4.13 0.000275      0.548       20 Biological…</span></span>
<span><span class="co">#&gt; 2 Factor 3         GO:0045109      3.59 0.000713      0.711       25 Biological…</span></span>
<span><span class="co">#&gt; 3 Factor 3         GO:0030855      2.56 0.00648       0.767       56 Biological…</span></span>
<span><span class="co">#&gt; 4 Factor 3         GO:0031069      2.07 0.0246        0.767       25 Biological…</span></span>
<span><span class="co">#&gt; 5 Factor 3         GO:0030216      1.89 0.0320        0.767       49 Biological…</span></span>
<span><span class="co">#&gt; 6 Factor 3         GO:0098609      1.82 0.0350        0.767      120 Biological…</span></span>
<span><span class="co">#&gt; # ℹ 1 more variable: go_name &lt;chr&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see for example that MOFA factor 3 seems enriched in genes associated with the keratinization GO term (<code>GO:0031424</code>). For factor 1 (the factor separating the healthy and infected animals), the keratinization term also yields the smallest p-value, together with several GO terms related to inflammation (even though after correction for multiple testing these p-values are no longer significant):</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mofa_enrichment_results</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">latent_dimension</span> <span class="op">==</span> <span class="st">"Factor 1"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 8</span></span>
<span><span class="co">#&gt;   latent_dimension set_id     stat_mean  pvalue adj_pvalue set_size go_domain   </span></span>
<span><span class="co">#&gt;   &lt;fct&gt;            &lt;chr&gt;          &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;       </span></span>
<span><span class="co">#&gt; 1 Factor 1         GO:0031424      2.75 0.00461      0.786       20 Biological …</span></span>
<span><span class="co">#&gt; 2 Factor 1         GO:0045109      2.47 0.00867      0.786       25 Biological …</span></span>
<span><span class="co">#&gt; 3 Factor 1         GO:0010951      2.15 0.0208       0.786       25 Biological …</span></span>
<span><span class="co">#&gt; 4 Factor 1         GO:0006954      1.86 0.0317       0.786      165 Biological …</span></span>
<span><span class="co">#&gt; 5 Factor 1         GO:0050727      1.75 0.0424       0.786       56 Biological …</span></span>
<span><span class="co">#&gt; 6 Factor 1         GO:0007166      1.62 0.0534       0.786      138 Biological …</span></span>
<span><span class="co">#&gt; # ℹ 1 more variable: go_name &lt;chr&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="plotting-features-weight-for-feature-sets" class="level3" data-number="13.3.5"><h3 data-number="13.3.5" class="anchored" data-anchor-id="plotting-features-weight-for-feature-sets">
<span class="header-section-number">13.3.5</span> Plotting features weight for feature sets</h3>
<p>We can dig into the results of the enrichment by visualising the distribution of features weight or importance score between features that are in a specific features set vs features that are not in the set. For example, to continue with the MOFA example, let’s look at the importance score of genes associated with the GO term <code>GO:0031424</code>. We’ll only look at factors 1 and 3 for this example:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_set.html">plot_features_weight_set</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  <span class="va">go_sets_filtered</span><span class="op">[[</span><span class="st">"GO:0031424"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  set_name <span class="op">=</span> <span class="st">"GO:0031424 (keratinization)"</span>,</span>
<span>  features_metric <span class="op">=</span> <span class="st">"importance"</span>,</span>
<span>  datasets <span class="op">=</span> <span class="st">"rnaseq"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  point_alpha <span class="op">=</span> <span class="fl">0.2</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="evaluation_files/figure-html/plot-features-weight-set-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>In the resulting plot, we can see for each dataset (in this case, we’re only looking at the transcriptomics dataset since we have gene sets) and latent dimensions (rows of the plot) the distribution of features importance for genes in the set vs not in the set. We can see that the GO term investigated contains 20 genes, and in total, 994 genes have been assigned a weight with MOFA (because we have performed some pre-filtering on the transcriptomics dataset prior to running MOFA). Therefore, out of all the genes in the set, 9 of them have no weight (they did not pass the pre-filtering step). We can add these features that did not make it into the MOFA input data in the same way as we did for the enrichment function, namely by setting the <code>add_missing_features</code> argument to TRUE and passing the non-filtering <code>MultiDataSet</code> object through <code>mo_data</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_set.html">plot_features_weight_set</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  <span class="va">go_sets_filtered</span><span class="op">[[</span><span class="st">"GO:0031424"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  set_name <span class="op">=</span> <span class="st">"GO:0031424 (keratinization)"</span>,</span>
<span>  features_metric <span class="op">=</span> <span class="st">"importance"</span>,</span>
<span>  datasets <span class="op">=</span> <span class="st">"rnaseq"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  point_alpha <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  add_missing_features <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="evaluation_files/figure-html/plot-features-weight-set-with-missing-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>This is the distribution of values that were used for the enrichment test, but it’s not very useful to look at since the vast majority of the genes have a weight of zero.</p>
</section></section><section id="assessing-samples-clustering" class="level2" data-number="13.4"><h2 data-number="13.4" class="anchored" data-anchor-id="assessing-samples-clustering">
<span class="header-section-number">13.4</span> Assessing samples clustering</h2>
<p>Another aspect that we can evaluate is whether the integration method was able to separate different groups of samples, for example to identify differences between control and treatment samples. This is done by comparing the coordinates of the samples in the space spanned by the latent dimensions created (i.e.&nbsp;the samples score) to a samples label that is obtained from the samples metadata. We can use for that the silhouette score, which is a metric that quantifies how well-defined different clusters are in a given space, i.e.&nbsp;how well-grouped points from a same cluster (here samples group) are, and how separate they are from points from other clusters. First, a silhouette score (or silhouette width) is computed for each individual point (here sample). Then, these values are averaged over all points in a given cluster to yield a cluster silhouette score. The resulting values, both at the point- and cluster- level, can range from -1 to 1. Negative values indicate that the point is grouped with points from another cluster, while positive values indicate that the point is grouped with other points from its cluster. Values close to 0 show that the clustering is ambiguous, while scores close to 1 are indicative of well-defined clusters.</p>
<p>The <code><a href="https://rdrr.io/pkg/moiraine/man/compute_samples_silhouette.html">compute_samples_silhouette()</a></code> function relies on the <code><a href="https://rdrr.io/pkg/cluster/man/silhouette.html">cluster::silhouette()</a></code> function from the <code>cluster</code> package to compute silhouette score. It takes as input the integration results in a standardised format, as well as the <code>MultiDataSet</code> object (to access the samples metadata) and the name of the column in the samples metadata table that contains the samples grouping information. In addition, it is possible to specify the distance metric that should be used to compute distances between the samples (through the <code>distance_metric</code> argument). By default, the function uses euclidean distances, but other options are available.</p>
<p>Here, we will check how well DIABLO managed to separate the bruising groups:</p>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_silhouette</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/compute_samples_silhouette.html">compute_samples_silhouette</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_output</span>,</span>
<span>    <span class="va">mo_set_complete</span>,</span>
<span>    <span class="st">"status"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_silhouette</span><span class="op">)</span></span>
<span></span>
<span><span class="va">diablo_silhouette</span></span>
<span><span class="co">#&gt; $samples_silhouette</span></span>
<span><span class="co">#&gt; # A tibble: 135 × 4</span></span>
<span><span class="co">#&gt;    sample_id group   neighbour_group silhouette_width</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;fct&gt;   &lt;fct&gt;                      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 G1979     BRD     Control                    0.362</span></span>
<span><span class="co">#&gt;  2 G2500     BRD     Control                    0.267</span></span>
<span><span class="co">#&gt;  3 G3030     BRD     Control                    0.325</span></span>
<span><span class="co">#&gt;  4 G3068     BRD     Control                    0.373</span></span>
<span><span class="co">#&gt;  5 G3121     BRD     Control                    0.448</span></span>
<span><span class="co">#&gt;  6 G3315     Control BRD                        0.526</span></span>
<span><span class="co">#&gt;  7 G3473     Control BRD                        0.444</span></span>
<span><span class="co">#&gt;  8 G3474     Control BRD                        0.497</span></span>
<span><span class="co">#&gt;  9 G3550     BRD     Control                    0.411</span></span>
<span><span class="co">#&gt; 10 G3594     Control BRD                        0.542</span></span>
<span><span class="co">#&gt; # ℹ 125 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $groups_average_silhouette</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 2</span></span>
<span><span class="co">#&gt;   group   group_average_width</span></span>
<span><span class="co">#&gt;   &lt;fct&gt;             &lt;dbl[1d]&gt;</span></span>
<span><span class="co">#&gt; 1 BRD                   0.317</span></span>
<span><span class="co">#&gt; 2 Control               0.482</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function returns a list with two elements: a tibble of samples silhouette widths (<code>samples_silhouette</code>), and a tibble of groups average silhouette (<code>groups_average_silhouette</code>). In the first table, the <code>group</code> column indicates for each sample the group to which it is assigned (based on the samples metadata), and the <code>neighbour_group</code> column indicates the closest other group in the latent dimensions space (which is not very useful in this case since we only have two groups). The samples silhouette width is shown in the <code>silhouette_width</code> column. This is very useful to identify “outlier” samples that were not grouped as expected:</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">diablo_silhouette</span><span class="op">$</span><span class="va">samples_silhouette</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">silhouette_width</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 1 × 4</span></span>
<span><span class="co">#&gt;   sample_id group neighbour_group silhouette_width</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;     &lt;fct&gt; &lt;fct&gt;                      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 P4735     BRD   Control                   -0.118</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, there is one sample from an infected animal that clusters with the control samples in the latent dimensions space.</p>
<p>The average silhouette score for the two disease status groups are both positive, although not close to 1, indicating that DIABLO did a reasonable job of identifying differences between the two groups:</p>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">diablo_silhouette</span><span class="op">$</span><span class="va">groups_average_silhouette</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 2</span></span>
<span><span class="co">#&gt;   group   group_average_width</span></span>
<span><span class="co">#&gt;   &lt;fct&gt;             &lt;dbl[1d]&gt;</span></span>
<span><span class="co">#&gt; 1 BRD                   0.317</span></span>
<span><span class="co">#&gt; 2 Control               0.482</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is expected, since DIABLO is a supervised integration method which tries to maximise the separation between groups of samples.</p>
<p>We can repeat the operation for the MOFA results. By default, the <code><a href="https://rdrr.io/pkg/moiraine/man/compute_samples_silhouette.html">compute_samples_silhouette()</a></code> function uses all latent dimensions to compute distances between samples, but it is possible to specify which latent dimensions should be considered through the <code>latent_dimensions</code> argument. For example, we will see how the first three MOFA factors separate the disease status groups:</p>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_silhouette</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/compute_samples_silhouette.html">compute_samples_silhouette</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_output</span>,</span>
<span>    <span class="va">mo_set_complete</span>,</span>
<span>    <span class="st">"status"</span>,</span>
<span>    latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mofa_silhouette</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mofa_silhouette</span><span class="op">$</span><span class="va">groups_average_silhouette</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 2</span></span>
<span><span class="co">#&gt;   group   group_average_width</span></span>
<span><span class="co">#&gt;   &lt;fct&gt;             &lt;dbl[1d]&gt;</span></span>
<span><span class="co">#&gt; 1 BRD                   0.256</span></span>
<span><span class="co">#&gt; 2 Control               0.522</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Both groups’ silhouette scores are positive, although the BRD group silhouette is not very high. Also, more samples have a silhouette width below 0.</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mofa_silhouette</span><span class="op">$</span><span class="va">samples_silhouette</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">silhouette_width</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 11 × 4</span></span>
<span><span class="co">#&gt;    sample_id group neighbour_group silhouette_width</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;fct&gt; &lt;fct&gt;                      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 O3720     BRD   Control                  -0.432 </span></span>
<span><span class="co">#&gt;  2 O4492     BRD   Control                  -0.249 </span></span>
<span><span class="co">#&gt;  3 O4919     BRD   Control                  -0.442 </span></span>
<span><span class="co">#&gt;  4 O5003     BRD   Control                  -0.243 </span></span>
<span><span class="co">#&gt;  5 P4666     BRD   Control                  -0.0993</span></span>
<span><span class="co">#&gt;  6 P4735     BRD   Control                  -0.431 </span></span>
<span><span class="co">#&gt;  7 R50       BRD   Control                  -0.168 </span></span>
<span><span class="co">#&gt;  8 R5979     BRD   Control                  -0.0102</span></span>
<span><span class="co">#&gt;  9 R9887     BRD   Control                  -0.0682</span></span>
<span><span class="co">#&gt; 10 Y3059     BRD   Control                  -0.0110</span></span>
<span><span class="co">#&gt; 11 Y9745     BRD   Control                  -0.0340</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that with MOFA, since the different factors represent different trends within the data, it is possible that using more factors will yield smaller average silhouette groups. For example if we used all constructed factors:</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/compute_samples_silhouette.html">compute_samples_silhouette</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_output</span>,</span>
<span>    <span class="va">mo_set_complete</span>,</span>
<span>    <span class="st">"status"</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">groups_average_silhouette</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 2</span></span>
<span><span class="co">#&gt;   group   group_average_width</span></span>
<span><span class="co">#&gt;   &lt;fct&gt;             &lt;dbl[1d]&gt;</span></span>
<span><span class="co">#&gt; 1 BRD                  0.0671</span></span>
<span><span class="co">#&gt; 2 Control              0.272</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In such cases, it makes more sense to evaluate the ability of one or two latent dimensions to separate the samples group.</p>
</section><section id="recap-targets-list" class="level1" data-number="14"><h1 data-number="14">
<span class="header-section-number">14</span> Recap – targets list</h1>
<p>For convenience, here is the list of targets that we created in this section:</p>
<details><summary>
Targets list for evaluating the output of an integration method:
</summary><div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Evaluating DIABLO selected features against single-omics results</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_selected_vs_singleomics_table</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/evaluate_feature_selection_table.html">evaluate_feature_selection_table</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_output</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"snps"</span> <span class="op">=</span> <span class="st">"qtl_type"</span>,</span>
<span>        <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_signif"</span>,</span>
<span>        <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"de_signif"</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plotting DIABLO features weight against single-omics results</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_features_weight_vs_singleomics_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_covariate.html">plot_features_weight_covariate</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_output</span>, </span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>, </span>
<span>      covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"snps"</span> <span class="op">=</span> <span class="st">"qtl_type"</span>, </span>
<span>        <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_status"</span>, </span>
<span>        <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"de_status"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      remove_null_weight <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Genes GO annotation file </span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_go_terms_file</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span></span>
<span>      <span class="st">"extdata/transcriptomics_go_annotation.csv"</span>, </span>
<span>      package <span class="op">=</span> <span class="st">"moiraine"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    format <span class="op">=</span> <span class="st">"file"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Genes GO annotation data-frame </span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_go_df</span>,</span>
<span>    <span class="fu">read_csv</span><span class="op">(</span><span class="va">rnaseq_go_terms_file</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">go_domain</span> <span class="op">==</span> <span class="st">"Biological process"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## GO term sets</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">go_sets</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/make_feature_sets_from_df.html">make_feature_sets_from_df</a></span><span class="op">(</span></span>
<span>      <span class="va">rnaseq_go_df</span>,</span>
<span>      col_id <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>      col_set <span class="op">=</span> <span class="st">"go_id"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Filtering GO term sets against measured features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">go_sets_filtered</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/reduce_feature_sets_data.html">reduce_feature_sets_data</a></span><span class="op">(</span><span class="va">go_sets</span>, <span class="va">mo_set_complete</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Checking genes GO term sets against datasets</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">go_sets_check</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/check_feature_sets.html">check_feature_sets</a></span><span class="op">(</span></span>
<span>      <span class="va">go_sets_filtered</span>, </span>
<span>      <span class="va">mo_set_complete</span>,</span>
<span>      datasets <span class="op">=</span> <span class="st">"rnaseq"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Table of information about GO terms</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">go_sets_info</span>,</span>
<span>    <span class="va">rnaseq_go_df</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">go_id</span>, <span class="va">go_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## MOFA latent components enrichment analysis</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_enrichment_results</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/evaluate_method_enrichment.html">evaluate_method_enrichment</a></span><span class="op">(</span></span>
<span>      <span class="va">mofa_output</span>,</span>
<span>      <span class="va">go_sets_filtered</span>,</span>
<span>      datasets <span class="op">=</span> <span class="st">"rnaseq"</span>,</span>
<span>      latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>      use_abs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      min_set_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>      add_missing_features <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      sets_info_df <span class="op">=</span> <span class="va">go_sets_info</span>,</span>
<span>      col_set <span class="op">=</span> <span class="st">"go_id"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plotting features weight for GO term 'GO:0031424'</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_enrichment_go0031424_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_set.html">plot_features_weight_set</a></span><span class="op">(</span></span>
<span>      <span class="va">mofa_output</span>,</span>
<span>      <span class="va">go_sets_filtered</span><span class="op">[[</span><span class="st">"GO:0031424"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>      set_name <span class="op">=</span> <span class="st">"GO:0031424 (keratinization)"</span>,</span>
<span>      features_metric <span class="op">=</span> <span class="st">"importance"</span>,</span>
<span>      datasets <span class="op">=</span> <span class="st">"rnaseq"</span>,</span>
<span>      latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      point_alpha <span class="op">=</span> <span class="fl">0.2</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Assessing DIABLO samples clustering</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_silhouette</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/compute_samples_silhouette.html">compute_samples_silhouette</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_output</span>,</span>
<span>      <span class="va">mo_set_complete</span>,</span>
<span>      <span class="st">"status"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>

<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography" style="display: none">
<div id="ref-luo2009" class="csl-entry" role="doc-biblioentry">
Luo, Weijun, Michael S. Friedman, Kerby Shedden, Kurt D. Hankenson, and Peter J. Woolf. 2009. <span>“<span>GAGE</span>: Generally Applicable Gene Set Enrichment for Pathway Analysis.”</span> <em>BMC Bioinformatics</em> 10 (1): 161. <a href="https://doi.org/10.1186/1471-2105-10-161">https://doi.org/10.1186/1471-2105-10-161</a>.
</div>
<div id="ref-timmons2015" class="csl-entry" role="doc-biblioentry">
Timmons, James A., Krzysztof J. Szkop, and Iain J. Gallagher. 2015. <span>“Multiple Sources of Bias Confound Functional Enrichment Analysis of Global -Omics Data.”</span> <em>Genome Biology</em> 16 (1): 186. <a href="https://doi.org/10.1186/s13059-015-0761-7">https://doi.org/10.1186/s13059-015-0761-7</a>.
</div>
<div id="ref-zhao2023" class="csl-entry" role="doc-biblioentry">
Zhao, Kangmei, and Seung Yon Rhee. 2023. <span>“Interpreting Omics Data with Pathway Enrichment Analysis.”</span> <em>Trends in Genetics</em> 39 (4): 308–19. <a href="https://doi.org/10.1016/j.tig.2023.01.003">https://doi.org/10.1016/j.tig.2023.01.003</a>.
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./interpretation.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>