<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>The moiraine R package user manual - 11&nbsp; Interpreting the integration results</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./evaluation.html" rel="next">
<link href="./diablo.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The moiraine R package user manual</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/PlantandFoodResearch/moiraine" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="https://twitter.com/intent/tweet?url=%7Curl%7C" title="Twitter" class="sidebar-tool px-1"><i class="bi bi-twitter"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./example_dataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">The example dataset</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Data preparation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_import.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Importing data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inspecting_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Inspecting the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./modifying_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocessing.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data pre-processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prefiltering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Dataset pre-filtering</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Unsupervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Integration with sPLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./so2pls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Integration with sO2PLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mofa.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Integration with MOFA</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Supervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./diablo.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Integration with DIABLO</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Results interpretation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./interpretation.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evaluation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./comparison.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Comparing the integration results</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#sec-interpretation-dim-reduction" id="toc-sec-interpretation-dim-reduction" class="nav-link active" data-scroll-target="#sec-interpretation-dim-reduction"><span class="toc-section-number">11.1</span>  Dimension reduction methods</a></li>
  <li>
<a href="#sec-interpretation-standard-output" id="toc-sec-interpretation-standard-output" class="nav-link" data-scroll-target="#sec-interpretation-standard-output"><span class="toc-section-number">11.2</span>  Generating a standardised output</a>
  <ul class="collapse">
<li><a href="#get_output-function" id="toc-get_output-function" class="nav-link" data-scroll-target="#get_output-function"><span class="toc-section-number">11.2.1</span>  <code>get_output</code> function</a></li>
  <li><a href="#averaging-latent-dimensions-over-datasets" id="toc-averaging-latent-dimensions-over-datasets" class="nav-link" data-scroll-target="#averaging-latent-dimensions-over-datasets"><span class="toc-section-number">11.2.2</span>  Averaging latent dimensions over datasets</a></li>
  </ul>
</li>
  <li><a href="#percentage-of-variance-explained" id="toc-percentage-of-variance-explained" class="nav-link" data-scroll-target="#percentage-of-variance-explained"><span class="toc-section-number">11.3</span>  Percentage of variance explained</a></li>
  <li>
<a href="#sec-interpretation-samples-scores" id="toc-sec-interpretation-samples-scores" class="nav-link" data-scroll-target="#sec-interpretation-samples-scores"><span class="toc-section-number">11.4</span>  Samples plot</a>
  <ul class="collapse">
<li><a href="#sec-interpretation-samples-ggpairs" id="toc-sec-interpretation-samples-ggpairs" class="nav-link" data-scroll-target="#sec-interpretation-samples-ggpairs"><span class="toc-section-number">11.4.1</span>  Matrix of scatter plots</a></li>
  <li><a href="#sec-interpretation-samples-pair" id="toc-sec-interpretation-samples-pair" class="nav-link" data-scroll-target="#sec-interpretation-samples-pair"><span class="toc-section-number">11.4.2</span>  Scatter plot for pair of latent dimensions</a></li>
  <li><a href="#sec-interpretation-samples-covariate" id="toc-sec-interpretation-samples-covariate" class="nav-link" data-scroll-target="#sec-interpretation-samples-covariate"><span class="toc-section-number">11.4.3</span>  Samples score vs covariate</a></li>
  </ul>
</li>
  <li>
<a href="#feature-plots" id="toc-feature-plots" class="nav-link" data-scroll-target="#feature-plots"><span class="toc-section-number">11.5</span>  Feature plots</a>
  <ul class="collapse">
<li><a href="#sec-interpretation-features-distr" id="toc-sec-interpretation-features-distr" class="nav-link" data-scroll-target="#sec-interpretation-features-distr"><span class="toc-section-number">11.5.1</span>  Features weight distribution</a></li>
  <li><a href="#sec-interpretation-features-pair" id="toc-sec-interpretation-features-pair" class="nav-link" data-scroll-target="#sec-interpretation-features-pair"><span class="toc-section-number">11.5.2</span>  Comparing features importance between latent components</a></li>
  <li><a href="#sec-interpretation-features-top-plot" id="toc-sec-interpretation-features-top-plot" class="nav-link" data-scroll-target="#sec-interpretation-features-top-plot"><span class="toc-section-number">11.5.3</span>  Plotting top features weight</a></li>
  <li><a href="#sec-interpretation-features-top-table" id="toc-sec-interpretation-features-top-table" class="nav-link" data-scroll-target="#sec-interpretation-features-top-table"><span class="toc-section-number">11.5.4</span>  Extracting top features</a></li>
  <li><a href="#sec-interpretation-features-selected" id="toc-sec-interpretation-features-selected" class="nav-link" data-scroll-target="#sec-interpretation-features-selected"><span class="toc-section-number">11.5.5</span>  Extracting the selected features</a></li>
  <li><a href="#sec-interpretation-features-meas" id="toc-sec-interpretation-features-meas" class="nav-link" data-scroll-target="#sec-interpretation-features-meas"><span class="toc-section-number">11.5.6</span>  Features measurements</a></li>
  </ul>
</li>
  <li><a href="#recap-targets-list" id="toc-recap-targets-list" class="nav-link" data-scroll-target="#recap-targets-list"><span class="toc-section-number">11.6</span>  Recap â€“ targets list</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-interpretation" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/targets/">targets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PlantandFoodResearch/moiraine">moiraine</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For working with lists</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For data-frames manipulation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For colour palettes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize">circlize</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For manipulating patchworks of plots</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the previous chapters, we have applied different integration tools to our example multi-omics dataset. In each of these chapters, we have shown some method-specific visualisations that could be used to interpret the integration results. With the <code>moiraine</code> package, it is possible to use method-agnostic functions to inspect and interpret these results. These functionalities make use of the metadata in the <code>MultiDataSet</code> object to enrich plots and other summary outputs with information about the samples and the omics features, and allow users to obtain consistent visualisations across the integration methods used.</p>
<details><summary><code>_targets.R</code> script
</summary><div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(targets)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tarchetypes)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(moiraine)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">tar_option_set</span>(</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">packages =</span> <span class="fu">c</span>(</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"moiraine"</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MOFA2"</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"mixOmics"</span>,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"readr"</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"tibble"</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"tidyr"</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"dplyr"</span>,</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ggplot2"</span>,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">"patchwork"</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="fu">list</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Importing data ---------------------------------------------------------------</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Data import using a target factory</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_dataset_csv_factory</span>(</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">c</span>(</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/genomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/transcriptomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/metabolomics_dataset.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>)</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="fu">c</span>(<span class="st">"marker"</span>, <span class="st">"gene_id"</span>, <span class="st">"sample_id"</span>),</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">features_as_rowss =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>),</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="fu">c</span>(<span class="st">"geno"</span>, <span class="st">"transcripto"</span>, <span class="st">"metabo"</span>)</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Genomics features metadata file</span></span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>    fmetadata_file_geno,</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system.file</span>(<span class="st">"extdata/genomics_features_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"file"</span></span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Genomics features metadata import</span></span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>    fmetadata_geno,</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">import_fmetadata_csv</span>(</span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>      fmetadata_file_geno,</span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_id =</span> <span class="st">"marker"</span>,</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>      <span class="at">col_types =</span> <span class="fu">c</span>(<span class="st">"chromosome"</span> <span class="ot">=</span> <span class="st">"c"</span>)</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Metabolomics features metadata import</span></span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_fmetadata_csv_factory</span>(</span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">c</span>(</span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a>      <span class="fu">system.file</span>(<span class="st">"extdata/metabolomics_features_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>)</span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="fu">c</span>(<span class="st">"feature_id"</span>),</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="fu">c</span>(<span class="st">"metabo"</span>)</span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-61"><a href="#cb2-61" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Transcriptomics features metadata import</span></span>
<span id="cb2-62"><a href="#cb2-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_fmetadata_gff_factory</span>(</span>
<span id="cb2-63"><a href="#cb2-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">system.file</span>(<span class="st">"extdata/bos_taurus_gene_model.gff3"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-64"><a href="#cb2-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">feature_types =</span> <span class="st">"genes"</span>,</span>
<span id="cb2-65"><a href="#cb2-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">add_fieldss =</span> <span class="fu">c</span>(<span class="st">"Name"</span>, <span class="st">"description"</span>),</span>
<span id="cb2-66"><a href="#cb2-66" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="st">"transcripto"</span></span>
<span id="cb2-67"><a href="#cb2-67" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-68"><a href="#cb2-68" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-69"><a href="#cb2-69" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Samples metadata import</span></span>
<span id="cb2-70"><a href="#cb2-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">import_smetadata_csv_factory</span>(</span>
<span id="cb2-71"><a href="#cb2-71" aria-hidden="true" tabindex="-1"></a>    <span class="at">files =</span> <span class="fu">system.file</span>(<span class="st">"extdata/samples_info.csv"</span>, <span class="at">package =</span> <span class="st">"moiraine"</span>),</span>
<span id="cb2-72"><a href="#cb2-72" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_ids =</span> <span class="st">"animal_id"</span>,</span>
<span id="cb2-73"><a href="#cb2-73" aria-hidden="true" tabindex="-1"></a>    <span class="at">target_name_suffixes =</span> <span class="st">"all"</span></span>
<span id="cb2-74"><a href="#cb2-74" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating omics sets for each dataset</span></span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a>  <span class="fu">create_omics_set_factory</span>(</span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a>    <span class="at">datasets =</span> <span class="fu">c</span>(data_geno, data_transcripto, data_metabo),</span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a>    <span class="at">omics_types =</span> <span class="fu">c</span>(<span class="st">"genomics"</span>, <span class="st">"transcriptomics"</span>, <span class="st">"metabolomics"</span>),</span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a>    <span class="at">features_metadatas =</span> <span class="fu">c</span>(fmetadata_geno, fmetadata_transcripto, fmetadata_metabo),</span>
<span id="cb2-81"><a href="#cb2-81" aria-hidden="true" tabindex="-1"></a>    <span class="at">samples_metadatas =</span> <span class="fu">c</span>(smetadata_all, smetadata_all, smetadata_all)</span>
<span id="cb2-82"><a href="#cb2-82" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-83"><a href="#cb2-83" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-84"><a href="#cb2-84" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating the MultiDataSet object</span></span>
<span id="cb2-85"><a href="#cb2-85" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-86"><a href="#cb2-86" aria-hidden="true" tabindex="-1"></a>    mo_set,</span>
<span id="cb2-87"><a href="#cb2-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">create_multiomics_set</span>(</span>
<span id="cb2-88"><a href="#cb2-88" aria-hidden="true" tabindex="-1"></a>      <span class="fu">list</span>(set_geno,</span>
<span id="cb2-89"><a href="#cb2-89" aria-hidden="true" tabindex="-1"></a>           set_transcripto,</span>
<span id="cb2-90"><a href="#cb2-90" aria-hidden="true" tabindex="-1"></a>           set_metabo)</span>
<span id="cb2-91"><a href="#cb2-91" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-92"><a href="#cb2-92" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-93"><a href="#cb2-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-94"><a href="#cb2-94" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspecting the `MultiDataSet` object -----------------------------------------</span></span>
<span id="cb2-95"><a href="#cb2-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-96"><a href="#cb2-96" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating a density plot for each dataset</span></span>
<span id="cb2-97"><a href="#cb2-97" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-98"><a href="#cb2-98" aria-hidden="true" tabindex="-1"></a>    density_plots,</span>
<span id="cb2-99"><a href="#cb2-99" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_density_data</span>(</span>
<span id="cb2-100"><a href="#cb2-100" aria-hidden="true" tabindex="-1"></a>      mo_set,</span>
<span id="cb2-101"><a href="#cb2-101" aria-hidden="true" tabindex="-1"></a>      <span class="at">combined =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-102"><a href="#cb2-102" aria-hidden="true" tabindex="-1"></a>      <span class="at">scales =</span> <span class="st">"free"</span></span>
<span id="cb2-103"><a href="#cb2-103" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-104"><a href="#cb2-104" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-105"><a href="#cb2-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-106"><a href="#cb2-106" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the relationship between features mean and standard deviation</span></span>
<span id="cb2-107"><a href="#cb2-107" aria-hidden="true" tabindex="-1"></a>  <span class="do">## for each dataset</span></span>
<span id="cb2-108"><a href="#cb2-108" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-109"><a href="#cb2-109" aria-hidden="true" tabindex="-1"></a>    mean_sd_plots,</span>
<span id="cb2-110"><a href="#cb2-110" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_meansd_data</span>(mo_set)</span>
<span id="cb2-111"><a href="#cb2-111" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-112"><a href="#cb2-112" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-113"><a href="#cb2-113" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Assessing missing values</span></span>
<span id="cb2-114"><a href="#cb2-114" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-115"><a href="#cb2-115" aria-hidden="true" tabindex="-1"></a>    n_missing_values,</span>
<span id="cb2-116"><a href="#cb2-116" aria-hidden="true" tabindex="-1"></a>    <span class="fu">check_missing_values</span>(mo_set)</span>
<span id="cb2-117"><a href="#cb2-117" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-118"><a href="#cb2-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-119"><a href="#cb2-119" aria-hidden="true" tabindex="-1"></a><span class="co"># Modifying the `MultiDataSet` object ------------------------------------------</span></span>
<span id="cb2-120"><a href="#cb2-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-121"><a href="#cb2-121" aria-hidden="true" tabindex="-1"></a>  <span class="do">## RNAseq differential expression results file</span></span>
<span id="cb2-122"><a href="#cb2-122" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-123"><a href="#cb2-123" aria-hidden="true" tabindex="-1"></a>    rnaseq_de_res_file,</span>
<span id="cb2-124"><a href="#cb2-124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">system.file</span>(</span>
<span id="cb2-125"><a href="#cb2-125" aria-hidden="true" tabindex="-1"></a>      <span class="st">"extdata/transcriptomics_de_results.csv"</span>,</span>
<span id="cb2-126"><a href="#cb2-126" aria-hidden="true" tabindex="-1"></a>      <span class="at">package =</span> <span class="st">"moiraine"</span></span>
<span id="cb2-127"><a href="#cb2-127" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb2-128"><a href="#cb2-128" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"file"</span></span>
<span id="cb2-129"><a href="#cb2-129" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-130"><a href="#cb2-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-131"><a href="#cb2-131" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Reading the RNAseq differential expression results</span></span>
<span id="cb2-132"><a href="#cb2-132" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-133"><a href="#cb2-133" aria-hidden="true" tabindex="-1"></a>    rnaseq_de_res_df,</span>
<span id="cb2-134"><a href="#cb2-134" aria-hidden="true" tabindex="-1"></a>    <span class="fu">read_csv</span>(rnaseq_de_res_file) <span class="sc">|&gt;</span></span>
<span id="cb2-135"><a href="#cb2-135" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rename</span>(<span class="at">feature_id =</span> gene_id) <span class="sc">|&gt;</span></span>
<span id="cb2-136"><a href="#cb2-136" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">dataset =</span> <span class="st">"rnaseq"</span>)</span>
<span id="cb2-137"><a href="#cb2-137" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-138"><a href="#cb2-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-139"><a href="#cb2-139" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Adding the differential expression results to the MultiDataSet object</span></span>
<span id="cb2-140"><a href="#cb2-140" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-141"><a href="#cb2-141" aria-hidden="true" tabindex="-1"></a>    mo_set_de,</span>
<span id="cb2-142"><a href="#cb2-142" aria-hidden="true" tabindex="-1"></a>    <span class="fu">add_features_metadata</span>(mo_set, rnaseq_de_res_df)</span>
<span id="cb2-143"><a href="#cb2-143" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-144"><a href="#cb2-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-145"><a href="#cb2-145" aria-hidden="true" tabindex="-1"></a><span class="co"># Data pre-processing ----------------------------------------------------------</span></span>
<span id="cb2-146"><a href="#cb2-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-147"><a href="#cb2-147" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Applying transformations to the datasets</span></span>
<span id="cb2-148"><a href="#cb2-148" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transformation_datasets_factory</span>(</span>
<span id="cb2-149"><a href="#cb2-149" aria-hidden="true" tabindex="-1"></a>    mo_set_de,</span>
<span id="cb2-150"><a href="#cb2-150" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"vst-deseq2"</span>,</span>
<span id="cb2-151"><a href="#cb2-151" aria-hidden="true" tabindex="-1"></a>      <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"logx"</span>),</span>
<span id="cb2-152"><a href="#cb2-152" aria-hidden="true" tabindex="-1"></a>    <span class="at">log_bases =</span> <span class="dv">2</span>,</span>
<span id="cb2-153"><a href="#cb2-153" aria-hidden="true" tabindex="-1"></a>    <span class="at">pre_log_functions =</span> zero_to_half_min,</span>
<span id="cb2-154"><a href="#cb2-154" aria-hidden="true" tabindex="-1"></a>    <span class="at">transformed_data_name =</span> <span class="st">"mo_set_transformed"</span></span>
<span id="cb2-155"><a href="#cb2-155" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-156"><a href="#cb2-156" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-157"><a href="#cb2-157" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Density plot for each transformed dataset</span></span>
<span id="cb2-158"><a href="#cb2-158" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-159"><a href="#cb2-159" aria-hidden="true" tabindex="-1"></a>    density_plots_transformed,</span>
<span id="cb2-160"><a href="#cb2-160" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_density_data</span>(</span>
<span id="cb2-161"><a href="#cb2-161" aria-hidden="true" tabindex="-1"></a>      mo_set_transformed,</span>
<span id="cb2-162"><a href="#cb2-162" aria-hidden="true" tabindex="-1"></a>      <span class="at">combined =</span> <span class="cn">FALSE</span>,</span>
<span id="cb2-163"><a href="#cb2-163" aria-hidden="true" tabindex="-1"></a>      <span class="at">scales =</span> <span class="st">"free"</span></span>
<span id="cb2-164"><a href="#cb2-164" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-165"><a href="#cb2-165" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-166"><a href="#cb2-166" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-167"><a href="#cb2-167" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the mean-SD trend for transformed each dataset</span></span>
<span id="cb2-168"><a href="#cb2-168" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-169"><a href="#cb2-169" aria-hidden="true" tabindex="-1"></a>    mean_sd_plots_transformed,</span>
<span id="cb2-170"><a href="#cb2-170" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_meansd_data</span>(mo_set_transformed)</span>
<span id="cb2-171"><a href="#cb2-171" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-172"><a href="#cb2-172" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-173"><a href="#cb2-173" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Summary table of the transformations applied</span></span>
<span id="cb2-174"><a href="#cb2-174" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-175"><a href="#cb2-175" aria-hidden="true" tabindex="-1"></a>    transformation_summary,</span>
<span id="cb2-176"><a href="#cb2-176" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_table_transformations</span>(transformations_runs_list)</span>
<span id="cb2-177"><a href="#cb2-177" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-178"><a href="#cb2-178" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-179"><a href="#cb2-179" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Running a PCA on each dataset</span></span>
<span id="cb2-180"><a href="#cb2-180" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pca_complete_data_factory</span>(</span>
<span id="cb2-181"><a href="#cb2-181" aria-hidden="true" tabindex="-1"></a>    mo_set_transformed,</span>
<span id="cb2-182"><a href="#cb2-182" aria-hidden="true" tabindex="-1"></a>    <span class="at">complete_data_name =</span> <span class="st">"mo_set_complete"</span></span>
<span id="cb2-183"><a href="#cb2-183" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-184"><a href="#cb2-184" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-185"><a href="#cb2-185" aria-hidden="true" tabindex="-1"></a>  <span class="do">## PCA screeplots</span></span>
<span id="cb2-186"><a href="#cb2-186" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-187"><a href="#cb2-187" aria-hidden="true" tabindex="-1"></a>    pca_screeplots,</span>
<span id="cb2-188"><a href="#cb2-188" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_screeplot_pca</span>(pca_runs_list)</span>
<span id="cb2-189"><a href="#cb2-189" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-190"><a href="#cb2-190" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-191"><a href="#cb2-191" aria-hidden="true" tabindex="-1"></a>  <span class="do">## PCA sample plots</span></span>
<span id="cb2-192"><a href="#cb2-192" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-193"><a href="#cb2-193" aria-hidden="true" tabindex="-1"></a>    pca_sample_plots,</span>
<span id="cb2-194"><a href="#cb2-194" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_samples_coordinates_pca</span>(</span>
<span id="cb2-195"><a href="#cb2-195" aria-hidden="true" tabindex="-1"></a>      pca_runs_list,</span>
<span id="cb2-196"><a href="#cb2-196" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="st">"snps"</span>,</span>
<span id="cb2-197"><a href="#cb2-197" aria-hidden="true" tabindex="-1"></a>      <span class="at">pcs =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb2-198"><a href="#cb2-198" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-199"><a href="#cb2-199" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span id="cb2-200"><a href="#cb2-200" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-201"><a href="#cb2-201" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-202"><a href="#cb2-202" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-203"><a href="#cb2-203" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-204"><a href="#cb2-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-205"><a href="#cb2-205" aria-hidden="true" tabindex="-1"></a><span class="co"># Dataset pre-filtering --------------------------------------------------------</span></span>
<span id="cb2-206"><a href="#cb2-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-207"><a href="#cb2-207" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Unsupervised feature selection based on MAD score</span></span>
<span id="cb2-208"><a href="#cb2-208" aria-hidden="true" tabindex="-1"></a>  <span class="fu">feature_preselection_mad_factory</span>(</span>
<span id="cb2-209"><a href="#cb2-209" aria-hidden="true" tabindex="-1"></a>    mo_set_complete,</span>
<span id="cb2-210"><a href="#cb2-210" aria-hidden="true" tabindex="-1"></a>    <span class="at">to_keep_ns =</span> <span class="fu">c</span>(<span class="st">"snps"</span> <span class="ot">=</span> <span class="dv">1000</span>, <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="dv">1000</span>),</span>
<span id="cb2-211"><a href="#cb2-211" aria-hidden="true" tabindex="-1"></a>    <span class="at">with_ties =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-212"><a href="#cb2-212" aria-hidden="true" tabindex="-1"></a>    <span class="at">filtered_set_target_name =</span> <span class="st">"mo_presel_unsupervised"</span></span>
<span id="cb2-213"><a href="#cb2-213" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-214"><a href="#cb2-214" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-215"><a href="#cb2-215" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Diagnostic plot for MAD-based feature selection</span></span>
<span id="cb2-216"><a href="#cb2-216" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-217"><a href="#cb2-217" aria-hidden="true" tabindex="-1"></a>    preselection_mad_plot,</span>
<span id="cb2-218"><a href="#cb2-218" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_feature_preselection_mad</span>(individual_mad_values)</span>
<span id="cb2-219"><a href="#cb2-219" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-220"><a href="#cb2-220" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-221"><a href="#cb2-221" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Supervised feature selection based on bruising groups</span></span>
<span id="cb2-222"><a href="#cb2-222" aria-hidden="true" tabindex="-1"></a>  <span class="fu">feature_preselection_splsda_factory</span>(</span>
<span id="cb2-223"><a href="#cb2-223" aria-hidden="true" tabindex="-1"></a>    mo_set_complete,</span>
<span id="cb2-224"><a href="#cb2-224" aria-hidden="true" tabindex="-1"></a>    <span class="at">group =</span> <span class="st">"status"</span>,</span>
<span id="cb2-225"><a href="#cb2-225" aria-hidden="true" tabindex="-1"></a>    <span class="at">to_keep_ns =</span> <span class="fu">c</span>(<span class="st">"snps"</span> <span class="ot">=</span> <span class="dv">1000</span>, <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="dv">1000</span>),</span>
<span id="cb2-226"><a href="#cb2-226" aria-hidden="true" tabindex="-1"></a>    <span class="at">filtered_set_target_name =</span> <span class="st">"mo_presel_supervised"</span></span>
<span id="cb2-227"><a href="#cb2-227" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-228"><a href="#cb2-228" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-229"><a href="#cb2-229" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Diagnostic plot for sPLS-DA based feature selection</span></span>
<span id="cb2-230"><a href="#cb2-230" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-231"><a href="#cb2-231" aria-hidden="true" tabindex="-1"></a>    preselection_splsda_plot,</span>
<span id="cb2-232"><a href="#cb2-232" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_feature_preselection_splsda</span>(individual_splsda_perf)</span>
<span id="cb2-233"><a href="#cb2-233" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-234"><a href="#cb2-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-235"><a href="#cb2-235" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with sPLS --------------------------------------------------------</span></span>
<span id="cb2-236"><a href="#cb2-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-237"><a href="#cb2-237" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating sPLS input</span></span>
<span id="cb2-238"><a href="#cb2-238" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-239"><a href="#cb2-239" aria-hidden="true" tabindex="-1"></a>    spls_input,</span>
<span id="cb2-240"><a href="#cb2-240" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_spls</span>(</span>
<span id="cb2-241"><a href="#cb2-241" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-242"><a href="#cb2-242" aria-hidden="true" tabindex="-1"></a>      <span class="at">mode =</span> <span class="st">"canonical"</span>,</span>
<span id="cb2-243"><a href="#cb2-243" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="fu">c</span>(<span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span>)</span>
<span id="cb2-244"><a href="#cb2-244" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-245"><a href="#cb2-245" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-246"><a href="#cb2-246" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-247"><a href="#cb2-247" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Initial PLS run with no feature selection and large number of components</span></span>
<span id="cb2-248"><a href="#cb2-248" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-249"><a href="#cb2-249" aria-hidden="true" tabindex="-1"></a>    spls_novarsel,</span>
<span id="cb2-250"><a href="#cb2-250" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_run</span>(</span>
<span id="cb2-251"><a href="#cb2-251" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-252"><a href="#cb2-252" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> <span class="dv">4</span></span>
<span id="cb2-253"><a href="#cb2-253" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-254"><a href="#cb2-254" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-255"><a href="#cb2-255" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-256"><a href="#cb2-256" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of components</span></span>
<span id="cb2-257"><a href="#cb2-257" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-258"><a href="#cb2-258" aria-hidden="true" tabindex="-1"></a>    spls_perf_res,</span>
<span id="cb2-259"><a href="#cb2-259" aria-hidden="true" tabindex="-1"></a>    mixOmics<span class="sc">::</span><span class="fu">perf</span>(</span>
<span id="cb2-260"><a href="#cb2-260" aria-hidden="true" tabindex="-1"></a>      spls_novarsel,</span>
<span id="cb2-261"><a href="#cb2-261" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-262"><a href="#cb2-262" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-263"><a href="#cb2-263" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">10</span>,</span>
<span id="cb2-264"><a href="#cb2-264" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-265"><a href="#cb2-265" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-266"><a href="#cb2-266" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-267"><a href="#cb2-267" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-268"><a href="#cb2-268" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of components)</span></span>
<span id="cb2-269"><a href="#cb2-269" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Can try criterion = 'Q2.total', 'cor.tpred', 'cor.upred', 'RSS.tpred',</span></span>
<span id="cb2-270"><a href="#cb2-270" aria-hidden="true" tabindex="-1"></a>  <span class="do">## 'RSS.upred' (but avoid 'RSS' and 'PRESS')</span></span>
<span id="cb2-271"><a href="#cb2-271" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-272"><a href="#cb2-272" aria-hidden="true" tabindex="-1"></a>    spls_perf_plot,</span>
<span id="cb2-273"><a href="#cb2-273" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(spls_perf_res, <span class="at">criterion =</span> <span class="st">"Q2.total"</span>)</span>
<span id="cb2-274"><a href="#cb2-274" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-275"><a href="#cb2-275" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-276"><a href="#cb2-276" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Selected value for ncomp</span></span>
<span id="cb2-277"><a href="#cb2-277" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-278"><a href="#cb2-278" aria-hidden="true" tabindex="-1"></a>    spls_optim_ncomp,</span>
<span id="cb2-279"><a href="#cb2-279" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_get_optim_ncomp</span>(spls_perf_res, <span class="at">min_ncomp =</span> <span class="dv">2</span>)</span>
<span id="cb2-280"><a href="#cb2-280" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-281"><a href="#cb2-281" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-282"><a href="#cb2-282" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of features to retain</span></span>
<span id="cb2-283"><a href="#cb2-283" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-284"><a href="#cb2-284" aria-hidden="true" tabindex="-1"></a>    spls_tune_res,</span>
<span id="cb2-285"><a href="#cb2-285" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_tune</span>(</span>
<span id="cb2-286"><a href="#cb2-286" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-287"><a href="#cb2-287" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> spls_optim_ncomp,</span>
<span id="cb2-288"><a href="#cb2-288" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepX =</span> <span class="fu">seq</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">10</span>),</span>
<span id="cb2-289"><a href="#cb2-289" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepY =</span> <span class="fu">seq</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">10</span>),</span>
<span id="cb2-290"><a href="#cb2-290" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-291"><a href="#cb2-291" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-292"><a href="#cb2-292" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">5</span>,</span>
<span id="cb2-293"><a href="#cb2-293" aria-hidden="true" tabindex="-1"></a>      <span class="at">measure =</span> <span class="st">"cor"</span>,</span>
<span id="cb2-294"><a href="#cb2-294" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-295"><a href="#cb2-295" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-296"><a href="#cb2-296" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-297"><a href="#cb2-297" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-298"><a href="#cb2-298" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of features)</span></span>
<span id="cb2-299"><a href="#cb2-299" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-300"><a href="#cb2-300" aria-hidden="true" tabindex="-1"></a>    spls_tune_plot,</span>
<span id="cb2-301"><a href="#cb2-301" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_plot_tune</span>(spls_tune_res)</span>
<span id="cb2-302"><a href="#cb2-302" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-303"><a href="#cb2-303" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-304"><a href="#cb2-304" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Final sPLS run</span></span>
<span id="cb2-305"><a href="#cb2-305" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-306"><a href="#cb2-306" aria-hidden="true" tabindex="-1"></a>    spls_final_run,</span>
<span id="cb2-307"><a href="#cb2-307" aria-hidden="true" tabindex="-1"></a>    <span class="fu">spls_run</span>(</span>
<span id="cb2-308"><a href="#cb2-308" aria-hidden="true" tabindex="-1"></a>      spls_input,</span>
<span id="cb2-309"><a href="#cb2-309" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> spls_optim_ncomp,</span>
<span id="cb2-310"><a href="#cb2-310" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepX =</span> spls_tune_res<span class="sc">$</span>choice.keepX,</span>
<span id="cb2-311"><a href="#cb2-311" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepY =</span> spls_tune_res<span class="sc">$</span>choice.keepY</span>
<span id="cb2-312"><a href="#cb2-312" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-313"><a href="#cb2-313" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-314"><a href="#cb2-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-315"><a href="#cb2-315" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with sO2PLS ------------------------------------------------------</span></span>
<span id="cb2-316"><a href="#cb2-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-317"><a href="#cb2-317" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating sO2PLS input</span></span>
<span id="cb2-318"><a href="#cb2-318" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-319"><a href="#cb2-319" aria-hidden="true" tabindex="-1"></a>    omicspls_input,</span>
<span id="cb2-320"><a href="#cb2-320" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_omicspls</span>(</span>
<span id="cb2-321"><a href="#cb2-321" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-322"><a href="#cb2-322" aria-hidden="true" tabindex="-1"></a>      <span class="at">datasets =</span> <span class="fu">c</span>(<span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span>)</span>
<span id="cb2-323"><a href="#cb2-323" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-324"><a href="#cb2-324" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-325"><a href="#cb2-325" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-326"><a href="#cb2-326" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Adjusted cross-validation for number of components</span></span>
<span id="cb2-327"><a href="#cb2-327" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-328"><a href="#cb2-328" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj,</span>
<span id="cb2-329"><a href="#cb2-329" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_o2m_adjR2</span>(</span>
<span id="cb2-330"><a href="#cb2-330" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-331"><a href="#cb2-331" aria-hidden="true" tabindex="-1"></a>      <span class="at">a =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,</span>
<span id="cb2-332"><a href="#cb2-332" aria-hidden="true" tabindex="-1"></a>      <span class="at">ax =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="dv">2</span>),</span>
<span id="cb2-333"><a href="#cb2-333" aria-hidden="true" tabindex="-1"></a>      <span class="at">ay =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="at">by =</span> <span class="dv">2</span>),</span>
<span id="cb2-334"><a href="#cb2-334" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-335"><a href="#cb2-335" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_cores =</span> <span class="dv">6</span>,</span>
<span id="cb2-336"><a href="#cb2-336" aria-hidden="true" tabindex="-1"></a>      <span class="at">seed =</span> <span class="dv">127</span></span>
<span id="cb2-337"><a href="#cb2-337" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-338"><a href="#cb2-338" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-339"><a href="#cb2-339" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-340"><a href="#cb2-340" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj_res,</span>
<span id="cb2-341"><a href="#cb2-341" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_ncomp_adj</span>(so2pls_cv_adj)</span>
<span id="cb2-342"><a href="#cb2-342" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-343"><a href="#cb2-343" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-344"><a href="#cb2-344" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting adjusted cross-validation results</span></span>
<span id="cb2-345"><a href="#cb2-345" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-346"><a href="#cb2-346" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_adj_plot,</span>
<span id="cb2-347"><a href="#cb2-347" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv_adj</span>(so2pls_cv_adj)</span>
<span id="cb2-348"><a href="#cb2-348" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-349"><a href="#cb2-349" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-350"><a href="#cb2-350" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Standard cross-validation for number of components</span></span>
<span id="cb2-351"><a href="#cb2-351" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-352"><a href="#cb2-352" aria-hidden="true" tabindex="-1"></a>    so2pls_cv,</span>
<span id="cb2-353"><a href="#cb2-353" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_o2m</span>(</span>
<span id="cb2-354"><a href="#cb2-354" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-355"><a href="#cb2-355" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_adj,</span>
<span id="cb2-356"><a href="#cb2-356" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-357"><a href="#cb2-357" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_cores =</span> <span class="dv">6</span>,</span>
<span id="cb2-358"><a href="#cb2-358" aria-hidden="true" tabindex="-1"></a>      <span class="at">seed =</span> <span class="dv">356</span></span>
<span id="cb2-359"><a href="#cb2-359" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-360"><a href="#cb2-360" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-361"><a href="#cb2-361" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-362"><a href="#cb2-362" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_res,</span>
<span id="cb2-363"><a href="#cb2-363" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_ncomp</span>(so2pls_cv)</span>
<span id="cb2-364"><a href="#cb2-364" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-365"><a href="#cb2-365" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-366"><a href="#cb2-366" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting standard cross-validation results</span></span>
<span id="cb2-367"><a href="#cb2-367" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-368"><a href="#cb2-368" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_plot,</span>
<span id="cb2-369"><a href="#cb2-369" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv</span>(so2pls_cv)</span>
<span id="cb2-370"><a href="#cb2-370" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-371"><a href="#cb2-371" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-372"><a href="#cb2-372" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for sparsity parameters</span></span>
<span id="cb2-373"><a href="#cb2-373" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-374"><a href="#cb2-374" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity,</span>
<span id="cb2-375"><a href="#cb2-375" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_crossval_sparsity</span>(</span>
<span id="cb2-376"><a href="#cb2-376" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-377"><a href="#cb2-377" aria-hidden="true" tabindex="-1"></a>      <span class="at">n =</span> so2pls_cv_res[<span class="st">"n"</span>],</span>
<span id="cb2-378"><a href="#cb2-378" aria-hidden="true" tabindex="-1"></a>      <span class="at">nx =</span> so2pls_cv_res[<span class="st">"nx"</span>],</span>
<span id="cb2-379"><a href="#cb2-379" aria-hidden="true" tabindex="-1"></a>      <span class="at">ny =</span> so2pls_cv_res[<span class="st">"ny"</span>],</span>
<span id="cb2-380"><a href="#cb2-380" aria-hidden="true" tabindex="-1"></a>      <span class="at">nr_folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-381"><a href="#cb2-381" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepx_seq =</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">5</span>, <span class="dv">30</span>, <span class="dv">5</span>), <span class="fu">seq</span>(<span class="dv">40</span>, <span class="dv">100</span>, <span class="dv">10</span>)),</span>
<span id="cb2-382"><a href="#cb2-382" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepy_seq =</span> <span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">5</span>, <span class="dv">40</span>, <span class="dv">5</span>))</span>
<span id="cb2-383"><a href="#cb2-383" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-384"><a href="#cb2-384" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-385"><a href="#cb2-385" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-386"><a href="#cb2-386" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_res,</span>
<span id="cb2-387"><a href="#cb2-387" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_get_optim_keep</span>(so2pls_cv_sparsity)</span>
<span id="cb2-388"><a href="#cb2-388" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-389"><a href="#cb2-389" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-390"><a href="#cb2-390" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting the results of the cross-validation for the number of features</span></span>
<span id="cb2-391"><a href="#cb2-391" aria-hidden="true" tabindex="-1"></a>  <span class="do">## to retain from each dataset for the different joint components</span></span>
<span id="cb2-392"><a href="#cb2-392" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-393"><a href="#cb2-393" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_plot,</span>
<span id="cb2-394"><a href="#cb2-394" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_cv_sparsity</span>(so2pls_cv_sparsity)</span>
<span id="cb2-395"><a href="#cb2-395" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-396"><a href="#cb2-396" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-397"><a href="#cb2-397" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Extracting sparsity results in table format</span></span>
<span id="cb2-398"><a href="#cb2-398" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-399"><a href="#cb2-399" aria-hidden="true" tabindex="-1"></a>    so2pls_cv_sparsity_table,</span>
<span id="cb2-400"><a href="#cb2-400" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_print_cv_sparsity</span>(so2pls_cv_sparsity_res)</span>
<span id="cb2-401"><a href="#cb2-401" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-402"><a href="#cb2-402" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-403"><a href="#cb2-403" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Final sO2PLS run</span></span>
<span id="cb2-404"><a href="#cb2-404" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-405"><a href="#cb2-405" aria-hidden="true" tabindex="-1"></a>    so2pls_final_run,</span>
<span id="cb2-406"><a href="#cb2-406" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_o2m</span>(</span>
<span id="cb2-407"><a href="#cb2-407" aria-hidden="true" tabindex="-1"></a>      omicspls_input,</span>
<span id="cb2-408"><a href="#cb2-408" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_res,</span>
<span id="cb2-409"><a href="#cb2-409" aria-hidden="true" tabindex="-1"></a>      so2pls_cv_sparsity_res</span>
<span id="cb2-410"><a href="#cb2-410" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-411"><a href="#cb2-411" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-412"><a href="#cb2-412" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-413"><a href="#cb2-413" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Summary plot of percentage of variance explained</span></span>
<span id="cb2-414"><a href="#cb2-414" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-415"><a href="#cb2-415" aria-hidden="true" tabindex="-1"></a>    so2pls_summary_plot,</span>
<span id="cb2-416"><a href="#cb2-416" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_summary</span>(so2pls_final_run)</span>
<span id="cb2-417"><a href="#cb2-417" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-418"><a href="#cb2-418" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-419"><a href="#cb2-419" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Screeplot</span></span>
<span id="cb2-420"><a href="#cb2-420" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-421"><a href="#cb2-421" aria-hidden="true" tabindex="-1"></a>    so2pls_screeplot,</span>
<span id="cb2-422"><a href="#cb2-422" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_screeplot</span>(so2pls_final_run)</span>
<span id="cb2-423"><a href="#cb2-423" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-424"><a href="#cb2-424" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-425"><a href="#cb2-425" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Comparison of samples score for joint components</span></span>
<span id="cb2-426"><a href="#cb2-426" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-427"><a href="#cb2-427" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_comparison_plot,</span>
<span id="cb2-428"><a href="#cb2-428" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_compare_samples_joint_components</span>(</span>
<span id="cb2-429"><a href="#cb2-429" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-430"><a href="#cb2-430" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-431"><a href="#cb2-431" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_by =</span> <span class="st">"status"</span>,</span>
<span id="cb2-432"><a href="#cb2-432" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_by =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-433"><a href="#cb2-433" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-434"><a href="#cb2-434" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-435"><a href="#cb2-435" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-436"><a href="#cb2-436" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Coefficient plot for joint components</span></span>
<span id="cb2-437"><a href="#cb2-437" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-438"><a href="#cb2-438" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_coefficients_plot,</span>
<span id="cb2-439"><a href="#cb2-439" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_joint_components_coefficients</span>(so2pls_final_run)</span>
<span id="cb2-440"><a href="#cb2-440" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-441"><a href="#cb2-441" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-442"><a href="#cb2-442" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Joint component samples score plot</span></span>
<span id="cb2-443"><a href="#cb2-443" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-444"><a href="#cb2-444" aria-hidden="true" tabindex="-1"></a>    so2pls_joint_components_samples_score_plot,</span>
<span id="cb2-445"><a href="#cb2-445" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_samples_joint_components</span>(</span>
<span id="cb2-446"><a href="#cb2-446" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-447"><a href="#cb2-447" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-448"><a href="#cb2-448" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"status"</span>,</span>
<span id="cb2-449"><a href="#cb2-449" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>),</span>
<span id="cb2-450"><a href="#cb2-450" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"feedlot"</span></span>
<span id="cb2-451"><a href="#cb2-451" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb2-452"><a href="#cb2-452" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>)</span>
<span id="cb2-453"><a href="#cb2-453" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-454"><a href="#cb2-454" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-455"><a href="#cb2-455" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Specific components samples score plot</span></span>
<span id="cb2-456"><a href="#cb2-456" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-457"><a href="#cb2-457" aria-hidden="true" tabindex="-1"></a>    so2pls_specific_components_samples_score_plot,</span>
<span id="cb2-458"><a href="#cb2-458" aria-hidden="true" tabindex="-1"></a>    <span class="fu">so2pls_plot_samples_specific_components</span>(</span>
<span id="cb2-459"><a href="#cb2-459" aria-hidden="true" tabindex="-1"></a>      so2pls_final_run,</span>
<span id="cb2-460"><a href="#cb2-460" aria-hidden="true" tabindex="-1"></a>      <span class="at">mo_data =</span> mo_set_de,</span>
<span id="cb2-461"><a href="#cb2-461" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_upper =</span> <span class="st">"feedlot"</span>,</span>
<span id="cb2-462"><a href="#cb2-462" aria-hidden="true" tabindex="-1"></a>      <span class="at">scale_colour_upper =</span> <span class="fu">scale_colour_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>),</span>
<span id="cb2-463"><a href="#cb2-463" aria-hidden="true" tabindex="-1"></a>      <span class="at">colour_lower =</span> <span class="st">"rnaseq_batch"</span>,</span>
<span id="cb2-464"><a href="#cb2-464" aria-hidden="true" tabindex="-1"></a>      <span class="at">shape_upper =</span> <span class="st">"gender"</span></span>
<span id="cb2-465"><a href="#cb2-465" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">|&gt;</span> </span>
<span id="cb2-466"><a href="#cb2-466" aria-hidden="true" tabindex="-1"></a>      <span class="fu">map</span>(\(x) x <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.box =</span> <span class="st">"vertical"</span>))</span>
<span id="cb2-467"><a href="#cb2-467" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-468"><a href="#cb2-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-469"><a href="#cb2-469" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with MOFA --------------------------------------------------------</span></span>
<span id="cb2-470"><a href="#cb2-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-471"><a href="#cb2-471" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating MOFA input</span></span>
<span id="cb2-472"><a href="#cb2-472" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-473"><a href="#cb2-473" aria-hidden="true" tabindex="-1"></a>    mofa_input,</span>
<span id="cb2-474"><a href="#cb2-474" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_mofa</span>(</span>
<span id="cb2-475"><a href="#cb2-475" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-476"><a href="#cb2-476" aria-hidden="true" tabindex="-1"></a>      <span class="at">options_list =</span> <span class="fu">list</span>(</span>
<span id="cb2-477"><a href="#cb2-477" aria-hidden="true" tabindex="-1"></a>        <span class="at">data_options =</span> <span class="fu">list</span>(<span class="at">scale_views =</span> <span class="cn">TRUE</span>),</span>
<span id="cb2-478"><a href="#cb2-478" aria-hidden="true" tabindex="-1"></a>        <span class="at">model_options =</span> <span class="fu">list</span>(<span class="at">likelihoods =</span> <span class="fu">c</span>(</span>
<span id="cb2-479"><a href="#cb2-479" aria-hidden="true" tabindex="-1"></a>          <span class="st">"snps"</span> <span class="ot">=</span> <span class="st">"poisson"</span>,</span>
<span id="cb2-480"><a href="#cb2-480" aria-hidden="true" tabindex="-1"></a>          <span class="st">"rnaseq"</span> <span class="ot">=</span> <span class="st">"gaussian"</span>,</span>
<span id="cb2-481"><a href="#cb2-481" aria-hidden="true" tabindex="-1"></a>          <span class="st">"metabolome"</span> <span class="ot">=</span> <span class="st">"gaussian"</span>)</span>
<span id="cb2-482"><a href="#cb2-482" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb2-483"><a href="#cb2-483" aria-hidden="true" tabindex="-1"></a>        <span class="at">training_options =</span> <span class="fu">list</span>(<span class="at">seed =</span> <span class="dv">43</span>)</span>
<span id="cb2-484"><a href="#cb2-484" aria-hidden="true" tabindex="-1"></a>      ),</span>
<span id="cb2-485"><a href="#cb2-485" aria-hidden="true" tabindex="-1"></a>      <span class="at">only_common_samples =</span> <span class="cn">FALSE</span></span>
<span id="cb2-486"><a href="#cb2-486" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-487"><a href="#cb2-487" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-488"><a href="#cb2-488" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-489"><a href="#cb2-489" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Overview plot of the samples in each dataset</span></span>
<span id="cb2-490"><a href="#cb2-490" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-491"><a href="#cb2-491" aria-hidden="true" tabindex="-1"></a>    mofa_input_plot,</span>
<span id="cb2-492"><a href="#cb2-492" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_data_overview</span>(mofa_input)</span>
<span id="cb2-493"><a href="#cb2-493" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-494"><a href="#cb2-494" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-495"><a href="#cb2-495" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Training MOFA model</span></span>
<span id="cb2-496"><a href="#cb2-496" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-497"><a href="#cb2-497" aria-hidden="true" tabindex="-1"></a>    mofa_trained,</span>
<span id="cb2-498"><a href="#cb2-498" aria-hidden="true" tabindex="-1"></a>    <span class="fu">run_mofa</span>(</span>
<span id="cb2-499"><a href="#cb2-499" aria-hidden="true" tabindex="-1"></a>      mofa_input,</span>
<span id="cb2-500"><a href="#cb2-500" aria-hidden="true" tabindex="-1"></a>      <span class="at">save_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb2-501"><a href="#cb2-501" aria-hidden="true" tabindex="-1"></a>      <span class="at">use_basilisk =</span> <span class="cn">TRUE</span></span>
<span id="cb2-502"><a href="#cb2-502" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-503"><a href="#cb2-503" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-504"><a href="#cb2-504" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-505"><a href="#cb2-505" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Formatting MOFA output</span></span>
<span id="cb2-506"><a href="#cb2-506" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-507"><a href="#cb2-507" aria-hidden="true" tabindex="-1"></a>    mofa_output,</span>
<span id="cb2-508"><a href="#cb2-508" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_output</span>(mofa_trained)</span>
<span id="cb2-509"><a href="#cb2-509" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-510"><a href="#cb2-510" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-511"><a href="#cb2-511" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plots of variance explained</span></span>
<span id="cb2-512"><a href="#cb2-512" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-513"><a href="#cb2-513" aria-hidden="true" tabindex="-1"></a>    mofa_var_explained_plot,</span>
<span id="cb2-514"><a href="#cb2-514" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(</span>
<span id="cb2-515"><a href="#cb2-515" aria-hidden="true" tabindex="-1"></a>      mofa_trained,</span>
<span id="cb2-516"><a href="#cb2-516" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"view"</span>,  <span class="do">## datasets on the x-axis</span></span>
<span id="cb2-517"><a href="#cb2-517" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"factor"</span> <span class="do">## factors on the y-axis</span></span>
<span id="cb2-518"><a href="#cb2-518" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-519"><a href="#cb2-519" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-520"><a href="#cb2-520" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-521"><a href="#cb2-521" aria-hidden="true" tabindex="-1"></a>    mofa_total_var_explained_plot,</span>
<span id="cb2-522"><a href="#cb2-522" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_variance_explained</span>(</span>
<span id="cb2-523"><a href="#cb2-523" aria-hidden="true" tabindex="-1"></a>      mofa_trained,</span>
<span id="cb2-524"><a href="#cb2-524" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"view"</span>,</span>
<span id="cb2-525"><a href="#cb2-525" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"factor"</span>,</span>
<span id="cb2-526"><a href="#cb2-526" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot_total =</span> <span class="cn">TRUE</span></span>
<span id="cb2-527"><a href="#cb2-527" aria-hidden="true" tabindex="-1"></a>    )[[<span class="dv">2</span>]]</span>
<span id="cb2-528"><a href="#cb2-528" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-529"><a href="#cb2-529" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-530"><a href="#cb2-530" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plot of factors correlation with covariates</span></span>
<span id="cb2-531"><a href="#cb2-531" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-532"><a href="#cb2-532" aria-hidden="true" tabindex="-1"></a>    mofa_factors_covariates_cor_plot,</span>
<span id="cb2-533"><a href="#cb2-533" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mofa_plot_cor_covariates</span>(mofa_trained)</span>
<span id="cb2-534"><a href="#cb2-534" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-535"><a href="#cb2-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-536"><a href="#cb2-536" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration with DIABLO ------------------------------------------------------</span></span>
<span id="cb2-537"><a href="#cb2-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-538"><a href="#cb2-538" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Creating the DIABLO input</span></span>
<span id="cb2-539"><a href="#cb2-539" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-540"><a href="#cb2-540" aria-hidden="true" tabindex="-1"></a>    diablo_input,</span>
<span id="cb2-541"><a href="#cb2-541" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_input_mixomics_supervised</span>(</span>
<span id="cb2-542"><a href="#cb2-542" aria-hidden="true" tabindex="-1"></a>      mo_presel_supervised,</span>
<span id="cb2-543"><a href="#cb2-543" aria-hidden="true" tabindex="-1"></a>      <span class="at">group =</span> <span class="st">"status"</span></span>
<span id="cb2-544"><a href="#cb2-544" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-545"><a href="#cb2-545" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-546"><a href="#cb2-546" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-547"><a href="#cb2-547" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Running sPLS on each dataset to construct the design matrix</span></span>
<span id="cb2-548"><a href="#cb2-548" aria-hidden="true" tabindex="-1"></a>  <span class="fu">diablo_pairwise_pls_factory</span>(diablo_input),</span>
<span id="cb2-549"><a href="#cb2-549" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-550"><a href="#cb2-550" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Initial DIABLO run with no feature selection and large number of components</span></span>
<span id="cb2-551"><a href="#cb2-551" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-552"><a href="#cb2-552" aria-hidden="true" tabindex="-1"></a>    diablo_novarsel,</span>
<span id="cb2-553"><a href="#cb2-553" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_run</span>(</span>
<span id="cb2-554"><a href="#cb2-554" aria-hidden="true" tabindex="-1"></a>      diablo_input,</span>
<span id="cb2-555"><a href="#cb2-555" aria-hidden="true" tabindex="-1"></a>      diablo_design_matrix,</span>
<span id="cb2-556"><a href="#cb2-556" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> <span class="dv">7</span></span>
<span id="cb2-557"><a href="#cb2-557" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-558"><a href="#cb2-558" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-559"><a href="#cb2-559" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-560"><a href="#cb2-560" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of components</span></span>
<span id="cb2-561"><a href="#cb2-561" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-562"><a href="#cb2-562" aria-hidden="true" tabindex="-1"></a>    diablo_perf_res,</span>
<span id="cb2-563"><a href="#cb2-563" aria-hidden="true" tabindex="-1"></a>    mixOmics<span class="sc">::</span><span class="fu">perf</span>(</span>
<span id="cb2-564"><a href="#cb2-564" aria-hidden="true" tabindex="-1"></a>      diablo_novarsel,</span>
<span id="cb2-565"><a href="#cb2-565" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-566"><a href="#cb2-566" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-567"><a href="#cb2-567" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">10</span>,</span>
<span id="cb2-568"><a href="#cb2-568" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-569"><a href="#cb2-569" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-570"><a href="#cb2-570" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-571"><a href="#cb2-571" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-572"><a href="#cb2-572" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of components)</span></span>
<span id="cb2-573"><a href="#cb2-573" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-574"><a href="#cb2-574" aria-hidden="true" tabindex="-1"></a>    diablo_perf_plot,</span>
<span id="cb2-575"><a href="#cb2-575" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_plot_perf</span>(diablo_perf_res)</span>
<span id="cb2-576"><a href="#cb2-576" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-577"><a href="#cb2-577" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-578"><a href="#cb2-578" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Selected value for ncomp</span></span>
<span id="cb2-579"><a href="#cb2-579" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-580"><a href="#cb2-580" aria-hidden="true" tabindex="-1"></a>    diablo_optim_ncomp,</span>
<span id="cb2-581"><a href="#cb2-581" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_get_optim_ncomp</span>(diablo_perf_res)</span>
<span id="cb2-582"><a href="#cb2-582" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-583"><a href="#cb2-583" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-584"><a href="#cb2-584" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cross-validation for number of features to retain</span></span>
<span id="cb2-585"><a href="#cb2-585" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-586"><a href="#cb2-586" aria-hidden="true" tabindex="-1"></a>    diablo_tune_res,</span>
<span id="cb2-587"><a href="#cb2-587" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_tune</span>(</span>
<span id="cb2-588"><a href="#cb2-588" aria-hidden="true" tabindex="-1"></a>      diablo_input,</span>
<span id="cb2-589"><a href="#cb2-589" aria-hidden="true" tabindex="-1"></a>      diablo_design_matrix,</span>
<span id="cb2-590"><a href="#cb2-590" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> diablo_optim_ncomp,</span>
<span id="cb2-591"><a href="#cb2-591" aria-hidden="true" tabindex="-1"></a>      <span class="at">validation =</span> <span class="st">"Mfold"</span>,</span>
<span id="cb2-592"><a href="#cb2-592" aria-hidden="true" tabindex="-1"></a>      <span class="at">folds =</span> <span class="dv">10</span>,</span>
<span id="cb2-593"><a href="#cb2-593" aria-hidden="true" tabindex="-1"></a>      <span class="at">nrepeat =</span> <span class="dv">5</span>,</span>
<span id="cb2-594"><a href="#cb2-594" aria-hidden="true" tabindex="-1"></a>      <span class="at">dist =</span> <span class="st">"centroids.dist"</span>,</span>
<span id="cb2-595"><a href="#cb2-595" aria-hidden="true" tabindex="-1"></a>      <span class="at">cpus =</span> <span class="dv">3</span></span>
<span id="cb2-596"><a href="#cb2-596" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-597"><a href="#cb2-597" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-598"><a href="#cb2-598" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-599"><a href="#cb2-599" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Plotting cross-validation results (for number of features)</span></span>
<span id="cb2-600"><a href="#cb2-600" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-601"><a href="#cb2-601" aria-hidden="true" tabindex="-1"></a>    diablo_tune_plot,</span>
<span id="cb2-602"><a href="#cb2-602" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_plot_tune</span>(diablo_tune_res)</span>
<span id="cb2-603"><a href="#cb2-603" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb2-604"><a href="#cb2-604" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-605"><a href="#cb2-605" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Final DIABLO run</span></span>
<span id="cb2-606"><a href="#cb2-606" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tar_target</span>(</span>
<span id="cb2-607"><a href="#cb2-607" aria-hidden="true" tabindex="-1"></a>    diablo_final_run,</span>
<span id="cb2-608"><a href="#cb2-608" aria-hidden="true" tabindex="-1"></a>    <span class="fu">diablo_run</span>(</span>
<span id="cb2-609"><a href="#cb2-609" aria-hidden="true" tabindex="-1"></a>      diablo_input,</span>
<span id="cb2-610"><a href="#cb2-610" aria-hidden="true" tabindex="-1"></a>      diablo_design_matrix,</span>
<span id="cb2-611"><a href="#cb2-611" aria-hidden="true" tabindex="-1"></a>      <span class="at">ncomp =</span> diablo_optim_ncomp,</span>
<span id="cb2-612"><a href="#cb2-612" aria-hidden="true" tabindex="-1"></a>      <span class="at">keepX =</span> diablo_tune_res<span class="sc">$</span>choice.keepX</span>
<span id="cb2-613"><a href="#cb2-613" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb2-614"><a href="#cb2-614" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-615"><a href="#cb2-615" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details><section id="sec-interpretation-dim-reduction" class="level2" data-number="11.1"><h2 data-number="11.1" class="anchored" data-anchor-id="sec-interpretation-dim-reduction">
<span class="header-section-number">11.1</span> Dimension reduction methods</h2>
<p>Despite relying on very different statistical approaches, the different integration methods included in the pipeline all perform dimension reduction of the omics datasets through feature extraction. That is, they construct a small number of latent components/variables/dimensions (that we refer to as <strong>latent dimensions</strong> in the <code>moiraine</code> package) that capture as much information from the original datasets as possible. A dimension reduction approach typically returns, for each latent dimension constructed, two sets of values:</p>
<ul>
<li><p><strong>Features weight</strong>: the contribution of the features from the different omics datasets to the latent dimension. All methods included in the pipeline construct latent dimensions as linear combinations of the original features, and therefore the features contribution is quantified by their weight in the linear combination for the corresponding latent dimension.</p></li>
<li><p><strong>Samples score</strong>: the projection of the samples onto the latent dimension.</p></li>
</ul>
<p>In addition, the fraction or percentage of variance that each latent dimension explains in the different omics datasets is usually calculated.</p>
<p>Interpreting the results of a dimension reduction method involves:</p>
<ul>
<li><p>Understanding the source of the variation captured by each latent dimension: is a given latent dimension representing an important source of biological variation, such as effect of a treatment, or age of the samples? Or do they show a source of technical variation, for example highlighting a group of outlier samples with different omics profiles from the rest of the observations? Answering these questions allows us to identify which latent dimensions capture the biological phenomenon investigated, or whether there are some sources of noise that should be accounted for in follow-up experiments.</p></li>
<li><p>Investigating which omics features are driving the latent dimensions: once we have identified some latent dimensions of interest, we can look at the features that contribute the most to these dimensions, in order to understand the molecular mechanisms or pathways involved. This is typically done after looking into the phenomenon captured by the latent dimensions, but can also help to identify it.</p></li>
</ul></section><section id="sec-interpretation-standard-output" class="level2" data-number="11.2"><h2 data-number="11.2" class="anchored" data-anchor-id="sec-interpretation-standard-output">
<span class="header-section-number">11.2</span> Generating a standardised output</h2>
<section id="get_output-function" class="level3" data-number="11.2.1"><h3 data-number="11.2.1" class="anchored" data-anchor-id="get_output-function">
<span class="header-section-number">11.2.1</span> <code>get_output</code> function</h3>
<p>In the <code>moiraine</code> package, the output of the different integration methods can be converted to a standardised output containing the pieces of information (features weight, samples score and percentage of variance explained) characteristic of dimension reduction methods, stored in a consistent format. This enables us to use functions for visualisation or analysis that can be applied to the results of any integration method, rather than having to implement one for each object type returned by the different integration packages.</p>
<p>The <code><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output()</a></code> function transforms the output from any integration package included in <code>moiraine</code> into an <code>output_dimension_reduction</code> object, which is a list with three tibbles: <code>features_weight</code>, <code>samples_score</code> and <code>variance_explained</code>:</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" aria-current="page">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-4" role="tab" aria-controls="tabset-1-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">spls_final_run</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">)</span></span>
<span><span class="va">spls_output</span></span>
<span><span class="co">#&gt; $features_weight</span></span>
<span><span class="co">#&gt; # A tibble: 2,098 Ã— 5</span></span>
<span><span class="co">#&gt;    feature_id         dataset latent_dimension weight importance</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;              &lt;fct&gt;   &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 ENSBTAG00000000020 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt;  2 ENSBTAG00000000046 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt;  3 ENSBTAG00000000056 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt;  4 ENSBTAG00000000061 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt;  5 ENSBTAG00000000113 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt;  6 ENSBTAG00000000149 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt;  7 ENSBTAG00000000164 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt;  8 ENSBTAG00000000205 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt;  9 ENSBTAG00000000212 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt; 10 ENSBTAG00000000289 rnaseq  Component 1           0          0</span></span>
<span><span class="co">#&gt; # â„¹ 2,088 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $samples_score</span></span>
<span><span class="co">#&gt; # A tibble: 278 Ã— 3</span></span>
<span><span class="co">#&gt;    sample_id latent_dimension score</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;fct&gt;            &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 G1979     Component 1      -3.63</span></span>
<span><span class="co">#&gt;  2 G2500     Component 1      -2.46</span></span>
<span><span class="co">#&gt;  3 G3030     Component 1      -6.42</span></span>
<span><span class="co">#&gt;  4 G3068     Component 1      -6.16</span></span>
<span><span class="co">#&gt;  5 G3121     Component 1      -3.39</span></span>
<span><span class="co">#&gt;  6 G3315     Component 1       4.98</span></span>
<span><span class="co">#&gt;  7 G3473     Component 1       5.29</span></span>
<span><span class="co">#&gt;  8 G3474     Component 1       4.97</span></span>
<span><span class="co">#&gt;  9 G3550     Component 1      -4.17</span></span>
<span><span class="co">#&gt; 10 G3594     Component 1       4.71</span></span>
<span><span class="co">#&gt; # â„¹ 268 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $variance_explained</span></span>
<span><span class="co">#&gt; # A tibble: 4 Ã— 3</span></span>
<span><span class="co">#&gt;   latent_dimension dataset    prop_var_expl</span></span>
<span><span class="co">#&gt;   &lt;fct&gt;            &lt;fct&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 Component 1      rnaseq             0.205</span></span>
<span><span class="co">#&gt; 2 Component 1      metabolome         0.179</span></span>
<span><span class="co">#&gt; 3 Component 2      rnaseq             0.147</span></span>
<span><span class="co">#&gt; 4 Component 2      metabolome         0.108</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "output_dimension_reduction"</span></span>
<span><span class="co">#&gt; attr(,"method")</span></span>
<span><span class="co">#&gt; [1] "sPLS"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">so2pls_final_run</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">so2pls_output</span><span class="op">)</span></span>
<span><span class="va">so2pls_output</span></span>
<span><span class="co">#&gt; $features_weight</span></span>
<span><span class="co">#&gt; # A tibble: 2,318 Ã— 5</span></span>
<span><span class="co">#&gt;    feature_id         dataset latent_dimension  weight importance</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;              &lt;fct&gt;   &lt;fct&gt;              &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 ENSBTAG00000000020 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt;  2 ENSBTAG00000000046 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt;  3 ENSBTAG00000000056 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt;  4 ENSBTAG00000000061 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt;  5 ENSBTAG00000000113 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt;  6 ENSBTAG00000000149 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt;  7 ENSBTAG00000000164 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt;  8 ENSBTAG00000000205 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt;  9 ENSBTAG00000000212 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt; 10 ENSBTAG00000000289 rnaseq  joint component 1      0          0</span></span>
<span><span class="co">#&gt; # â„¹ 2,308 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $samples_score</span></span>
<span><span class="co">#&gt; # A tibble: 973 Ã— 3</span></span>
<span><span class="co">#&gt;    sample_id latent_dimension   score</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;fct&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 G1979     joint component 1  -8.12</span></span>
<span><span class="co">#&gt;  2 G2500     joint component 1  -9.81</span></span>
<span><span class="co">#&gt;  3 G3030     joint component 1  -9.93</span></span>
<span><span class="co">#&gt;  4 G3068     joint component 1 -14.7 </span></span>
<span><span class="co">#&gt;  5 G3121     joint component 1  -5.01</span></span>
<span><span class="co">#&gt;  6 G3315     joint component 1   9.66</span></span>
<span><span class="co">#&gt;  7 G3473     joint component 1   9.79</span></span>
<span><span class="co">#&gt;  8 G3474     joint component 1   9.99</span></span>
<span><span class="co">#&gt;  9 G3550     joint component 1  -9.70</span></span>
<span><span class="co">#&gt; 10 G3594     joint component 1   8.37</span></span>
<span><span class="co">#&gt; # â„¹ 963 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $variance_explained</span></span>
<span><span class="co">#&gt; # A tibble: 8 Ã— 3</span></span>
<span><span class="co">#&gt;   latent_dimension                dataset    prop_var_expl</span></span>
<span><span class="co">#&gt;   &lt;fct&gt;                           &lt;fct&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 joint component 1               rnaseq            0.327 </span></span>
<span><span class="co">#&gt; 2 joint component 1               metabolome        0.150 </span></span>
<span><span class="co">#&gt; 3 rnaseq specific component 1     rnaseq            0.123 </span></span>
<span><span class="co">#&gt; 4 metabolome specific component 1 metabolome        0.107 </span></span>
<span><span class="co">#&gt; 5 metabolome specific component 2 metabolome        0.0767</span></span>
<span><span class="co">#&gt; 6 metabolome specific component 3 metabolome        0.0613</span></span>
<span><span class="co">#&gt; 7 metabolome specific component 4 metabolome        0.0605</span></span>
<span><span class="co">#&gt; 8 metabolome specific component 5 metabolome        0.0376</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "output_dimension_reduction"</span></span>
<span><span class="co">#&gt; attr(,"method")</span></span>
<span><span class="co">#&gt; [1] "sO2PLS"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">mofa_trained</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mofa_output</span><span class="op">)</span></span>
<span><span class="va">mofa_output</span></span>
<span><span class="co">#&gt; $features_weight</span></span>
<span><span class="co">#&gt; # A tibble: 30,735 Ã— 5</span></span>
<span><span class="co">#&gt;    feature_id                  dataset latent_dimension     weight importance</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                       &lt;fct&gt;   &lt;fct&gt;                 &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 21-25977541-C-T-rs41974686  snps    Factor 1          0.000374      0.139 </span></span>
<span><span class="co">#&gt;  2 22-51403583-A-C-rs210306176 snps    Factor 1          0.0000535     0.0199</span></span>
<span><span class="co">#&gt;  3 24-12959068-G-T-rs381471286 snps    Factor 1          0.000268      0.0999</span></span>
<span><span class="co">#&gt;  4 8-85224224-T-C-rs43565287   snps    Factor 1         -0.000492      0.183 </span></span>
<span><span class="co">#&gt;  5 ARS-BFGL-BAC-16973          snps    Factor 1         -0.000883      0.329 </span></span>
<span><span class="co">#&gt;  6 ARS-BFGL-BAC-19403          snps    Factor 1         -0.000500      0.186 </span></span>
<span><span class="co">#&gt;  7 ARS-BFGL-BAC-2450           snps    Factor 1          0.000555      0.207 </span></span>
<span><span class="co">#&gt;  8 ARS-BFGL-BAC-2600           snps    Factor 1         -0.0000325     0.0121</span></span>
<span><span class="co">#&gt;  9 ARS-BFGL-BAC-27911          snps    Factor 1         -0.000568      0.212 </span></span>
<span><span class="co">#&gt; 10 ARS-BFGL-BAC-35925          snps    Factor 1          0.000803      0.299 </span></span>
<span><span class="co">#&gt; # â„¹ 30,725 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $samples_score</span></span>
<span><span class="co">#&gt; # A tibble: 2,160 Ã— 3</span></span>
<span><span class="co">#&gt;    sample_id latent_dimension score</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;fct&gt;            &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 G1979     Factor 1          1.95</span></span>
<span><span class="co">#&gt;  2 G2500     Factor 1          1.98</span></span>
<span><span class="co">#&gt;  3 G3030     Factor 1          3.39</span></span>
<span><span class="co">#&gt;  4 G3068     Factor 1          3.65</span></span>
<span><span class="co">#&gt;  5 G3121     Factor 1          1.57</span></span>
<span><span class="co">#&gt;  6 G3315     Factor 1         -1.81</span></span>
<span><span class="co">#&gt;  7 G3473     Factor 1         -2.63</span></span>
<span><span class="co">#&gt;  8 G3474     Factor 1         -2.30</span></span>
<span><span class="co">#&gt;  9 G3550     Factor 1          2.98</span></span>
<span><span class="co">#&gt; 10 G3594     Factor 1         -1.88</span></span>
<span><span class="co">#&gt; # â„¹ 2,150 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $variance_explained</span></span>
<span><span class="co">#&gt; # A tibble: 45 Ã— 3</span></span>
<span><span class="co">#&gt;    latent_dimension dataset    prop_var_expl</span></span>
<span><span class="co">#&gt;    &lt;fct&gt;            &lt;fct&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 Factor 1         snps           0.000313 </span></span>
<span><span class="co">#&gt;  2 Factor 1         rnaseq         0.496    </span></span>
<span><span class="co">#&gt;  3 Factor 1         metabolome     0.137    </span></span>
<span><span class="co">#&gt;  4 Factor 2         snps           0.239    </span></span>
<span><span class="co">#&gt;  5 Factor 2         rnaseq         0.000197 </span></span>
<span><span class="co">#&gt;  6 Factor 2         metabolome     0.0108   </span></span>
<span><span class="co">#&gt;  7 Factor 3         snps           0.000108 </span></span>
<span><span class="co">#&gt;  8 Factor 3         rnaseq         0.199    </span></span>
<span><span class="co">#&gt;  9 Factor 3         metabolome     0.0000872</span></span>
<span><span class="co">#&gt; 10 Factor 4         snps           0.0000927</span></span>
<span><span class="co">#&gt; # â„¹ 35 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "output_dimension_reduction"</span></span>
<span><span class="co">#&gt; attr(,"method")</span></span>
<span><span class="co">#&gt; [1] "MOFA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-1-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-4-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">diablo_final_run</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_output</span><span class="op">)</span></span>
<span><span class="va">diablo_output</span></span>
<span><span class="co">#&gt; $features_weight</span></span>
<span><span class="co">#&gt; # A tibble: 8,196 Ã— 5</span></span>
<span><span class="co">#&gt;    feature_id                  dataset latent_dimension weight importance</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                       &lt;fct&gt;   &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 21-25977541-C-T-rs41974686  snps    Component 1           0          0</span></span>
<span><span class="co">#&gt;  2 22-51403583-A-C-rs210306176 snps    Component 1           0          0</span></span>
<span><span class="co">#&gt;  3 24-12959068-G-T-rs381471286 snps    Component 1           0          0</span></span>
<span><span class="co">#&gt;  4 8-85224224-T-C-rs43565287   snps    Component 1           0          0</span></span>
<span><span class="co">#&gt;  5 ARS-BFGL-BAC-16973          snps    Component 1           0          0</span></span>
<span><span class="co">#&gt;  6 ARS-BFGL-BAC-19403          snps    Component 1           0          0</span></span>
<span><span class="co">#&gt;  7 ARS-BFGL-BAC-2450           snps    Component 1           0          0</span></span>
<span><span class="co">#&gt;  8 ARS-BFGL-BAC-2600           snps    Component 1           0          0</span></span>
<span><span class="co">#&gt;  9 ARS-BFGL-BAC-27911          snps    Component 1           0          0</span></span>
<span><span class="co">#&gt; 10 ARS-BFGL-BAC-35925          snps    Component 1           0          0</span></span>
<span><span class="co">#&gt; # â„¹ 8,186 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $samples_score</span></span>
<span><span class="co">#&gt; # A tibble: 540 Ã— 3</span></span>
<span><span class="co">#&gt;    sample_id latent_dimension score</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;     &lt;fct&gt;            &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 G1979     Component 1       1.97</span></span>
<span><span class="co">#&gt;  2 G2500     Component 1       1.50</span></span>
<span><span class="co">#&gt;  3 G3030     Component 1       3.57</span></span>
<span><span class="co">#&gt;  4 G3068     Component 1       3.50</span></span>
<span><span class="co">#&gt;  5 G3121     Component 1       2.13</span></span>
<span><span class="co">#&gt;  6 G3315     Component 1      -3.26</span></span>
<span><span class="co">#&gt;  7 G3473     Component 1      -3.97</span></span>
<span><span class="co">#&gt;  8 G3474     Component 1      -3.20</span></span>
<span><span class="co">#&gt;  9 G3550     Component 1       2.46</span></span>
<span><span class="co">#&gt; 10 G3594     Component 1      -2.38</span></span>
<span><span class="co">#&gt; # â„¹ 530 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $variance_explained</span></span>
<span><span class="co">#&gt; # A tibble: 12 Ã— 3</span></span>
<span><span class="co">#&gt;    latent_dimension dataset    prop_var_expl</span></span>
<span><span class="co">#&gt;    &lt;fct&gt;            &lt;fct&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 Component 1      snps             0.0698 </span></span>
<span><span class="co">#&gt;  2 Component 1      rnaseq           0.201  </span></span>
<span><span class="co">#&gt;  3 Component 1      metabolome       0.171  </span></span>
<span><span class="co">#&gt;  4 Component 2      snps             0.0213 </span></span>
<span><span class="co">#&gt;  5 Component 2      rnaseq           0.0801 </span></span>
<span><span class="co">#&gt;  6 Component 2      metabolome       0.0521 </span></span>
<span><span class="co">#&gt;  7 Component 3      snps             0.00898</span></span>
<span><span class="co">#&gt;  8 Component 3      rnaseq           0.149  </span></span>
<span><span class="co">#&gt;  9 Component 3      metabolome       0.0634 </span></span>
<span><span class="co">#&gt; 10 Component 4      snps             0.0136 </span></span>
<span><span class="co">#&gt; 11 Component 4      rnaseq           0.0601 </span></span>
<span><span class="co">#&gt; 12 Component 4      metabolome       0.0399 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attr(,"class")</span></span>
<span><span class="co">#&gt; [1] "output_dimension_reduction"</span></span>
<span><span class="co">#&gt; attr(,"method")</span></span>
<span><span class="co">#&gt; [1] "DIABLO"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>The <code>features_weight</code> tibble contains one row per combination of feature and latent dimension. The ID of the features and the name of the dataset from which they originate are stored in the <code>feature_id</code> and <code>dataset</code> columns, respectively. The <code>latent_dimension</code> column gives the name of the latent dimension; this is a factor column. For each feature and latent dimension, the <code>weight</code> column shows the weight that was attributed to the feature for the corresponding latent dimension. In addition, the <code>importance</code> column contains the features <strong>importance score</strong>, which is computed as the absolute value of the features weight, divided by the maximum absolute weight across all features from the same omics dataset for the corresponding latent dimension. This importance score allows us to compare the contribution of the features across latent dimensions or integration methods, as the weight can be on different scales and thus cannot be directly compared. The importance scores range from 0 to 1. For any method performing feature selection (e.g.&nbsp;sPLS or DIABLO), features that were not selected for a given latent dimension are assigned a weight and importance score of 0.</p>
<p>The <code>samples score</code> tibble contains for each sample (<code>sample_id</code>) and latent dimension (<code>latent_dimension</code>) the sampleâ€™s coordinate for the corresponding latent dimension.</p>
<p>The <code>variance_explained</code> tibble gives for each latent dimension (<code>latent_dimension</code>) the proportion of variance explained (<code>prop_var_expl</code>) for each dataset (<code>dataset</code>). The values in <code>prop_var_expl</code> are between 0 and 1.</p>
<p>In addition, the name of the integration method used to obtain these results is stored in the <code>method</code> attribute of the <code>output_dimension_reduction</code> object:</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-3" role="tab" aria-controls="tabset-2-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-4" role="tab" aria-controls="tabset-2-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">spls_output</span>, <span class="st">"method"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "sPLS"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">so2pls_output</span>, <span class="st">"method"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "sO2PLS"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-2-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-3-tab">
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">mofa_output</span>, <span class="st">"method"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "MOFA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-2-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-4-tab">
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">diablo_output</span>, <span class="st">"method"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "DIABLO"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>For convenience, the <code><a href="https://rdrr.io/pkg/moiraine/man/get_latent_dimensions.html">get_latent_dimensions()</a></code> function can be used on an <code>output_dimension_reduction</code> object to see the names of the latent dimensions (the levels used for the <code>latent_dimension</code> column in each tibble):</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-2" role="tab" aria-controls="tabset-3-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-3" role="tab" aria-controls="tabset-3-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-4" role="tab" aria-controls="tabset-3-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_latent_dimensions.html">get_latent_dimensions</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Component 1" "Component 2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-3-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-2-tab">
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_latent_dimensions.html">get_latent_dimensions</a></span><span class="op">(</span><span class="va">so2pls_output</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "joint component 1"               "rnaseq specific component 1"    </span></span>
<span><span class="co">#&gt; [3] "metabolome specific component 1" "metabolome specific component 2"</span></span>
<span><span class="co">#&gt; [5] "metabolome specific component 3" "metabolome specific component 4"</span></span>
<span><span class="co">#&gt; [7] "metabolome specific component 5"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-3-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-3-tab">
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_latent_dimensions.html">get_latent_dimensions</a></span><span class="op">(</span><span class="va">mofa_output</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Factor 1"  "Factor 2"  "Factor 3"  "Factor 4"  "Factor 5"  "Factor 6" </span></span>
<span><span class="co">#&gt;  [7] "Factor 7"  "Factor 8"  "Factor 9"  "Factor 10" "Factor 11" "Factor 12"</span></span>
<span><span class="co">#&gt; [13] "Factor 13" "Factor 14" "Factor 15"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-3-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-4-tab">
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_latent_dimensions.html">get_latent_dimensions</a></span><span class="op">(</span><span class="va">diablo_output</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Component 1" "Component 2" "Component 3" "Component 4"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Other methods covered by <code>get_output</code>
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that both PCA and sPLS-DA (the method used for supervised features preselection in <a href="prefiltering.html#sec-prefiltering-supervised"><span>Section&nbsp;6.3.2</span></a>) are also dimension reduction methods. Therefore, the <code><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output()</a></code> function also converts <code>pcaRes</code> objects (from <code><a href="https://rdrr.io/pkg/moiraine/man/run_pca.html">run_pca()</a></code> or <code><a href="https://rdrr.io/pkg/pcaMethods/man/pca.html">pcaMethods::pca()</a></code>) and <code>mixo_splsda</code> objects (from <code><a href="https://rdrr.io/pkg/moiraine/man/run_splsda.html">run_splsda()</a></code> or <code><a href="https://rdrr.io/pkg/mixOmics/man/splsda.html">mixOmics::splsda()</a></code>) into <code>output_dimension_reduction</code> objects.</p>
</div>
</div>
</section><section id="averaging-latent-dimensions-over-datasets" class="level3" data-number="11.2.2"><h3 data-number="11.2.2" class="anchored" data-anchor-id="averaging-latent-dimensions-over-datasets">
<span class="header-section-number">11.2.2</span> Averaging latent dimensions over datasets</h3>
<p>While MOFA computes one score per sample for each latent dimension created, sPLS, sO2PLS and DIABLO all compute one score per dataset for each sample and latent dimension. For each latent dimension, the samples score obtained for the different datasets are then compared, to assess the agreement or covariation between datasets. Ideally, these scores should be highly correlated across datasets, since the methods aim at maximising the variation between datasets, but it is not always the case. However, when they are highly correlated, it becomes redundant to interpret each dataset-specific version of the latent dimensions.</p>
<p>Instead, the <code>mixOmics</code> authors proposed a solution for <code>DIABLO</code>, which is to construct a weighted average space in which to represent the samples: for each latent component, the samples score are averaged over the different datasets. The weight attributed to each dataset is determined by how well the corresponding dataset discriminates between the samples group of interest. This way, rather than looking at samples score for each dataset for any given latent component, we can look at an average of them.</p>
<p>The <code><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output()</a></code> function uses this idea to construct, for the output of sPLS, sO2PLS and DIABLO a set of average samples score for each latent dimension, rather than returning a set of samples score per dataset. For DIABLO, the average is weighted as explained above, while for sPLS and sO2PLS each dataset is given equal weight in the average. This calculation can be disabled in the <code><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output()</a></code> function to extract the dataset-specific samples score, by setting the <code>use_average_dimensions</code> parameter to <code>FALSE</code>. Note that this only affects the <code>samples_score</code> tibble in terms of dimensions, but the name of the latent dimensions will change to reflect the dataset to which they refer.</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-2" role="tab" aria-controls="tabset-4-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-3" role="tab" aria-controls="tabset-4-3" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output_no_average</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">spls_final_run</span>, use_average_dimensions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">spls_output_no_average</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_latent_dimensions.html">get_latent_dimensions</a></span><span class="op">(</span><span class="va">spls_output_no_average</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "rnaseq Component 1"     "metabolome Component 1" "rnaseq Component 2"    </span></span>
<span><span class="co">#&gt; [4] "metabolome Component 2"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">$</span><span class="va">samples_score</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 278</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">spls_output_no_average</span><span class="op">$</span><span class="va">samples_score</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 556</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-4-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-2-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output_no_average</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">so2pls_final_run</span>, use_average_dimensions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">so2pls_output_no_average</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_latent_dimensions.html">get_latent_dimensions</a></span><span class="op">(</span><span class="va">so2pls_output_no_average</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "rnaseq joint component 1"        "metabolome joint component 1"   </span></span>
<span><span class="co">#&gt; [3] "rnaseq specific component 1"     "metabolome specific component 1"</span></span>
<span><span class="co">#&gt; [5] "metabolome specific component 2" "metabolome specific component 3"</span></span>
<span><span class="co">#&gt; [7] "metabolome specific component 4" "metabolome specific component 5"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">so2pls_output</span><span class="op">$</span><span class="va">samples_score</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 973</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">so2pls_output_no_average</span><span class="op">$</span><span class="va">samples_score</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1112</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-4-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-3-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output_no_average</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">diablo_final_run</span>, use_average_dimensions <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">diablo_output_no_average</span><span class="op">)</span></span>
<span>  </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_latent_dimensions.html">get_latent_dimensions</a></span><span class="op">(</span><span class="va">diablo_output_no_average</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "snps Component 1"       "rnaseq Component 1"     "metabolome Component 1"</span></span>
<span><span class="co">#&gt;  [4] "snps Component 2"       "rnaseq Component 2"     "metabolome Component 2"</span></span>
<span><span class="co">#&gt;  [7] "snps Component 3"       "rnaseq Component 3"     "metabolome Component 3"</span></span>
<span><span class="co">#&gt; [10] "snps Component 4"       "rnaseq Component 4"     "metabolome Component 4"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">diablo_output</span><span class="op">$</span><span class="va">samples_score</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 540</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">diablo_output_no_average</span><span class="op">$</span><span class="va">samples_score</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1620</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section></section><section id="percentage-of-variance-explained" class="level2" data-number="11.3"><h2 data-number="11.3" class="anchored" data-anchor-id="percentage-of-variance-explained">
<span class="header-section-number">11.3</span> Percentage of variance explained</h2>
<p>The first step in interpreting the results of a dimension reduction method is to assess how much variance is explained by the different latent dimensions for each dataset. The function <code><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained()</a></code> takes as input a <code>output_dimension_reduction</code> object and constructs a dataset-specific screeplot:</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-5-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-1" role="tab" aria-controls="tabset-5-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-2" role="tab" aria-controls="tabset-5-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-3" role="tab" aria-controls="tabset-5-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-4" role="tab" aria-controls="tabset-5-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-5-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-5-1-tab">
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/plot-variance-explained-spls-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-5-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-2-tab">
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">so2pls_output</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">9</span>, angle <span class="op">=</span> <span class="fl">30</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/plot-variance-explained-so2pls-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-5-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-3-tab">
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">mofa_output</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">9</span>, angle <span class="op">=</span> <span class="fl">30</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/plot-variance-explained-mofa-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-5-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-4-tab">
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">diablo_output</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/plot-variance-explained-diablo-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
</div>
</div>
</section><section id="sec-interpretation-samples-scores" class="level2" data-number="11.4"><h2 data-number="11.4" class="anchored" data-anchor-id="sec-interpretation-samples-scores">
<span class="header-section-number">11.4</span> Samples plot</h2>
<p>To understand the phenomenon driving the variation represented by the different latent components, we can investigate the coordinates of the samples in the space spanned by these latent components, in combination with information that we have about the samples.</p>
<p>We will start by loading the <code>MultiDataSet</code> object and generating some custom colour palettes for different samples covariates of interest, that we will use in the rest of the section.</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mo_set_complete</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Choosing colour palettes for the samples covariates</span></span>
<span><span class="va">palette_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span>
<span><span class="va">palette_feedlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span></span>
<span><span class="va">palette_geno_comp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></span>
<span><span class="va">palette_rnaseq_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For some plots will need both colour and fill</span></span>
<span><span class="va">palette_status_fill</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span></span>
<span>  palette <span class="op">=</span> <span class="st">"Set1"</span>, </span>
<span>  aesthetics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"colour"</span>, <span class="st">"fill"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="sec-interpretation-samples-ggpairs" class="level3" data-number="11.4.1"><h3 data-number="11.4.1" class="anchored" data-anchor-id="sec-interpretation-samples-ggpairs">
<span class="header-section-number">11.4.1</span> Matrix of scatter plots</h3>
<p>The <code><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score()</a></code> function shows the samples score in a matrix of scatter plots that represent all possible two-by-two combinations of the latent dimensions. The plots in the resulting matrix are redundant: the lower plots (below the diagonal) are just a rotated version of the upper plots (above the diagonal). The function takes as input an <code>output_dimension_reduction</code> object. In addition, it accepts a <code>MultiDataSet</code> object to extract information about the samples that can then be used to customise the plot. Since the plots are redundant, different properties of the samples can be shown in the upper and lower plots. For example, we will colour the samples according to disease status in the upper plots, while representing some other covariate in the lower plots. We will use the shape of the points to illustrate gender in both upper and lower plots. By default, all latent dimensions are included in the plots, but it is possible to select some of them by passing their names to the <code>latent_dimensions</code> argument. There are more options to further customise the plot, which are shown in the functionâ€™s help.</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-6-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-1" role="tab" aria-controls="tabset-6-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-2" role="tab" aria-controls="tabset-6-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-3" role="tab" aria-controls="tabset-6-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-4" role="tab" aria-controls="tabset-6-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-6-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-6-1-tab">
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  scale_colour_upper <span class="op">=</span> <span class="va">palette_status</span>,</span>
<span>  shape_upper <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>  colour_lower <span class="op">=</span> <span class="st">"feedlot"</span>,</span>
<span>  scale_colour_lower <span class="op">=</span> <span class="va">palette_feedlot</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-plot-samples-scores-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Component 1 clearly splits the control and BRD animals, while component 2 seems to separate two samples from the rest. There is no clear separation of the samples by feedlot or gender.</p>
</div>
<div id="tabset-6-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-2-tab">
<p>Here we will focus on the joint components. Since there is only one, instead of a matrix of scatterplot, the function will display the samples score with a violin plot, and the properties selected for the upper plots will be used to customise the points:</p>
<div class="cell">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"joint component 1"</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  scale_colour_upper <span class="op">=</span> <span class="va">palette_status</span>,</span>
<span>  shape_upper <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>  colour_lower <span class="op">=</span> <span class="st">"feedlot"</span>,</span>
<span>  scale_colour_lower <span class="op">=</span> <span class="va">palette_feedlot</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in plot_samples_score(so2pls_output, latent_dimensions = "joint</span></span>
<span><span class="co">#&gt; component 1", : Only one latent dimension to plot; 'colour_diag',</span></span>
<span><span class="co">#&gt; 'colour_lower' and 'shape_lower' argument will be ignored.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-plot-samples-scores-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Joint component 1 separates the BRD and control animals, although some of them overlap.</p>
</div>
<div id="tabset-6-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-3-tab">
<p>For clarity we will focus here on the first four factors:</p>
<div class="cell">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  scale_colour_upper <span class="op">=</span> <span class="va">palette_status</span>,</span>
<span>  shape_upper <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>  colour_lower <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span>  scale_colour_lower <span class="op">=</span> <span class="va">palette_geno_comp</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 2 rows containing missing values (`geom_point()`).</span></span>
<span><span class="co">#&gt; Removed 2 rows containing missing values (`geom_point()`).</span></span>
<span><span class="co">#&gt; Removed 2 rows containing missing values (`geom_point()`).</span></span>
<span><span class="co">#&gt; Removed 2 rows containing missing values (`geom_point()`).</span></span>
<span><span class="co">#&gt; Removed 2 rows containing missing values (`geom_point()`).</span></span>
<span><span class="co">#&gt; Removed 2 rows containing missing values (`geom_point()`).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-plot-samples-scores-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Factor 1 clearly separates the control and BRD animals. Factor 2 separates samples from the genomics composition cluster K1 from the other two clusters. None of the factors are separating the animals by gender, indicating that this covariate does not have a strong effect in the omics measurements.</p>
</div>
<div id="tabset-6-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-4-tab">
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  scale_colour_upper <span class="op">=</span> <span class="va">palette_status</span>,</span>
<span>  shape_upper <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>  colour_lower <span class="op">=</span> <span class="st">"rnaseq_batch"</span>,</span>
<span>  scale_colour_lower <span class="op">=</span> <span class="va">palette_rnaseq_batch</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-plot-samples-scores-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>As DIABLO is a supervised method, it is no surprise that the first component clearly separates the two disease status groups. The second latent component separates samples according to the RNAseq batch. From this plot it is not clear what components 3 and 4 represent.</p>
</div>
</div>
</div>
</section><section id="sec-interpretation-samples-pair" class="level3" data-number="11.4.2"><h3 data-number="11.4.2" class="anchored" data-anchor-id="sec-interpretation-samples-pair">
<span class="header-section-number">11.4.2</span> Scatter plot for pair of latent dimensions</h3>
<p>We can also focus on a specific pair of latent dimensions and represent the samples in the space spanned by these two dimensions, with the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair()</a></code> function. The two latent dimensions to plot are selected by passing their names to the function. This function also accepts a <code>MultiDataSet</code> object, whose samples metadata will be used to customise the plot.</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-7-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-1" role="tab" aria-controls="tabset-7-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-2" role="tab" aria-controls="tabset-7-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-3" role="tab" aria-controls="tabset-7-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-4" role="tab" aria-controls="tabset-7-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-7-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-7-1-tab">
<div class="cell">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"gender"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-plot-samples-score-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>The two disease status groups are well separated in this space.</p>
</div>
<div id="tabset-7-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-2-tab">
<p>To showcase this function, we will look at the first two metabolomics-specific latent components:</p>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"metabolome specific component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"gender"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-plot-samples-score-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>These two specific components are not related to disease status or gender.</p>
</div>
<div id="tabset-7-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-3-tab">
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"geno_comp_cluster"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status</span></span>
<span><span class="co">#&gt; Warning: Removed 2 rows containing missing values (`geom_point()`).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-plot-samples-score-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We see that these first two factors discriminate the disease status groups and the genomics composition clusters.</p>
</div>
<div id="tabset-7-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-4-tab">
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_pair.html">plot_samples_score_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"rnaseq_batch"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-plot-samples-score-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We see that these first two components discriminate the disease status groups and the RNAseq batches.</p>
</div>
</div>
</div>
</section><section id="sec-interpretation-samples-covariate" class="level3" data-number="11.4.3"><h3 data-number="11.4.3" class="anchored" data-anchor-id="sec-interpretation-samples-covariate">
<span class="header-section-number">11.4.3</span> Samples score vs covariate</h3>
<p>Lastly, we can plot the samples score for each latent dimension against a samples covariate of interest, with the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate()</a></code> function. This is useful to focus on one dimension at a time rather than pairs of them. By default, the function displays all latent dimensions, but we can focus on a subset of them through the <code>latent_dimensions</code> arguments. The function needs a <code>MultiDataSet</code> object to extract information about the samples. The plot generated will depend on whether the covariate of interest is categorical or continuous. We will show both options below. In addition to the main covariate of interest that will be displayed on the x-axis, the colour and shape of the points can be further customised according to other samples properties.</p>
<p>First, with a categorical covariate:</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-8-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-1" role="tab" aria-controls="tabset-8-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-8-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-2" role="tab" aria-controls="tabset-8-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-8-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-3" role="tab" aria-controls="tabset-8-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-8-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-4" role="tab" aria-controls="tabset-8-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-8-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-8-1-tab">
<div class="cell">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"status"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status_fill</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-plot-samples-score-covariate-cat-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-8-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-8-2-tab">
<p>For clarity we will focus on the first joint component only.</p>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"status"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"joint component 1"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status_fill</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-plot-samples-score-covariate-cat-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-8-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-8-3-tab">
<p>For clarity we will focus on the first two factors only.</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"status"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status_fill</span></span>
<span><span class="co">#&gt; Warning: Removed 4 rows containing missing values (`geom_point()`).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-plot-samples-score-covariate-cat-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-8-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-8-4-tab">
<p>For clarity we will focus on the first two components only.</p>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"status"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"rnaseq_batch"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status_fill</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-plot-samples-score-covariate-cat-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
</div>
</div>
<p>Now, with a continuous covariate:</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-9-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-1" role="tab" aria-controls="tabset-9-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-9-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-2" role="tab" aria-controls="tabset-9-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-9-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-3" role="tab" aria-controls="tabset-9-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-9-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-4" role="tab" aria-controls="tabset-9-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-9-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-9-1-tab">
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"day_on_feed"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status_fill</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-plot-samples-score-covariate-cont-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-9-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-9-2-tab">
<p>For clarity we will focus on the first joint component only.</p>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"day_on_feed"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"joint component 1"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status_fill</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-plot-samples-score-covariate-cont-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-9-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-9-3-tab">
<p>For clarity we will focus on the first two factors only.</p>
<div class="cell">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"day_on_feed"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status_fill</span></span>
<span><span class="co">#&gt; Warning: Removed 4 rows containing non-finite values (`stat_smooth()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 4 rows containing missing values (`geom_point()`).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-plot-samples-score-covariate-cont-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-9-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-9-4-tab">
<p>For clarity we will focus on the first two components only.</p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"day_on_feed"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  shape_by <span class="op">=</span> <span class="st">"rnaseq_batch"</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="va">palette_status_fill</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-plot-samples-score-covariate-cont-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="feature-plots" class="level2" data-number="11.5"><h2 data-number="11.5" class="anchored" data-anchor-id="feature-plots">
<span class="header-section-number">11.5</span> Feature plots</h2>
<p>Next, we can have a look at the features contributing to each latent dimension.</p>
<section id="sec-interpretation-features-distr" class="level3" data-number="11.5.1"><h3 data-number="11.5.1" class="anchored" data-anchor-id="sec-interpretation-features-distr">
<span class="header-section-number">11.5.1</span> Features weight distribution</h3>
<p>We can start by looking at the distribution of features weight for each latent dimension across the datasets, with the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr()</a></code> function. Again, the function takes an <code>output_dimension_reduction</code> object as first input; by default it will show all latent dimensions and datasets, but they can be specified through the <code>latent_dimensions</code> and <code>datasets</code> arguments. The function shows by default the distribution of the â€œsignedâ€ importance scores, i.e.&nbsp;the features weights normalised by the highest absolute weight value for the corresponding latent dimension and dataset, but we can also show the distribution of importance scores (which are absolute values) and non-transformed features weight (through the <code>features_metric</code> argument). Note that the function returns a <a href="https://patchwork.data-imaginist.com/">patchwork</a> of ggplots.</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-10-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-1" role="tab" aria-controls="tabset-10-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-10-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-2" role="tab" aria-controls="tabset-10-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-10-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-3" role="tab" aria-controls="tabset-10-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-10-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-4" role="tab" aria-controls="tabset-10-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-10-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-10-1-tab">
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-plot-features-weight-distr-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>As sPLS performs feature selection, most of the weights are equal to 0.</p>
</div>
<div id="tabset-10-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-10-2-tab">
<div class="cell">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr</a></span><span class="op">(</span><span class="va">so2pls_output</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">## showing two plots per column (from patchwork package)</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-plot-features-weight-distr-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>As sO2PLS performs feature selection for the joint components, most of the weights are equal to 0 for joint component 1.</p>
</div>
<div id="tabset-10-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-10-3-tab">
<p>We will only look at factors 1 to 4 for clarity:</p>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">## showing one plot per column (from patchwork package)</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-plot-features-weight-distr-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We can see for example for factor 1 that only a couple of metabolites have an importance score above 0.5, this small number of features are driving the variation captured by factor 1.</p>
</div>
<div id="tabset-10-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-10-4-tab">
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">## showing one plot per column (from patchwork package)</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-plot-features-weight-distr-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>As DIABLO performs feature selection, most of the weights are equal to zero.</p>
</div>
</div>
</div>
</section><section id="sec-interpretation-features-pair" class="level3" data-number="11.5.2"><h3 data-number="11.5.2" class="anchored" data-anchor-id="sec-interpretation-features-pair">
<span class="header-section-number">11.5.2</span> Comparing features importance between latent components</h3>
<p>We can also compare the importance score given to the features between any two latent dimensions with the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair()</a></code> function.</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-11-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-1" role="tab" aria-controls="tabset-11-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-11-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-2" role="tab" aria-controls="tabset-11-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-11-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-3" role="tab" aria-controls="tabset-11-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-11-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-4" role="tab" aria-controls="tabset-11-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-11-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-11-1-tab">
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-features-weight-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>All genes and most metabolites selected for the two components are different.</p>
</div>
<div id="tabset-11-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-11-2-tab">
<div class="cell">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"metabolome specific component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-features-weight-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>The plot shows an interesting pattern: the weights of the compound for these two specific components are negatively correlated, except for a couple of outliers.</p>
</div>
<div id="tabset-11-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-11-3-tab">
<div class="cell">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-features-weight-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We can see that the genomic markers are mostly given weights of opposite signs for the first two MOFA factors (e.g.&nbsp;positive weight for factor 1 and negative weight for factor 2). For the transcriptomics and metabolomics datasets, the importance score are mostly uncorrelated between the first two factors.</p>
</div>
<div id="tabset-11-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-11-4-tab">
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-features-weight-pair-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We can see that different features were selected for the first two latent components from each omics dataset.</p>
</div>
</div>
</div>
<p>These plots allow us to quickly check whether different latent dimensions capture different or related aspects of the data. The top 5 features according to their consensus importance score (see <a href="comparison.html#sec-comparison-consensus"><span>Section&nbsp;13.9</span></a>) are highlighted in red.</p>
<p>By default, the function shows the signed importance score of the features, i.e.&nbsp;their importance score to which the sign of their weight was added. This can be changed through the <code>features_metric</code> argument of the function.</p>
</section><section id="sec-interpretation-features-top-plot" class="level3" data-number="11.5.3"><h3 data-number="11.5.3" class="anchored" data-anchor-id="sec-interpretation-features-top-plot">
<span class="header-section-number">11.5.3</span> Plotting top features weight</h3>
<p>We can then investigate which features are assigned the highest weights for each latent dimension. These are the features driving the variation captured by the latent components. This is done with the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features()</a></code> function.</p>
<p>By default, the feature IDs will be used as labels in the plot, but by passing a <code>MultiDataSet</code> object to the function we can use a column from the features metadata table instead. To do so, we need to pass a named list to the <code>label_cols</code> argument: the names of the elements in the list should correspond to dataset names in the <code>MultiDataSet</code> object, and the value should be the name of the column in the features metadata table of the corresponding dataset to use as feature label. The number of features displayed is controlled through the <code>top_n</code> argument (the default value is 20). Note that if a lower number of features is selected for a certain dataset and latent dimension, only the selected features will be displayed. Again, the function returns a <a href="https://patchwork.data-imaginist.com/">patchwork</a> of ggplots.</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-12-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-12-1" role="tab" aria-controls="tabset-12-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-12-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-12-2" role="tab" aria-controls="tabset-12-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-12-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-12-3" role="tab" aria-controls="tabset-12-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-12-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-12-4" role="tab" aria-controls="tabset-12-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-12-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-12-1-tab">
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-plot-top-features-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We can see that citric acid and D-mannose have an importance score close to 1 for component 1. Citric acid has a negative weight, meaning that its abundance is negatively correlated with component 1, while D-mannose has a positive weight, so its abundance is positively correlated with component 1. For the second component, only two genes have an importance score above 0.5.</p>
</div>
<div id="tabset-12-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-12-2-tab">
<p>We will focus on joint component 1:</p>
<div class="cell">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"joint component 1"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-plot-top-features-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>While all top 20 genes have an importance score above 0.5 for joint component 1, only citric acid is given a high importance score for joint component 1 in the metabolomics dataset.</p>
</div>
<div id="tabset-12-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-12-3-tab">
<p>We will focus on the first two factors here:</p>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-plot-top-features-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>For factor 1, citric acid is the only metabolite with an importance score above 0.75, while in the genomics dataset four markers have a high importance score, and most of the top 20 genes have an importance score aroung 0.7 or higher.</p>
</div>
<div id="tabset-12-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-12-4-tab">
<p>For clarity, we will focus on the first two latent components:</p>
<div class="cell">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-plot-top-features-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>For component 1, we can see three markers from the genomics dataset and three genes with a high importance score. From the metabolomics dataset, less than 20 features were retained.</p>
</div>
</div>
</div>
</section><section id="sec-interpretation-features-top-table" class="level3" data-number="11.5.4"><h3 data-number="11.5.4" class="anchored" data-anchor-id="sec-interpretation-features-top-table">
<span class="header-section-number">11.5.4</span> Extracting top features</h3>
<p>It is also possible to extract information about the top contributing features as a table with the function <code><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features()</a></code>. The function offers two options:</p>
<ul>
<li><p>Return for each dataset the top N features contributing to the latent dimensions of interest. N is specified through the <code>n_features</code> argument;</p></li>
<li><p>Return for each dataset features whose importance score is at least equal to a certain threshold. The threshold is specified through the <code>min_importance</code> argument.</p></li>
</ul>
<p>The first option is preferred when many features contribute to a given factor. The second option has been implemented for cases where some latent dimensions are driven by a small number of features (so for example in the top 10 contributing features, only 3 features would actually have a high importance score). By default, the function returns the top contributing features from all datasets for all latent dimensions, but we can focus on some datasets and/or latent dimensions of interest by passing their names to the <code>datasets</code> and <code>latent_dimensions</code> arguments of the function. Additionally, we can pass to the function a <code>MultiDataSet</code> object (through the <code>mo_data</code> argument); the function will extract information about the features from the features metadata.</p>
<p>We illustrate the two possible options below. For the second option, we add features information to the table:</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-13-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-13-1" role="tab" aria-controls="tabset-13-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-13-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-13-2" role="tab" aria-controls="tabset-13-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-13-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-13-3" role="tab" aria-controls="tabset-13-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-13-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-13-4" role="tab" aria-controls="tabset-13-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-13-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-13-1-tab">
<div class="cell">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span><span class="va">spls_output</span>, n_features <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 5</span></span>
<span><span class="co">#&gt;   feature_id         dataset    latent_dimension weight importance</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;fct&gt;      &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ENSBTAG00000022715 rnaseq     Component 1      -0.318      1    </span></span>
<span><span class="co">#&gt; 2 ENSBTAG00000050618 rnaseq     Component 1      -0.302      0.950</span></span>
<span><span class="co">#&gt; 3 ENSBTAG00000018223 rnaseq     Component 1      -0.291      0.916</span></span>
<span><span class="co">#&gt; 4 HMDB00094          metabolome Component 1       0.451      1    </span></span>
<span><span class="co">#&gt; 5 HMDB00169          metabolome Component 1      -0.327      0.726</span></span>
<span><span class="co">#&gt; 6 HMDB00159          metabolome Component 1      -0.312      0.692</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  min_importance <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 31</span></span>
<span><span class="co">#&gt;   feature_id      dataset latent_dimension weight importance chromosome  p_value</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;           &lt;fct&gt;   &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ENSBTAG0000002â€¦ rnaseq  Component 1      -0.318      1     26         1.79e-33</span></span>
<span><span class="co">#&gt; 2 ENSBTAG0000005â€¦ rnaseq  Component 1      -0.302      0.950 26         2.54e-33</span></span>
<span><span class="co">#&gt; 3 ENSBTAG0000001â€¦ rnaseq  Component 1      -0.291      0.916 16         7.14e-31</span></span>
<span><span class="co">#&gt; 4 ENSBTAG0000004â€¦ rnaseq  Component 1      -0.266      0.837 15         4.14e-36</span></span>
<span><span class="co">#&gt; 5 ENSBTAG0000003â€¦ rnaseq  Component 1      -0.264      0.831 7          1.22e-35</span></span>
<span><span class="co">#&gt; 6 HMDB00094       metaboâ€¦ Component 1       0.451      1     &lt;NA&gt;       2.48e-41</span></span>
<span><span class="co">#&gt; # â„¹ 24 more variables: fdr &lt;dbl&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;,</span></span>
<span><span class="co">#&gt; #   strand &lt;fct&gt;, Name &lt;chr&gt;, description &lt;chr&gt;, log_fc &lt;dbl&gt;, log_cpm &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   f &lt;dbl&gt;, de_signif &lt;chr&gt;, de_status &lt;chr&gt;, hmdb_id &lt;chr&gt;, name &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   chemical_formula &lt;chr&gt;, monisotopic_molecular_weight &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   cas_registry_number &lt;chr&gt;, smiles &lt;chr&gt;, inchikey &lt;chr&gt;, kegg_id &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   direct_parent &lt;chr&gt;, super_class &lt;chr&gt;, t_value &lt;dbl&gt;, padj &lt;dbl&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-13-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-13-2-tab">
<div class="cell">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span><span class="va">so2pls_output</span>, n_features <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 5</span></span>
<span><span class="co">#&gt;   feature_id         dataset    latent_dimension  weight importance</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;fct&gt;      &lt;fct&gt;              &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ENSBTAG00000048835 rnaseq     joint component 1 -0.268      1    </span></span>
<span><span class="co">#&gt; 2 ENSBTAG00000049569 rnaseq     joint component 1 -0.266      0.993</span></span>
<span><span class="co">#&gt; 3 ENSBTAG00000039037 rnaseq     joint component 1 -0.248      0.925</span></span>
<span><span class="co">#&gt; 4 HMDB00094          metabolome joint component 1  0.751      1    </span></span>
<span><span class="co">#&gt; 5 HMDB00042          metabolome joint component 1  0.335      0.445</span></span>
<span><span class="co">#&gt; 6 HMDB00357          metabolome joint component 1  0.319      0.424</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  min_importance <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 31</span></span>
<span><span class="co">#&gt;   feature_id      dataset latent_dimension weight importance chromosome  p_value</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;           &lt;fct&gt;   &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ENSBTAG0000004â€¦ rnaseq  joint componentâ€¦ -0.268      1     24         5.43e-27</span></span>
<span><span class="co">#&gt; 2 ENSBTAG0000004â€¦ rnaseq  joint componentâ€¦ -0.266      0.993 NKLS02001â€¦ 1.14e-27</span></span>
<span><span class="co">#&gt; 3 ENSBTAG0000003â€¦ rnaseq  joint componentâ€¦ -0.248      0.925 NKLS02001â€¦ 1.45e-28</span></span>
<span><span class="co">#&gt; 4 ENSBTAG0000003â€¦ rnaseq  joint componentâ€¦ -0.235      0.876 7          1.22e-35</span></span>
<span><span class="co">#&gt; 5 ENSBTAG0000005â€¦ rnaseq  joint componentâ€¦ -0.227      0.847 NKLS02000â€¦ 1.31e-22</span></span>
<span><span class="co">#&gt; 6 ENSBTAG0000002â€¦ rnaseq  joint componentâ€¦ -0.216      0.805 26         1.79e-33</span></span>
<span><span class="co">#&gt; # â„¹ 24 more variables: fdr &lt;dbl&gt;, start &lt;int&gt;, end &lt;int&gt;, width &lt;int&gt;,</span></span>
<span><span class="co">#&gt; #   strand &lt;fct&gt;, Name &lt;chr&gt;, description &lt;chr&gt;, log_fc &lt;dbl&gt;, log_cpm &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   f &lt;dbl&gt;, de_signif &lt;chr&gt;, de_status &lt;chr&gt;, hmdb_id &lt;chr&gt;, name &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   chemical_formula &lt;chr&gt;, monisotopic_molecular_weight &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   cas_registry_number &lt;chr&gt;, smiles &lt;chr&gt;, inchikey &lt;chr&gt;, kegg_id &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   direct_parent &lt;chr&gt;, super_class &lt;chr&gt;, t_value &lt;dbl&gt;, padj &lt;dbl&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-13-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-13-3-tab">
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span><span class="va">mofa_output</span>, n_features <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 5</span></span>
<span><span class="co">#&gt;   feature_id             dataset latent_dimension   weight importance</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                  &lt;fct&gt;   &lt;fct&gt;               &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 BovineHD0100032240     snps    Factor 1         -0.00268      1    </span></span>
<span><span class="co">#&gt; 2 Hapmap55381-rs29025399 snps    Factor 1         -0.00268      1    </span></span>
<span><span class="co">#&gt; 3 BovineHD0100032254     snps    Factor 1          0.00249      0.927</span></span>
<span><span class="co">#&gt; 4 ENSBTAG00000049569     rnaseq  Factor 1          1.89         1    </span></span>
<span><span class="co">#&gt; 5 ENSBTAG00000048835     rnaseq  Factor 1          1.88         0.999</span></span>
<span><span class="co">#&gt; 6 ENSBTAG00000039037     rnaseq  Factor 1          1.81         0.961</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  min_importance <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 40</span></span>
<span><span class="co">#&gt;   feature_id    dataset latent_dimension   weight importance chromosome position</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;         &lt;fct&gt;   &lt;fct&gt;               &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 BovineHD0100â€¦ snps    Factor 1         -0.00268      1     1            1.13e8</span></span>
<span><span class="co">#&gt; 2 Hapmap55381-â€¦ snps    Factor 1         -0.00268      1     1            1.13e8</span></span>
<span><span class="co">#&gt; 3 BovineHD0100â€¦ snps    Factor 1          0.00249      0.927 1            1.13e8</span></span>
<span><span class="co">#&gt; 4 BovineHD2700â€¦ snps    Factor 1         -0.00225      0.839 27           3.57e7</span></span>
<span><span class="co">#&gt; 5 ENSBTAG00000â€¦ rnaseq  Factor 1          1.89         1     NKLS02001â€¦  NA     </span></span>
<span><span class="co">#&gt; 6 ENSBTAG00000â€¦ rnaseq  Factor 1          1.88         0.999 24          NA     </span></span>
<span><span class="co">#&gt; # â„¹ 33 more variables: gen_train_score &lt;dbl&gt;, ref &lt;chr&gt;, alt &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   ilmn_strand &lt;chr&gt;, customer_strand &lt;chr&gt;, norm_id &lt;dbl&gt;, qtl_type &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   qtl_effect &lt;dbl&gt;, p_value &lt;dbl&gt;, fdr &lt;dbl&gt;, start &lt;int&gt;, end &lt;int&gt;,</span></span>
<span><span class="co">#&gt; #   width &lt;int&gt;, strand &lt;fct&gt;, Name &lt;chr&gt;, description &lt;chr&gt;, log_fc &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   log_cpm &lt;dbl&gt;, f &lt;dbl&gt;, de_signif &lt;chr&gt;, de_status &lt;chr&gt;, hmdb_id &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   name &lt;chr&gt;, chemical_formula &lt;chr&gt;, monisotopic_molecular_weight &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   cas_registry_number &lt;chr&gt;, smiles &lt;chr&gt;, inchikey &lt;chr&gt;, kegg_id &lt;chr&gt;, â€¦</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-13-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-13-4-tab">
<div class="cell">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span><span class="va">diablo_output</span>, n_features <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 5</span></span>
<span><span class="co">#&gt;   feature_id         dataset latent_dimension weight importance</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;fct&gt;   &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ARS-BFGL-NGS-27468 snps    Component 1      -0.632      1    </span></span>
<span><span class="co">#&gt; 2 BovineHD2300010006 snps    Component 1       0.509      0.807</span></span>
<span><span class="co">#&gt; 3 BovineHD0300000351 snps    Component 1       0.337      0.534</span></span>
<span><span class="co">#&gt; 4 ENSBTAG00000022715 rnaseq  Component 1       0.465      1    </span></span>
<span><span class="co">#&gt; 5 ENSBTAG00000050618 rnaseq  Component 1       0.378      0.813</span></span>
<span><span class="co">#&gt; 6 ENSBTAG00000049888 rnaseq  Component 1       0.305      0.655</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  min_importance <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_complete</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 40</span></span>
<span><span class="co">#&gt;   feature_id      dataset latent_dimension weight importance chromosome position</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;           &lt;fct&gt;   &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ARS-BFGL-NGS-2â€¦ snps    Component 1      -0.632      1     1            1.34e8</span></span>
<span><span class="co">#&gt; 2 BovineHD230001â€¦ snps    Component 1       0.509      0.807 23           3.43e7</span></span>
<span><span class="co">#&gt; 3 ENSBTAG0000002â€¦ rnaseq  Component 1       0.465      1     26          NA     </span></span>
<span><span class="co">#&gt; 4 ENSBTAG0000005â€¦ rnaseq  Component 1       0.378      0.813 26          NA     </span></span>
<span><span class="co">#&gt; 5 HMDB00094       metaboâ€¦ Component 1      -0.716      1     &lt;NA&gt;        NA     </span></span>
<span><span class="co">#&gt; 6 BovineHD070000â€¦ snps    Component 2      -0.543      1     7            2.11e5</span></span>
<span><span class="co">#&gt; # â„¹ 33 more variables: gen_train_score &lt;dbl&gt;, ref &lt;chr&gt;, alt &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   ilmn_strand &lt;chr&gt;, customer_strand &lt;chr&gt;, norm_id &lt;dbl&gt;, qtl_type &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   qtl_effect &lt;dbl&gt;, p_value &lt;dbl&gt;, fdr &lt;dbl&gt;, start &lt;int&gt;, end &lt;int&gt;,</span></span>
<span><span class="co">#&gt; #   width &lt;int&gt;, strand &lt;fct&gt;, Name &lt;chr&gt;, description &lt;chr&gt;, log_fc &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   log_cpm &lt;dbl&gt;, f &lt;dbl&gt;, de_signif &lt;chr&gt;, de_status &lt;chr&gt;, hmdb_id &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   name &lt;chr&gt;, chemical_formula &lt;chr&gt;, monisotopic_molecular_weight &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   cas_registry_number &lt;chr&gt;, smiles &lt;chr&gt;, inchikey &lt;chr&gt;, kegg_id &lt;chr&gt;, â€¦</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="sec-interpretation-features-selected" class="level3" data-number="11.5.5"><h3 data-number="11.5.5" class="anchored" data-anchor-id="sec-interpretation-features-selected">
<span class="header-section-number">11.5.5</span> Extracting the selected features</h3>
<p>For methods that perform feature selection (i.e.&nbsp;sPLS, sO2PLS and DIABLO), we can extract information about the selected features with the <code><a href="https://rdrr.io/pkg/moiraine/man/get_selected_features.html">get_selected_features()</a></code> function. As with <code><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features()</a></code>, if a <code>MultiDataSet</code> object is passed to the function, it will extract information about the features from the features metadata tables and return them. If applied to the results of a method that does not perform feature selection, all features will be returned.</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-14-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-14-1" role="tab" aria-controls="tabset-14-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-14-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-14-2" role="tab" aria-controls="tabset-14-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-14-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-14-3" role="tab" aria-controls="tabset-14-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-14-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-14-4" role="tab" aria-controls="tabset-14-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-14-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-14-1-tab">
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">selected_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_selected_features.html">get_selected_features</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">selected_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 120   5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">selected_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 5</span></span>
<span><span class="co">#&gt;   feature_id         dataset latent_dimension weight importance</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;fct&gt;   &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ENSBTAG00000022715 rnaseq  Component 1      -0.318      1    </span></span>
<span><span class="co">#&gt; 2 ENSBTAG00000050618 rnaseq  Component 1      -0.302      0.950</span></span>
<span><span class="co">#&gt; 3 ENSBTAG00000018223 rnaseq  Component 1      -0.291      0.916</span></span>
<span><span class="co">#&gt; 4 ENSBTAG00000049888 rnaseq  Component 1      -0.266      0.837</span></span>
<span><span class="co">#&gt; 5 ENSBTAG00000031647 rnaseq  Component 1      -0.264      0.831</span></span>
<span><span class="co">#&gt; 6 ENSBTAG00000046158 rnaseq  Component 1      -0.248      0.781</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-14-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-14-2-tab">
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">selected_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_selected_features.html">get_selected_features</a></span><span class="op">(</span><span class="va">so2pls_output</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">selected_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1394    5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">selected_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 5</span></span>
<span><span class="co">#&gt;   feature_id         dataset latent_dimension  weight importance</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;fct&gt;   &lt;fct&gt;              &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ENSBTAG00000048835 rnaseq  joint component 1 -0.268      1    </span></span>
<span><span class="co">#&gt; 2 ENSBTAG00000049569 rnaseq  joint component 1 -0.266      0.993</span></span>
<span><span class="co">#&gt; 3 ENSBTAG00000039037 rnaseq  joint component 1 -0.248      0.925</span></span>
<span><span class="co">#&gt; 4 ENSBTAG00000031647 rnaseq  joint component 1 -0.235      0.876</span></span>
<span><span class="co">#&gt; 5 ENSBTAG00000052012 rnaseq  joint component 1 -0.227      0.847</span></span>
<span><span class="co">#&gt; 6 ENSBTAG00000022715 rnaseq  joint component 1 -0.216      0.805</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-14-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-14-3-tab">
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">selected_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_selected_features.html">get_selected_features</a></span><span class="op">(</span><span class="va">mofa_output</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">selected_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 30735     5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">selected_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 5</span></span>
<span><span class="co">#&gt;   feature_id             dataset latent_dimension   weight importance</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                  &lt;fct&gt;   &lt;fct&gt;               &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 BovineHD0100032240     snps    Factor 1         -0.00268      1    </span></span>
<span><span class="co">#&gt; 2 Hapmap55381-rs29025399 snps    Factor 1         -0.00268      1    </span></span>
<span><span class="co">#&gt; 3 BovineHD0100032254     snps    Factor 1          0.00249      0.927</span></span>
<span><span class="co">#&gt; 4 BovineHD2700010051     snps    Factor 1         -0.00225      0.839</span></span>
<span><span class="co">#&gt; 5 Hapmap44084-BTA-48504  snps    Factor 1         -0.00194      0.724</span></span>
<span><span class="co">#&gt; 6 BovineHD2100007115     snps    Factor 1          0.00190      0.709</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-14-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-14-4-tab">
<div class="cell">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">selected_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_selected_features.html">get_selected_features</a></span><span class="op">(</span><span class="va">diablo_output</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">selected_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 200   5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">selected_features</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 5</span></span>
<span><span class="co">#&gt;   feature_id         dataset latent_dimension weight importance</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;              &lt;fct&gt;   &lt;fct&gt;             &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 ARS-BFGL-NGS-27468 snps    Component 1      -0.632      1    </span></span>
<span><span class="co">#&gt; 2 BovineHD2300010006 snps    Component 1       0.509      0.807</span></span>
<span><span class="co">#&gt; 3 BovineHD0300000351 snps    Component 1       0.337      0.534</span></span>
<span><span class="co">#&gt; 4 BovineHD1900011146 snps    Component 1      -0.210      0.333</span></span>
<span><span class="co">#&gt; 5 BovineHD0900026231 snps    Component 1      -0.191      0.302</span></span>
<span><span class="co">#&gt; 6 BovineHD1100030384 snps    Component 1      -0.181      0.286</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="sec-interpretation-features-meas" class="level3" data-number="11.5.6"><h3 data-number="11.5.6" class="anchored" data-anchor-id="sec-interpretation-features-meas">
<span class="header-section-number">11.5.6</span> Features measurements</h3>
<p>Finally, we can investigate further some features of interest using the functions shown in <a href="inspecting_multidataset.html#sec-inspecting-plot-data"><span>Section&nbsp;3.5</span></a>. For the example, we will extract the top 3 contributing features from each dataset for the first latent dimension generated with the methods:</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-15-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-15-1" role="tab" aria-controls="tabset-15-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-15-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-15-2" role="tab" aria-controls="tabset-15-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-15-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-15-3" role="tab" aria-controls="tabset-15-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-15-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-15-4" role="tab" aria-controls="tabset-15-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-15-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-15-1-tab">
<div class="cell">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">spls_top_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">spls_output</span>,</span>
<span>  n_features <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"Component 1"</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">feature_id</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spls_top_features</span></span>
<span><span class="co">#&gt; [1] "ENSBTAG00000022715" "ENSBTAG00000050618" "ENSBTAG00000018223"</span></span>
<span><span class="co">#&gt; [4] "HMDB00094"          "HMDB00169"          "HMDB00159"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-15-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-15-2-tab">
<div class="cell">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">so2pls_top_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">so2pls_output</span>,</span>
<span>  n_features <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"joint component 1"</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">feature_id</span><span class="op">)</span></span>
<span></span>
<span><span class="va">so2pls_top_features</span></span>
<span><span class="co">#&gt; [1] "ENSBTAG00000048835" "ENSBTAG00000049569" "ENSBTAG00000039037"</span></span>
<span><span class="co">#&gt; [4] "HMDB00094"          "HMDB00042"          "HMDB00357"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-15-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-15-3-tab">
<div class="cell">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mofa_top_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,</span>
<span>  n_features <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"Factor 1"</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">feature_id</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mofa_top_features</span></span>
<span><span class="co">#&gt; [1] "BovineHD0100032240"     "Hapmap55381-rs29025399" "BovineHD0100032254"    </span></span>
<span><span class="co">#&gt; [4] "ENSBTAG00000049569"     "ENSBTAG00000048835"     "ENSBTAG00000039037"    </span></span>
<span><span class="co">#&gt; [7] "HMDB00094"              "HMDB00042"              "HMDB00357"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-15-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-15-4-tab">
<div class="cell">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">diablo_top_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,</span>
<span>  n_features <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"Component 1"</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">feature_id</span><span class="op">)</span></span>
<span></span>
<span><span class="va">diablo_top_features</span></span>
<span><span class="co">#&gt; [1] "ARS-BFGL-NGS-27468" "BovineHD2300010006" "BovineHD0300000351"</span></span>
<span><span class="co">#&gt; [4] "ENSBTAG00000022715" "ENSBTAG00000050618" "ENSBTAG00000049888"</span></span>
<span><span class="co">#&gt; [7] "HMDB00094"          "HMDB00042"          "HMDB00169"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>We can use the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_data_heatmap.html">plot_data_heatmap()</a></code> function to visualise the measurements for the features across the samples as a heatmap. By passing a <code>MultiDataSet</code> object to the function, we can add information about the samples and the features around the heatmap, and change the label used for the features.</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-16-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-16-1" role="tab" aria-controls="tabset-16-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-16-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-16-2" role="tab" aria-controls="tabset-16-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-16-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-16-3" role="tab" aria-controls="tabset-16-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-16-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-16-4" role="tab" aria-controls="tabset-16-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-16-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-16-1-tab">
<div class="cell">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_heatmap.html">plot_data_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="va">spls_top_features</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  samples_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"status"</span>, <span class="st">"day_on_feed"</span><span class="op">)</span>,</span>
<span>  features_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chromosome"</span><span class="op">)</span>,</span>
<span>  colours_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"status"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"gold"</span>, <span class="st">"BRD"</span> <span class="op">=</span> <span class="st">"lightblue"</span><span class="op">)</span>,</span>
<span>    <span class="st">"day_on_feed"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">70</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"pink3"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  truncate <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-plot-top-features-heatmap-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</div>
<div id="tabset-16-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-16-2-tab">
<div class="cell">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_heatmap.html">plot_data_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="va">so2pls_top_features</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  samples_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"status"</span>, <span class="st">"day_on_feed"</span><span class="op">)</span>,</span>
<span>  features_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chromosome"</span><span class="op">)</span>,</span>
<span>  colours_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"status"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"gold"</span>, <span class="st">"BRD"</span> <span class="op">=</span> <span class="st">"lightblue"</span><span class="op">)</span>,</span>
<span>    <span class="st">"day_on_feed"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">70</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"pink3"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  truncate <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-plot-top-features-heatmap-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</div>
<div id="tabset-16-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-16-3-tab">
<div class="cell">
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_heatmap.html">plot_data_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="va">mofa_top_features</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  samples_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"status"</span>, <span class="st">"day_on_feed"</span><span class="op">)</span>,</span>
<span>  features_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chromosome"</span><span class="op">)</span>,</span>
<span>  colours_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"status"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"gold"</span>, <span class="st">"BRD"</span> <span class="op">=</span> <span class="st">"lightblue"</span><span class="op">)</span>,</span>
<span>    <span class="st">"day_on_feed"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">70</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"pink3"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  truncate <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-plot-top-features-heatmap-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</div>
<div id="tabset-16-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-16-4-tab">
<div class="cell">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_heatmap.html">plot_data_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="va">diablo_top_features</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  samples_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"status"</span>, <span class="st">"day_on_feed"</span><span class="op">)</span>,</span>
<span>  features_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chromosome"</span><span class="op">)</span>,</span>
<span>  colours_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"status"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"gold"</span>, <span class="st">"BRD"</span> <span class="op">=</span> <span class="st">"lightblue"</span><span class="op">)</span>,</span>
<span>    <span class="st">"day_on_feed"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">70</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"pink3"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  truncate <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-plot-top-features-heatmap-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</div>
</div>
</div>
<p>Alternatively, we can display the featuresâ€™ measurements against some samples covariate, with the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_data_covariate.html">plot_data_covariate()</a></code> function:</p>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-17-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-17-1" role="tab" aria-controls="tabset-17-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-17-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-17-2" role="tab" aria-controls="tabset-17-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-17-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-17-3" role="tab" aria-controls="tabset-17-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-17-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-17-4" role="tab" aria-controls="tabset-17-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-17-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-17-1-tab">
<div class="cell">
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_covariate.html">plot_data_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"status"</span>,</span>
<span>  <span class="va">spls_top_features</span>,</span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>, </span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/spls-plot-top-features-covariate-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-17-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-17-2-tab">
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_covariate.html">plot_data_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"status"</span>,</span>
<span>  <span class="va">so2pls_top_features</span>,</span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>, </span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/so2pls-plot-top-features-covariate-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-17-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-17-3-tab">
<div class="cell">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_covariate.html">plot_data_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"status"</span>,</span>
<span>  <span class="va">mofa_top_features</span>,</span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>, </span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/mofa-plot-top-features-covariate-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-17-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-17-4-tab">
<div class="cell">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_covariate.html">plot_data_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,</span>
<span>  <span class="st">"status"</span>,</span>
<span>  <span class="va">diablo_top_features</span>,</span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>, </span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="interpretation_files/figure-html/diablo-plot-top-features-covariate-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="recap-targets-list" class="level2" data-number="11.6"><h2 data-number="11.6" class="anchored" data-anchor-id="recap-targets-list">
<span class="header-section-number">11.6</span> Recap â€“ targets list</h2>
<p>For convenience, here is the list of targets that we created in this section:</p>
<details><summary>
Targets list for results interpretation
</summary><p>In the targets script, before the targets list, we will add the following code to create the colour palettes used in some of the plots:</p>
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## In the preamble of the _targets.R script</span></span>
<span></span>
<span><span class="co">## Choosing colour palettes for the samples covariates</span></span>
<span><span class="va">palette_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span>
<span><span class="va">palette_feedlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span></span>
<span><span class="va">palette_geno_comp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></span>
<span><span class="va">palette_rnaseq_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="panel-tabset" data-group="method">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-18-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-18-1" role="tab" aria-controls="tabset-18-1" aria-selected="true">sPLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-18-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-18-2" role="tab" aria-controls="tabset-18-2" aria-selected="false">sO2PLS</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-18-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-18-3" role="tab" aria-controls="tabset-18-3" aria-selected="false">MOFA</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-18-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-18-4" role="tab" aria-controls="tabset-18-4" aria-selected="false">DIABLO</a></li>
</ul>
<div class="tab-content" data-group="method">
<div id="tabset-18-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-18-1-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Generating standardised output</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">spls_output</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">spls_final_run</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Percentage of variance explained</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">spls_plot_variance_explained</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Samples score matrix plot</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">spls_samples_scores_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>      <span class="va">spls_output</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      colour_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>      scale_colour_upper <span class="op">=</span> <span class="va">palette_status</span>,</span>
<span>      shape_upper <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>      colour_lower <span class="op">=</span> <span class="st">"feedlot"</span>,</span>
<span>      scale_colour_lower <span class="op">=</span> <span class="va">palette_feedlot</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Distribution of features weight</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">spls_features_weight_distribution</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plot of top contributing features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">spls_top_features_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>      <span class="va">spls_output</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>        <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Table of selected features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">spls_selected_features</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_selected_features.html">get_selected_features</a></span><span class="op">(</span><span class="va">spls_output</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="tabset-18-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-18-2-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Generating standardised output</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">so2pls_output</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">so2pls_final_run</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Percentage of variance explained</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">so2pls_plot_variance_explained</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">so2pls_output</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">9</span>, angle <span class="op">=</span> <span class="fl">30</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Samples score matrix plot</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">so2pls_samples_scores_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>      <span class="va">so2pls_output</span>,</span>
<span>      latent_dimensions <span class="op">=</span> <span class="st">"joint component 1"</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      colour_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>      scale_colour_upper <span class="op">=</span> <span class="va">palette_status</span>,</span>
<span>      shape_upper <span class="op">=</span> <span class="st">"gender"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Distribution of features weight</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">so2pls_features_weight_distribution</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr</a></span><span class="op">(</span><span class="va">so2pls_output</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plot of top contributing features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">so2pls_top_features_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>      <span class="va">so2pls_output</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>        <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Table of selected features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">so2pls_selected_features</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_selected_features.html">get_selected_features</a></span><span class="op">(</span></span>
<span>      <span class="va">so2pls_output</span>,</span>
<span>      latent_dimensions <span class="op">=</span> <span class="st">"joint component 1"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="tabset-18-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-18-3-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Generating standardised output</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_output</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">mofa_trained</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Percentage of variance explained</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_plot_variance_explained</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">mofa_output</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">9</span>, angle <span class="op">=</span> <span class="fl">30</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Samples score matrix plot</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_samples_scores_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>      <span class="va">mofa_output</span>,</span>
<span>      latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      colour_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>      scale_colour_upper <span class="op">=</span> <span class="va">palette_status</span>,</span>
<span>      shape_upper <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>      colour_lower <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span>      scale_colour_lower <span class="op">=</span> <span class="va">palette_geno_comp</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Distribution of features weight</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_features_weight_distribution</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr</a></span><span class="op">(</span></span>
<span>      <span class="va">mofa_output</span>,</span>
<span>      latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plot of top contributing features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_top_features_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>      <span class="va">mofa_output</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>        <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Table of top contributing features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mofa_top_features</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_top_features.html">get_top_features</a></span><span class="op">(</span></span>
<span>      <span class="va">mofa_output</span>,</span>
<span>      min_importance <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="tabset-18-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-18-4-tab">
<div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Generating standardised output</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_output</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">diablo_final_run</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Percentage of variance explained</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_plot_variance_explained</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">diablo_output</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Samples score matrix plot</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_samples_scores_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_output</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      colour_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>      scale_colour_upper <span class="op">=</span> <span class="va">palette_status</span>,</span>
<span>      shape_upper <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>      colour_lower <span class="op">=</span> <span class="st">"rnaseq_batch"</span>,</span>
<span>      scale_colour_lower <span class="op">=</span> <span class="va">palette_rnaseq_batch</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Distribution of features weight</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_features_weight_distribution</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_distr.html">plot_features_weight_distr</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_output</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>      <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plot of top contributing features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_top_features_plot</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>      <span class="va">diablo_output</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_complete</span>,</span>
<span>      label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>        <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Component"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Table of top contributing features</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">diablo_selected_features</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_selected_features.html">get_selected_features</a></span><span class="op">(</span><span class="va">diablo_output</span><span class="op">)</span> </span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</details>

</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./diablo.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Integration with DIABLO</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./evaluation.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>