<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>The moiraine R package user manual - 5&nbsp; Modifying the MultiDataSet object</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./preprocessing.html" rel="next">
<link href="./inspecting_multidataset.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The moiraine R package user manual</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./example_dataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">The example dataset</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Data preparation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_import.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Importing data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inspecting_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Inspecting the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./modifying_multidataset.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocessing.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data pre-processing</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#modifying-a-dataset-matrix" id="toc-modifying-a-dataset-matrix" class="nav-link active" data-scroll-target="#modifying-a-dataset-matrix"><span class="toc-section-number">5.1</span>  Modifying a dataset matrix</a></li>
  <li><a href="#adding-information-to-features-metadata" id="toc-adding-information-to-features-metadata" class="nav-link" data-scroll-target="#adding-information-to-features-metadata"><span class="toc-section-number">5.2</span>  Adding information to features metadata</a></li>
  <li><a href="#adding-information-to-samples-metadata" id="toc-adding-information-to-samples-metadata" class="nav-link" data-scroll-target="#adding-information-to-samples-metadata"><span class="toc-section-number">5.3</span>  Adding information to samples metadata</a></li>
  <li><a href="#recap-targets-list" id="toc-recap-targets-list" class="nav-link" data-scroll-target="#recap-targets-list"><span class="toc-section-number">5.4</span>  Recap – targets list</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-modifying-multidataset" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/targets/">targets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PlantandFoodResearch/moiraine">moiraine</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">MultiDataSet</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For reading data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span></span>
<span><span class="co">## For creating data-frames</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="co">## For data-frame manipulation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">## For working with lists</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this chapter, we will show how to modify the <code>MultiDataSet</code> object that we created in <a href="data_import.html"><span>Chapter&nbsp;3</span></a>.</p>
<p>As a reminder, here is what the <code>_targets.R</code> script should look like so far:</p>
<details><summary><code>_targets.R</code> script
</summary><div class="cell">
<div class="cell">

</div>
</div>
</details><p>At this stage, our different omics datasets (and associated metadata) are stored in a <code>MultiDataSet</code> object, saved in the target called <code>mo_set</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mo_set</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mo_set</span></span>
<span><span class="co">#&gt; Object of class 'MultiDataSet'</span></span>
<span><span class="co">#&gt;  . assayData: 3 elements</span></span>
<span><span class="co">#&gt;     . snps: 23036 features, 139 samples </span></span>
<span><span class="co">#&gt;     . rnaseq: 20335 features, 143 samples </span></span>
<span><span class="co">#&gt;     . metabolome: 55 features, 139 samples </span></span>
<span><span class="co">#&gt;  . featureData:</span></span>
<span><span class="co">#&gt;     . snps: 23036 rows, 13 cols (feature_id, ..., p_value)</span></span>
<span><span class="co">#&gt;     . rnaseq: 20335 rows, 8 cols (feature_id, ..., Name)</span></span>
<span><span class="co">#&gt;     . metabolome: 55 rows, 16 cols (feature_id, ..., de_signif)</span></span>
<span><span class="co">#&gt;  . rowRanges:</span></span>
<span><span class="co">#&gt;     . snps: YES</span></span>
<span><span class="co">#&gt;     . rnaseq: YES</span></span>
<span><span class="co">#&gt;     . metabolome: NO</span></span>
<span><span class="co">#&gt;  . phenoData:</span></span>
<span><span class="co">#&gt;     . snps: 139 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span><span class="co">#&gt;     . rnaseq: 143 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span><span class="co">#&gt;     . metabolome: 139 samples, 10 cols (id, ..., geno_comp_3)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are several reasons why we would want to modify the information contained in this object after it has been created. For example, we might want to apply a specific transformation to the omics dataset, using a custom function not implemented in the <code>moiraine</code> package. In that case, it would be preferable to apply the transformation on the datasets after the <code>MultiDataSet</code> object has been created and exploratory analyses have been performed, rather than on the raw data before importing it into the <code>moiraine</code> workflow.</p>
<p>In another scenario, after having created the <code>MultiDataSet</code> object, we might obtain more information about the samples and/or features that we want to add to the object’s features or samples metadata, to be used in visualisations. In that case, it is recommended to include this information in the samples or features metadata files and re-run the integration pipeline; however functions to add information to the metadata tables are included for convenience. This can also be useful if, as with this example data, the features metadata was extracted from a specialised format (e.g.&nbsp;a GFF file for the RNAseq dataset) and we want to add in some information from another source.</p>
<section id="modifying-a-dataset-matrix" class="level2" data-number="5.1"><h2 data-number="5.1" class="anchored" data-anchor-id="modifying-a-dataset-matrix">
<span class="header-section-number">5.1</span> Modifying a dataset matrix</h2>
<p>Let us imagine that, prior to performing the data transformation step (which we will see in <a href="preprocessing.html"><span>Chapter&nbsp;6</span></a>), we want to replace all zero values in the transcriptomics dataset with a small value, to avoid issues during the log-transformation process (note that this is only to illustrate this functionality, for the actual analysis we will use a different transformation that handles zero values). We will pick this small value as half of the non-null minimum value in the dataset (which will be 0.5, since we are working with count data). We can compute the new matrix of RNAseq counts:</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">rnaseq_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_datasets.html">get_datasets</a></span><span class="op">(</span><span class="va">mo_set</span><span class="op">)</span><span class="op">[[</span><span class="st">"rnaseq"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">rnaseq_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                    R9497 R5969 R5327 R5979 R9504</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005   733  1407  2919   872   740</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008     6     7     3    10    20</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009     0     1     9     0     0</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010  2693  2212  2937  2000  2345</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011     0     1     1     1     0</span></span>
<span></span>
<span><span class="va">small_val</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">rnaseq_mat</span><span class="op">[</span><span class="va">rnaseq_mat</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">small_val</span></span>
<span><span class="co">#&gt; [1] 0.5</span></span>
<span></span>
<span><span class="va">new_mat</span> <span class="op">&lt;-</span> <span class="va">rnaseq_mat</span></span>
<span><span class="va">new_mat</span><span class="op">[</span><span class="va">new_mat</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">small_val</span></span>
<span></span>
<span><span class="va">new_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                     R9497 R5969 R5327  R5979  R9504</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005  733.0  1407  2919  872.0  740.0</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008    6.0     7     3   10.0   20.0</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009    0.5     1     9    0.5    0.5</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010 2693.0  2212  2937 2000.0 2345.0</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011    0.5     1     1    1.0    0.5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In order to replace the rnaseq dataset stored in the <code>MultiDataSet</code> object with this new matrix, we pass both the object and the new matrix to the <code><a href="https://rdrr.io/pkg/moiraine/man/replace_dataset.html">replace_dataset()</a></code> function, along with the name of the omics dataset whose matrix should be replaced:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mo_set_modif</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/replace_dataset.html">replace_dataset</a></span><span class="op">(</span><span class="va">mo_set</span>, <span class="st">"rnaseq"</span>, <span class="va">new_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Checking that the replacement has been done</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_datasets.html">get_datasets</a></span><span class="op">(</span><span class="va">mo_set_modif</span><span class="op">)</span><span class="op">[[</span><span class="st">"rnaseq"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                     R9497 R5969 R5327  R5979  R9504</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005  733.0  1407  2919  872.0  740.0</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008    6.0     7     3   10.0   20.0</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009    0.5     1     9    0.5    0.5</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010 2693.0  2212  2937 2000.0 2345.0</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011    0.5     1     1    1.0    0.5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that this only works for modifying the values within an omics dataset, and not for filtering, since both features and samples number and IDs in the new dataset matrix should match the ones in the original matrix:</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mo_set_modif</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/replace_dataset.html">replace_dataset</a></span><span class="op">(</span><span class="va">mo_set</span>, <span class="st">"rnaseq"</span>, <span class="va">new_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; Error in replace_dataset(mo_set, "rnaseq", new_mat[1:10, 1:10]): 'new_data' argument has incorrect dimensions. Should have 20335 rows (features) and 143 columns (samples).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details><summary>
Click here to see a targets version of the code.
</summary><div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Replacing zero values in RNAseq dataset</span></span>
<span>  <span class="co">## (note that it is more tidy to write a function for that and call it here)</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_mat_nozero</span>,</span>
<span>    <span class="op">{</span></span>
<span>      <span class="va">rnaseq_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_datasets.html">get_datasets</a></span><span class="op">(</span><span class="va">mo_set</span><span class="op">)</span><span class="op">[[</span><span class="st">"rnaseq"</span><span class="op">]</span><span class="op">]</span></span>
<span>      <span class="va">small_val</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">rnaseq_mat</span><span class="op">[</span><span class="va">rnaseq_mat</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">]</span><span class="op">)</span></span>
<span>      <span class="va">new_mat</span> <span class="op">&lt;-</span> <span class="va">rnaseq_mat</span></span>
<span>      <span class="va">new_mat</span><span class="op">[</span><span class="va">new_mat</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">small_val</span></span>
<span></span>
<span>      <span class="va">new_mat</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">)</span>,</span>
<span></span>
<span>  <span class="co">## Replacing RNAseq dataset in MultiDataSet object</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mo_set_rnaseq_nozero</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/replace_dataset.html">replace_dataset</a></span><span class="op">(</span><span class="va">mo_set</span>, <span class="st">"rnaseq"</span>, <span class="va">rnaseq_mat_nozero</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details></section><section id="adding-information-to-features-metadata" class="level2" data-number="5.2"><h2 data-number="5.2" class="anchored" data-anchor-id="adding-information-to-features-metadata">
<span class="header-section-number">5.2</span> Adding information to features metadata</h2>
<p>In the case of the transcriptomics dataset, we extracted the features metadata directly from a GFF file, which provides information about the genome annotation used. However, we might want to add information about the genes from a different source. We could add this information to the data-frame generated with <code><a href="https://rdrr.io/pkg/moiraine/man/import_fmetadata_gff.html">import_fmetadata_gff()</a></code> (see <a href="data_import.html#sec-import-fmeta-gff"><span>Section&nbsp;3.3.3</span></a>) before creating the<code>MultiDataSet</code> object, but we will demonstrate here how to add information once we’ve already created the object. Note that we will use targets for this example, as we will incorporate these changes in our analysis pipeline.</p>
<p>Let’s start by reading in the differential expression results:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_de_res_file</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span></span>
<span>      <span class="st">"extdata/transcriptomics_de_results.csv"</span>,</span>
<span>      package <span class="op">=</span> <span class="st">"moiraine"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    format <span class="op">=</span> <span class="st">"file"</span></span>
<span>  <span class="op">)</span>,</span>
<span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_de_res_df</span>,</span>
<span>    <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">rnaseq_de_res_file</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>feature_id <span class="op">=</span> <span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"rnaseq"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Notice that in the results file, the gene IDs are stored in the <code>gene_id</code> column. Here, we rename this column as <code>feature_id</code>, which is required for adding it to the features metadata. In addition, we create a <code>dataset</code> column which contains the name of the dataset in the <code>MultiDataSet</code> object to which the features belong. This is also necessary.</p>
<p>The differential results look like this:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">rnaseq_de_res_df</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 9</span></span>
<span><span class="co">#&gt;   feature_id  log_fc log_cpm     f  p_value      fdr de_signif de_status dataset</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;        &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;  </span></span>
<span><span class="co">#&gt; 1 ENSBTAG000…   4.61    4.19  358. 5.32e-40 5.93e-36 DE        upregula… rnaseq </span></span>
<span><span class="co">#&gt; 2 ENSBTAG000…   3.80    6.83  356. 6.33e-40 5.93e-36 DE        upregula… rnaseq </span></span>
<span><span class="co">#&gt; 3 ENSBTAG000…   5.41    2.24  347. 2.41e-39 1.50e-35 DE        upregula… rnaseq </span></span>
<span><span class="co">#&gt; 4 ENSBTAG000…   4.34    3.52  344. 3.40e-39 1.59e-35 DE        upregula… rnaseq </span></span>
<span><span class="co">#&gt; 5 ENSBTAG000…   2.18    6.74  327. 4.05e-38 1.52e-34 DE        upregula… rnaseq </span></span>
<span><span class="co">#&gt; 6 ENSBTAG000…  -1.31    2.64  316. 2.39e-37 7.48e-34 Not DE    Not DE    rnaseq</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now use the <code><a href="https://rdrr.io/pkg/moiraine/man/add_features_metadata.html">add_features_metadata()</a></code> function to add this table to the features metadata of the transcriptomics dataset. The new <code>MultiDataSet</code> object that includes information about the differential expression results will be saved in the <code>mo_set_de</code> target:</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_de</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/add_features_metadata.html">add_features_metadata</a></span><span class="op">(</span><span class="va">mo_set</span>, <span class="va">rnaseq_de_res_df</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The new information has been added to the features metadata of the transcriptomics dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">mo_set_de</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_features_metadata.html">get_features_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html">pluck</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                            feature_id chromosome    start      end  width</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005 ENSBTAG00000000005         17 65389743 65505336 115594</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008 ENSBTAG00000000008         29 32214439 32244810  30372</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009 ENSBTAG00000000009         18 12338037 12342272   4236</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010 ENSBTAG00000000010         21 34209956 34223394  13439</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011 ENSBTAG00000000011          8  7950815  7971600  20786</span></span>
<span><span class="co">#&gt; ENSBTAG00000000012 ENSBTAG00000000012         20 33708626 33732944  24319</span></span>
<span><span class="co">#&gt;                    strand  Name</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005      +  GRK3</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008      - KCNJ1</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009      + FOXF1</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010      +  UBL7</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011      -   TDH</span></span>
<span><span class="co">#&gt; ENSBTAG00000000012      + TTC33</span></span>
<span><span class="co">#&gt;                                                                                                       description</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005                        G protein-coupled receptor kinase 3 [Source:VGNC Symbol;Acc:VGNC:53904]</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008 potassium inwardly rectifying channel subfamily J member 1 [Source:VGNC Symbol;Acc:VGNC:30453]</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009                                            forkhead box F1 [Source:VGNC Symbol;Acc:VGNC:29084]</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010                                           ubiquitin like 7 [Source:VGNC Symbol;Acc:VGNC:50128]</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011                                 L-threonine dehydrogenase [Source:VGNC Symbol;Acc:VGNC:108945]</span></span>
<span><span class="co">#&gt; ENSBTAG00000000012                         tetratricopeptide repeat domain 33 [Source:VGNC Symbol;Acc:VGNC:36471]</span></span>
<span><span class="co">#&gt;                         log_fc    log_cpm          f      p_value          fdr</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005  0.13603041  5.9051177  4.1626465 4.324808e-02 7.170137e-02</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008 -0.12965356 -0.7437052  1.0676845 3.032916e-01 3.938643e-01</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009  1.27141158 -2.5627934 23.9562951 2.731328e-06 9.665375e-06</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010  0.39567729  6.2563594 60.5303416 1.581955e-12 1.521160e-11</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011  0.07766873 -2.7608361  0.1418457 7.070363e-01 7.773874e-01</span></span>
<span><span class="co">#&gt; ENSBTAG00000000012  0.15756169  3.6628775 11.0448775 1.141125e-03 2.623062e-03</span></span>
<span><span class="co">#&gt;                    de_signif de_status</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011    Not DE    Not DE</span></span>
<span><span class="co">#&gt; ENSBTAG00000000012    Not DE    Not DE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that with this function, we can add information about features from different datasets at once, which is why there needs to be a <code>dataset</code> column in the data-frame to add, indicating the dataset to which each feature belongs. Also, not all features from a given dataset need to be present in this new data-frame; it is possible to add information for only a subset of them. In that case, the function will throw a warning giving the number of features from the corresponding dataset missing from the new data-frame, and the new columns in the corresponding features metadata will be filled with <code>NA</code> for features not present. However, it is only possible to add columns that do not already exist in the corresponding features metadata.</p>
</section><section id="adding-information-to-samples-metadata" class="level2" data-number="5.3"><h2 data-number="5.3" class="anchored" data-anchor-id="adding-information-to-samples-metadata">
<span class="header-section-number">5.3</span> Adding information to samples metadata</h2>
<p>Similarly, we can add a data-frame of information to the samples metadata. Here, we will create a table that contains “new” simulated information about the samples that we want to incorporate in our <code>MultiDataSet</code> object.</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Getting the list of samples ID across the datasets</span></span>
<span><span class="va">samples_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_samples.html">get_samples</a></span><span class="op">(</span><span class="va">mo_set</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unname.html">unname</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Simulating new information table, with new samples grouping</span></span>
<span><span class="va">new_samples_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>id <span class="op">=</span> <span class="va">samples_list</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>new_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">new_samples_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 2</span></span>
<span><span class="co">#&gt;   id    new_group</span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;    </span></span>
<span><span class="co">#&gt; 1 R21   b        </span></span>
<span><span class="co">#&gt; 2 Y3660 b        </span></span>
<span><span class="co">#&gt; 3 Y3243 c        </span></span>
<span><span class="co">#&gt; 4 R5764 a        </span></span>
<span><span class="co">#&gt; 5 P4669 a        </span></span>
<span><span class="co">#&gt; 6 R5452 b</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that the sample IDs must be stored in a column named <code>id</code>. We will use the <code><a href="https://rdrr.io/pkg/moiraine/man/add_samples_metadata.html">add_samples_metadata()</a></code> function to add this new data-frame to the samples metadata in our <code>MultiDataSet</code> object. There are several options for which samples metadata tables should be modified; this is controlled through the <code>datasets</code> argument of the function. By default, the information is added to the samples metadata table of all omics datasets (case when <code>datasets = NULL</code>):</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mo_set_new_samples_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/add_samples_metadata.html">add_samples_metadata</a></span><span class="op">(</span><span class="va">mo_set</span>, <span class="va">new_samples_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: snps dataset: 5 sample IDs not in 'mo_data', will be removed from</span></span>
<span><span class="co">#&gt; samples metadata.</span></span>
<span><span class="co">#&gt; Warning: rnaseq dataset: 1 sample IDs not in 'mo_data', will be removed from</span></span>
<span><span class="co">#&gt; samples metadata.</span></span>
<span><span class="co">#&gt; Warning: metabolome dataset: 5 sample IDs not in 'mo_data', will be removed</span></span>
<span><span class="co">#&gt; from samples metadata.</span></span>
<span></span>
<span><span class="va">mo_set_new_samples_info</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_samples_metadata.html">get_samples_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">head</span><span class="op">)</span></span>
<span><span class="co">#&gt; $snps</span></span>
<span><span class="co">#&gt;          id feedlot gender  status day_on_feed rnaseq_batch geno_comp_1</span></span>
<span><span class="co">#&gt; R21     R21      F1 female Control          31           B2    0.007853</span></span>
<span><span class="co">#&gt; Y3660 Y3660      F1   male Control          19           B2    0.852220</span></span>
<span><span class="co">#&gt; Y3243 Y3243      F1   male Control          16           B2    0.044171</span></span>
<span><span class="co">#&gt; R5764 R5764      F2   male Control          46           B1    0.213094</span></span>
<span><span class="co">#&gt; P4669 P4669      F3   male     BRD          35           B2    0.389393</span></span>
<span><span class="co">#&gt; R5452 R5452      F2   male Control          49           B1    0.265998</span></span>
<span><span class="co">#&gt;       geno_comp_2 geno_comp_3 geno_comp_cluster new_group</span></span>
<span><span class="co">#&gt; R21      0.820691    0.171456                K3         b</span></span>
<span><span class="co">#&gt; Y3660    0.093585    0.054195                K2         b</span></span>
<span><span class="co">#&gt; Y3243    0.190262    0.765567                K1         c</span></span>
<span><span class="co">#&gt; R5764    0.676299    0.110607                K3         a</span></span>
<span><span class="co">#&gt; P4669    0.503471    0.107136                K3         a</span></span>
<span><span class="co">#&gt; R5452    0.733992    0.000010                K3         b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $rnaseq</span></span>
<span><span class="co">#&gt;          id feedlot gender status day_on_feed rnaseq_batch geno_comp_1</span></span>
<span><span class="co">#&gt; R9497 R9497      F2   male    BRD          35           B1    0.153716</span></span>
<span><span class="co">#&gt; R5969 R5969      F2   male    BRD          24           B1    0.110663</span></span>
<span><span class="co">#&gt; R5327 R5327      F2   male    BRD          38           B1    0.140730</span></span>
<span><span class="co">#&gt; R5979 R5979      F2   male    BRD          30           B1    0.286733</span></span>
<span><span class="co">#&gt; R9504 R9504      F2   male    BRD          31           B1    0.000010</span></span>
<span><span class="co">#&gt; R5994 R5994      F2   male    BRD          24           B1    0.129271</span></span>
<span><span class="co">#&gt;       geno_comp_2 geno_comp_3 geno_comp_cluster new_group</span></span>
<span><span class="co">#&gt; R9497    0.744505    0.101779                K3         a</span></span>
<span><span class="co">#&gt; R5969    0.625823    0.263514                K3         c</span></span>
<span><span class="co">#&gt; R5327    0.809396    0.049874                K3         c</span></span>
<span><span class="co">#&gt; R5979    0.107794    0.605473                K1         b</span></span>
<span><span class="co">#&gt; R9504    0.998913    0.001077                K3         a</span></span>
<span><span class="co">#&gt; R5994    0.034351    0.836377                K1         a</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $metabolome</span></span>
<span><span class="co">#&gt;          id feedlot gender  status day_on_feed rnaseq_batch geno_comp_1</span></span>
<span><span class="co">#&gt; R21     R21      F1 female Control          31           B2    0.007853</span></span>
<span><span class="co">#&gt; Y3660 Y3660      F1   male Control          19           B2    0.852220</span></span>
<span><span class="co">#&gt; Y3243 Y3243      F1   male Control          16           B2    0.044171</span></span>
<span><span class="co">#&gt; R5764 R5764      F2   male Control          46           B1    0.213094</span></span>
<span><span class="co">#&gt; P4669 P4669      F3   male     BRD          35           B2    0.389393</span></span>
<span><span class="co">#&gt; R5452 R5452      F2   male Control          49           B1    0.265998</span></span>
<span><span class="co">#&gt;       geno_comp_2 geno_comp_3 geno_comp_cluster new_group</span></span>
<span><span class="co">#&gt; R21      0.820691    0.171456                K3         b</span></span>
<span><span class="co">#&gt; Y3660    0.093585    0.054195                K2         b</span></span>
<span><span class="co">#&gt; Y3243    0.190262    0.765567                K1         c</span></span>
<span><span class="co">#&gt; R5764    0.676299    0.110607                K3         a</span></span>
<span><span class="co">#&gt; P4669    0.503471    0.107136                K3         a</span></span>
<span><span class="co">#&gt; R5452    0.733992    0.000010                K3         b</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, it is also possible to specify for which dataset(s) the changes should be made, by passing their name to the <code>datasets</code> argument.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mo_set_new_samples_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/add_samples_metadata.html">add_samples_metadata</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set</span>, </span>
<span>  <span class="va">new_samples_df</span>, </span>
<span>  datasets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: rnaseq dataset: 1 sample IDs not in 'mo_data', will be removed from</span></span>
<span><span class="co">#&gt; samples metadata.</span></span>
<span><span class="co">#&gt; Warning: metabolome dataset: 5 sample IDs not in 'mo_data', will be removed</span></span>
<span><span class="co">#&gt; from samples metadata.</span></span>
<span></span>
<span><span class="va">mo_set_new_samples_info</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_samples_metadata.html">get_samples_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">head</span><span class="op">)</span></span>
<span><span class="co">#&gt; $snps</span></span>
<span><span class="co">#&gt;          id feedlot gender  status day_on_feed rnaseq_batch geno_comp_1</span></span>
<span><span class="co">#&gt; R21     R21      F1 female Control          31           B2    0.007853</span></span>
<span><span class="co">#&gt; Y3660 Y3660      F1   male Control          19           B2    0.852220</span></span>
<span><span class="co">#&gt; Y3243 Y3243      F1   male Control          16           B2    0.044171</span></span>
<span><span class="co">#&gt; R5764 R5764      F2   male Control          46           B1    0.213094</span></span>
<span><span class="co">#&gt; P4669 P4669      F3   male     BRD          35           B2    0.389393</span></span>
<span><span class="co">#&gt; R5452 R5452      F2   male Control          49           B1    0.265998</span></span>
<span><span class="co">#&gt;       geno_comp_2 geno_comp_3 geno_comp_cluster</span></span>
<span><span class="co">#&gt; R21      0.820691    0.171456                K3</span></span>
<span><span class="co">#&gt; Y3660    0.093585    0.054195                K2</span></span>
<span><span class="co">#&gt; Y3243    0.190262    0.765567                K1</span></span>
<span><span class="co">#&gt; R5764    0.676299    0.110607                K3</span></span>
<span><span class="co">#&gt; P4669    0.503471    0.107136                K3</span></span>
<span><span class="co">#&gt; R5452    0.733992    0.000010                K3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $rnaseq</span></span>
<span><span class="co">#&gt;          id feedlot gender status day_on_feed rnaseq_batch geno_comp_1</span></span>
<span><span class="co">#&gt; R9497 R9497      F2   male    BRD          35           B1    0.153716</span></span>
<span><span class="co">#&gt; R5969 R5969      F2   male    BRD          24           B1    0.110663</span></span>
<span><span class="co">#&gt; R5327 R5327      F2   male    BRD          38           B1    0.140730</span></span>
<span><span class="co">#&gt; R5979 R5979      F2   male    BRD          30           B1    0.286733</span></span>
<span><span class="co">#&gt; R9504 R9504      F2   male    BRD          31           B1    0.000010</span></span>
<span><span class="co">#&gt; R5994 R5994      F2   male    BRD          24           B1    0.129271</span></span>
<span><span class="co">#&gt;       geno_comp_2 geno_comp_3 geno_comp_cluster new_group</span></span>
<span><span class="co">#&gt; R9497    0.744505    0.101779                K3         a</span></span>
<span><span class="co">#&gt; R5969    0.625823    0.263514                K3         c</span></span>
<span><span class="co">#&gt; R5327    0.809396    0.049874                K3         c</span></span>
<span><span class="co">#&gt; R5979    0.107794    0.605473                K1         b</span></span>
<span><span class="co">#&gt; R9504    0.998913    0.001077                K3         a</span></span>
<span><span class="co">#&gt; R5994    0.034351    0.836377                K1         a</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $metabolome</span></span>
<span><span class="co">#&gt;          id feedlot gender  status day_on_feed rnaseq_batch geno_comp_1</span></span>
<span><span class="co">#&gt; R21     R21      F1 female Control          31           B2    0.007853</span></span>
<span><span class="co">#&gt; Y3660 Y3660      F1   male Control          19           B2    0.852220</span></span>
<span><span class="co">#&gt; Y3243 Y3243      F1   male Control          16           B2    0.044171</span></span>
<span><span class="co">#&gt; R5764 R5764      F2   male Control          46           B1    0.213094</span></span>
<span><span class="co">#&gt; P4669 P4669      F3   male     BRD          35           B2    0.389393</span></span>
<span><span class="co">#&gt; R5452 R5452      F2   male Control          49           B1    0.265998</span></span>
<span><span class="co">#&gt;       geno_comp_2 geno_comp_3 geno_comp_cluster new_group</span></span>
<span><span class="co">#&gt; R21      0.820691    0.171456                K3         b</span></span>
<span><span class="co">#&gt; Y3660    0.093585    0.054195                K2         b</span></span>
<span><span class="co">#&gt; Y3243    0.190262    0.765567                K1         c</span></span>
<span><span class="co">#&gt; R5764    0.676299    0.110607                K3         a</span></span>
<span><span class="co">#&gt; P4669    0.503471    0.107136                K3         a</span></span>
<span><span class="co">#&gt; R5452    0.733992    0.000010                K3         b</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In both cases, the function throws some warnings to alert about samples missing from this new table, or samples that are not present in the original samples metadata table. These warnings should be checked to avoid issues due to typos, etc.</p>
<p>As with the <code><a href="https://rdrr.io/pkg/moiraine/man/add_features_metadata.html">add_features_metadata()</a></code> function, it is possible to add information about only a subset of the samples; however the columns in the new data-frame must not already be present in the features metadata tables to which it will be added.</p>
</section><section id="recap-targets-list" class="level2" data-number="5.4"><h2 data-number="5.4" class="anchored" data-anchor-id="recap-targets-list">
<span class="header-section-number">5.4</span> Recap – targets list</h2>
<p>For convenience, here is the list of targets that we created in this section:</p>
<details><summary>
Targets list for modifying a MultiDataSet object
</summary><div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## RNAseq differential expression results file</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_de_res_file</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span></span>
<span>      <span class="st">"extdata/transcriptomics_de_results.csv"</span>,</span>
<span>      package <span class="op">=</span> <span class="st">"moiraine"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    format <span class="op">=</span> <span class="st">"file"</span></span>
<span>  <span class="op">)</span>,</span>
<span></span>
<span>  <span class="co">## Reading the RNAseq differential expression results</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">rnaseq_de_res_df</span>,</span>
<span>    <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">rnaseq_de_res_file</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>feature_id <span class="op">=</span> <span class="va">gene_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"rnaseq"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span></span>
<span>  <span class="co">## Adding the differential expression results to the MultiDataSet object</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mo_set_de</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/add_features_metadata.html">add_features_metadata</a></span><span class="op">(</span><span class="va">mo_set</span>, <span class="va">rnaseq_de_res_df</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>

</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./inspecting_multidataset.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Inspecting the <code>MultiDataSet</code> object</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./preprocessing.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data pre-processing</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>