<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>The moiraine R package user manual - 1&nbsp; Overview of moiraine</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./example_dataset.html" rel="next">
<link href="./index.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Overview of <code>moiraine</code></span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The <code>moiraine</code> R package user manual</a> 
        <div class="sidebar-tools-main">
    <a href="" title="" id="sidebar-tool-dropdown-0" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-github"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-0">
<li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/PlantandFoodResearch/moiraine-manual">
          Manual source code
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/PlantandFoodResearch/moiraine">
          Package source code
          </a>
        </li>
    </ul>
<a href="https://twitter.com/intent/tweet?url=%7Curl%7C" title="Twitter" class="sidebar-tool px-1"><i class="bi bi-twitter"></i></a>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Introduction</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./overview.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Overview of <code>moiraine</code></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./example_dataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">The example dataset</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Data preparation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_import.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Importing data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inspecting_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Inspecting the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./modifying_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocessing.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data pre-processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prefiltering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Dataset pre-filtering</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Unsupervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./spls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Integration with sPLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./so2pls.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Integration with sO2PLS</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mofa.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Integration with MOFA</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">Supervised integration</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./diablo.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Integration with DIABLO</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">Results interpretation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./interpretation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Interpreting the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./evaluation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Evaluating the integration results</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./comparison.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Comparing the integration results</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#input-data" id="toc-input-data" class="nav-link active" data-scroll-target="#input-data"><span class="toc-section-number">1.1</span>  Input data</a></li>
  <li><a href="#data-pre-processing" id="toc-data-pre-processing" class="nav-link" data-scroll-target="#data-pre-processing"><span class="toc-section-number">1.2</span>  Data pre-processing</a></li>
  <li><a href="#data-pre-filtering" id="toc-data-pre-filtering" class="nav-link" data-scroll-target="#data-pre-filtering"><span class="toc-section-number">1.3</span>  Data pre-filtering</a></li>
  <li><a href="#multi-omics-data-integration" id="toc-multi-omics-data-integration" class="nav-link" data-scroll-target="#multi-omics-data-integration"><span class="toc-section-number">1.4</span>  Multi-omics data integration</a></li>
  <li><a href="#interpreting-the-integration-results" id="toc-interpreting-the-integration-results" class="nav-link" data-scroll-target="#interpreting-the-integration-results"><span class="toc-section-number">1.5</span>  Interpreting the integration results</a></li>
  <li><a href="#evaluating-the-integration-results" id="toc-evaluating-the-integration-results" class="nav-link" data-scroll-target="#evaluating-the-integration-results"><span class="toc-section-number">1.6</span>  Evaluating the integration results</a></li>
  <li><a href="#comparison-different-integration-results" id="toc-comparison-different-integration-results" class="nav-link" data-scroll-target="#comparison-different-integration-results"><span class="toc-section-number">1.7</span>  Comparison different integration results</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-overview" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Overview of <code>moiraine</code></span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">

</div>
<p>In this chapter, we provide an overview of the capabilities of <code>moiraine</code>. Some code is presented for illustration, but for more details readers will be directed towards the corresponding chapter in the manual.</p>
<div class="cell">
<details><summary>Loading packages</summary><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PlantandFoodResearch/moiraine">moiraine</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For custom colour palettes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize">circlize</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For working with lists</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For visualising sO2PLS summary</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">OmicsPLS</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="input-data" class="level2" data-number="1.1"><h2 data-number="1.1" class="anchored" data-anchor-id="input-data">
<span class="header-section-number">1.1</span> Input data</h2>
<p>The first step is to import the omics datasets and associated information into R. For each omics dataset, the <code>moiraine</code> package expects three pieces of information, which are read from csv files (or other specific formats when possible such as gtf or gff3):</p>
<ul>
<li><p>the measurements of omics features across the samples (the dataset)</p></li>
<li><p>information about the omics features measured (the features metadata)</p></li>
<li><p>information about the samples measured (the samples metadata)</p></li>
</ul>
<p>An example of input files is shown below:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="images/input_data_example.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">An example of input files for <code>moiraine</code> (first 10 rows and 4 columns only): A) RNAseq read counts across genes (rows) and samples (columns). B) Table of features metadata, providing information about the genes measured. c) Table of samples metadata, i.e.&nbsp;information about the animals that were observed.</figcaption><p></p>
</figure>
</div>
<p><code>moiraine</code> uses the <a href="https://bioconductor.org/packages/release/bioc/html/MultiDataSet.html"><code>MultiDataSet</code> package</a> to store the omics datasets and associated metadata into a single R object, which can be used as input for the different <code>moiraine</code> functions (see <a href="data_import.html"><span>Chapter&nbsp;3</span></a> for details about data import). This ensures that these functions can be used regardless of the number or type of omics datasets to analyse. At any point, the datasets, features and samples metadata can easily be extracted from the <code>MultiDataSet</code> object through <code><a href="https://rdrr.io/pkg/moiraine/man/get_datasets.html">get_datasets()</a></code>, <code><a href="https://rdrr.io/pkg/moiraine/man/get_features_metadata.html">get_features_metadata()</a></code> and <code><a href="https://rdrr.io/pkg/moiraine/man/get_samples_metadata.html">get_samples_metadata()</a></code>.</p>
<p>Here is an example of a <code>MultiDataSet</code> object:</p>
<div class="cell">
<pre><code>#&gt; Object of class 'MultiDataSet'
#&gt;  . assayData: 3 elements
#&gt;     . snps: 23036 features, 139 samples 
#&gt;     . rnaseq: 20335 features, 143 samples 
#&gt;     . metabolome: 55 features, 139 samples 
#&gt;  . featureData:
#&gt;     . snps: 23036 rows, 13 cols (feature_id, ..., p_value)
#&gt;     . rnaseq: 20335 rows, 15 cols (feature_id, ..., de_signif)
#&gt;     . metabolome: 55 rows, 16 cols (feature_id, ..., de_signif)
#&gt;  . rowRanges:
#&gt;     . snps: YES
#&gt;     . rnaseq: YES
#&gt;     . metabolome: NO
#&gt;  . phenoData:
#&gt;     . snps: 139 samples, 10 cols (id, ..., geno_comp_3)
#&gt;     . rnaseq: 143 samples, 10 cols (id, ..., geno_comp_3)
#&gt;     . metabolome: 139 samples, 10 cols (id, ..., geno_comp_3)</code></pre>
</div>
<p>Importantly, this means that all the information that we have about the omics features and samples are available to <code>moiraine</code> for customising plots or outputs. For example, it is possible to quickly generate a heatmap displaying the measurements for specific features of interest, and add information about the features and samples to facilitate the interpretation of the plot:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">interesting_features</span><span class="op">)</span> <span class="co">## vector of feature IDs</span></span>
<span><span class="co">#&gt; [1] "ARS-BFGL-NGS-27468" "BovineHD2300010006" "BovineHD0300000351"</span></span>
<span><span class="co">#&gt; [4] "BovineHD1900011146" "BovineHD0900026231" "ENSBTAG00000022715"</span></span>
<span></span>
<span><span class="va">colours_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>                       </span>
<span>  <span class="st">"status"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"gold"</span>, <span class="st">"BRD"</span> <span class="op">=</span> <span class="st">"lightblue"</span><span class="op">)</span>,</span>
<span>  <span class="st">"day_on_feed"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">70</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"pink3"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="st">"de_status"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"downregulated"</span> <span class="op">=</span> <span class="st">"deepskyblue"</span>, </span>
<span>                  <span class="st">"upregulated"</span> <span class="op">=</span> <span class="st">"chartreuse3"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_data_heatmap.html">plot_data_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set</span>,                                    <span class="co"># the MultiDataSet object</span></span>
<span>  <span class="va">interesting_features</span>,                      <span class="co"># vector of feature IDs of interest</span></span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,                             <span class="co"># centering and scaling data for</span></span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,                              <span class="co">#   easier visualisation</span></span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,                 <span class="co"># hide sample IDs</span></span>
<span>  only_common_samples <span class="op">=</span> <span class="cn">TRUE</span>,                <span class="co"># only samples present in all omics</span></span>
<span>  samples_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"status"</span>, <span class="st">"day_on_feed"</span><span class="op">)</span>, <span class="co"># add info about samples</span></span>
<span>  features_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"de_status"</span><span class="op">)</span>,            <span class="co"># add info about features</span></span>
<span>  colours_list <span class="op">=</span> <span class="va">colours_list</span>,               <span class="co"># customise colours</span></span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>                         <span class="co"># specify features label</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,                       <span class="co">#   from features metadata</span></span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/plot-data-heatmap-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>Similarly, a number of functions allow to quickly summarise different aspects of the multi-omics dataset, such as creating an upset plot to compare the samples present in each omics dataset (<code><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_upset.html">plot_samples_upset()</a></code>), or generating a density plot for each omics dataset (<code><a href="https://rdrr.io/pkg/moiraine/man/plot_density_data.html">plot_density_data()</a></code>). See <a href="inspecting_multidataset.html"><span>Chapter&nbsp;4</span></a> for more details about the different visualisations and summary functions implemented.</p>
</section><section id="data-pre-processing" class="level2" data-number="1.2"><h2 data-number="1.2" class="anchored" data-anchor-id="data-pre-processing">
<span class="header-section-number">1.2</span> Data pre-processing</h2>
<p>Target factories have been implemented to facilitate the application of similar tasks across the different omics datasets. For example, the <code><a href="https://rdrr.io/pkg/moiraine/man/transformation_datasets_factory.html">transformation_datasets_factory()</a></code> function generates a sequence of targets to apply one of many possible transformations (from the <a href="https://bioconductor.org/packages/release/bioc/html/vsn.html"><code>vsn</code></a>, <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html"><code>DESeq2</code></a>, or <a href="https://cran.r-project.org/web/packages/bestNormalize/index.html"><code>bestNormalize</code></a> packages, for example) on each omics dataset, store information about each transformation performed, and generate a new <code>MultiDataSet</code> object in which the omics measurements have been transformed:</p>
<details><summary>
Code
</summary><div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/transformation_datasets_factory.html">transformation_datasets_factory</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set</span>,                                      <span class="co"># MultiDataSet object</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"vst-deseq2"</span>,                   <span class="co"># VST through DESeq2 for RNAseq</span></span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"logx"</span><span class="op">)</span>,                    <span class="co"># log2-transf. for NMR dataset</span></span>
<span>  log_bases <span class="op">=</span> <span class="fl">2</span>,                               <span class="co"># Base for log transformation</span></span>
<span>  pre_log_functions <span class="op">=</span> <span class="va">zero_to_half_min</span>,        <span class="co"># Handling 0s in log2-transf.</span></span>
<span>  transformed_data_name <span class="op">=</span> <span class="st">"mo_set_transformed"</span> <span class="co"># New MultiDataSet object</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details><p><img src="images/dag_transformation.png" class="img-fluid"></p>
<p>Note that there is also the option for users to apply their own custom transformations to the datasets (see <a href="modifying_multidataset.html"><span>Chapter&nbsp;5</span></a>).</p>
<p>Similarly, the <code>pca_complete_data_factory</code> generates a list of targets to run a PCA on each omics dataset via the <a href="https://bioconductor.org/packages/release/bioc/html/pcaMethods.html"><code>pcaMethods</code> package</a>, and if necessary imputes missing values through NIPALS-PCA. The PCA results can be easily visualised for all or specific omics datasets:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_screeplot_pca.html">plot_screeplot_pca</a></span><span class="op">(</span><span class="va">pca_runs_list</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/pca-screeplot-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_coordinates_pca.html">plot_samples_coordinates_pca</a></span><span class="op">(</span></span>
<span>  <span class="va">pca_runs_list</span>,                              <span class="co"># List of PCA results</span></span>
<span>  datasets <span class="op">=</span> <span class="st">"snps"</span>,                          <span class="co"># Dataset to plot</span></span>
<span>  pcs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,                                  <span class="co"># Principal components to display</span></span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set</span>,                           <span class="co"># MultiDataSet object</span></span>
<span>  colour_upper <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>,         <span class="co"># Samples covariate</span></span>
<span>  shape_upper <span class="op">=</span> <span class="st">"status"</span>,                     <span class="co"># Samples covariate</span></span>
<span>  colour_lower <span class="op">=</span> <span class="st">"feedlot"</span>,                   <span class="co"># Samples covariate</span></span>
<span>  scale_colour_lower <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="co"># Custom palette</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span>              <span class="co"># Plot legend vertically</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/show-pca-res-snps-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>More information about data pre-processing can be found in <a href="preprocessing.html"><span>Chapter&nbsp;6</span></a>.</p>
</section><section id="data-pre-filtering" class="level2" data-number="1.3"><h2 data-number="1.3" class="anchored" data-anchor-id="data-pre-filtering">
<span class="header-section-number">1.3</span> Data pre-filtering</h2>
<p>The created <code>MultiDataSet</code> object can be filtered, both in terms of samples and features, by passing a list of sample or feature IDs to retain, or by using logical tests on samples or features metadata. In addition, we implement target factories to retain only the most variable features in each omics dataset –unsupervised filtering–, or to retain the features most associated with an outcome of interest, via sPLS-DA from <a href="http://mixomics.org/"><code>mixOmics</code></a> –supervised filtering– (see <a href="prefiltering.html"><span>Chapter&nbsp;7</span></a>). This pre-filtering step is essential to reduce the size of the datasets prior to multi-omics integration.</p>
<details><summary>
Code
</summary><div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/feature_preselection_splsda_factory.html">feature_preselection_splsda_factory</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_complete</span>,            <span class="co"># A MultiDataSet object</span></span>
<span>  group <span class="op">=</span> <span class="st">"status"</span>,           <span class="co"># Sample covariate to use for supervised filtering</span></span>
<span>  to_keep_ns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>             <span class="co"># Number of features to retain per dataset</span></span>
<span>    <span class="st">"snps"</span> <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="fl">1000</span></span>
<span>  <span class="op">)</span>, </span>
<span>  filtered_set_target_name <span class="op">=</span> <span class="st">"mo_presel_supervised"</span> <span class="co"># Name of filtered object</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details><div class="cell">
<pre><code>#&gt; Object of class 'MultiDataSet'
#&gt;  . assayData: 3 elements
#&gt;     . snps: 1000 features, 139 samples 
#&gt;     . rnaseq: 994 features, 143 samples 
#&gt;     . metabolome: 55 features, 139 samples 
#&gt;  . featureData:
#&gt;     . snps: 1000 rows, 13 cols (feature_id, ..., p_value)
#&gt;     . rnaseq: 994 rows, 15 cols (feature_id, ..., de_signif)
#&gt;     . metabolome: 55 rows, 16 cols (feature_id, ..., de_signif)
#&gt;  . rowRanges:
#&gt;     . snps: YES
#&gt;     . rnaseq: YES
#&gt;     . metabolome: NO
#&gt;  . phenoData:
#&gt;     . snps: 139 samples, 10 cols (id, ..., geno_comp_3)
#&gt;     . rnaseq: 143 samples, 10 cols (id, ..., geno_comp_3)
#&gt;     . metabolome: 139 samples, 10 cols (id, ..., geno_comp_3)</code></pre>
</div>
</section><section id="multi-omics-data-integration" class="level2" data-number="1.4"><h2 data-number="1.4" class="anchored" data-anchor-id="multi-omics-data-integration">
<span class="header-section-number">1.4</span> Multi-omics data integration</h2>
<p>Currently, <code>moiraine</code> provides functions and target factories to facilitate the use of five integration methods: sPLS and DIABLO from the <code>mixOmics</code> package, sO2PLS from <a href="https://cran.r-project.org/web/packages/OmicsPLS/index.html"><code>OmicsPLS</code></a>, as well as <code>MOFA</code> and <code>MEFISTO</code> from <a href="https://biofam.github.io/MOFA2/"><code>MOFA2</code></a>.</p>
<p>This includes functions that transform a <code>MultiDataSet</code> object into the required input format for each integration method; for example for sPLS (only top of the matrices shown):</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_input_spls.html">get_input_spls</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_presel_supervised</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"canonical"</span>,</span>
<span>  datasets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rnaseq"</span>, <span class="st">"metabolome"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<pre><code>#&gt; $rnaseq
#&gt;       ENSBTAG00000000020 ENSBTAG00000000046 ENSBTAG00000000056
#&gt; R9497           3.486141           7.211634           10.25302
#&gt; R5969           3.486141           7.436865           10.45008
#&gt; R5327           4.005139           6.731310           10.72323
#&gt; R5979           3.486141           7.538956           10.46189
#&gt; R9504           4.252314           7.493829           10.20723
#&gt;       ENSBTAG00000000061 ENSBTAG00000000113
#&gt; R9497           5.028442           13.49233
#&gt; R5969           4.285078           12.95106
#&gt; R5327           5.313842           13.14427
#&gt; R5979           4.333030           12.55548
#&gt; R9504           4.784574           12.71693
#&gt; 
#&gt; $metabolome
#&gt;       HMDB00001 HMDB00008 HMDB00042 HMDB00043 HMDB00060
#&gt; R9497  3.397217  3.405992  9.654099  7.559186 0.5849625
#&gt; R5969  3.318827  2.137504  7.639522  6.539159 2.5849625
#&gt; R5327  3.721137  6.270529  6.963474  6.024586 2.6322682
#&gt; R5979  3.326626  2.232661  8.401306  7.394891 2.5849625
#&gt; R9504  3.603548  5.675251  7.879583  7.669594 0.8479969</code></pre>
</div>
<p><code>moiraine</code> also offers helper functions and target factories to facilitate the application of these integration tools. For example, the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_predefined_design_matrix.html">diablo_predefined_design_matrix()</a></code> function generates, for a given DIABLO input object, one of the three recommended design matrices for DIABLO (null, full or weighted full), while the <code><a href="https://rdrr.io/pkg/moiraine/man/diablo_pairwise_pls_factory.html">diablo_pairwise_pls_factory()</a></code> factory creates a list of targets to estimate the optimal design matrix to use for DIABLO based on datasets pairwise correlations estimated using PLS:</p>
<details><summary>
Code
</summary><div class="targets-chunk">
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="fu">tar_target</span><span class="op">(</span></span>
<span>    <span class="va">diablo_input</span>,                            <span class="co"># DIABLO input object</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_input_mixomics_supervised.html">get_input_mixomics_supervised</a></span><span class="op">(</span></span>
<span>      <span class="va">mo_presel_supervised</span>,                  <span class="co"># MultiDataSet object (prefiltered)</span></span>
<span>      group <span class="op">=</span> <span class="st">"status"</span>                       <span class="co"># Samples covariate of interest</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_pairwise_pls_factory.html">diablo_pairwise_pls_factory</a></span><span class="op">(</span><span class="va">diablo_input</span><span class="op">)</span>  <span class="co"># Target factory for design matrix</span></span>
<span>                                             <span class="co">#   estimation</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details><p><img src="images/dag_diablo_design_matrix.png" class="img-fluid"></p>
<p>In addition, a number of plotting functions have been implemented e.g.&nbsp;to visualise different aspects of the integration process: e.g.&nbsp;<code><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_tune.html">diablo_plot_tune()</a></code> to show the results of model tuning in DIABLO or <code><a href="https://rdrr.io/pkg/moiraine/man/so2pls_plot_summary.html">so2pls_plot_summary()</a></code> (shown below) to display the percentage of variance explained by each latent component constructed by sO2PLS:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/so2pls_plot_summary.html">so2pls_plot_summary</a></span><span class="op">(</span><span class="va">so2pls_final_run</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/so2pls-plot-summary-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Wouldn’t it be nice to have informative labels for the features in DIABLO’s circos plots? With <code>moiraine</code>, it is possible to use information from the features metadata provided as labels for the features in the plots. So, we can go from:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">mixOmics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/circosPlot.html">circosPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_final_run</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  size.variables <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  comp <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/diablo-circos-plot-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>to:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/diablo_plot_circos.html">diablo_plot_circos</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_final_run</span>,</span>
<span>  <span class="va">mo_set</span>,</span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  size.variables <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  comp <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/moiraine-circos-plot-1.png" class="img-fluid" width="576"></p>
</div>
</div>
</div>
</div>
<p>More details about how to use these integration tools through <code>moiraine</code> can be found in <a href="spls.html">Chapters&nbsp;<span>8</span></a> to <a href="diablo.html"><span>11</span></a>.</p>
</section><section id="interpreting-the-integration-results" class="level2" data-number="1.5"><h2 data-number="1.5" class="anchored" data-anchor-id="interpreting-the-integration-results">
<span class="header-section-number">1.5</span> Interpreting the integration results</h2>
<p>One of the main goals of <code>moiraine</code> is to facilitate the interpretation of the omics integration results. To this end, the outcome of any of the supported integration methods can be converted to a standardised integration output format, e.g.:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_output.html">get_output</a></span><span class="op">(</span><span class="va">mofa_trained</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<pre><code>#&gt; $features_weight
#&gt; # A tibble: 8,196 × 5
#&gt;    feature_id                  dataset latent_dimension     weight importance
#&gt;    &lt;chr&gt;                       &lt;fct&gt;   &lt;fct&gt;                 &lt;dbl&gt;      &lt;dbl&gt;
#&gt;  1 21-25977541-C-T-rs41974686  snps    Factor 1          0.000374      0.139 
#&gt;  2 22-51403583-A-C-rs210306176 snps    Factor 1          0.0000535     0.0199
#&gt;  3 24-12959068-G-T-rs381471286 snps    Factor 1          0.000268      0.0999
#&gt;  4 8-85224224-T-C-rs43565287   snps    Factor 1         -0.000492      0.183 
#&gt;  5 ARS-BFGL-BAC-16973          snps    Factor 1         -0.000883      0.329 
#&gt;  6 ARS-BFGL-BAC-19403          snps    Factor 1         -0.000500      0.186 
#&gt;  7 ARS-BFGL-BAC-2450           snps    Factor 1          0.000555      0.207 
#&gt;  8 ARS-BFGL-BAC-2600           snps    Factor 1         -0.0000325     0.0121
#&gt;  9 ARS-BFGL-BAC-27911          snps    Factor 1         -0.000568      0.212 
#&gt; 10 ARS-BFGL-BAC-35925          snps    Factor 1          0.000803      0.299 
#&gt; # ℹ 8,186 more rows
#&gt; 
#&gt; $samples_score
#&gt; # A tibble: 576 × 3
#&gt;    sample_id latent_dimension score
#&gt;    &lt;chr&gt;     &lt;fct&gt;            &lt;dbl&gt;
#&gt;  1 G1979     Factor 1          1.95
#&gt;  2 G2500     Factor 1          1.98
#&gt;  3 G3030     Factor 1          3.39
#&gt;  4 G3068     Factor 1          3.65
#&gt;  5 G3121     Factor 1          1.57
#&gt;  6 G3315     Factor 1         -1.81
#&gt;  7 G3473     Factor 1         -2.63
#&gt;  8 G3474     Factor 1         -2.30
#&gt;  9 G3550     Factor 1          2.98
#&gt; 10 G3594     Factor 1         -1.88
#&gt; # ℹ 566 more rows
#&gt; 
#&gt; $variance_explained
#&gt; # A tibble: 12 × 3
#&gt;    latent_dimension dataset    prop_var_expl
#&gt;    &lt;fct&gt;            &lt;fct&gt;              &lt;dbl&gt;
#&gt;  1 Factor 1         snps           0.000313 
#&gt;  2 Factor 1         rnaseq         0.496    
#&gt;  3 Factor 1         metabolome     0.137    
#&gt;  4 Factor 2         snps           0.239    
#&gt;  5 Factor 2         rnaseq         0.000197 
#&gt;  6 Factor 2         metabolome     0.0108   
#&gt;  7 Factor 3         snps           0.000108 
#&gt;  8 Factor 3         rnaseq         0.199    
#&gt;  9 Factor 3         metabolome     0.0000872
#&gt; 10 Factor 4         snps           0.0000927
#&gt; 11 Factor 4         rnaseq         0.0759   
#&gt; 12 Factor 4         metabolome     0.0194   
#&gt; 
#&gt; attr(,"class")
#&gt; [1] "output_dimension_reduction"
#&gt; attr(,"method")
#&gt; [1] "MOFA"</code></pre>
</div>
<p>This object can then be used to visualise the integration results in a number of ways, including:</p>
<ul>
<li>percentage of variance explained:</li>
</ul>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_variance_explained.html">plot_variance_explained</a></span><span class="op">(</span><span class="va">mofa_output</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/mofa-variance-explained-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<ul>
<li>Sample scores as pairwise scatterplots:</li>
</ul>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,                                        <span class="co"># MOFA standardised output</span></span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>,           <span class="co"># MOFA factors to display</span></span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set</span>,                                   <span class="co"># MultiDataSet object</span></span>
<span>  colour_upper <span class="op">=</span> <span class="st">"status"</span>,                            <span class="co"># Sample covariate</span></span>
<span>  scale_colour_upper <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span>, <span class="co"># Custom palette</span></span>
<span>  shape_upper <span class="op">=</span> <span class="st">"gender"</span>,                             <span class="co"># Sample covariate</span></span>
<span>  colour_lower <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>                  <span class="co"># Sample covariate</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/mofa-samples-score-ggpairs-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<ul>
<li>Sample scores against a sample covariate of interest (either categorical or continuous):</li>
</ul>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score_covariate.html">plot_samples_score_covariate</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,                             <span class="co"># MOFA standardised output</span></span>
<span>  <span class="va">mo_set</span>,                                  <span class="co"># MultiDataSet object</span></span>
<span>  <span class="st">"status"</span>,                                <span class="co"># Sample covariate of interest</span></span>
<span>  colour_by <span class="op">=</span> <span class="st">"status"</span>,                    <span class="co"># Other sample covariate</span></span>
<span>  shape_by <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>,          <span class="co"># Other sample covariate</span></span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="co"># MOFA factors to display</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/mofa-samples-score-covariate-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<ul>
<li>Top contributing features with their importance:</li>
</ul>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_top_features.html">plot_top_features</a></span><span class="op">(</span></span>
<span>  <span class="va">mofa_output</span>,                             <span class="co"># MOFA standardised output</span></span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set</span>,                        <span class="co"># MultiDataSet object</span></span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>                       <span class="co"># Custom labels for features from</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,                     <span class="co">#   features metadata</span></span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  truncate <span class="op">=</span> <span class="fl">25</span>,                           <span class="co"># truncate long feature labels</span></span>
<span>  latent_dimensions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Factor"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="co"># MOFA factors to display</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<p><img src="overview_files/figure-html/mofa-top-features-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>More details can be found in <a href="interpretation.html"><span>Chapter&nbsp;12</span></a>.</p>
</section><section id="evaluating-the-integration-results" class="level2" data-number="1.6"><h2 data-number="1.6" class="anchored" data-anchor-id="evaluating-the-integration-results">
<span class="header-section-number">1.6</span> Evaluating the integration results</h2>
<p>With <code>moiraine</code>, it is possible to evaluate the results of a given integration tool against prior information that we have about the features (e.g.&nbsp;knowledge about biological functions or results from a single-omics analysis) or samples (e.g.&nbsp;to assess the success of samples clustering into meaningful groups). For example, we can compare the feature selection performed by DIABLO to results from differential expression analyses performed on the omics datasets:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/evaluate_feature_selection_table.html">evaluate_feature_selection_table</a></span><span class="op">(</span></span>
<span>  <span class="va">diablo_output</span>,                                <span class="co"># DIABLO standardised output</span></span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set</span>,                             <span class="co"># MultiDataSet object</span></span>
<span>  col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>                             <span class="co"># Columns from features metadata</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"de_signif"</span>,                     <span class="co">#   containing DE outcome</span></span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"de_signif"</span> </span>
<span>  <span class="op">)</span>,</span>
<span>  latent_dimensions <span class="op">=</span> <span class="st">"Component 1"</span>             <span class="co"># Latent component to focus on</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>#&gt; # A tibble: 4 × 6
#&gt;   method latent_dimension dataset    feature_label selected not_selected
#&gt;   &lt;chr&gt;  &lt;fct&gt;            &lt;fct&gt;      &lt;chr&gt;            &lt;int&gt;        &lt;int&gt;
#&gt; 1 DIABLO Component 1      rnaseq     DE                  23           52
#&gt; 2 DIABLO Component 1      rnaseq     Not DE               7          912
#&gt; 3 DIABLO Component 1      metabolome DE                  10           20
#&gt; 4 DIABLO Component 1      metabolome Not DE              NA           25</code></pre>
</div>
</div>
<p>In addition, a number of functions are provided to help with features set enrichment (such as over-representation analysis or gene set enrichment analysis), e.g.&nbsp;by generating feature sets from features metadata through <code><a href="https://rdrr.io/pkg/moiraine/man/make_feature_sets_from_fm.html">make_feature_sets_from_fm()</a></code>, or by ensuring that a proper background set is used in the enrichment analysis, with <code><a href="https://rdrr.io/pkg/moiraine/man/reduce_feature_sets_data.html">reduce_feature_sets_data()</a></code>. Information about evaluation of integration results is presented in <a href="evaluation.html"><span>Chapter&nbsp;13</span></a>.</p>
</section><section id="comparison-different-integration-results" class="level2" data-number="1.7"><h2 data-number="1.7" class="anchored" data-anchor-id="comparison-different-integration-results">
<span class="header-section-number">1.7</span> Comparison different integration results</h2>
<p>Lastly, <code>moiraine</code> facilitates the comparison of the results from different integration methods, or from a same integration method but with different pre-processing options or parameter choices. It is possible to visualise the correlation between the results of different methods, in terms of the latent dimensions they constructed:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/comparison_heatmap_corr.html">comparison_heatmap_corr</a></span><span class="op">(</span></span>
<span>  <span class="va">output_list</span>,              <span class="co"># List of integration results (standardised format)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img src="overview_files/figure-html/show-comp-heatmap-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Or compare the samples score or features weight from two latent dimensions created by two integration methods:</p>
<div class="cell">
<details><summary>Code</summary><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_features_weight_pair.html">plot_features_weight_pair</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">mofa_output</span>, <span class="va">diablo_output</span><span class="op">)</span>,  <span class="co"># Integration results (stand. format)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>                              <span class="co"># Latent dimensions to compare</span></span>
<span>    <span class="st">"MOFA"</span> <span class="op">=</span> <span class="st">"Factor 1"</span>, </span>
<span>    <span class="st">"DIABLO"</span> <span class="op">=</span> <span class="st">"Component 1"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set</span>,                 <span class="co"># MultiDataSet object</span></span>
<span>  features_metric <span class="op">=</span> <span class="st">"importance"</span>,   <span class="co"># Plot absolute or signed importance score</span></span>
<span>  label_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>                <span class="co"># Columns from features metadata to use</span></span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"Name"</span>,              <span class="co">#  as labels in plot</span></span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"name"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img src="overview_files/figure-html/show-features-weight-comp-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>It is also possible to compute a consensus importance score for each feature, which summarises the contribution of the feature to different latent dimensions, and thus assess which features are highlighted across several integration methods. More details are provided in <a href="comparison.html"><span>Chapter&nbsp;14</span></a>.</p>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Introduction</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./example_dataset.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">The example dataset</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>