<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>The moiraine R package user manual - 6&nbsp; Data pre-processing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./prefiltering.html" rel="next">
<link href="./modifying_multidataset.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data pre-processing</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">The moiraine R package user manual</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./example_dataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">The example dataset</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Data preparation</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_import.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Importing data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inspecting_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Inspecting the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./modifying_multidataset.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocessing.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data pre-processing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./prefiltering.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Dataset pre-filtering</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#datasets-transformations" id="toc-datasets-transformations" class="nav-link active" data-scroll-target="#datasets-transformations"><span class="toc-section-number">6.1</span>  Datasets transformations</a>
  <ul class="collapse">
<li><a href="#transforming-a-single-dataset" id="toc-transforming-a-single-dataset" class="nav-link" data-scroll-target="#transforming-a-single-dataset"><span class="toc-section-number">6.1.1</span>  Transforming a single dataset</a></li>
  <li><a href="#transformation-target-factory" id="toc-transformation-target-factory" class="nav-link" data-scroll-target="#transformation-target-factory"><span class="toc-section-number">6.1.2</span>  Transformation target factory</a></li>
  </ul>
</li>
  <li>
<a href="#running-a-pca-on-each-dataset" id="toc-running-a-pca-on-each-dataset" class="nav-link" data-scroll-target="#running-a-pca-on-each-dataset"><span class="toc-section-number">6.2</span>  Running a PCA on each dataset</a>
  <ul class="collapse">
<li><a href="#running-the-pcas" id="toc-running-the-pcas" class="nav-link" data-scroll-target="#running-the-pcas"><span class="toc-section-number">6.2.1</span>  Running the PCAs</a></li>
  <li><a href="#visualising-the-pca-results" id="toc-visualising-the-pca-results" class="nav-link" data-scroll-target="#visualising-the-pca-results"><span class="toc-section-number">6.2.2</span>  Visualising the PCA results</a></li>
  <li><a href="#missing-values-imputation" id="toc-missing-values-imputation" class="nav-link" data-scroll-target="#missing-values-imputation"><span class="toc-section-number">6.2.3</span>  Missing values imputation</a></li>
  </ul>
</li>
  <li><a href="#recap-targets-list" id="toc-recap-targets-list" class="nav-link" data-scroll-target="#recap-targets-list"><span class="toc-section-number">6.3</span>  Recap – targets list</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-preprocessing" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data pre-processing</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/targets/">targets</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/PlantandFoodResearch/moiraine">moiraine</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">MultiDataSet</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For PCA results</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hredestig/pcamethods">pcaMethods</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For working with lists</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="co">## For custom colour palettes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For displaying documentation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/S4Vectors">S4Vectors</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">stats4</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once each omics dataset has been imported into R with associated metadata and combined into a <code>MultiDataSet</code> object, there is a number of pre-processing steps that should be considered. In this chapter, we will show how to apply different transformations to the datasets, as well as how to run a PCA on each dataset and impute missing values if needed.</p>
<p>As a reminder, here is what the <code>_targets.R</code> script should look like so far:</p>
<details><summary><code>_targets.R</code> script
</summary><div class="cell">
<div class="cell">

</div>
</div>
</details><section id="datasets-transformations" class="level2" data-number="6.1"><h2 data-number="6.1" class="anchored" data-anchor-id="datasets-transformations">
<span class="header-section-number">6.1</span> Datasets transformations</h2>
<p>After inspection of the density plots for the different datasets (see <a href="inspecting_multidataset.html#sec-inspecting-multidataset-summary-plots"><span>Section&nbsp;4.3</span></a>), it might be necessary to normalise or transform some or all datasets. This is necessary to mitigate the mean-variance trend that occurs in RNAseq data, for example, or simply to bring the different features to a comparable scale. Transformation here refers to applying a function to each feature (i.e.&nbsp;each row) within a dataset that will transform the measurement values for the feature.</p>
<p><code>moiraine</code> implements several options to transform an omics dataset:</p>
<ul>
<li><p>Variance Stabilising Normalisation (VSN) through the <code>vsn</code> package – recommended for metabolomics datasets or other continuous datasets with a strong mean-variance trend;</p></li>
<li><p>Variance Stabilising Transformation (VST) through the <code>DESeq2</code> package – recommended for RNAseq data or any raw read count-type data;</p></li>
<li><p>Automatic selection of the best normalisation method for each feature through the <code>bestNormalize</code> package – recommended for phenotype data, and when the number of features is small (note that the selection of the normalisation method is done independently for each feature, so the same transformation might not be applied to all features);</p></li>
<li><p>A selection of common normalisation methods through the <code>bestNormalize</code> package, including center/scale, log, exponential, square-root, arcsinh, Box Cox, Yeo-Johnson and ordered quantile transformations (see details in the <a href="https://cran.r-project.org/web/packages/bestNormalize/vignettes/bestNormalize.html">bestNormalize vignette</a>) – recommended when applying the same transformation to all features, e.g.&nbsp;log2 transformation or centering.</p></li>
</ul>
<section id="transforming-a-single-dataset" class="level3" data-number="6.1.1"><h3 data-number="6.1.1" class="anchored" data-anchor-id="transforming-a-single-dataset">
<span class="header-section-number">6.1.1</span> Transforming a single dataset</h3>
<p>The transformation of one dataset is done through the <code><a href="https://rdrr.io/pkg/moiraine/man/transform_dataset.html">transform_dataset()</a></code> function, which takes as input a <code>MultiDataSet</code> object, the name of the dataset to transform, and the name of the transformation to be applied, which should be one of <code>vsn</code>, <code>vst-deseq2</code>, <code>best-normalize-auto</code> or <code>best-normalize-manual</code>. For the latter, the name of the normalisation method from the <code>BestNormalize</code> package to use must also be supplied through the <code>method</code> argument.</p>
<p>The <code>return_multidataset</code> argument determines whether a <code>MultiDataSet</code> object with the corresponding dataset transformed should be returned. If it is set to <code>FALSE</code>, the function will instead return a list with the transformed dataset as a matrix as well as other useful information returned by the transformation function applied. It is possible to only return the transformed matrix, by setting <code>return_matrix_only</code> to <code>TRUE</code>. This can be useful to quickly assess the effects of the transformation outside of the analysis pipeline.</p>
<p>For example, we can apply the Variance Stabilising Transformation to the transcriptomics dataset:</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mo_set_de</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rnaseq_vst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/transform_dataset.html">transform_dataset</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_de</span>,</span>
<span>  <span class="st">"rnaseq"</span>,</span>
<span>  <span class="st">"vst-deseq2"</span>,</span>
<span>  return_multidataset <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Applying Variance Stabilising Transformation (DESeq2) to rnaseq dataset.</span></span>
<span><span class="co">#&gt; converting counts to integer mode</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function returns a list, with the transformed dataset as matrix in the <code>transformed_data</code> element. Information generated during the transformation by the <code>DESeq2</code> package is stored in the <code>info_transformation</code> element. The name of the transformation applied is stored in the <code>transformation</code> element:</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">rnaseq_vst</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "transformed_data"    "info_transformation" "transformation"</span></span>
<span></span>
<span><span class="va">rnaseq_vst</span><span class="op">$</span><span class="va">transformed_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                        R9497     R5969     R5327     R5979     R9504</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005 10.357314 11.289047 12.070536 10.201227 10.251942</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008  4.836495  4.937948  4.375764  4.986154  5.710581</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009  3.486141  4.054636  4.983434  3.486141  3.486141</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010 12.216690 11.937084 12.079359 11.383360 11.897780</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011  3.486141  4.054636  4.005139  3.979732  3.486141</span></span>
<span></span>
<span><span class="va">rnaseq_vst</span><span class="op">$</span><span class="va">info_transformation</span></span>
<span><span class="co">#&gt; class: DESeqTransform </span></span>
<span><span class="co">#&gt; dim: 20335 143 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(1): ''</span></span>
<span><span class="co">#&gt; rownames(20335): ENSBTAG00000000005 ENSBTAG00000000008 ...</span></span>
<span><span class="co">#&gt;   ENSBTAG00000055312 ENSBTAG00000055314</span></span>
<span><span class="co">#&gt; rowData names(6): baseMean baseVar ... dispGeneIter dispFit</span></span>
<span><span class="co">#&gt; colnames(143): R9497 R5969 ... Y9747 Y9816</span></span>
<span><span class="co">#&gt; colData names(1): sizeFactor</span></span>
<span></span>
<span><span class="va">rnaseq_vst</span><span class="op">$</span><span class="va">transformation</span></span>
<span><span class="co">#&gt; [1] "vst-deseq2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If we instead want to apply a log2 transformation to the dataset, we will use the <code>best-normalize-manual</code> transformation option instead, and we have to specify the transformation to use through the <code>method</code> argument; in our case, we will use the <code>log_x</code> method.</p>
<p>The <code><a href="https://petersonr.github.io/bestNormalize/reference/log_x.html">bestNormalize::log_x()</a></code> function uses by default a log base 10, but this can be changed by passing the <code>b</code> argument (the log base to use) to <code>2</code>. We will also set <code>a</code> (the offset to use before log-transformation) to <code>0.5</code> – see the <code><a href="https://petersonr.github.io/bestNormalize/reference/log_x.html">bestNormalize::log_x()</a></code> function help for information about the parameters. By default, all <code>bestNormalize</code> functions standardise the transformed datasets, which we can prevent by setting <code>standardize = FALSE</code>:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">rnaseq_log2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/transform_dataset.html">transform_dataset</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_de</span>,</span>
<span>  <span class="st">"rnaseq"</span>,</span>
<span>  <span class="st">"best-normalize-manual"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"log_x"</span>,</span>
<span>  return_multidataset <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  <span class="co">## Arguments passed to BestNormalize::log_x():</span></span>
<span>  a <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  b <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Applying log_x transformation (bestNormalize) to rnaseq dataset.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In that case, we asked the function to return a <code>MultiDataSet</code> object, in which the <code>rnaseq</code> dataset has been transformed:</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">rnaseq_log2</span></span>
<span><span class="co">#&gt; Object of class 'MultiDataSet'</span></span>
<span><span class="co">#&gt;  . assayData: 3 elements</span></span>
<span><span class="co">#&gt;     . snps: 23036 features, 139 samples </span></span>
<span><span class="co">#&gt;     . rnaseq: 20335 features, 143 samples </span></span>
<span><span class="co">#&gt;     . metabolome: 55 features, 139 samples </span></span>
<span><span class="co">#&gt;  . featureData:</span></span>
<span><span class="co">#&gt;     . snps: 23036 rows, 13 cols (feature_id, ..., p_value)</span></span>
<span><span class="co">#&gt;     . rnaseq: 20335 rows, 15 cols (feature_id, ..., de_signif)</span></span>
<span><span class="co">#&gt;     . metabolome: 55 rows, 16 cols (feature_id, ..., de_signif)</span></span>
<span><span class="co">#&gt;  . rowRanges:</span></span>
<span><span class="co">#&gt;     . snps: YES</span></span>
<span><span class="co">#&gt;     . rnaseq: YES</span></span>
<span><span class="co">#&gt;     . metabolome: NO</span></span>
<span><span class="co">#&gt;  . phenoData:</span></span>
<span><span class="co">#&gt;     . snps: 139 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span><span class="co">#&gt;     . rnaseq: 143 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span><span class="co">#&gt;     . metabolome: 139 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_dataset_matrix.html">get_dataset_matrix</a></span><span class="op">(</span><span class="va">rnaseq_log2</span>, <span class="st">"rnaseq"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                        R9497      R5969      R5327      R5979     R9504</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005  9.518653 10.4589192 11.5115056  9.7690113  9.532356</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008  2.700440  2.9068906  1.8073549  3.3923174  4.357552</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009 -1.000000  0.5849625  3.2479275 -1.0000000 -1.000000</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010 11.395266 11.1114617 11.5203731 10.9661449 11.195680</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011 -1.000000  0.5849625  0.5849625  0.5849625 -1.000000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="transformation-target-factory" class="level3" data-number="6.1.2"><h3 data-number="6.1.2" class="anchored" data-anchor-id="transformation-target-factory">
<span class="header-section-number">6.1.2</span> Transformation target factory</h3>
<p>The target factory function <code><a href="https://rdrr.io/pkg/moiraine/man/transformation_datasets_factory.html">transformation_datasets_factory()</a></code> provides a wrapper to apply (potentially different) transformations to several datasets at once. The function takes as input the <code>MultiDataSet</code> object as well as a named character vector, in which each element corresponds to a transformation that should be applied to a specific dataset. If a dataset is not present in the transformation vector, it will not be transformed (but it will still be present in the resulting <code>MultiDataSet</code> object).</p>
<p>Here, we would like to apply Variance Stabilising Transformation to the transcriptomics dataset, and a log2 transformation to the metabolomics dataset. Note that the VST and VSN transformations are very close to the log2 transformation, especially for features with high means.</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/transformation_datasets_factory.html">transformation_datasets_factory</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_de</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"vst-deseq2"</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"best-normalize-manual"</span><span class="op">)</span>,</span>
<span>  methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"log_x"</span><span class="op">)</span>,</span>
<span>  a <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  b <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  transformed_data_name <span class="op">=</span> <span class="st">"mo_set_transformed"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/moiraine/man/transformation_datasets_factory.html">transformation_datasets_factory()</a></code> function works as follows:</p>
<ul>
<li>It creates a grouped tibble listing the transformation to apply to each dataset, stored in the <code>transformations_spec</code> target;</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">transformations_spec</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 4</span></span>
<span><span class="co">#&gt;   dsn        transf                meth  tar_group</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt;                 &lt;chr&gt;     &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 rnaseq     vst-deseq2            &lt;NA&gt;          2</span></span>
<span><span class="co">#&gt; 2 metabolome best-normalize-manual log_x         1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>It performs the required transformation on each dataset via <a href="https://books.ropensci.org/targets/dynamic.html">dynamic branching</a>. This is done through a call to the <code><a href="https://rdrr.io/pkg/moiraine/man/transform_dataset.html">transform_dataset()</a></code> function. The transformed datasets are stored in a list, in the <code>transformations_runs_list</code> target. Note that by default the function will store all details of the transformations, which can be useful for later inspection, but can be memory-intensive. It is possible to only store the transformed datasets instead, by setting the <code>return_matrix_only</code> argument to <code>TRUE</code> in the <code><a href="https://rdrr.io/pkg/moiraine/man/transformation_datasets_factory.html">transformation_datasets_factory()</a></code> call.</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">transformations_runs_list</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">transformations_runs_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "transformations_runs_list_7a466037" "transformations_runs_list_a1c8db41"</span></span>
<span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_chr</a></span><span class="op">(</span><span class="va">transformations_runs_list</span>, <span class="va">attr</span>, <span class="st">"dataset_name"</span><span class="op">)</span></span>
<span><span class="co">#&gt; transformations_runs_list_7a466037 transformations_runs_list_a1c8db41 </span></span>
<span><span class="co">#&gt;                       "metabolome"                           "rnaseq"</span></span>
<span></span>
<span><span class="va">transformations_runs_list</span><span class="op">[[</span><span class="st">"transformations_runs_list_a1c8db41"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "transformed_data"    "info_transformation" "transformation"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>It creates a new <code>MultiDataSet</code> object, with the transformed version of the datasets. By default, this new <code>MultiDataSet</code> object is stored in a target called <code>transformed_set</code>, but a different name can be specified via the <code>transformed_data_name</code> argument (here we called it <code>mo_set_transformed</code>).</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mo_set_transformed</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mo_set_transformed</span></span>
<span><span class="co">#&gt; Object of class 'MultiDataSet'</span></span>
<span><span class="co">#&gt;  . assayData: 3 elements</span></span>
<span><span class="co">#&gt;     . snps: 23036 features, 139 samples </span></span>
<span><span class="co">#&gt;     . rnaseq: 20335 features, 143 samples </span></span>
<span><span class="co">#&gt;     . metabolome: 55 features, 139 samples </span></span>
<span><span class="co">#&gt;  . featureData:</span></span>
<span><span class="co">#&gt;     . snps: 23036 rows, 13 cols (feature_id, ..., p_value)</span></span>
<span><span class="co">#&gt;     . rnaseq: 20335 rows, 15 cols (feature_id, ..., de_signif)</span></span>
<span><span class="co">#&gt;     . metabolome: 55 rows, 16 cols (feature_id, ..., de_signif)</span></span>
<span><span class="co">#&gt;  . rowRanges:</span></span>
<span><span class="co">#&gt;     . snps: YES</span></span>
<span><span class="co">#&gt;     . rnaseq: YES</span></span>
<span><span class="co">#&gt;     . metabolome: NO</span></span>
<span><span class="co">#&gt;  . phenoData:</span></span>
<span><span class="co">#&gt;     . snps: 139 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span><span class="co">#&gt;     . rnaseq: 143 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span><span class="co">#&gt;     . metabolome: 139 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_dataset_matrix.html">get_dataset_matrix</a></span><span class="op">(</span><span class="va">mo_set_de</span>, <span class="st">"metabolome"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt;             R21 Y3660 Y3243</span></span>
<span><span class="co">#&gt; HMDB00001   9.1   9.7   9.3</span></span>
<span><span class="co">#&gt; HMDB00008  58.2  22.8   9.1</span></span>
<span><span class="co">#&gt; HMDB00042 403.0 392.0 606.0</span></span>
<span><span class="co">#&gt; HMDB00043 172.6 163.1 165.2</span></span>
<span><span class="co">#&gt; HMDB00060   0.7   1.5   1.4</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_dataset_matrix.html">get_dataset_matrix</a></span><span class="op">(</span><span class="va">mo_set_de</span>, <span class="st">"metabolome"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt;             R21 Y3660 Y3243</span></span>
<span><span class="co">#&gt; HMDB00001   9.1   9.7   9.3</span></span>
<span><span class="co">#&gt; HMDB00008  58.2  22.8   9.1</span></span>
<span><span class="co">#&gt; HMDB00042 403.0 392.0 606.0</span></span>
<span><span class="co">#&gt; HMDB00043 172.6 163.1 165.2</span></span>
<span><span class="co">#&gt; HMDB00060   0.7   1.5   1.4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can assess the effect of the transformations by generating density and mean-sd plots for the transformed datasets:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_density_data.html">plot_density_data</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_transformed</span>,</span>
<span>  combined <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  scales <span class="op">=</span> <span class="st">"free"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="preprocessing_files/figure-html/plot-density-transformed-data-1.png" class="img-fluid" width="864"></p>
</div>
</div>
<p>Note how the relationship between features mean and standard deviation has been reduced in both transformed datasets:</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_meansd_data.html">plot_meansd_data</a></span><span class="op">(</span><span class="va">mo_set_transformed</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="preprocessing_files/figure-html/plot-meansd-transformed-data-1.png" class="img-fluid" width="864"></p>
</div>
</div>
<p>Finally, it can be useful to summarise which transformations have been applied to the datasets, for example when creating a report. The function <code>get_table_transformation()</code> is here for that. It takes as an input the <code>transformations_runs_list</code> target generated by <code><a href="https://rdrr.io/pkg/moiraine/man/transformation_datasets_factory.html">transformation_datasets_factory()</a></code>, and returns a tibble indicating the transformation applied to each dataset:</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_table_transformations.html">get_table_transformations</a></span><span class="op">(</span><span class="va">transformations_runs_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 2 × 2</span></span>
<span><span class="co">#&gt;   Dataset    Transformation                                                 </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt;                                                          </span></span>
<span><span class="co">#&gt; 1 metabolome Non-Standardized Log_2(x + 0.01) transformation (bestNormalize)</span></span>
<span><span class="co">#&gt; 2 rnaseq     Variance Stabilising Transformation (DESeq2)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="running-a-pca-on-each-dataset" class="level2" data-number="6.2"><h2 data-number="6.2" class="anchored" data-anchor-id="running-a-pca-on-each-dataset">
<span class="header-section-number">6.2</span> Running a PCA on each dataset</h2>
<p>It is always best practice to run some exploratory analysis on a dataset prior to running analyses. This is largely outside the scope of this package, and we assume that any input dataset has been properly assessed before turning to the integration pipeline. However, running a Principal Component Analysis (PCA) on each of the omics datasets within the integration pipeline serves two purposes:</p>
<ul>
<li><p>as a last check to ensure that there are no obvious batch effects or problematic samples that should be addressed,</p></li>
<li><p>as a missing data imputation method.</p></li>
</ul>
<p>The <code>moiraine</code> package relies on the Bioconductor <a href="https://bioconductor.org/packages/release/bioc/html/pcaMethods.html"><code>pcaMethods</code> package</a> to perform the PCA. In particular, the <code>pcaMethods</code> package implements a NIPALS (non-linear iterative partial least squares) method for PCA, which allows for missing values in the input dataset, and imputes missing values based on the results of the PCA.</p>
<section id="running-the-pcas" class="level3" data-number="6.2.1"><h3 data-number="6.2.1" class="anchored" data-anchor-id="running-the-pcas">
<span class="header-section-number">6.2.1</span> Running the PCAs</h3>
<p>The <code><a href="https://rdrr.io/pkg/moiraine/man/pca_complete_data_factory.html">pca_complete_data_factory()</a></code> function uses dynamic branching to perform a PCA on each omics dataset within a <code>MultiDataSet</code> object. It takes as input the <code>MultiDataSet</code> object (in our case, <code>mo_set_transformed</code>), and, optionally, the names of the datasets on which a PCA should be run. This is useful if one dataset is very large and has no missing values, and we want to avoid running a PCA on it. It then proceeds as follows:</p>
<ul>
<li><p>It creates a target called <code>dataset_names_pca</code>, which stores a vector of dataset names on which a PCA should be applied;</p></li>
<li><p>For each value in <code>dataset_names_pca</code>, it extracts the omics dataset as a matrix with features as rows and samples as columns, using the <code><a href="https://rdrr.io/pkg/moiraine/man/get_dataset_matrix.html">get_dataset_matrix()</a></code> function. This is done via dynamic branching, and the results are stored as a list in the <code>pca_mats_list</code> target. Note that the names of this list are not meaningful; to check which element of the list corresponds to which dataset, you can run <code>map_chr(pca_mats_list, attr, "dataset_name")</code>;</p></li>
<li><p>For each matrix in <code>pca_mats_list</code>, it applies the <code><a href="https://rdrr.io/pkg/moiraine/man/run_pca_matrix.html">run_pca_matrix()</a></code> function to the corresponding dataset. This is done via dynamic branching; it results in a list where each element is the PCA result (i.e.&nbsp;a <code><a href="https://rdrr.io/pkg/pcaMethods/man/pcaRes.html">pcaMethods::pcaRes</a></code> object) for a given dataset. This list is stored in the <code>pca_runs_list</code> target. Note that the names of this list are not meaningful; to check which element of the list corresponds to which dataset, you can run <code>map_chr(pca_runs_list, attr, "dataset_name")</code>;</p></li>
<li><p>It extracts from the result of each PCA the complete dataset, i.e.&nbsp;with missing values imputed, and uses this information to construct a new <code>MultiDataSet</code> object, in which the datasets are complete (i.e.&nbsp;no missing value). This is done by calling the <code><a href="https://rdrr.io/pkg/moiraine/man/get_complete_data.html">get_complete_data()</a></code> function. If no PCA was run on a dataset, the dataset will still be present in the new <code>MultiDataSet</code> object, but its missing values will not be imputed. The resulting complete <code>MultiDataSet</code> object is stored by default in a target called <code>complete_set</code>; this name can be changed via the <code>complete_data_name</code> argument.</p></li>
</ul>
<p>Let’s apply this to our multi-omics dataset:</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/pca_complete_data_factory.html">pca_complete_data_factory</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_transformed</span>,</span>
<span>  complete_data_name <span class="op">=</span> <span class="st">"mo_set_complete"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can have a look at the different targets constructed. By default, a PCA was run on all datasets:</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">dataset_names_pca</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "snps"       "rnaseq"     "metabolome"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">pca_mats_list</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_chr</a></span><span class="op">(</span><span class="va">pca_mats_list</span>, <span class="va">attr</span>, <span class="st">"dataset_name"</span><span class="op">)</span></span>
<span><span class="co">#&gt; pca_mats_list_302d7473 pca_mats_list_84d36937 pca_mats_list_64d37e6c </span></span>
<span><span class="co">#&gt;                 "snps"               "rnaseq"           "metabolome"</span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">pca_mats_list</span>, <span class="op">~</span><span class="va">.x</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; $pca_mats_list_302d7473</span></span>
<span><span class="co">#&gt;                             R21 Y3660 Y3243 R5764 P4669</span></span>
<span><span class="co">#&gt; 1_41768691                    1     0     2     2     1</span></span>
<span><span class="co">#&gt; 10-27008241-A-C-rs42918694    2     2     2     1     2</span></span>
<span><span class="co">#&gt; 10-37505419-T-C-rs136559242   0     1     0     2     0</span></span>
<span><span class="co">#&gt; 10-49904259-G-A-rs471723345   1     2     2     2     2</span></span>
<span><span class="co">#&gt; 1-109550832-G-A-rs209732846   2     2     1     2     2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $pca_mats_list_84d36937</span></span>
<span><span class="co">#&gt;                        R9497     R5969     R5327     R5979     R9504</span></span>
<span><span class="co">#&gt; ENSBTAG00000000005 10.357314 11.289047 12.070536 10.201227 10.251942</span></span>
<span><span class="co">#&gt; ENSBTAG00000000008  4.836495  4.937948  4.375764  4.986154  5.710581</span></span>
<span><span class="co">#&gt; ENSBTAG00000000009  3.486141  4.054636  4.983434  3.486141  3.486141</span></span>
<span><span class="co">#&gt; ENSBTAG00000000010 12.216690 11.937084 12.079359 11.383360 11.897780</span></span>
<span><span class="co">#&gt; ENSBTAG00000000011  3.486141  4.054636  4.005139  3.979732  3.486141</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $pca_mats_list_64d37e6c</span></span>
<span><span class="co">#&gt;                  R21     Y3660     Y3243    R5764     P4669</span></span>
<span><span class="co">#&gt; HMDB00001  3.1874511 3.2794713 3.2187812       NA 3.6334312</span></span>
<span><span class="co">#&gt; HMDB00008  5.8631951 4.5115945 3.1874511 2.657640 4.1626934</span></span>
<span><span class="co">#&gt; HMDB00042  8.6546718 8.6147466 9.2431978 8.655388 7.9009272</span></span>
<span><span class="co">#&gt; HMDB00043  7.4313722 7.3497014 7.3681572 7.707428 6.4027563</span></span>
<span><span class="co">#&gt; HMDB00060 -0.4941091 0.5945485 0.4956952 2.908813 0.6870607</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">pca_runs_list</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pca_runs_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "pca_runs_list_74d71ae8" "pca_runs_list_71fd94b4" "pca_runs_list_559272f0"</span></span>
<span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_chr</a></span><span class="op">(</span><span class="va">pca_runs_list</span>, <span class="va">attr</span>, <span class="st">"dataset_name"</span><span class="op">)</span></span>
<span><span class="co">#&gt; pca_runs_list_74d71ae8 pca_runs_list_71fd94b4 pca_runs_list_559272f0 </span></span>
<span><span class="co">#&gt;                 "snps"               "rnaseq"           "metabolome"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The result of the PCA run on the genomics dataset looks like this:</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_read.html">tar_read</a></span><span class="op">(</span><span class="va">pca_runs_list_74d71ae8</span><span class="op">)</span></span>
<span><span class="co">#&gt; nipals calculated PCA</span></span>
<span><span class="co">#&gt; Importance of component(s):</span></span>
<span><span class="co">#&gt;                   PC1     PC2     PC3     PC4     PC5     PC6      PC7      PC8</span></span>
<span><span class="co">#&gt; R2            0.05578 0.02881 0.01305 0.01196 0.01168 0.01005 0.009848 0.009565</span></span>
<span><span class="co">#&gt; Cumulative R2 0.05578 0.08459 0.09765 0.10960 0.12128 0.13133 0.141175 0.150739</span></span>
<span><span class="co">#&gt;                    PC9     PC10</span></span>
<span><span class="co">#&gt; R2            0.009286 0.008915</span></span>
<span><span class="co">#&gt; Cumulative R2 0.160026 0.168941</span></span>
<span><span class="co">#&gt; 23036    Variables</span></span>
<span><span class="co">#&gt; 139  Samples</span></span>
<span><span class="co">#&gt; 9615     NAs ( 0.3 %)</span></span>
<span><span class="co">#&gt; 10   Calculated component(s)</span></span>
<span><span class="co">#&gt; Data was mean centered before running PCA </span></span>
<span><span class="co">#&gt; Data was NOT scaled before running PCA </span></span>
<span><span class="co">#&gt; Scores structure:</span></span>
<span><span class="co">#&gt; [1] 139  10</span></span>
<span><span class="co">#&gt; Loadings structure:</span></span>
<span><span class="co">#&gt; [1] 23036    10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can notice that there is some information about the number of principal components computed, and whether the dataset was centred and scaled before applying the PCA. This is handled by the default arguments of <code><a href="https://rdrr.io/pkg/moiraine/man/run_pca_matrix.html">run_pca_matrix()</a></code>, but can be specified by passing the corresponding arguments to <code><a href="https://rdrr.io/pkg/moiraine/man/pca_complete_data_factory.html">pca_complete_data_factory()</a></code>. For example, to scale the datasets before performing a PCA, we could use:</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/pca_complete_data_factory.html">pca_complete_data_factory</a></span><span class="op">(</span></span>
<span>  <span class="va">mo_set_transformed</span>,</span>
<span>  complete_data_name <span class="op">=</span> <span class="st">"mo_set_complete"</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For convenience, the <code><a href="https://rdrr.io/pkg/moiraine/man/run_pca.html">run_pca()</a></code> function can be used to run a PCA on one of the omics datasets directly from a <code>MultiDataSet</code> object. It is a wrapper around the <code><a href="https://rdrr.io/pkg/moiraine/man/run_pca_matrix.html">run_pca_matrix()</a></code> function, and takes as input a <code>MultiDataSet</code> object as well as the name of the omics dataset on which a PCA should be run, e.g.:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/run_pca.html">run_pca</a></span><span class="op">(</span><span class="va">mo_set_de</span>, <span class="st">"rnaseq"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="visualising-the-pca-results" class="level3" data-number="6.2.2"><h3 data-number="6.2.2" class="anchored" data-anchor-id="visualising-the-pca-results">
<span class="header-section-number">6.2.2</span> Visualising the PCA results</h3>
<p>It is possible to get an overview of the results of each PCA. First, the function <code><a href="https://rdrr.io/pkg/moiraine/man/plot_screeplot_pca.html">plot_screeplot_pca()</a></code> displays the percentage of variance explained by the principal components computed for each dataset. It takes as input the <code>pca_runs_list</code> target constructed in the previous step. Note that by default, 10 components are computed for each dataset.</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_screeplot_pca.html">plot_screeplot_pca</a></span><span class="op">(</span><span class="va">pca_runs_list</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="preprocessing_files/figure-html/plot-screeplot-pca-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>In addition, the <code>plot_samples_coordinates_pca</code> allows us to display the samples in the reduced principal components space (the common PCA sample plot). The function returns a list of plots (one plot per dataset). By default, it shows all principal components computed for each dataset, but for clarity we will only look at the first three:</p>
<div class="cell">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_coordinates_pca.html">plot_samples_coordinates_pca</a></span><span class="op">(</span></span>
<span>  <span class="va">pca_runs_list</span>,</span>
<span>  pcs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; $snps</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="preprocessing_files/figure-html/plot-samples-coordinates-pca-1.png" class="img-fluid" width="768"></p>
</div>
<pre><code>#&gt; 
#&gt; $rnaseq</code></pre>
<div class="cell-output-display">
<p><img src="preprocessing_files/figure-html/plot-samples-coordinates-pca-2.png" class="img-fluid" width="768"></p>
</div>
<pre><code>#&gt; 
#&gt; $metabolome</code></pre>
<div class="cell-output-display">
<p><img src="preprocessing_files/figure-html/plot-samples-coordinates-pca-3.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Note that it is possible to look at a different set of principal components for each dataset. For that, the index of the principal components should be passed to the <code>pcs</code> argument as a named list (where the name of each element corresponds to a dataset name), e.g.:</p>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_coordinates_pca.html">plot_samples_coordinates_pca</a></span><span class="op">(</span></span>
<span>  <span class="va">pca_runs_list</span>,</span>
<span>  pcs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"snps"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,</span>
<span>    <span class="st">"rnaseq"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,</span>
<span>    <span class="st">"metabolome"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default, the points in the sample plots are not coloured. It is however possible to colour the samples according to the information contained in the sample metadata tables available through the <code>MultiDataset</code> object. We can set different colours and shapes for the upper and lower plots in the scatterplot matrix, see the <code><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_score.html">plot_samples_score()</a></code> function for more information. For example, we can assess whether the first three principal components show any clustering of the samples according to their cluster computed from genomics similarity, disease status or feedlot (we’ll only show the results for the SNPs dataset here):</p>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_coordinates_pca.html">plot_samples_coordinates_pca</a></span><span class="op">(</span></span>
<span>  <span class="va">pca_runs_list</span>,</span>
<span>  datasets <span class="op">=</span> <span class="st">"snps"</span>,</span>
<span>  pcs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>  mo_data <span class="op">=</span> <span class="va">mo_set_de</span>,</span>
<span>  colour_upper <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span>  shape_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  colour_lower <span class="op">=</span> <span class="st">"feedlot"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.box <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="preprocessing_files/figure-html/plot-samples-coordinates-pca-colours-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section><section id="missing-values-imputation" class="level3" data-number="6.2.3"><h3 data-number="6.2.3" class="anchored" data-anchor-id="missing-values-imputation">
<span class="header-section-number">6.2.3</span> Missing values imputation</h3>
<p>We can check that the complete multi-omics set constructed has no more missing values:</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_load.html">tar_load</a></span><span class="op">(</span><span class="va">mo_set_complete</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mo_set_complete</span></span>
<span><span class="co">#&gt; Object of class 'MultiDataSet'</span></span>
<span><span class="co">#&gt;  . assayData: 3 elements</span></span>
<span><span class="co">#&gt;     . snps: 23036 features, 139 samples </span></span>
<span><span class="co">#&gt;     . rnaseq: 20335 features, 143 samples </span></span>
<span><span class="co">#&gt;     . metabolome: 55 features, 139 samples </span></span>
<span><span class="co">#&gt;  . featureData:</span></span>
<span><span class="co">#&gt;     . snps: 23036 rows, 13 cols (feature_id, ..., p_value)</span></span>
<span><span class="co">#&gt;     . rnaseq: 20335 rows, 15 cols (feature_id, ..., de_signif)</span></span>
<span><span class="co">#&gt;     . metabolome: 55 rows, 16 cols (feature_id, ..., de_signif)</span></span>
<span><span class="co">#&gt;  . rowRanges:</span></span>
<span><span class="co">#&gt;     . snps: YES</span></span>
<span><span class="co">#&gt;     . rnaseq: YES</span></span>
<span><span class="co">#&gt;     . metabolome: NO</span></span>
<span><span class="co">#&gt;  . phenoData:</span></span>
<span><span class="co">#&gt;     . snps: 139 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span><span class="co">#&gt;     . rnaseq: 143 samples, 10 cols (id, ..., geno_comp_3)</span></span>
<span><span class="co">#&gt;     . metabolome: 139 samples, 10 cols (id, ..., geno_comp_3)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/check_missing_values.html">check_missing_values</a></span><span class="op">(</span><span class="va">mo_set_complete</span><span class="op">)</span></span>
<span><span class="co">#&gt; No missing values in snps dataset.</span></span>
<span><span class="co">#&gt; No missing values in rnaseq dataset.</span></span>
<span><span class="co">#&gt; No missing values in metabolome dataset.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="recap-targets-list" class="level2" data-number="6.3"><h2 data-number="6.3" class="anchored" data-anchor-id="recap-targets-list">
<span class="header-section-number">6.3</span> Recap – targets list</h2>
<p>For convenience, here is the list of targets that we created in this section:</p>
<details><summary>
Targets list for datasets preprocessing
</summary><div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="co">## Applying transformations to the datasets</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/transformation_datasets_factory.html">transformation_datasets_factory</a></span><span class="op">(</span></span>
<span>    <span class="va">mo_set_de</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"rnaseq"</span> <span class="op">=</span> <span class="st">"vst-deseq2"</span>,</span>
<span>      <span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"best-normalize-manual"</span><span class="op">)</span>,</span>
<span>    methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"metabolome"</span> <span class="op">=</span> <span class="st">"log_x"</span><span class="op">)</span>,</span>
<span>    a <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>    b <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    standardize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    transformed_data_name <span class="op">=</span> <span class="st">"mo_set_transformed"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Density plot for each transformed dataset</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">density_plots_transformed</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_density_data.html">plot_density_data</a></span><span class="op">(</span></span>
<span>      <span class="va">mo_set_transformed</span>,</span>
<span>      combined <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>      scales <span class="op">=</span> <span class="st">"free"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Plotting the mean-SD trend for transformed each dataset</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">mean_sd_plots_transformed</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_meansd_data.html">plot_meansd_data</a></span><span class="op">(</span><span class="va">mo_set_transformed</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Summary table of the transformations applied</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">transformation_summary</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/get_table_transformations.html">get_table_transformations</a></span><span class="op">(</span><span class="va">transformations_runs_list</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## Running a PCA on each dataset</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/pca_complete_data_factory.html">pca_complete_data_factory</a></span><span class="op">(</span></span>
<span>    <span class="va">mo_set_transformed</span>,</span>
<span>    complete_data_name <span class="op">=</span> <span class="st">"mo_set_complete"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## PCA screeplots</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">pca_screeplots</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_screeplot_pca.html">plot_screeplot_pca</a></span><span class="op">(</span><span class="va">pca_runs_list</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co">## PCA sample plots</span></span>
<span>  <span class="fu"><a href="https://docs.ropensci.org/targets/reference/tar_target.html">tar_target</a></span><span class="op">(</span></span>
<span>    <span class="va">pca_sample_plots</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/moiraine/man/plot_samples_coordinates_pca.html">plot_samples_coordinates_pca</a></span><span class="op">(</span></span>
<span>      <span class="va">pca_runs_list</span>,</span>
<span>      datasets <span class="op">=</span> <span class="st">"snps"</span>,</span>
<span>      pcs <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,</span>
<span>      mo_data <span class="op">=</span> <span class="va">mo_set_de</span>,</span>
<span>      colour_upper <span class="op">=</span> <span class="st">"geno_comp_cluster"</span>,</span>
<span>      shape_upper <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>      colour_lower <span class="op">=</span> <span class="st">"feedlot"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details>

</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./modifying_multidataset.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Modifying the <code>MultiDataSet</code> object</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./prefiltering.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Dataset pre-filtering</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>